{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add some additional layers of information to the hierarchy\n",
    "\n",
    "Requires DDOT for ontology parsing https://github.com/michaelkyu/ddot/blob/master/examples/Tutorial.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load required packages\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import random\n",
    "\n",
    "import ndex2\n",
    "import getpass\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "# latex rendering of text in graphs\n",
    "import matplotlib as mpl\n",
    "mpl.rc('text', usetex = False)\n",
    "mpl.rc('font', family = 'serif')\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'sans-serif'\n",
    "rcParams['font.sans-serif'] = ['Arial']\n",
    "\n",
    "sns.set(font_scale=1.4)\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.set_style(\"ticks\", {\"xtick.major.size\": 15, \"ytick.major.size\": 15})\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "import sys\n",
    "\n",
    "# % matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need ddot to parse the ontology\n",
    "import ddot\n",
    "from ddot import Ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.11'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ddot only compatible with nx 1.11....\n",
    "nx.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load interactome (needs to be the same one used in previous analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['UBE2Q1', 'RNF14', 'UBE2Q2', 'TMCO1', 'UBAC1']\n",
      "number of nodes:\n",
      "18820\n",
      "\n",
      "number of edges:\n",
      "2693250\n"
     ]
    }
   ],
   "source": [
    "\n",
    "interactome_uuid='4de852d9-9908-11e9-bcaf-0ac135e8bacf'\n",
    "ndex_server='public.ndexbio.org'\n",
    "ndex_user=None\n",
    "ndex_password=None\n",
    "G_PC = ndex2.create_nice_cx_from_server(\n",
    "            ndex_server, \n",
    "            username=ndex_user, \n",
    "            password=ndex_password, \n",
    "            uuid=interactome_uuid\n",
    "        ).to_networkx()\n",
    "\n",
    "# relabel nodes.... this is a nx 1.11 thing...\n",
    "G_PC_data = pd.DataFrame.from_dict(dict(G_PC.nodes(data=True)), orient='index')\n",
    "G_PC = nx.relabel_nodes(G_PC,dict(zip(G_PC_data.index.tolist(),G_PC_data['name'].tolist())))\n",
    "print(G_PC.nodes()[0:5])\n",
    "nodes = G_PC.nodes()\n",
    "\n",
    "# print out interactome num nodes and edges for diagnostic purposes\n",
    "print('number of nodes:')\n",
    "print(len(G_PC.nodes()))\n",
    "print('\\nnumber of edges:')\n",
    "print(len(G_PC.edges()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load ASD rare-common hierarchy from ndex\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDEx password: ········\n",
      "number of nodes:\n",
      "73\n",
      "\n",
      "number of edges:\n",
      "74\n"
     ]
    }
   ],
   "source": [
    "ASD_uuid='0ecb653c-9405-11eb-9e72-0ac135e8bacf'\n",
    "ndex_server='public.ndexbio.org'\n",
    "ndex_user='sbrosenthal@health.ucsd.edu'\n",
    "ndex_password=getpass.getpass('NDEx password: ')\n",
    "G_ASD = ndex2.create_nice_cx_from_server(\n",
    "            ndex_server, \n",
    "            username=ndex_user, \n",
    "            password=ndex_password, \n",
    "            uuid=ASD_uuid\n",
    "        ).to_networkx()\n",
    "nodes = G_ASD.nodes()\n",
    "\n",
    "# print out interactome num nodes and edges for diagnostic purposes\n",
    "print('number of nodes:')\n",
    "print(len(G_ASD.nodes()))\n",
    "print('\\nnumber of edges:')\n",
    "print(len(G_ASD.edges()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(6548,\n",
       "  {'CD_MemberList': 'NSG2 PSD2 SLC17A6 STXBP5L',\n",
       "   'CD_AnnotatedMembers_Size': '1',\n",
       "   'CD_AnnotatedMembers_Pvalue': '0.007995642253764',\n",
       "   'CD_AnnotatedMembers': 'SLC17A6',\n",
       "   'CD_CommunityName': 'Organic anion transporters Homo sapiens R-HSA-428643',\n",
       "   'CD_AnnotatedMembers_Overlap': '0.25',\n",
       "   'HiDeF_persistence': '5',\n",
       "   'CD_MemberList_LogSize': '2.0',\n",
       "   'CD_Labeled': 'true',\n",
       "   'CD_AnnotatedAlgorithm': 'Annotated by Enrichr [Docker: coleslawndex/cdenrichrgenestoterm:0.4.0] {{--maxpval=0.05, --genesets=GO_Biological_Process_2018,GO_Cellular_Component_2018,GO_Molecular_Function_2018,KEGG_2019_Human,Reactome_2016}} via CyCommunityDetection Cytoscape App (1.11.0)',\n",
       "   'CD_NonAnnotatedMembers': 'NSG2 PSD2 STXBP5L',\n",
       "   'CD_AnnotatedMembers_SourceDB': 'Reactome_2016',\n",
       "   'CD_MemberList_Size': '4',\n",
       "   'name': 'C435'}),\n",
       " (6547,\n",
       "  {'CD_MemberList': 'EN1 GSX2 IRX3 NKX6-2',\n",
       "   'CD_AnnotatedMembers_Size': '1',\n",
       "   'CD_AnnotatedMembers_Pvalue': '0.0249546954540196',\n",
       "   'CD_AnnotatedMembers': 'NKX6-2',\n",
       "   'CD_CommunityName': 'regulation of myelination (GO:0031641)',\n",
       "   'CD_AnnotatedMembers_Overlap': '0.25',\n",
       "   'HiDeF_persistence': '5',\n",
       "   'CD_MemberList_LogSize': '2.0',\n",
       "   'CD_Labeled': 'true',\n",
       "   'CD_AnnotatedAlgorithm': 'Annotated by Enrichr [Docker: coleslawndex/cdenrichrgenestoterm:0.4.0] {{--maxpval=0.05, --genesets=GO_Biological_Process_2018,GO_Cellular_Component_2018,GO_Molecular_Function_2018,KEGG_2019_Human,Reactome_2016}} via CyCommunityDetection Cytoscape App (1.11.0)',\n",
       "   'CD_NonAnnotatedMembers': 'EN1 GSX2 IRX3',\n",
       "   'CD_AnnotatedMembers_SourceDB': 'GO_Biological_Process_2018',\n",
       "   'CD_MemberList_Size': '4',\n",
       "   'name': 'C434'})]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G_ASD.nodes(data=True)[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_AnnotatedAlgorithm</th>\n",
       "      <th>CD_NonAnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_SourceDB</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C435</th>\n",
       "      <td>NSG2 PSD2 SLC17A6 STXBP5L</td>\n",
       "      <td>1</td>\n",
       "      <td>0.007995642253764</td>\n",
       "      <td>SLC17A6</td>\n",
       "      <td>Organic anion transporters Homo sapiens R-HSA-...</td>\n",
       "      <td>0.25</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>NSG2 PSD2 STXBP5L</td>\n",
       "      <td>Reactome_2016</td>\n",
       "      <td>4</td>\n",
       "      <td>C435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C434</th>\n",
       "      <td>EN1 GSX2 IRX3 NKX6-2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0249546954540196</td>\n",
       "      <td>NKX6-2</td>\n",
       "      <td>regulation of myelination (GO:0031641)</td>\n",
       "      <td>0.25</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>EN1 GSX2 IRX3</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>4</td>\n",
       "      <td>C434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C433</th>\n",
       "      <td>CALM3 FSD1 GNAZ SRC</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0034578149777732</td>\n",
       "      <td>SRC CALM3</td>\n",
       "      <td>GnRH signaling pathway</td>\n",
       "      <td>0.5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>FSD1 GNAZ</td>\n",
       "      <td>KEGG_2019_Human</td>\n",
       "      <td>4</td>\n",
       "      <td>C433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C432</th>\n",
       "      <td>ABCC8 CDK5R2 KIF1A MKRN3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0137533444679581</td>\n",
       "      <td>ABCC8</td>\n",
       "      <td>ABC transporters</td>\n",
       "      <td>0.25</td>\n",
       "      <td>10</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>CDK5R2 KIF1A MKRN3</td>\n",
       "      <td>KEGG_2019_Human</td>\n",
       "      <td>4</td>\n",
       "      <td>C432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C431</th>\n",
       "      <td>KIF1A MKRN3 TBC1D14 VPS41</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0016952190930884</td>\n",
       "      <td>VPS41 TBC1D14</td>\n",
       "      <td>regulation of vesicle fusion (GO:0031338)</td>\n",
       "      <td>0.5</td>\n",
       "      <td>7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>KIF1A MKRN3</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>4</td>\n",
       "      <td>C431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CD_MemberList CD_AnnotatedMembers_Size  \\\n",
       "name                                                       \n",
       "C435  NSG2 PSD2 SLC17A6 STXBP5L                        1   \n",
       "C434       EN1 GSX2 IRX3 NKX6-2                        1   \n",
       "C433        CALM3 FSD1 GNAZ SRC                        2   \n",
       "C432   ABCC8 CDK5R2 KIF1A MKRN3                        1   \n",
       "C431  KIF1A MKRN3 TBC1D14 VPS41                        2   \n",
       "\n",
       "     CD_AnnotatedMembers_Pvalue CD_AnnotatedMembers  \\\n",
       "name                                                  \n",
       "C435          0.007995642253764             SLC17A6   \n",
       "C434         0.0249546954540196              NKX6-2   \n",
       "C433         0.0034578149777732           SRC CALM3   \n",
       "C432         0.0137533444679581               ABCC8   \n",
       "C431         0.0016952190930884       VPS41 TBC1D14   \n",
       "\n",
       "                                       CD_CommunityName  \\\n",
       "name                                                      \n",
       "C435  Organic anion transporters Homo sapiens R-HSA-...   \n",
       "C434             regulation of myelination (GO:0031641)   \n",
       "C433                             GnRH signaling pathway   \n",
       "C432                                   ABC transporters   \n",
       "C431          regulation of vesicle fusion (GO:0031338)   \n",
       "\n",
       "     CD_AnnotatedMembers_Overlap  HiDeF_persistence CD_MemberList_LogSize  \\\n",
       "name                                                                        \n",
       "C435                        0.25                  5                   2.0   \n",
       "C434                        0.25                  5                   2.0   \n",
       "C433                         0.5                  5                   2.0   \n",
       "C432                        0.25                 10                   2.0   \n",
       "C431                         0.5                  7                   2.0   \n",
       "\n",
       "     CD_Labeled                              CD_AnnotatedAlgorithm  \\\n",
       "name                                                                 \n",
       "C435       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C434       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C433       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C432       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C431       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "\n",
       "     CD_NonAnnotatedMembers CD_AnnotatedMembers_SourceDB  CD_MemberList_Size  \\\n",
       "name                                                                           \n",
       "C435      NSG2 PSD2 STXBP5L                Reactome_2016                   4   \n",
       "C434          EN1 GSX2 IRX3   GO_Biological_Process_2018                   4   \n",
       "C433              FSD1 GNAZ              KEGG_2019_Human                   4   \n",
       "C432     CDK5R2 KIF1A MKRN3              KEGG_2019_Human                   4   \n",
       "C431            KIF1A MKRN3   GO_Biological_Process_2018                   4   \n",
       "\n",
       "      name  \n",
       "name        \n",
       "C435  C435  \n",
       "C434  C434  \n",
       "C433  C433  \n",
       "C432  C432  \n",
       "C431  C431  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add node attributes to dataframe for easier access\n",
    "ASD_clust_df = pd.DataFrame.from_dict(dict(G_ASD.nodes(data=True)), orient='index')\n",
    "# relabel nodes in G_ASD\n",
    "G_ASD=nx.relabel_nodes(G_ASD,dict(ASD_clust_df['name']))\n",
    "ASD_clust_df.index = ASD_clust_df['name']\n",
    "# some columns are not the right type\n",
    "ASD_clust_df['CD_MemberList_Size']=[int(x) for x in ASD_clust_df['CD_MemberList_Size'].tolist()]\n",
    "ASD_clust_df['HiDeF_persistence']=[int(x) for x in ASD_clust_df['HiDeF_persistence'].tolist()]\n",
    "ASD_clust_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add some annotation the the systems \n",
    "\n",
    "- Fraction rare/common/neither seeds\n",
    "\n",
    "\n",
    "## Do we need to pull down the other network for this?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_AnnotatedAlgorithm</th>\n",
       "      <th>CD_NonAnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_SourceDB</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C435</th>\n",
       "      <td>NSG2 PSD2 SLC17A6 STXBP5L</td>\n",
       "      <td>1</td>\n",
       "      <td>0.007995642253764</td>\n",
       "      <td>SLC17A6</td>\n",
       "      <td>Organic anion transporters Homo sapiens R-HSA-...</td>\n",
       "      <td>0.25</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>NSG2 PSD2 STXBP5L</td>\n",
       "      <td>Reactome_2016</td>\n",
       "      <td>4</td>\n",
       "      <td>C435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C434</th>\n",
       "      <td>EN1 GSX2 IRX3 NKX6-2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0249546954540196</td>\n",
       "      <td>NKX6-2</td>\n",
       "      <td>regulation of myelination (GO:0031641)</td>\n",
       "      <td>0.25</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>EN1 GSX2 IRX3</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>4</td>\n",
       "      <td>C434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C433</th>\n",
       "      <td>CALM3 FSD1 GNAZ SRC</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0034578149777732</td>\n",
       "      <td>SRC CALM3</td>\n",
       "      <td>GnRH signaling pathway</td>\n",
       "      <td>0.5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>FSD1 GNAZ</td>\n",
       "      <td>KEGG_2019_Human</td>\n",
       "      <td>4</td>\n",
       "      <td>C433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C432</th>\n",
       "      <td>ABCC8 CDK5R2 KIF1A MKRN3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0137533444679581</td>\n",
       "      <td>ABCC8</td>\n",
       "      <td>ABC transporters</td>\n",
       "      <td>0.25</td>\n",
       "      <td>10</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>CDK5R2 KIF1A MKRN3</td>\n",
       "      <td>KEGG_2019_Human</td>\n",
       "      <td>4</td>\n",
       "      <td>C432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C431</th>\n",
       "      <td>KIF1A MKRN3 TBC1D14 VPS41</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0016952190930884</td>\n",
       "      <td>VPS41 TBC1D14</td>\n",
       "      <td>regulation of vesicle fusion (GO:0031338)</td>\n",
       "      <td>0.5</td>\n",
       "      <td>7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>KIF1A MKRN3</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>4</td>\n",
       "      <td>C431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CD_MemberList CD_AnnotatedMembers_Size  \\\n",
       "name                                                       \n",
       "C435  NSG2 PSD2 SLC17A6 STXBP5L                        1   \n",
       "C434       EN1 GSX2 IRX3 NKX6-2                        1   \n",
       "C433        CALM3 FSD1 GNAZ SRC                        2   \n",
       "C432   ABCC8 CDK5R2 KIF1A MKRN3                        1   \n",
       "C431  KIF1A MKRN3 TBC1D14 VPS41                        2   \n",
       "\n",
       "     CD_AnnotatedMembers_Pvalue CD_AnnotatedMembers  \\\n",
       "name                                                  \n",
       "C435          0.007995642253764             SLC17A6   \n",
       "C434         0.0249546954540196              NKX6-2   \n",
       "C433         0.0034578149777732           SRC CALM3   \n",
       "C432         0.0137533444679581               ABCC8   \n",
       "C431         0.0016952190930884       VPS41 TBC1D14   \n",
       "\n",
       "                                       CD_CommunityName  \\\n",
       "name                                                      \n",
       "C435  Organic anion transporters Homo sapiens R-HSA-...   \n",
       "C434             regulation of myelination (GO:0031641)   \n",
       "C433                             GnRH signaling pathway   \n",
       "C432                                   ABC transporters   \n",
       "C431          regulation of vesicle fusion (GO:0031338)   \n",
       "\n",
       "     CD_AnnotatedMembers_Overlap  HiDeF_persistence CD_MemberList_LogSize  \\\n",
       "name                                                                        \n",
       "C435                        0.25                  5                   2.0   \n",
       "C434                        0.25                  5                   2.0   \n",
       "C433                         0.5                  5                   2.0   \n",
       "C432                        0.25                 10                   2.0   \n",
       "C431                         0.5                  7                   2.0   \n",
       "\n",
       "     CD_Labeled                              CD_AnnotatedAlgorithm  \\\n",
       "name                                                                 \n",
       "C435       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C434       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C433       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C432       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "C431       true  Annotated by Enrichr [Docker: coleslawndex/cde...   \n",
       "\n",
       "     CD_NonAnnotatedMembers CD_AnnotatedMembers_SourceDB  CD_MemberList_Size  \\\n",
       "name                                                                           \n",
       "C435      NSG2 PSD2 STXBP5L                Reactome_2016                   4   \n",
       "C434          EN1 GSX2 IRX3   GO_Biological_Process_2018                   4   \n",
       "C433              FSD1 GNAZ              KEGG_2019_Human                   4   \n",
       "C432     CDK5R2 KIF1A MKRN3              KEGG_2019_Human                   4   \n",
       "C431            KIF1A MKRN3   GO_Biological_Process_2018                   4   \n",
       "\n",
       "      name  \n",
       "name        \n",
       "C435  C435  \n",
       "C434  C434  \n",
       "C433  C433  \n",
       "C432  C432  \n",
       "C431  C431  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ASD_clust_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# System validation with mammalian phenotype ontology + mouse KO data\n",
    "\n",
    "Parse the ontology, data from http://www.informatics.jax.org/vocab/mp_ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import hypergeom\n",
    "from statsmodels.stats import contingency_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI_Allele_Accession_ID</th>\n",
       "      <th>Allele symbol</th>\n",
       "      <th>involves</th>\n",
       "      <th>MP</th>\n",
       "      <th>PMID</th>\n",
       "      <th>MGI_marker_accession_ID</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000600</td>\n",
       "      <td>12529408</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001716</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001698</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001092</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000961</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           MGI_Allele_Accession_ID Allele symbol               involves  \\\n",
       "gene_name                                                                 \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "\n",
       "                   MP      PMID MGI_marker_accession_ID gene_name  \n",
       "gene_name                                                          \n",
       "Rb1        MP:0000600  12529408               MGI:97874       Rb1  \n",
       "Rb1        MP:0001716  16449662               MGI:97874       Rb1  \n",
       "Rb1        MP:0001698  16449662               MGI:97874       Rb1  \n",
       "Rb1        MP:0001092  16449662               MGI:97874       Rb1  \n",
       "Rb1        MP:0000961  16449662               MGI:97874       Rb1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_df = pd.read_csv('../data/MGI/MGI_PhenoGenoMP.rpt',sep='\\t',\n",
    "                    names=['MGI_Allele_Accession_ID','Allele symbol','involves','MP','PMID','MGI_marker_accession_ID'])\n",
    "gene_name = [a.split('<')[0] for a in mgi_df['Allele symbol'].tolist()]\n",
    "mgi_df['gene_name']=gene_name\n",
    "mgi_df.index=mgi_df['gene_name']\n",
    "mgi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddot.parse_obo('../data/MGI/mp.obo',\n",
    "               '../data/MGI/parsed_mp.txt',\n",
    "              '../data/MGI/id2name_mp.txt',\n",
    "              '../data/MGI/id2namespace_mp.txt',\n",
    "              '../data/MGI/altID_mp.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14338\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0000001</th>\n",
       "      <td>mammalian phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000002</th>\n",
       "      <td>obsolete Morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000003</th>\n",
       "      <td>abnormal adipose tissue morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000011</th>\n",
       "      <td>abnormal adipose tissue morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000005</th>\n",
       "      <td>increased brown adipose tissue amount</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      description\n",
       "MP                                               \n",
       "MP:0000001                    mammalian phenotype\n",
       "MP:0000002                    obsolete Morphology\n",
       "MP:0000003     abnormal adipose tissue morphology\n",
       "MP:0000011     abnormal adipose tissue morphology\n",
       "MP:0000005  increased brown adipose tissue amount"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP2desc = pd.read_csv('../data/MGI/id2name_mp.txt',sep='\\t',\n",
    "                      names=['MP','description'],index_col='MP')\n",
    "\n",
    "MP2desc=MP2desc.loc[MP2desc.index.dropna()] # drop NAN from index\n",
    "print(len(MP2desc))\n",
    "\n",
    "\n",
    "MP2desc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Child</th>\n",
       "      <th>Relation</th>\n",
       "      <th>Namespace</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MP:0005375</td>\n",
       "      <td>MP:0000003</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MP:0001778</td>\n",
       "      <td>MP:0000005</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MP:0001781</td>\n",
       "      <td>MP:0000008</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MP:0005334</td>\n",
       "      <td>MP:0000010</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MP:0000003</td>\n",
       "      <td>MP:0000013</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Parent       Child Relation        Namespace\n",
       "0  MP:0005375  MP:0000003     is_a  MPheno.ontology\n",
       "1  MP:0001778  MP:0000005     is_a  MPheno.ontology\n",
       "2  MP:0001781  MP:0000008     is_a  MPheno.ontology\n",
       "3  MP:0005334  MP:0000010     is_a  MPheno.ontology\n",
       "4  MP:0000003  MP:0000013     is_a  MPheno.ontology"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "hierarchy = pd.read_table('../data/MGI/parsed_mp.txt',\n",
    "                          sep='\\t',\n",
    "                          header=None,\n",
    "                          names=['Parent', 'Child', 'Relation', 'Namespace'])\n",
    "\n",
    "hierarchy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0 genes, 13205 terms, 0 gene-term relations, 16647 term-term relations\n",
       "node_attributes: []\n",
       "edge_attributes: ['Relation', 'Namespace']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MPO = Ontology.from_table(\n",
    "    table=hierarchy,\n",
    "    parent='Parent',\n",
    "    child='Child',\n",
    "    add_root_name='MP:00SUPER',\n",
    "    ignore_orphan_terms=True)\n",
    "#MPO.clear_node_attr()\n",
    "#MPO.clear_edge_attr()\n",
    "MPO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13205\n"
     ]
    }
   ],
   "source": [
    "print(len(MPO.node_attr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abnormal nervous system tract morphology\n",
      "hippocampal neuron degeneration\n",
      "abnormal motor neuron morphology\n",
      "motor neuron degeneration\n",
      "decreased motor neuron number\n",
      "abnormal motor neuron innervation pattern\n",
      "peripheral nervous system degeneration\n",
      "abnormal sensory neuron morphology\n",
      "decreased sensory neuron number\n",
      "abnormal sensory neuron innervation pattern\n",
      "abnormal hair guard neuron morphology\n",
      "absent hair guard neurons\n",
      "abnormal hair-tylotrich neuron morphology\n",
      "abnormal hair-down neuron morphology\n",
      "absent hair-down neurons\n",
      "abnormal sympathetic neuron morphology\n",
      "abnormal adrenergic neuron morphology\n",
      "abnormal peptidergic neuron morphology\n",
      "abnormal parasympathetic neuron morphology\n",
      "abnormal cholinergic neuron morphology\n",
      "abnormal enteric nervous system morphology\n",
      "abnormal enteric neuron morphology\n",
      "abnormal enteric cholinergic neuron morphology\n",
      "absent enteric neurons\n",
      "failure of neuromuscular synapse presynaptic differentiation\n",
      "failure of neuromuscular synapse postsynaptic differentiation\n",
      "impaired synaptic plasticity\n",
      "absence of NMDA-mediated synaptic currents\n",
      "reduced NMDA-mediated synaptic currents\n",
      "reduced NMDA receptor mediated synaptic activity in barrel cortex\n",
      "abnormal CNS synaptic transmission\n",
      "abnormal nervous system electrophysiology\n",
      "abnormal autonomic nervous system morphology\n",
      "abnormal somatic nervous system morphology\n",
      "abnormal neuron morphology\n",
      "abnormal AMPA-mediated synaptic currents\n",
      "abnormal NMDA-mediated synaptic currents\n",
      "abnormal excitatory postsynaptic currents\n",
      "abnormal inhibitory postsynaptic potential\n",
      "abnormal excitatory postsynaptic potential\n",
      "abnormal PNS synaptic transmission\n",
      "abnormal synaptic depression\n",
      "increased synaptic depression\n",
      "decreased synaptic depression\n",
      "abnormal inhibitory postsynaptic currents\n",
      "abnormal neuron specification\n",
      "abnormal neuron apoptosis\n",
      "increased neuron apoptosis\n",
      "decreased neuron apoptosis\n",
      "neuron degeneration\n",
      "enhanced NMDA-mediated synaptic currents\n",
      "loss of hippocampal neurons\n",
      "loss of cortex neurons\n",
      "loss of basal ganglia neurons\n",
      "abnormal dopaminergic neuron morphology\n",
      "loss of dopaminergic neurons\n",
      "abnormal GABAergic neuron morphology\n",
      "loss of GABAergic neurons\n",
      "abnormal glutaminergic neuron morphology\n",
      "loss of glutamate neurons\n",
      "abnormal central nervous system regeneration\n",
      "premature neuronal precursor differentiation\n",
      "nervous system phenotype\n",
      "abnormal nervous system morphology\n",
      "abnormal nervous system physiology\n",
      "abnormal synaptic transmission\n",
      "abnormal nervous system development\n",
      "abnormal serotonergic neuron morphology\n",
      "abnormal spinal cord interneuron morphology\n",
      "abnormal brain interneuron morphology\n",
      "abnormal sympathetic nervous system physiology\n",
      "neuronal intranuclear inclusions\n",
      "abnormal peripheral nervous system regeneration\n",
      "abnormal proprioceptive neuron morphology\n",
      "abnormal synaptic glutamate release\n",
      "decreased synaptic glutamate release\n",
      "abnormal inner hair cell synaptic ribbon morphology\n",
      "absent active-zone-anchored inner hair cell synaptic ribbon\n",
      "abnormal miniature excitatory postsynaptic currents\n",
      "abnormal synaptic vesicle morphology\n",
      "abnormal synaptic vesicle recycling\n",
      "abnormal synaptic vesicle number\n",
      "abnormal synaptic vesicle clustering\n",
      "abnormal neuron physiology\n",
      "abnormal synaptic acetylcholine release\n",
      "abnormal nervous system regeneration\n",
      "abnormal synaptic plasticity\n",
      "abnormal neuronal precursor proliferation\n",
      "absent neuronal precursor cells\n",
      "abnormal neuronal precursor cell number\n",
      "increased neuronal precursor cell number\n",
      "decreased neuronal precursor cell number\n",
      "increased motor neuron number\n",
      "abnormal somatic nervous system physiology\n",
      "abnormal synaptic epinephrine release\n",
      "abnormal synaptic norepinephrine release\n",
      "neuron hypertrophy\n",
      "increased sensory neuron number\n",
      "abnormal neuronal migration\n",
      "abnormal retinal neuronal layer morphology\n",
      "abnormal olfactory sensory neuron morphology\n",
      "abnormal autonomic nervous system physiology\n",
      "abnormal parasympathetic nervous system physiology\n",
      "increased susceptibility to neuronal excitotoxicity\n",
      "decreased susceptibility to neuronal excitotoxicity\n",
      "abnormal medium spiny neuron morphology\n",
      "absent proprioceptive neurons\n",
      "abnormal synaptic bouton morphology\n",
      "nervous\n",
      "abnormal neuron number\n",
      "increased neuron number\n",
      "decreased neuron number\n",
      "abnormal sympathetic neuron physiology\n",
      "abnormal sympathetic neuron innervation pattern\n",
      "abnormal miniature inhibitory postsynaptic currents\n",
      "abnormal synaptonemal complex\n",
      "abnormal neuron differentiation\n",
      "abnormal hippocampus neuron morphology\n",
      "abnormal olfactory bulb interneuron morphology\n",
      "abnormal neuron proliferation\n",
      "abnormal sensory neuron physiology\n",
      "abnormal synaptic dopamine release\n",
      "increased nervous system tumor incidence\n",
      "abnormal thalamus neuron morphology\n",
      "abnormal glycinergic neuron morphology\n",
      "absence of AMPA-mediated synaptic currents\n",
      "enhanced AMPA-mediated synaptic currents\n",
      "reduced AMPA-mediated synaptic currents\n",
      "abnormal excitatory postsynaptic current amplitude\n",
      "biphasic excitatory postsynaptic current amplitude\n",
      "increased excitatory postsynaptic current amplitude\n",
      "decreased excitatory postsynaptic current amplitude\n",
      "prolonged excitatory postsynaptic current rise time\n",
      "short excitatory postsynaptic current rise time\n",
      "prolonged excitatory postsynaptic current decay time\n",
      "short excitatory postsynaptic current decay time\n",
      "decreased dopaminergic neuron number\n",
      "increased dopaminergic neuron number\n",
      "ectopic dopaminergic neuron\n",
      "increased susceptibility to dopaminergic neuron neurotoxicity\n",
      "decreased susceptibility to dopaminergic neuron neurotoxicity\n",
      "abnormal olfactory sensory neuron physiology\n",
      "abnormal vomeronasal sensory neuron physiology\n",
      "abnormal vomeronasal sensory neuron morphology\n",
      "ectopic neuron\n",
      "ectopic cortical neuron\n",
      "neuronal leptomeningeal ectopia\n",
      "abnormal X-Y chromosome synaptonemal complex assembly during male meiosis\n",
      "impaired neuron differentiation\n",
      "neuronal cytoplasmic inclusions\n",
      "abnormal GABAergic neuron physiology\n",
      "abnormal olfactory neuron innervation pattern\n",
      "abnormal neuronal stem cell morphology\n",
      "abnormal neuronal stem cell physiology\n",
      "abnormal neuronal stem cell self-renewal\n",
      "increased neuronal stem cell self-renewal\n",
      "decreased neuronal stem cell self-renewal\n",
      "decreased nervous system tumor incidence\n",
      "nervous system inclusion bodies\n",
      "abnormal dentate gyrus neuron dendrite morphology\n",
      "abnormal motor neuron dendrite morphology\n",
      "abnormal pyramidal neuron dendrite morphology\n",
      "abnormal hippocampal pyramidal neuron dendrite morphology\n",
      "decreased olfactory sensory neuron number\n",
      "increased neuronal precursor proliferation\n",
      "decreased neuronal precursor proliferation\n",
      "abnormal postsynaptic density morphology\n",
      "abnormal neuron polarity\n",
      "abnormal nervous system dopamine level\n",
      "decreased nervous system dopamine level\n",
      "increased nervous system dopamine level\n",
      "impaired central nervous system regeneration\n",
      "improved central nervous system regeneration\n",
      "abnormal synaptic vesicle exocytosis\n",
      "abnormal neuronal precursor cell migration\n",
      "175\n"
     ]
    }
   ],
   "source": [
    "# find terms related to brain\n",
    "MP_focal_list = []\n",
    "for t in MPO.node_attr.index.tolist():\n",
    "    descr_temp = MPO.node_attr.loc[t]['description']\n",
    "    if descr_temp.find('nervous')>-1:\n",
    "        print(descr_temp)\n",
    "        MP_focal_list.append(t)\n",
    "    elif descr_temp.find('neuron')>-1:\n",
    "        print(descr_temp)\n",
    "        MP_focal_list.append(t)\n",
    "    elif descr_temp.find('synapt')>-1:\n",
    "        print(descr_temp)\n",
    "        MP_focal_list.append(t)\n",
    "        \n",
    "print(len(MP_focal_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal nervous system tract morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 188\n",
      "number overlapping genes = 15\n",
      "[[15, 173], [277, 18355]]\n",
      "2.335897013142731e-10\n",
      "(1.2076770372746504, 2.289119755421405)\n",
      "1.7483983963480276\n",
      "4.122755360272459e-08\n",
      "\n",
      "hippocampal neuron degeneration\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "[[1, 24], [291, 18504]]\n",
      "0.3405477911699869\n",
      "(-1.029363442050749, 2.9780936548344226)\n",
      "0.9743651063918368\n",
      "0.05694876545560805\n",
      "\n",
      "abnormal motor neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 166\n",
      "number overlapping genes = 10\n",
      "[[10, 156], [282, 18372]]\n",
      "1.6354482238281865e-05\n",
      "(0.7793266760396684, 2.079483444464491)\n",
      "1.4294050602520798\n",
      "5.8329668688649927e-05\n",
      "\n",
      "motor neuron degeneration\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 3\n",
      "[[3, 39], [289, 18489]]\n",
      "0.008125918692104933\n",
      "(0.41351907905777385, 2.7735913090549937)\n",
      "1.5935551940563837\n",
      "0.003996748903698174\n",
      "\n",
      "decreased motor neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 73\n",
      "number overlapping genes = 6\n",
      "[[6, 67], [286, 18461]]\n",
      "4.5515556678055635e-05\n",
      "(0.9111532384624295, 2.5978281951131827)\n",
      "1.754490716787806\n",
      "0.00013680884561816304\n",
      "\n",
      "abnormal motor neuron innervation pattern\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 4\n",
      "[[4, 40], [288, 18488]]\n",
      "0.000426546568415919\n",
      "(0.8249489825023488, 2.8937141751461013)\n",
      "1.859331578824225\n",
      "0.0005749613089311303\n",
      "\n",
      "abnormal sensory neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 465\n",
      "number overlapping genes = 20\n",
      "[[20, 445], [272, 18083]]\n",
      "3.7220482265705346e-06\n",
      "(0.6308584719642751, 1.5583084769891893)\n",
      "1.0945834744767322\n",
      "1.315413298254924e-05\n",
      "\n",
      "decreased sensory neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 173\n",
      "number overlapping genes = 6\n",
      "[[6, 167], [286, 18361]]\n",
      "0.046481217211136794\n",
      "(0.013026684152297019, 1.6584892656749726)\n",
      "0.8357579749136348\n",
      "0.01856692266611126\n",
      "\n",
      "abnormal sensory neuron innervation pattern\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 93\n",
      "number overlapping genes = 11\n",
      "[[11, 82], [281, 18446]]\n",
      "2.751843610957206e-11\n",
      "(1.5351511802507036, 2.815697179741369)\n",
      "2.1754241799960363\n",
      "2.112898275196241e-08\n",
      "\n",
      "abnormal sympathetic neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "[[0, 13], [292, 18515]]\n",
      "0.5365260701148018\n",
      "(-1.9354870751207875, 3.7184593812366846)\n",
      "0.8914861530579485\n",
      "0.18400477242536364\n",
      "\n",
      "abnormal cholinergic neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "[[0, 12], [292, 18516]]\n",
      "0.5012202081348538\n",
      "(-1.8597423250804863, 3.8029080641503543)\n",
      "0.971582869534934\n",
      "0.17113640958677187\n",
      "\n",
      "abnormal enteric nervous system morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 1\n",
      "[[1, 34], [291, 18494]]\n",
      "0.5382413825341976\n",
      "(-1.3664288203052954, 2.6174645050051533)\n",
      "0.6255178423499288\n",
      "0.10224364118157965\n",
      "\n",
      "abnormal enteric neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 0\n",
      "[[0, 27], [292, 18501]]\n",
      "0.9109044480571389\n",
      "(-2.639900132264657, 2.9595845619187493)\n",
      "0.1598422148270462\n",
      "0.3445898732963503\n",
      "\n",
      "impaired synaptic plasticity\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 2\n",
      "[[2, 45], [290, 18483]]\n",
      "0.15099770965027615\n",
      "(-0.379898076680528, 2.4623189516641006)\n",
      "1.0412104374917863\n",
      "0.03631382441942184\n",
      "\n",
      "abnormal CNS synaptic transmission\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 748\n",
      "number overlapping genes = 45\n",
      "[[45, 703], [247, 17825]]\n",
      "4.064209561976624e-20\n",
      "(1.2037808533999876, 1.8567681598283587)\n",
      "1.5302745066141732\n",
      "1.0445756312808588e-15\n",
      "\n",
      "abnormal nervous system electrophysiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 320\n",
      "number overlapping genes = 34\n",
      "[[34, 286], [258, 18242]]\n",
      "1.3760453133544927e-28\n",
      "(1.7527052894220914, 2.5050767818075004)\n",
      "2.128891035614796\n",
      "8.49558494868396e-20\n",
      "\n",
      "abnormal autonomic nervous system morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 94\n",
      "number overlapping genes = 1\n",
      "[[1, 93], [291, 18435]]\n",
      "0.7030465504051107\n",
      "(-2.357789317269975, 1.5899564179569667)\n",
      "-0.3839164496565042\n",
      "0.42976686242691314\n",
      "\n",
      "abnormal somatic nervous system morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 980\n",
      "number overlapping genes = 47\n",
      "[[47, 933], [245, 17595]]\n",
      "2.7690390522369767e-15\n",
      "(0.9668742468238668, 1.60483423332348)\n",
      "1.2858542400736734\n",
      "1.2073762371556662e-12\n",
      "\n",
      "abnormal neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 1472\n",
      "number overlapping genes = 69\n",
      "[[69, 1403], [223, 17125]]\n",
      "3.193956347682084e-21\n",
      "(1.0534308536340715, 1.6042917786961413)\n",
      "1.3288613161651064\n",
      "1.0766968590421017e-17\n",
      "\n",
      "abnormal AMPA-mediated synaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 1\n",
      "[[1, 30], [291, 18498]]\n",
      "0.460853435693958\n",
      "(-1.244827183183422, 2.7466216797445355)\n",
      "0.7508972482805567\n",
      "0.08313763931828339\n",
      "\n",
      "abnormal NMDA-mediated synaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 1\n",
      "[[1, 21], [291, 18507]]\n",
      "0.2797933789357876\n",
      "(-0.9013676455559336, 3.117484871521599)\n",
      "1.1080586129828327\n",
      "0.04519231859009093\n",
      "\n",
      "abnormal excitatory postsynaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 127\n",
      "number overlapping genes = 7\n",
      "[[7, 120], [285, 18408]]\n",
      "0.0007464658653372937\n",
      "(0.5554449064697811, 2.097494808969857)\n",
      "1.326469857719819\n",
      "0.000830854165117607\n",
      "\n",
      "abnormal excitatory postsynaptic potential\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 103\n",
      "number overlapping genes = 8\n",
      "[[8, 95], [284, 18433]]\n",
      "5.262667577612935e-06\n",
      "(0.9674934081085892, 2.4294830467199353)\n",
      "1.6984882274142623\n",
      "3.231079009871059e-05\n",
      "\n",
      "abnormal PNS synaptic transmission\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 36\n",
      "number overlapping genes = 3\n",
      "[[3, 33], [289, 18495]]\n",
      "0.0036588373717520327\n",
      "(0.5733336152701582, 2.948533871441345)\n",
      "1.7609337433557517\n",
      "0.002261841753386615\n",
      "\n",
      "abnormal synaptic depression\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 112\n",
      "number overlapping genes = 6\n",
      "[[6, 106], [286, 18422]]\n",
      "0.0022726866371396456\n",
      "(0.4628927161001034, 2.12436617558853)\n",
      "1.2936294458443167\n",
      "0.001823143556997791\n",
      "\n",
      "increased synaptic depression\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 2\n",
      "[[2, 13], [290, 18515]]\n",
      "0.0027104744399343477\n",
      "(0.7914420451857083, 3.7778647417454945)\n",
      "2.2846533934656015\n",
      "0.0014647520335292746\n",
      "\n",
      "decreased synaptic depression\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 1\n",
      "[[1, 13], [291, 18515]]\n",
      "0.12655656220219816\n",
      "(-0.44918035343213214, 3.625308090861498)\n",
      "1.5880638687146829\n",
      "0.019304507419965578\n",
      "\n",
      "abnormal inhibitory postsynaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 75\n",
      "number overlapping genes = 7\n",
      "[[7, 68], [285, 18460]]\n",
      "2.2832736952120862e-06\n",
      "(1.1105354423791176, 2.6840141009556797)\n",
      "1.8972747716673988\n",
      "2.0916536422439743e-05\n",
      "\n",
      "abnormal neuron specification\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 2\n",
      "[[2, 12], [290, 18516]]\n",
      "0.0020223405010326443\n",
      "(0.8633158438301471, 3.8661843760550267)\n",
      "2.364750109942587\n",
      "0.0011853434646458601\n",
      "\n",
      "abnormal neuron apoptosis\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 217\n",
      "number overlapping genes = 7\n",
      "[[7, 210], [285, 18318]]\n",
      "0.05003725053609512\n",
      "(-0.00012387101471900142, 1.524029671078973)\n",
      "0.761952900032127\n",
      "0.02031224124024791\n",
      "\n",
      "increased neuron apoptosis\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 154\n",
      "number overlapping genes = 5\n",
      "[[5, 149], [287, 18379]]\n",
      "0.0952531201296645\n",
      "(-0.13372971759737018, 1.6636764739863596)\n",
      "0.7649733781944947\n",
      "0.03320087811696227\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "decreased neuron apoptosis\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 2\n",
      "[[2, 68], [290, 18460]]\n",
      "0.3836661701815005\n",
      "(-0.7837931040950523, 2.038033225015376)\n",
      "0.6271200604601619\n",
      "0.09501981503733828\n",
      "\n",
      "neuron degeneration\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 387\n",
      "number overlapping genes = 16\n",
      "[[16, 371], [276, 18157]]\n",
      "7.082414538546484e-05\n",
      "(0.5284282899205166, 1.5571661784512119)\n",
      "1.0427972341858642\n",
      "0.0001233673637615258\n",
      "\n",
      "loss of hippocampal neurons\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "[[0, 15], [292, 18513]]\n",
      "0.6030145036163768\n",
      "(-2.0717188299891993, 3.568273396106678)\n",
      "0.7482772830587394\n",
      "0.20914723347278139\n",
      "\n",
      "loss of cortex neurons\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 17\n",
      "number overlapping genes = 1\n",
      "[[1, 16], [291, 18512]]\n",
      "0.18126925286086726\n",
      "(-0.6433378953410438, 3.403862815385411)\n",
      "1.3802624600221836\n",
      "0.027985015779562125\n",
      "\n",
      "abnormal dopaminergic neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 3\n",
      "[[3, 40], [289, 18488]]\n",
      "0.009135068709843869\n",
      "(0.389191029406843, 2.747175567385073)\n",
      "1.568183298395958\n",
      "0.004353738541050144\n",
      "\n",
      "loss of dopaminergic neurons\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 1\n",
      "[[1, 18], [291, 18510]]\n",
      "0.21994454227569638\n",
      "(-0.7546267908631523, 3.279369551862924)\n",
      "1.2623713804998857\n",
      "0.03448192672371518\n",
      "\n",
      "abnormal GABAergic neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 2\n",
      "[[2, 24], [290, 18504]]\n",
      "0.02363102106703002\n",
      "(0.22380425685364758, 3.1181050054318633)\n",
      "1.6709546311427554\n",
      "0.007380580131955909\n",
      "\n",
      "premature neuronal precursor differentiation\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 0\n",
      "[[0, 25], [292, 18503]]\n",
      "0.8683695309655081\n",
      "(-2.56486295827298, 3.038685663043557)\n",
      "0.2369113523852886\n",
      "0.32373972855184396\n",
      "\n",
      "nervous system phenotype\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 3524\n",
      "number overlapping genes = 132\n",
      "[[132, 3392], [160, 15136]]\n",
      "7.25250719536444e-28\n",
      "(1.0698376623887873, 1.5367311410760338)\n",
      "1.3032844017324106\n",
      "4.447756340771824e-26\n",
      "\n",
      "abnormal nervous system morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 2799\n",
      "number overlapping genes = 111\n",
      "[[111, 2688], [181, 15840]]\n",
      "8.548397463701224e-26\n",
      "(1.0449765170119085, 1.524571750531861)\n",
      "1.2847741337718848\n",
      "3.4384117797326163e-23\n",
      "\n",
      "abnormal nervous system physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 2009\n",
      "number overlapping genes = 92\n",
      "[[92, 1917], [200, 16611]]\n",
      "4.2431041820528264e-27\n",
      "(1.1313838784690902, 1.6341658674237827)\n",
      "1.3827748729464364\n",
      "3.872616263580695e-23\n",
      "\n",
      "abnormal synaptic transmission\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 940\n",
      "number overlapping genes = 51\n",
      "[[51, 889], [241, 17639]]\n",
      "1.0312937161448965e-19\n",
      "(1.125281302702863, 1.7443169019596239)\n",
      "1.4347991023312434\n",
      "8.886284517565308e-16\n",
      "\n",
      "abnormal nervous system development\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 1279\n",
      "number overlapping genes = 58\n",
      "[[58, 1221], [234, 17307]]\n",
      "4.567624308608452e-17\n",
      "(0.9632767793334284, 1.549848708672597)\n",
      "1.2565627440030127\n",
      "2.651335619385286e-14\n",
      "\n",
      "abnormal spinal cord interneuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 4\n",
      "[[4, 29], [288, 18499]]\n",
      "4.8044797685129616e-05\n",
      "(1.1296664541280206, 3.2333535589696982)\n",
      "2.1815100065488595\n",
      "0.00014440936708218013\n",
      "\n",
      "abnormal brain interneuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 38\n",
      "number overlapping genes = 5\n",
      "[[5, 33], [287, 18495]]\n",
      "2.4501166904840978e-06\n",
      "(1.3309210670586211, 3.2264866118904854)\n",
      "2.278703839474553\n",
      "2.4093025205109295e-05\n",
      "\n",
      "abnormal sympathetic nervous system physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 0\n",
      "[[0, 27], [292, 18501]]\n",
      "0.9109044480571389\n",
      "(-2.639900132264657, 2.9595845619187493)\n",
      "0.1598422148270462\n",
      "0.3445898732963503\n",
      "\n",
      "abnormal proprioceptive neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 17\n",
      "number overlapping genes = 0\n",
      "[[0, 17], [292, 18511]]\n",
      "0.6644143858331306\n",
      "(-2.191642901128336, 3.4376551052788193)\n",
      "0.6230061020752418\n",
      "0.23351757591027478\n",
      "\n",
      "abnormal synaptic glutamate release\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "[[0, 14], [292, 18514]]\n",
      "0.5704374208211285\n",
      "(-2.005913462941659, 3.6405618013090013)\n",
      "0.8173241691836712\n",
      "0.1966740235737086\n",
      "\n",
      "abnormal miniature excitatory postsynaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 116\n",
      "number overlapping genes = 8\n",
      "[[8, 108], [284, 18420]]\n",
      "2.3623764179492608e-05\n",
      "(0.8418703432690384, 2.2971864291383772)\n",
      "1.5695283862037077\n",
      "8.236626074733365e-05\n",
      "\n",
      "abnormal synaptic vesicle morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 5\n",
      "[[5, 42], [287, 18486]]\n",
      "1.9350536574848924e-05\n",
      "(1.1025242832054718, 2.9715858092261933)\n",
      "2.0370550462158326\n",
      "8.336177980361424e-05\n",
      "\n",
      "abnormal synaptic vesicle recycling\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 2\n",
      "[[2, 17], [290, 18511]]\n",
      "0.0071742151579365825\n",
      "(0.5464279532989562, 3.485918731666833)\n",
      "2.0161733424828947\n",
      "0.002979576375815365\n",
      "\n",
      "abnormal synaptic vesicle number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 40\n",
      "number overlapping genes = 5\n",
      "[[5, 35], [287, 18493]]\n",
      "4.076852225643973e-06\n",
      "(1.27548811042745, 3.1640222821124317)\n",
      "2.219755196269941\n",
      "3.263779244741189e-05\n",
      "\n",
      "abnormal neuron physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 598\n",
      "number overlapping genes = 37\n",
      "[[37, 561], [255, 17967]]\n",
      "2.174030198645884e-17\n",
      "(1.1813324253392754, 1.8911185404520745)\n",
      "1.536225482895675\n",
      "1.5203770620769945e-13\n",
      "\n",
      "abnormal nervous system regeneration\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 2\n",
      "[[2, 17], [290, 18511]]\n",
      "0.0071742151579365825\n",
      "(0.5464279532989562, 3.485918731666833)\n",
      "2.0161733424828947\n",
      "0.002979576375815365\n",
      "\n",
      "abnormal synaptic plasticity\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 69\n",
      "number overlapping genes = 2\n",
      "[[2, 67], [290, 18461]]\n",
      "0.3725924075633762\n",
      "(-0.7692226140646701, 2.053201245982729)\n",
      "0.6419893159590293\n",
      "0.09195043338645985\n",
      "\n",
      "abnormal neuronal precursor proliferation\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 80\n",
      "number overlapping genes = 4\n",
      "[[4, 76], [288, 18452]]\n",
      "0.018583450477685903\n",
      "(0.2033740511081381, 2.2276831196425118)\n",
      "1.215528585375325\n",
      "0.008104847148622037\n",
      "\n",
      "abnormal neuronal precursor cell number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 64\n",
      "number overlapping genes = 4\n",
      "[[4, 60], [288, 18468]]\n",
      "0.0051918118323693654\n",
      "(0.43399294432343627, 2.4715752604540127)\n",
      "1.4527841023887245\n",
      "0.0031389709584433555\n",
      "\n",
      "increased neuronal precursor cell number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 1\n",
      "[[1, 18], [291, 18510]]\n",
      "0.21994454227569638\n",
      "(-0.7546267908631523, 3.279369551862924)\n",
      "1.2623713804998857\n",
      "0.03448192672371518\n",
      "\n",
      "decreased neuronal precursor cell number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 1\n",
      "[[1, 42], [291, 18486]]\n",
      "0.6830966737813391\n",
      "(-1.5727613119538704, 2.4003134766935927)\n",
      "0.4137760823698611\n",
      "0.14343262071612706\n",
      "\n",
      "increased motor neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "[[1, 10], [291, 18518]]\n",
      "0.07812410747727289\n",
      "(-0.20829620754804745, 3.9094765092329546)\n",
      "1.8505901508424536\n",
      "0.012029927463292856\n",
      "\n",
      "abnormal somatic nervous system physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 72\n",
      "number overlapping genes = 5\n",
      "[[5, 67], [287, 18461]]\n",
      "0.0007902939838220029\n",
      "(0.6525892782854944, 2.4847682318226707)\n",
      "1.5686787550540826\n",
      "0.0008776924445438458\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "increased sensory neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 52\n",
      "number overlapping genes = 7\n",
      "[[7, 45], [285, 18483]]\n",
      "1.8188310568892242e-08\n",
      "(1.5064848946321163, 3.11624540276593)\n",
      "2.311365148699023\n",
      "1.2700627284175902e-06\n",
      "\n",
      "abnormal neuronal migration\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 127\n",
      "number overlapping genes = 7\n",
      "[[7, 120], [285, 18408]]\n",
      "0.0007464658653372937\n",
      "(0.5554449064697811, 2.097494808969857)\n",
      "1.326469857719819\n",
      "0.000830854165117607\n",
      "\n",
      "abnormal retinal neuronal layer morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 350\n",
      "number overlapping genes = 11\n",
      "[[11, 339], [281, 18189]]\n",
      "0.017454251392781988\n",
      "(0.13020010517321612, 1.354025476474817)\n",
      "0.7421127908240166\n",
      "0.008807473221156468\n",
      "\n",
      "abnormal olfactory sensory neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 2\n",
      "[[2, 22], [290, 18506]]\n",
      "0.01765201029461214\n",
      "(0.3059049633566786, 3.2102432107034913)\n",
      "1.758074087030085\n",
      "0.005877954284737828\n",
      "\n",
      "abnormal autonomic nervous system physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 1\n",
      "[[1, 42], [291, 18486]]\n",
      "0.6830966737813391\n",
      "(-1.5727613119538704, 2.4003134766935927)\n",
      "0.4137760823698611\n",
      "0.14343262071612706\n",
      "\n",
      "abnormal parasympathetic nervous system physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "[[1, 11], [291, 18517]]\n",
      "0.09338245397860352\n",
      "(-0.2951619696292096, 3.805613905765116)\n",
      "1.7552259680679532\n",
      "0.014289294987732197\n",
      "\n",
      "increased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 30\n",
      "number overlapping genes = 1\n",
      "[[1, 29], [291, 18499]]\n",
      "0.4410848583645801\n",
      "(-1.2119774961998546, 2.7816832129866995)\n",
      "0.7848528583934224\n",
      "0.07855166747489066\n",
      "\n",
      "decreased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 32\n",
      "number overlapping genes = 0\n",
      "[[0, 32], [292, 18496]]\n",
      "0.9942235371127878\n",
      "(-2.8060965255624306, 2.7854422972507336)\n",
      "-0.010327114155848527\n",
      "0.3939548970095386\n",
      "\n",
      "abnormal synaptic bouton morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 1\n",
      "[[1, 19], [291, 18509]]\n",
      "0.23971122705428372\n",
      "(-0.8059616977389124, 3.2224619635764804)\n",
      "1.208250132918784\n",
      "0.037928197353114534\n",
      "\n",
      "abnormal neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 461\n",
      "number overlapping genes = 26\n",
      "[[26, 435], [266, 18093]]\n",
      "3.072541468680818e-11\n",
      "(0.9887312074399965, 1.8163379931003936)\n",
      "1.402534600270195\n",
      "3.6980289610671108e-09\n",
      "\n",
      "increased neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 85\n",
      "number overlapping genes = 9\n",
      "[[9, 76], [283, 18452]]\n",
      "1.0888408430210007e-08\n",
      "(1.343146789037268, 2.7447979791314565)\n",
      "2.043972384084362\n",
      "7.774316631466154e-07\n",
      "\n",
      "decreased neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 385\n",
      "number overlapping genes = 21\n",
      "[[21, 364], [271, 18164]]\n",
      "6.099369315238841e-09\n",
      "(0.8965226553258853, 1.8083706275237619)\n",
      "1.3524466414248235\n",
      "1.6775652849851536e-07\n",
      "\n",
      "abnormal miniature inhibitory postsynaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 48\n",
      "number overlapping genes = 4\n",
      "[[4, 44], [288, 18484]]\n",
      "0.0007913459318781332\n",
      "(0.733651238152577, 2.7939587999591033)\n",
      "1.76380501905584\n",
      "0.0008619335865148737\n",
      "\n",
      "abnormal synaptonemal complex\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 0\n",
      "[[0, 24], [292, 18504]]\n",
      "0.8459838428086519\n",
      "(-2.5251292523373507, 3.0807040338075558)\n",
      "0.2777873907351025\n",
      "0.31306809004780056\n",
      "\n",
      "abnormal neuron differentiation\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 309\n",
      "number overlapping genes = 19\n",
      "[[19, 290], [273, 18238]]\n",
      "1.5665971508704499e-09\n",
      "(0.997067782821248, 1.9556299561678945)\n",
      "1.4763488694945712\n",
      "8.426395655596928e-08\n",
      "\n",
      "abnormal hippocampus neuron morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 117\n",
      "number overlapping genes = 9\n",
      "[[9, 108], [283, 18420]]\n",
      "1.5672156037936002e-06\n",
      "(1.0007804831729024, 2.380897041583218)\n",
      "1.69083876237806\n",
      "1.4329542895840912e-05\n",
      "\n",
      "abnormal neuron proliferation\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 108\n",
      "number overlapping genes = 6\n",
      "[[6, 102], [286, 18426]]\n",
      "0.0016888721277607596\n",
      "(0.5007211968406827, 2.1639044727449344)\n",
      "1.3323128347928086\n",
      "0.0014785396152285437\n",
      "\n",
      "abnormal sensory neuron physiology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 61\n",
      "number overlapping genes = 4\n",
      "[[4, 57], [288, 18471]]\n",
      "0.003862350093390952\n",
      "(0.4837962512733036, 2.5246834021841247)\n",
      "1.504239826728714\n",
      "0.002543084126560324\n",
      "\n",
      "abnormal synaptic dopamine release\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "[[0, 12], [292, 18516]]\n",
      "0.5012202081348538\n",
      "(-1.8597423250804863, 3.8029080641503543)\n",
      "0.971582869534934\n",
      "0.17113640958677187\n",
      "\n",
      "increased nervous system tumor incidence\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 55\n",
      "number overlapping genes = 2\n",
      "[[2, 53], [290, 18475]]\n",
      "0.2248918547881984\n",
      "(-0.5394196291823741, 2.293715808869787)\n",
      "0.8771480898437063\n",
      "0.05373938663804336\n",
      "\n",
      "reduced AMPA-mediated synaptic currents\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "[[0, 16], [292, 18512]]\n",
      "0.6343194783497113\n",
      "(-2.1334719801344324, 3.500841468865331)\n",
      "0.6836847443654492\n",
      "0.2214274253135972\n",
      "\n",
      "abnormal excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 2\n",
      "[[2, 29], [290, 18499]]\n",
      "0.04340834697927194\n",
      "(0.043858446469202894, 2.919026319819479)\n",
      "1.481442383144341\n",
      "0.012055075756875042\n",
      "\n",
      "increased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "[[1, 10], [291, 18518]]\n",
      "0.07812410747727289\n",
      "(-0.20829620754804745, 3.9094765092329546)\n",
      "1.8505901508424536\n",
      "0.012029927463292856\n",
      "\n",
      "decreased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 1\n",
      "[[1, 20], [291, 18508]]\n",
      "0.2596792228087118\n",
      "(-0.8547980153282224, 3.168603633931037)\n",
      "1.1569028093014073\n",
      "0.04149989216729133\n",
      "\n",
      "decreased dopaminergic neuron number\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 3\n",
      "[[3, 21], [289, 18507]]\n",
      "0.0003570346787914264\n",
      "(0.9982845107759224, 3.428850450644083)\n",
      "2.2135674807100028\n",
      "0.00047214941286228743\n",
      "\n",
      "increased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "[[1, 12], [291, 18516]]\n",
      "0.10956727546390566\n",
      "(-0.37511833495000557, 3.711439505333342)\n",
      "1.6681605851916683\n",
      "0.016716055428094623\n",
      "\n",
      "decreased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "[[0, 12], [292, 18516]]\n",
      "0.5012202081348538\n",
      "(-1.8597423250804863, 3.8029080641503543)\n",
      "0.971582869534934\n",
      "0.17113640958677187\n",
      "\n",
      "ectopic neuron\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 64\n",
      "number overlapping genes = 6\n",
      "[[6, 58], [286, 18470]]\n",
      "1.1515646801391596e-05\n",
      "(1.050631914261966, 2.7478235278481726)\n",
      "1.8992277210550692\n",
      "5.873726098908741e-05\n",
      "\n",
      "impaired neuron differentiation\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "[[0, 12], [292, 18516]]\n",
      "0.5012202081348538\n",
      "(-1.8597423250804863, 3.8029080641503543)\n",
      "0.971582869534934\n",
      "0.17113640958677187\n",
      "\n",
      "nervous system inclusion bodies\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 2\n",
      "[[2, 55], [290, 18473]]\n",
      "0.24483496465154964\n",
      "(-0.5756385692642241, 2.2556356850754087)\n",
      "0.8399985579055924\n",
      "0.05861957081705705\n",
      "\n",
      "abnormal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "[[0, 14], [292, 18514]]\n",
      "0.5704374208211285\n",
      "(-2.005913462941659, 3.6405618013090013)\n",
      "0.8173241691836712\n",
      "0.1966740235737086\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "decreased neuronal precursor proliferation\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "[[0, 11], [292, 18517]]\n",
      "0.46446474434102136\n",
      "(-1.7778115670155863, 3.895108071838024)\n",
      "1.0586482524112188\n",
      "0.1580658166929844\n",
      "\n",
      "abnormal postsynaptic density morphology\n",
      "number of genes in root node = 292\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "[[0, 16], [292, 18512]]\n",
      "0.6343194783497113\n",
      "(-2.1334719801344324, 3.500841468865331)\n",
      "0.6836847443654492\n",
      "0.2214274253135972\n"
     ]
    }
   ],
   "source": [
    "# test for enrichment in root node\n",
    "OR_p_list,OR_CI_list,log_OR_list = [],[],[]\n",
    "\n",
    "MP_keep_list = []\n",
    "\n",
    "# root node is the largest node\n",
    "root_node = ASD_clust_df['CD_MemberList_Size'].sort_values(ascending=False).head(1).index.tolist()[0]\n",
    "for MP_focal in MP_focal_list:\n",
    "    MP_desc_focal = dict(MP2desc['description'])[MP_focal]\n",
    "\n",
    "    # focus the hierarchy on one branch, and look up all terms within that branch\n",
    "    if len(MPO.parent_2_child[MP_focal])>0:\n",
    "        MPO_focal = MPO.focus(MP_focal,verbose=False)\n",
    "        focal_terms = MPO_focal.terms\n",
    "    else: # if the term has no children, just look at that term\n",
    "        focal_terms=[MP_focal]\n",
    "\n",
    "\n",
    "    # check enrichment in root node\n",
    "    focal_genes = ASD_clust_df['CD_MemberList'].loc[root_node].split(' ')\n",
    "    \n",
    "\n",
    "    mgi_temp = mgi_df[mgi_df['MP'].isin(focal_terms)]\n",
    "    mgi_genes = list(np.unique(mgi_temp['gene_name']))\n",
    "    mgi_genes = [g.upper() for g in mgi_genes]\n",
    "    mgi_genes = list(np.intersect1d(mgi_genes,G_PC.nodes()))\n",
    "    \n",
    "    if len(mgi_genes)>10: # only test if there are at least 10 genes\n",
    "        print('\\n'+MP_desc_focal)\n",
    "        print('number of genes in root node = '+str(len(focal_genes)))\n",
    "        print('number of genes in focal MPO term = '+str(len(mgi_genes)))\n",
    "\n",
    "        q00 = len(np.intersect1d(mgi_genes,focal_genes))\n",
    "        print('number overlapping genes = '+str(q00))\n",
    "        q01 = len(mgi_genes)-q00\n",
    "\n",
    "        q10 = len(focal_genes)-q00\n",
    "        q11 = len(G_PC.nodes())-q00-q01-q10\n",
    "\n",
    "        table_temp = [[q00,q01],[q10,q11]]\n",
    "        print(table_temp)\n",
    "\n",
    "        CT= contingency_tables.Table2x2(table_temp)\n",
    "        OR_p_temp = CT.log_oddsratio_pvalue()\n",
    "        OR_CI_temp = CT.log_oddsratio_confint()\n",
    "        log_OR_temp = CT.log_oddsratio\n",
    "        print(OR_p_temp)\n",
    "        print(OR_CI_temp)\n",
    "        print(log_OR_temp)\n",
    "\n",
    "        OR_p_list.append(OR_p_temp)\n",
    "        OR_CI_list.append(OR_CI_temp)\n",
    "        log_OR_list.append(log_OR_temp)\n",
    "        \n",
    "        MP_keep_list.append(MP_focal)\n",
    "\n",
    "        print(hypergeom.sf(q00,len(G_PC.nodes()),len(focal_genes),len(mgi_genes)))\n",
    "    \n",
    "OR_CI_lower, OR_CI_upper = zip(*OR_CI_list)\n",
    "\n",
    "root_KO_df = pd.DataFrame({'OR_p':OR_p_list,'OR_CI_lower':OR_CI_lower,'OR_CI_upper':OR_CI_upper},\n",
    "                          index=MP_keep_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>OR_CI_lower</th>\n",
       "      <th>OR_CI_upper</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0002272</th>\n",
       "      <td>1.376045e-28</td>\n",
       "      <td>1.752705</td>\n",
       "      <td>2.505077</td>\n",
       "      <td>abnormal nervous system electrophysiology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003631</th>\n",
       "      <td>7.252507e-28</td>\n",
       "      <td>1.069838</td>\n",
       "      <td>1.536731</td>\n",
       "      <td>nervous system phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003633</th>\n",
       "      <td>4.243104e-27</td>\n",
       "      <td>1.131384</td>\n",
       "      <td>1.634166</td>\n",
       "      <td>abnormal nervous system physiology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003632</th>\n",
       "      <td>8.548397e-26</td>\n",
       "      <td>1.044977</td>\n",
       "      <td>1.524572</td>\n",
       "      <td>abnormal nervous system morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002882</th>\n",
       "      <td>3.193956e-21</td>\n",
       "      <td>1.053431</td>\n",
       "      <td>1.604292</td>\n",
       "      <td>abnormal neuron morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002206</th>\n",
       "      <td>4.064210e-20</td>\n",
       "      <td>1.203781</td>\n",
       "      <td>1.856768</td>\n",
       "      <td>abnormal CNS synaptic transmission</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003635</th>\n",
       "      <td>1.031294e-19</td>\n",
       "      <td>1.125281</td>\n",
       "      <td>1.744317</td>\n",
       "      <td>abnormal synaptic transmission</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0004811</th>\n",
       "      <td>2.174030e-17</td>\n",
       "      <td>1.181332</td>\n",
       "      <td>1.891119</td>\n",
       "      <td>abnormal neuron physiology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003861</th>\n",
       "      <td>4.567624e-17</td>\n",
       "      <td>0.963277</td>\n",
       "      <td>1.549849</td>\n",
       "      <td>abnormal nervous system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002752</th>\n",
       "      <td>2.769039e-15</td>\n",
       "      <td>0.966874</td>\n",
       "      <td>1.604834</td>\n",
       "      <td>abnormal somatic nervous system morphology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p  OR_CI_lower  OR_CI_upper  \\\n",
       "MP:0002272  1.376045e-28     1.752705     2.505077   \n",
       "MP:0003631  7.252507e-28     1.069838     1.536731   \n",
       "MP:0003633  4.243104e-27     1.131384     1.634166   \n",
       "MP:0003632  8.548397e-26     1.044977     1.524572   \n",
       "MP:0002882  3.193956e-21     1.053431     1.604292   \n",
       "MP:0002206  4.064210e-20     1.203781     1.856768   \n",
       "MP:0003635  1.031294e-19     1.125281     1.744317   \n",
       "MP:0004811  2.174030e-17     1.181332     1.891119   \n",
       "MP:0003861  4.567624e-17     0.963277     1.549849   \n",
       "MP:0002752  2.769039e-15     0.966874     1.604834   \n",
       "\n",
       "                                        MP_description  \n",
       "MP:0002272   abnormal nervous system electrophysiology  \n",
       "MP:0003631                    nervous system phenotype  \n",
       "MP:0003633          abnormal nervous system physiology  \n",
       "MP:0003632          abnormal nervous system morphology  \n",
       "MP:0002882                  abnormal neuron morphology  \n",
       "MP:0002206          abnormal CNS synaptic transmission  \n",
       "MP:0003635              abnormal synaptic transmission  \n",
       "MP:0004811                  abnormal neuron physiology  \n",
       "MP:0003861         abnormal nervous system development  \n",
       "MP:0002752  abnormal somatic nervous system morphology  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_df['MP_description']=root_KO_df.index.map(dict(MP2desc['description']))\n",
    "\n",
    "root_KO_df=root_KO_df.sort_values('OR_p')\n",
    "root_KO_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test for enrichment in all systems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abnormal nervous system electrophysiology\n",
      "Genes and Terms to keep: 34\n",
      "nervous system phenotype\n",
      "Genes and Terms to keep: 2042\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brinrosenthal/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:83: RuntimeWarning: divide by zero encountered in log10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abnormal nervous system physiology\n",
      "Genes and Terms to keep: 354\n",
      "abnormal nervous system morphology\n",
      "Genes and Terms to keep: 1687\n",
      "abnormal neuron morphology\n",
      "Genes and Terms to keep: 262\n",
      "abnormal CNS synaptic transmission\n",
      "Genes and Terms to keep: 43\n",
      "abnormal synaptic transmission\n",
      "Genes and Terms to keep: 132\n",
      "abnormal neuron physiology\n",
      "Genes and Terms to keep: 46\n",
      "abnormal nervous system development\n",
      "Genes and Terms to keep: 191\n",
      "abnormal somatic nervous system morphology\n",
      "Genes and Terms to keep: 356\n"
     ]
    }
   ],
   "source": [
    "MP_focal_top = root_KO_df.head(10).index.tolist()\n",
    "\n",
    "MP_full_results_df=pd.DataFrame(index=ASD_clust_df.index.tolist())\n",
    "\n",
    "for MP_focal in MP_focal_top:\n",
    "    MP_desc_focal = dict(MP2desc['description'])[MP_focal]\n",
    "    print(MP_desc_focal)\n",
    "\n",
    "    # focus the hierarchy on one branch, and look up all terms within that branch\n",
    "    if len(MPO.parent_2_child[MP_focal])>0:\n",
    "        MPO_focal = MPO.focus(MP_focal)\n",
    "        focal_terms = MPO_focal.terms\n",
    "    else: # if the term has no children, just look at that term\n",
    "        focal_terms=MP_focal\n",
    "\n",
    "\n",
    "    hyper_p_list = []\n",
    "    num_genes_list = []\n",
    "    genes_id_list = []\n",
    "\n",
    "    OR_p_list,OR_CI_list,log_OR_list=[],[],[]\n",
    "    for focal_cluster in ASD_clust_df.index.tolist():\n",
    "        #print(focal_cluster)\n",
    "        mFocal_genes = ASD_clust_df['CD_MemberList'].loc[focal_cluster].split(' ')\n",
    "\n",
    "\n",
    "        from scipy.stats import hypergeom\n",
    "        M=len(list(G_PC.nodes())) # only keep genes in PCnet\n",
    "        # Look up all entries matching focal_terms, and mFocal_genes\n",
    "        mgi_temp = mgi_df[mgi_df['MP'].isin(focal_terms)]\n",
    "        new_index=[g.upper() for g in mgi_temp.index.tolist()]\n",
    "        mgi_temp.index=new_index\n",
    "\n",
    "        N=len(np.intersect1d(list(np.unique(mgi_temp.index.tolist())),list(G_PC.nodes()))) # only keep genes in PCnet\n",
    "\n",
    "        if len(np.intersect1d(mFocal_genes,mgi_temp.index.tolist()))>0:\n",
    "            mgi_genes = mgi_temp.index.tolist()\n",
    "            mgi_genes = list(np.intersect1d(mgi_genes,list(G_PC.nodes())))\n",
    "            x = len(np.intersect1d(mFocal_genes,mgi_genes))\n",
    "            #print(x)\n",
    "            n=len(mFocal_genes)\n",
    "\n",
    "            #mgi_temp['MP_description']=mgi_temp['MP'].map(dict(MP2desc['description']))\n",
    "\n",
    "\n",
    "\n",
    "            #print(hypergeom.sf(x,M,n,N))\n",
    "            hyper_p_list.append(hypergeom.sf(x,M,n,N))\n",
    "            num_genes_list.append(x)\n",
    "            genes_id_list.append(' '.join(list(np.intersect1d(mFocal_genes,mgi_genes))))\n",
    "\n",
    "            q00 = len(np.intersect1d(mgi_genes,mFocal_genes))\n",
    "            q01 = len(mgi_genes)-q00\n",
    "\n",
    "            q10 = len(mFocal_genes)-q00\n",
    "            q11 = len(list(G_PC.nodes()))-q00-q01-q10\n",
    "\n",
    "            table_temp = [[q00,q01],[q10,q11]]\n",
    "\n",
    "            CT= contingency_tables.Table2x2(table_temp)\n",
    "            OR_p_temp = CT.log_oddsratio_pvalue()\n",
    "            OR_CI_temp = CT.log_oddsratio_confint()\n",
    "            log_OR_temp = CT.log_oddsratio\n",
    "            #print(OR_p_temp)\n",
    "\n",
    "\n",
    "            OR_p_list.append(OR_p_temp)\n",
    "            OR_CI_list.append(OR_CI_temp)\n",
    "            log_OR_list.append(log_OR_temp)\n",
    "\n",
    "\n",
    "        else:\n",
    "            hyper_p_list.append(1)\n",
    "            num_genes_list.append(0)\n",
    "            genes_id_list.append('')\n",
    "\n",
    "            OR_p_list.append(1)\n",
    "            OR_CI_list.append(0)\n",
    "            log_OR_list.append(0)\n",
    "\n",
    "\n",
    "\n",
    "    MP_focal_df = pd.DataFrame({MP_desc_focal+':-log(hyper_p)':-np.log10(hyper_p_list),\n",
    "                                MP_desc_focal+':-log(OR_p)':-np.log10(OR_p_list),\n",
    "                                MP_desc_focal+':log_OR':log_OR_list,\n",
    "                                MP_desc_focal+':num_genes':num_genes_list,\n",
    "                                MP_desc_focal+':gene_ids':genes_id_list},index=ASD_clust_df.index.tolist())\n",
    "\n",
    "    if MP_desc_focal+':-log(hyper_p)' not in MP_full_results_df.columns.tolist():\n",
    "        MP_full_results_df=MP_full_results_df.join(MP_focal_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abnormal nervous system electrophysiology:-log(hyper_p)</th>\n",
       "      <th>abnormal nervous system electrophysiology:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system electrophysiology:log_OR</th>\n",
       "      <th>abnormal nervous system electrophysiology:num_genes</th>\n",
       "      <th>abnormal nervous system electrophysiology:gene_ids</th>\n",
       "      <th>nervous system phenotype:-log(hyper_p)</th>\n",
       "      <th>nervous system phenotype:-log(OR_p)</th>\n",
       "      <th>nervous system phenotype:log_OR</th>\n",
       "      <th>nervous system phenotype:num_genes</th>\n",
       "      <th>nervous system phenotype:gene_ids</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal nervous system development:-log(hyper_p)</th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal somatic nervous system morphology:-log(hyper_p)</th>\n",
       "      <th>abnormal somatic nervous system morphology:-log(OR_p)</th>\n",
       "      <th>abnormal somatic nervous system morphology:log_OR</th>\n",
       "      <th>abnormal somatic nervous system morphology:num_genes</th>\n",
       "      <th>abnormal somatic nervous system morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C435</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>2.910936</td>\n",
       "      <td>1.581589</td>\n",
       "      <td>2.567393</td>\n",
       "      <td>3</td>\n",
       "      <td>PSD2 SLC17A6 STXBP5L</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C434</th>\n",
       "      <td>4.715884</td>\n",
       "      <td>4.303365</td>\n",
       "      <td>4.063367</td>\n",
       "      <td>2</td>\n",
       "      <td>EN1 NKX6-2</td>\n",
       "      <td>2.910936</td>\n",
       "      <td>1.581589</td>\n",
       "      <td>2.567393</td>\n",
       "      <td>3</td>\n",
       "      <td>EN1 GSX2 NKX6-2</td>\n",
       "      <td>...</td>\n",
       "      <td>4.672897</td>\n",
       "      <td>2.892220</td>\n",
       "      <td>3.719366</td>\n",
       "      <td>3</td>\n",
       "      <td>EN1 GSX2 NKX6-2</td>\n",
       "      <td>5.136098</td>\n",
       "      <td>3.276389</td>\n",
       "      <td>4.003268</td>\n",
       "      <td>3</td>\n",
       "      <td>EN1 GSX2 NKX6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C433</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0.791776</td>\n",
       "      <td>0.125505</td>\n",
       "      <td>0.369470</td>\n",
       "      <td>1</td>\n",
       "      <td>GNAZ</td>\n",
       "      <td>...</td>\n",
       "      <td>1.597754</td>\n",
       "      <td>0.725707</td>\n",
       "      <td>1.520461</td>\n",
       "      <td>1</td>\n",
       "      <td>GNAZ</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C432</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1.646664</td>\n",
       "      <td>0.847538</td>\n",
       "      <td>1.468432</td>\n",
       "      <td>2</td>\n",
       "      <td>CDK5R2 KIF1A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.597754</td>\n",
       "      <td>0.725707</td>\n",
       "      <td>1.520461</td>\n",
       "      <td>1</td>\n",
       "      <td>CDK5R2</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C431</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0.791776</td>\n",
       "      <td>0.125505</td>\n",
       "      <td>0.369470</td>\n",
       "      <td>1</td>\n",
       "      <td>KIF1A</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      abnormal nervous system electrophysiology:-log(hyper_p)  \\\n",
       "C435                                          -0.000000         \n",
       "C434                                           4.715884         \n",
       "C433                                          -0.000000         \n",
       "C432                                          -0.000000         \n",
       "C431                                          -0.000000         \n",
       "\n",
       "      abnormal nervous system electrophysiology:-log(OR_p)  \\\n",
       "C435                                          -0.000000      \n",
       "C434                                           4.303365      \n",
       "C433                                          -0.000000      \n",
       "C432                                          -0.000000      \n",
       "C431                                          -0.000000      \n",
       "\n",
       "      abnormal nervous system electrophysiology:log_OR  \\\n",
       "C435                                          0.000000   \n",
       "C434                                          4.063367   \n",
       "C433                                          0.000000   \n",
       "C432                                          0.000000   \n",
       "C431                                          0.000000   \n",
       "\n",
       "      abnormal nervous system electrophysiology:num_genes  \\\n",
       "C435                                                  0     \n",
       "C434                                                  2     \n",
       "C433                                                  0     \n",
       "C432                                                  0     \n",
       "C431                                                  0     \n",
       "\n",
       "     abnormal nervous system electrophysiology:gene_ids  \\\n",
       "C435                                                      \n",
       "C434                                         EN1 NKX6-2   \n",
       "C433                                                      \n",
       "C432                                                      \n",
       "C431                                                      \n",
       "\n",
       "      nervous system phenotype:-log(hyper_p)  \\\n",
       "C435                                2.910936   \n",
       "C434                                2.910936   \n",
       "C433                                0.791776   \n",
       "C432                                1.646664   \n",
       "C431                                0.791776   \n",
       "\n",
       "      nervous system phenotype:-log(OR_p)  nervous system phenotype:log_OR  \\\n",
       "C435                             1.581589                         2.567393   \n",
       "C434                             1.581589                         2.567393   \n",
       "C433                             0.125505                         0.369470   \n",
       "C432                             0.847538                         1.468432   \n",
       "C431                             0.125505                         0.369470   \n",
       "\n",
       "      nervous system phenotype:num_genes nervous system phenotype:gene_ids  \\\n",
       "C435                                   3              PSD2 SLC17A6 STXBP5L   \n",
       "C434                                   3                   EN1 GSX2 NKX6-2   \n",
       "C433                                   1                              GNAZ   \n",
       "C432                                   2                      CDK5R2 KIF1A   \n",
       "C431                                   1                             KIF1A   \n",
       "\n",
       "      ...  abnormal nervous system development:-log(hyper_p)  \\\n",
       "C435  ...                                          -0.000000   \n",
       "C434  ...                                           4.672897   \n",
       "C433  ...                                           1.597754   \n",
       "C432  ...                                           1.597754   \n",
       "C431  ...                                          -0.000000   \n",
       "\n",
       "      abnormal nervous system development:-log(OR_p)  \\\n",
       "C435                                       -0.000000   \n",
       "C434                                        2.892220   \n",
       "C433                                        0.725707   \n",
       "C432                                        0.725707   \n",
       "C431                                       -0.000000   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C435                                    0.000000   \n",
       "C434                                    3.719366   \n",
       "C433                                    1.520461   \n",
       "C432                                    1.520461   \n",
       "C431                                    0.000000   \n",
       "\n",
       "      abnormal nervous system development:num_genes  \\\n",
       "C435                                              0   \n",
       "C434                                              3   \n",
       "C433                                              1   \n",
       "C432                                              1   \n",
       "C431                                              0   \n",
       "\n",
       "     abnormal nervous system development:gene_ids  \\\n",
       "C435                                                \n",
       "C434                              EN1 GSX2 NKX6-2   \n",
       "C433                                         GNAZ   \n",
       "C432                                       CDK5R2   \n",
       "C431                                                \n",
       "\n",
       "      abnormal somatic nervous system morphology:-log(hyper_p)  \\\n",
       "C435                                          -0.000000          \n",
       "C434                                           5.136098          \n",
       "C433                                          -0.000000          \n",
       "C432                                          -0.000000          \n",
       "C431                                          -0.000000          \n",
       "\n",
       "      abnormal somatic nervous system morphology:-log(OR_p)  \\\n",
       "C435                                          -0.000000       \n",
       "C434                                           3.276389       \n",
       "C433                                          -0.000000       \n",
       "C432                                          -0.000000       \n",
       "C431                                          -0.000000       \n",
       "\n",
       "      abnormal somatic nervous system morphology:log_OR  \\\n",
       "C435                                           0.000000   \n",
       "C434                                           4.003268   \n",
       "C433                                           0.000000   \n",
       "C432                                           0.000000   \n",
       "C431                                           0.000000   \n",
       "\n",
       "      abnormal somatic nervous system morphology:num_genes  \\\n",
       "C435                                                  0      \n",
       "C434                                                  3      \n",
       "C433                                                  0      \n",
       "C432                                                  0      \n",
       "C431                                                  0      \n",
       "\n",
       "     abnormal somatic nervous system morphology:gene_ids  \n",
       "C435                                                      \n",
       "C434                                    EN1 GSX2 NKX6-2   \n",
       "C433                                                      \n",
       "C432                                                      \n",
       "C431                                                      \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update hierarchy on ndex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add new fields as node attributes to hierarchy\n",
    "MP_full_results_dict = MP_full_results_df.to_dict()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CD_MemberList\n",
      "CD_AnnotatedMembers_Size\n",
      "CD_AnnotatedMembers_Pvalue\n",
      "CD_AnnotatedMembers\n",
      "CD_CommunityName\n",
      "CD_AnnotatedMembers_Overlap\n",
      "HiDeF_persistence\n",
      "CD_MemberList_LogSize\n",
      "CD_Labeled\n",
      "CD_AnnotatedAlgorithm\n",
      "CD_NonAnnotatedMembers\n",
      "CD_AnnotatedMembers_SourceDB\n",
      "CD_MemberList_Size\n",
      "name\n",
      "abnormal nervous system electrophysiology:-log(hyper_p)\n",
      "abnormal nervous system electrophysiology:-log(OR_p)\n",
      "abnormal nervous system electrophysiology:log_OR\n",
      "abnormal nervous system electrophysiology:num_genes\n",
      "abnormal nervous system electrophysiology:gene_ids\n",
      "nervous system phenotype:-log(hyper_p)\n",
      "nervous system phenotype:-log(OR_p)\n",
      "nervous system phenotype:log_OR\n",
      "nervous system phenotype:num_genes\n",
      "nervous system phenotype:gene_ids\n",
      "abnormal nervous system physiology:-log(hyper_p)\n",
      "abnormal nervous system physiology:-log(OR_p)\n",
      "abnormal nervous system physiology:log_OR\n",
      "abnormal nervous system physiology:num_genes\n",
      "abnormal nervous system physiology:gene_ids\n",
      "abnormal nervous system morphology:-log(hyper_p)\n",
      "abnormal nervous system morphology:-log(OR_p)\n",
      "abnormal nervous system morphology:log_OR\n",
      "abnormal nervous system morphology:num_genes\n",
      "abnormal nervous system morphology:gene_ids\n",
      "abnormal neuron morphology:-log(hyper_p)\n",
      "abnormal neuron morphology:-log(OR_p)\n",
      "abnormal neuron morphology:log_OR\n",
      "abnormal neuron morphology:num_genes\n",
      "abnormal neuron morphology:gene_ids\n",
      "abnormal CNS synaptic transmission:-log(hyper_p)\n",
      "abnormal CNS synaptic transmission:-log(OR_p)\n",
      "abnormal CNS synaptic transmission:log_OR\n",
      "abnormal CNS synaptic transmission:num_genes\n",
      "abnormal CNS synaptic transmission:gene_ids\n",
      "abnormal synaptic transmission:-log(hyper_p)\n",
      "abnormal synaptic transmission:-log(OR_p)\n",
      "abnormal synaptic transmission:log_OR\n",
      "abnormal synaptic transmission:num_genes\n",
      "abnormal synaptic transmission:gene_ids\n",
      "abnormal neuron physiology:-log(hyper_p)\n",
      "abnormal neuron physiology:-log(OR_p)\n",
      "abnormal neuron physiology:log_OR\n",
      "abnormal neuron physiology:num_genes\n",
      "abnormal neuron physiology:gene_ids\n",
      "abnormal nervous system development:-log(hyper_p)\n",
      "abnormal nervous system development:-log(OR_p)\n",
      "abnormal nervous system development:log_OR\n",
      "abnormal nervous system development:num_genes\n",
      "abnormal nervous system development:gene_ids\n",
      "abnormal somatic nervous system morphology:-log(hyper_p)\n",
      "abnormal somatic nervous system morphology:-log(OR_p)\n",
      "abnormal somatic nervous system morphology:log_OR\n",
      "abnormal somatic nervous system morphology:num_genes\n",
      "abnormal somatic nervous system morphology:gene_ids\n"
     ]
    }
   ],
   "source": [
    "for key in MP_full_results_dict.keys():\n",
    "    print(key)\n",
    "    nx.set_node_attributes(G_ASD,key,MP_full_results_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_AnnotatedAlgorithm</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal nervous system development:-log(hyper_p)</th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal somatic nervous system morphology:-log(hyper_p)</th>\n",
       "      <th>abnormal somatic nervous system morphology:-log(OR_p)</th>\n",
       "      <th>abnormal somatic nervous system morphology:log_OR</th>\n",
       "      <th>abnormal somatic nervous system morphology:num_genes</th>\n",
       "      <th>abnormal somatic nervous system morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C435</th>\n",
       "      <td>NSG2 PSD2 SLC17A6 STXBP5L</td>\n",
       "      <td>1</td>\n",
       "      <td>0.007995642253764</td>\n",
       "      <td>SLC17A6</td>\n",
       "      <td>Organic anion transporters Homo sapiens R-HSA-...</td>\n",
       "      <td>0.25</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C367</th>\n",
       "      <td>ABCC8 ADARB2 BRINP3 C14orf132 CADM2 CALB1 CALB...</td>\n",
       "      <td>18</td>\n",
       "      <td>2.977322934006789E-14</td>\n",
       "      <td>GABRA2 UNC13B GABRB3 UNC13C GABRB2 CHRNA7 GAD1...</td>\n",
       "      <td>chemical synaptic transmission (GO:0007268)</td>\n",
       "      <td>0.228</td>\n",
       "      <td>17</td>\n",
       "      <td>6.304</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>...</td>\n",
       "      <td>4.744992</td>\n",
       "      <td>5.097764</td>\n",
       "      <td>1.256907</td>\n",
       "      <td>16</td>\n",
       "      <td>CDK5R2 CHRNA7 CNTN1 GFAP GNAZ LGI1 MAPT MPZ NC...</td>\n",
       "      <td>5.597954</td>\n",
       "      <td>6.389371</td>\n",
       "      <td>1.462644</td>\n",
       "      <td>15</td>\n",
       "      <td>CHRNA7 GABRB2 GABRB3 GRM7 KIF5C MAPT MOG MPZ N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C434</th>\n",
       "      <td>EN1 GSX2 IRX3 NKX6-2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0249546954540196</td>\n",
       "      <td>NKX6-2</td>\n",
       "      <td>regulation of myelination (GO:0031641)</td>\n",
       "      <td>0.25</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>...</td>\n",
       "      <td>4.672897</td>\n",
       "      <td>2.892220</td>\n",
       "      <td>3.719366</td>\n",
       "      <td>3</td>\n",
       "      <td>EN1 GSX2 NKX6-2</td>\n",
       "      <td>5.136098</td>\n",
       "      <td>3.276389</td>\n",
       "      <td>4.003268</td>\n",
       "      <td>3</td>\n",
       "      <td>EN1 GSX2 NKX6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C372</th>\n",
       "      <td>EN1 GSX2 HOXA4 HOXA7 IRX1 IRX3 IRX6 MEIS3P1 MK...</td>\n",
       "      <td>3</td>\n",
       "      <td>1.677832487479E-4</td>\n",
       "      <td>PKNOX1 PBX1 HOXA4</td>\n",
       "      <td>Activation of HOX genes during differentiation...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>16</td>\n",
       "      <td>3.907</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>...</td>\n",
       "      <td>3.544148</td>\n",
       "      <td>3.358930</td>\n",
       "      <td>1.928662</td>\n",
       "      <td>5</td>\n",
       "      <td>EN1 GSX2 IRX1 MNX1 NKX6-2</td>\n",
       "      <td>4.183997</td>\n",
       "      <td>4.259458</td>\n",
       "      <td>2.213053</td>\n",
       "      <td>5</td>\n",
       "      <td>EN1 GSX2 IRX6 MNX1 NKX6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C433</th>\n",
       "      <td>CALM3 FSD1 GNAZ SRC</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0034578149777732</td>\n",
       "      <td>SRC CALM3</td>\n",
       "      <td>GnRH signaling pathway</td>\n",
       "      <td>0.5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>true</td>\n",
       "      <td>Annotated by Enrichr [Docker: coleslawndex/cde...</td>\n",
       "      <td>...</td>\n",
       "      <td>1.597754</td>\n",
       "      <td>0.725707</td>\n",
       "      <td>1.520461</td>\n",
       "      <td>1</td>\n",
       "      <td>GNAZ</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CD_MemberList  \\\n",
       "C435                          NSG2 PSD2 SLC17A6 STXBP5L   \n",
       "C367  ABCC8 ADARB2 BRINP3 C14orf132 CADM2 CALB1 CALB...   \n",
       "C434                               EN1 GSX2 IRX3 NKX6-2   \n",
       "C372  EN1 GSX2 HOXA4 HOXA7 IRX1 IRX3 IRX6 MEIS3P1 MK...   \n",
       "C433                                CALM3 FSD1 GNAZ SRC   \n",
       "\n",
       "     CD_AnnotatedMembers_Size CD_AnnotatedMembers_Pvalue  \\\n",
       "C435                        1          0.007995642253764   \n",
       "C367                       18      2.977322934006789E-14   \n",
       "C434                        1         0.0249546954540196   \n",
       "C372                        3          1.677832487479E-4   \n",
       "C433                        2         0.0034578149777732   \n",
       "\n",
       "                                    CD_AnnotatedMembers  \\\n",
       "C435                                            SLC17A6   \n",
       "C367  GABRA2 UNC13B GABRB3 UNC13C GABRB2 CHRNA7 GAD1...   \n",
       "C434                                             NKX6-2   \n",
       "C372                                  PKNOX1 PBX1 HOXA4   \n",
       "C433                                          SRC CALM3   \n",
       "\n",
       "                                       CD_CommunityName  \\\n",
       "C435  Organic anion transporters Homo sapiens R-HSA-...   \n",
       "C367        chemical synaptic transmission (GO:0007268)   \n",
       "C434             regulation of myelination (GO:0031641)   \n",
       "C372  Activation of HOX genes during differentiation...   \n",
       "C433                             GnRH signaling pathway   \n",
       "\n",
       "     CD_AnnotatedMembers_Overlap  HiDeF_persistence CD_MemberList_LogSize  \\\n",
       "C435                        0.25                  5                   2.0   \n",
       "C367                       0.228                 17                 6.304   \n",
       "C434                        0.25                  5                   2.0   \n",
       "C372                         0.2                 16                 3.907   \n",
       "C433                         0.5                  5                   2.0   \n",
       "\n",
       "     CD_Labeled                              CD_AnnotatedAlgorithm  ...  \\\n",
       "C435       true  Annotated by Enrichr [Docker: coleslawndex/cde...  ...   \n",
       "C367       true  Annotated by Enrichr [Docker: coleslawndex/cde...  ...   \n",
       "C434       true  Annotated by Enrichr [Docker: coleslawndex/cde...  ...   \n",
       "C372       true  Annotated by Enrichr [Docker: coleslawndex/cde...  ...   \n",
       "C433       true  Annotated by Enrichr [Docker: coleslawndex/cde...  ...   \n",
       "\n",
       "     abnormal nervous system development:-log(hyper_p)  \\\n",
       "C435                                         -0.000000   \n",
       "C367                                          4.744992   \n",
       "C434                                          4.672897   \n",
       "C372                                          3.544148   \n",
       "C433                                          1.597754   \n",
       "\n",
       "     abnormal nervous system development:-log(OR_p)  \\\n",
       "C435                                      -0.000000   \n",
       "C367                                       5.097764   \n",
       "C434                                       2.892220   \n",
       "C372                                       3.358930   \n",
       "C433                                       0.725707   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C435                                    0.000000   \n",
       "C367                                    1.256907   \n",
       "C434                                    3.719366   \n",
       "C372                                    1.928662   \n",
       "C433                                    1.520461   \n",
       "\n",
       "     abnormal nervous system development:num_genes  \\\n",
       "C435                                             0   \n",
       "C367                                            16   \n",
       "C434                                             3   \n",
       "C372                                             5   \n",
       "C433                                             1   \n",
       "\n",
       "           abnormal nervous system development:gene_ids  \\\n",
       "C435                                                      \n",
       "C367  CDK5R2 CHRNA7 CNTN1 GFAP GNAZ LGI1 MAPT MPZ NC...   \n",
       "C434                                    EN1 GSX2 NKX6-2   \n",
       "C372                          EN1 GSX2 IRX1 MNX1 NKX6-2   \n",
       "C433                                               GNAZ   \n",
       "\n",
       "      abnormal somatic nervous system morphology:-log(hyper_p)  \\\n",
       "C435                                          -0.000000          \n",
       "C367                                           5.597954          \n",
       "C434                                           5.136098          \n",
       "C372                                           4.183997          \n",
       "C433                                          -0.000000          \n",
       "\n",
       "      abnormal somatic nervous system morphology:-log(OR_p)  \\\n",
       "C435                                          -0.000000       \n",
       "C367                                           6.389371       \n",
       "C434                                           3.276389       \n",
       "C372                                           4.259458       \n",
       "C433                                          -0.000000       \n",
       "\n",
       "      abnormal somatic nervous system morphology:log_OR  \\\n",
       "C435                                           0.000000   \n",
       "C367                                           1.462644   \n",
       "C434                                           4.003268   \n",
       "C372                                           2.213053   \n",
       "C433                                           0.000000   \n",
       "\n",
       "     abnormal somatic nervous system morphology:num_genes  \\\n",
       "C435                                                  0     \n",
       "C367                                                 15     \n",
       "C434                                                  3     \n",
       "C372                                                  5     \n",
       "C433                                                  0     \n",
       "\n",
       "      abnormal somatic nervous system morphology:gene_ids  \n",
       "C435                                                       \n",
       "C367  CHRNA7 GABRB2 GABRB3 GRM7 KIF5C MAPT MOG MPZ N...    \n",
       "C434                                    EN1 GSX2 NKX6-2    \n",
       "C372                          EN1 GSX2 IRX6 MNX1 NKX6-2    \n",
       "C433                                                       \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame.from_dict(dict(G_ASD.nodes(data=True)), orient='index').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "G_ASD_cx = ndex2.create_nice_cx_from_networkx(G_ASD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDEx password: ········\n",
      "Generating CX\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update the network on ndex\n",
    "ndex_password=getpass.getpass('NDEx password: ') # enter password again (not sure if necessary)\n",
    "G_ASD_cx.update_to(ASD_uuid,ndex_server,ndex_user,ndex_password)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0ecb653c-9405-11eb-9e72-0ac135e8bacf'"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ASD_uuid\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
