{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NetColoc analysis of rare variants in Autism spectrum disorder (ASD) and Congenital Heart Disease (CHD)\n",
    "\n",
    "Example of NetColoc workflow on genes associated with rare exome variants in ASD and CHD\n",
    "\n",
    "**Some background:**\n",
    "\n",
    "Here we introduce NetColoc, a tool which evaluates the extent to which two gene sets are related in network space, i.e. the extent to which they are colocalized in a molecular interaction network, and interrogates the underlying biological pathways and processes using multiscale community detection. This framework may be applied to any number of scenarios in which gene sets have been associated with a phenotype or condition, including rare and common variants within the same disease, genes associated with two comorbid diseases, genetically correlated GWAS phenotypes, GWAS across two different species, or gene expression changes after treatment with two different drugs, to name a few. NetColoc relies on a dual network propagation approach to identify the region of network space which is significantly proximal to both input gene sets, and as such is highly effective for small to medium input gene sets.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Obtain input gene sets and interactome_\n",
    "\n",
    "### 1. Load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load required packages\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import random\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "import getpass\n",
    "import ndex2\n",
    "\n",
    "import json\n",
    "import cdapsutil\n",
    "\n",
    "from gprofiler import GProfiler\n",
    "gp = GProfiler(\"MyToolName/0.1\")\n",
    "\n",
    "from scipy.stats import hypergeom\n",
    "from scipy.stats import norm\n",
    "\n",
    "# latex rendering of text in graphs\n",
    "import matplotlib as mpl\n",
    "mpl.rc('text', usetex = False)\n",
    "mpl.rc('font', family = 'serif')\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'sans-serif'\n",
    "rcParams['font.sans-serif'] = ['Arial']\n",
    "\n",
    "sns.set(font_scale=1.4)\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.set_style(\"ticks\", {\"xtick.major.size\": 15, \"ytick.major.size\": 15})\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "from datetime import datetime\n",
    "import sys\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# verify DDOT was installed\n",
    "import ddot\n",
    "\n",
    "from netcoloc import netprop_zscore\n",
    "from netcoloc import netprop\n",
    "from netcoloc import network_colocalization\n",
    "from netcoloc import validation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Select two gene sets of interest. Load gene sets from text files into python.\n",
    "\n",
    "\n",
    "Identify two gene sets of interest. Gene sets should come from experimental data (not manual curation) to avoid bias. \n",
    "\n",
    "**Usage Note**: gene sets should be < 500 genes (propagation algorithm breaks down if seeded with larger sets). If your gene set is larger, only use the top 500 as seeds to the network propagation.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set names of geneset 1 and geneset 2\n",
    "# ------ customize this section based on your gene sets and how they should be labeled -------\n",
    "d1_name='ASD'\n",
    "d2_name='CHD'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number ASD genes:\n",
      "102\n",
      "['CHD8', 'SCN2A', 'SYNGAP1', 'ADNP', 'FOXP1']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# ------ customize this section based your input genesets -------\n",
    "\n",
    "# load rare variants Satterstrom et al 2020; PMID: 31981491\n",
    "\n",
    "D1_df = pd.read_csv('data/Satterstrom--Top-102-ASD-genes--May2019.csv')\n",
    "D1_df.index = D1_df['gene']\n",
    "print('number '+d1_name+' genes:')\n",
    "print(len(D1_df))\n",
    "D1_genes = D1_df.index.tolist() # define rare variant genes to seed network propagation\n",
    "print(D1_genes[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number CHD genes:\n",
      "66\n",
      "['KMT2D', 'CHD7', 'PTPN11', 'NSD1', 'NOTCH1']\n"
     ]
    }
   ],
   "source": [
    "# ------ customize this section based your input genesets -------\n",
    "\n",
    "# load D2 rare variant genes from Jin et al 2017; PMID: 28991257\n",
    "D2_df = pd.read_csv('data/CHD_HC.tsv',sep='\\t')\n",
    "\n",
    "D2_genes = D2_df['0'].tolist()\n",
    "print('number '+d2_name+' genes:')\n",
    "print(len(D2_genes))\n",
    "print(D2_genes[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of ASD and CHD genes overlapping:\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "# Output the overlap between the two gene sets\n",
    "print('number of '+d1_name+' and '+d2_name+' genes overlapping:')\n",
    "print(len(np.intersect1d(D1_genes,D2_genes)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Select molecular interaction network to use for the analysis. \n",
    "\n",
    "Identify network UUID on  Ndex (Pratt et al. 2015) and use this to import to jupyter notebook. \n",
    "\n",
    "**Usage note**: PCnet is a general purpose interactome, a good starting place https://www.sciencedirect.com/science/article/pii/S2405471218300954.\n",
    "\n",
    "We recommend switching to “STRING high confidence” if using a machine with low memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nodes:\n",
      "18820\n",
      "\n",
      "number of edges:\n",
      "2693109\n"
     ]
    }
   ],
   "source": [
    "\n",
    "interactome_uuid='4de852d9-9908-11e9-bcaf-0ac135e8bacf' # for PCNet\n",
    "# interactome_uuid='275bd84e-3d18-11e8-a935-0ac135e8bacf' # for STRING high confidence\n",
    "ndex_server='public.ndexbio.org'\n",
    "ndex_user=None\n",
    "ndex_password=None\n",
    "G_int = ndex2.create_nice_cx_from_server(\n",
    "            ndex_server, \n",
    "            username=ndex_user, \n",
    "            password=ndex_password, \n",
    "            uuid=interactome_uuid\n",
    "        ).to_networkx()\n",
    "nodes = list(G_int.nodes)\n",
    "\n",
    "# remove self edges from network\n",
    "G_int.remove_edges_from(nx.selfloop_edges(G_int))\n",
    "\n",
    "# print out the numbers of nodes and edges in the interatome for diagnostic purposes:\n",
    "print('number of nodes:')\n",
    "print(len(G_int.nodes))\n",
    "print('\\nnumber of edges:')\n",
    "print(len(G_int.edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "int_nodes = list(G_int.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Identify network colocalized gene network_\n",
    "\n",
    "### 4. Precalculate matrices needed for propagation. This step should take a few minutes \n",
    "\n",
    "Network propagation from genes on selected interactome\n",
    "- Control for degree of input genes\n",
    "- Generate a proximity z-score, which defines genes which are closer to input set than expected by chance.\n",
    "- Repeat for rare and common variant genes, defined above\n",
    "\n",
    "Background on network propagation: https://www.nature.com/articles/nrg.2017.38.pdf?origin=ppub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "calculating w_prime\n",
      "\n",
      "calculating w_double_prime\n"
     ]
    }
   ],
   "source": [
    "# pre-calculate matrices used for network propagation. this step takes a few minutes, more for denser interactomes\n",
    "print('\\ncalculating w_prime')\n",
    "w_prime = netprop.get_normalized_adjacency_matrix(G_int, conserve_heat=True)\n",
    "\n",
    "print('\\ncalculating w_double_prime')\n",
    "w_double_prime = netprop.get_individual_heats_matrix(w_prime, .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Subset seed genes to those found in the selected interactome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102\n",
      "96\n",
      "66\n",
      "60\n"
     ]
    }
   ],
   "source": [
    "# subset seed genes to those found in interactome\n",
    "print(len(D1_genes))\n",
    "D1_genes = list(np.intersect1d(D1_genes,int_nodes))\n",
    "print(len(D1_genes))\n",
    "\n",
    "print(len(D2_genes))\n",
    "D2_genes = list(np.intersect1d(D2_genes,int_nodes))\n",
    "print(len(D2_genes))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. Compute network proximity scores from both seed gene sets independently. Network proximity scores correct for the degree distribution of the input gene sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating D1 variant z-scores: \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "417e51e7044c4123b1bd9101836ed6d9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1000 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TRIP12</th>\n",
       "      <td>23.674264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NUP155</th>\n",
       "      <td>20.643715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADNP</th>\n",
       "      <td>20.489814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TLK2</th>\n",
       "      <td>19.532815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CREBBP</th>\n",
       "      <td>18.945873</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                z\n",
       "TRIP12  23.674264\n",
       "NUP155  20.643715\n",
       "ADNP    20.489814\n",
       "TLK2    19.532815\n",
       "CREBBP  18.945873"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# D1 network propagation\n",
    "print('\\nCalculating D1 variant z-scores: ')\n",
    "z_D1, Fnew_D1, Fnew_rand_D1 = netprop_zscore.calculate_heat_zscores(w_double_prime, int_nodes, \n",
    "                                                                    dict(G_int.degree), \n",
    "                                                                    D1_genes, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_D1 = pd.DataFrame({'z':z_D1})\n",
    "\n",
    "z_D1.sort_values('z',ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating D2 variant z-scores: \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6b44fe0a0f26429b8d23d53fb38b672d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1000 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACTB</th>\n",
       "      <td>25.046474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPL5</th>\n",
       "      <td>24.320330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMAD2</th>\n",
       "      <td>22.308273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U2SURP</th>\n",
       "      <td>22.132597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAA15</th>\n",
       "      <td>20.724822</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                z\n",
       "ACTB    25.046474\n",
       "RPL5    24.320330\n",
       "SMAD2   22.308273\n",
       "U2SURP  22.132597\n",
       "NAA15   20.724822"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# D2 variant network propagation\n",
    "print('\\nCalculating D2 variant z-scores: ')\n",
    "z_D2, Fnew_D2, Fnew_rand_D2 = netprop_zscore.calculate_heat_zscores(w_double_prime, int_nodes, \n",
    "                                                                    dict(G_int.degree), \n",
    "                                                                    D2_genes, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_D2 = pd.DataFrame({'z':z_D2})\n",
    "z_D2.sort_values('z',ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. Compute the network colocalization enrichment for gene set pair. If gene sets are significantly colocalized, proceed with analysis. \n",
    "\n",
    "\n",
    "Size of network co-localization subnetwork compared to null model created by permuting individual propagation z-scores. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASD + CHD\n",
      "size of network intersection = 773\n",
      "observed size/ expected size = 3.013292792266012\n",
      "p = 0.0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "seed_dict = {d1_name:D1_genes,d2_name:D2_genes}\n",
    "# --------------------------------------------------------------------------------------------\n",
    "\n",
    "zthresh=3\n",
    "\n",
    "z_d1d2_size, high_z_rand = network_colocalization.calculate_expected_overlap(\n",
    "    z_D1['z'],\n",
    "    z_D2['z'],\n",
    "    plot=False,\n",
    "    num_reps=100,\n",
    "    z_score_threshold=zthresh,\n",
    "    z1_threshold=1.5,\n",
    "    z2_threshold=1.5\n",
    ")\n",
    "\n",
    "ztemp = (z_d1d2_size - np.mean(high_z_rand)) / np.std(high_z_rand)\n",
    "ptemp = norm.sf(ztemp)\n",
    "print(d1_name + ' + ' + d2_name)\n",
    "print('size of network intersection = ' + str(z_d1d2_size))\n",
    "obs_exp_temp = float(z_d1d2_size) / np.mean(high_z_rand)\n",
    "print('observed size/ expected size = ' + str(obs_exp_temp))\n",
    "print('p = ' + str(ptemp))\n",
    "\n",
    "# save the number of overlapping genes and overlap p-value\n",
    "network_num_overlap = z_d1d2_size\n",
    "network_pval_overlap = ptemp\n",
    "network_obs_exp = obs_exp_temp\n",
    "network_exp_mean_overlap = np.mean(high_z_rand)\n",
    "network_exp_std_overlap = np.std(high_z_rand)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8. Plot the sizes of the observed and expected NetColoc subnetworks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-1.0, 2.0)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAALQAAAFoCAYAAAD+RtHSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAfcklEQVR4nO3de7xc49n/8U8kiHOQIpFo0Ob7KNogUaRIqKK0WkU9oqUoKtVfCAkaQROkKnmclZRGiweP54dQv4bWIVHHNEKcLudDSuoQKakQkv3741pjTyazx5q9Z2bNXq7367Vfe+Zea81c+7Wvfe977nvd992lpaWFEPJihawDCKGWIqFDrkRCh1yJhA65EgkdciUSOuRKt0a/oaQVgdOBg4F1gNnAaDO7Pzk+ADgfGAi8DfyXmZ1XdP0KwGnAEUAPYDow3MxeTPn+KwODgDeAJR3/iUKDdQV6AY+Y2UelBxue0HgyHgYcArwIjAb+LGkz4EPgTuBW4GhgW+BSSe+Z2ZXJ9acCxwA/AV4FJgDTJG1R7gcsYxAwo4Y/T8jGjsB9pYVZJPT3gGvN7A4ASSPx2nYwsCnwMXCkmX0CPC3pS8DJwJWSVgJGAieb2W3J9Qfite3+wNUp3v8NgGuuuYYNNtiglj9XaIB58+YxbNgwSH6PpbJI6DeBvSVdCLwGHAksxpsehwIzkmQuuAcYI2lDYENgDeCuwkEze0/SLGBn0iX0EoANNtiAPn36dPRnCdkp21zMIqGPBW4AXsKDWgrsb2bPSuqDNzmKvZ587wv0Th7PLXNO39I3ktQDb2cXiyzOsSwSejPgX3jT4x/AUcAfJO0MrAqUtoMLz7snx2njnO5l3msE3mYPnxMNTWhJfYFrgD3M7O6keKakzfGej0XAyiWXFZ4vTI4XyhaXnLOwzFueB0wpKetDfCjMrUbX0F8HVgIeKSl/CPgW8AqtzYqC4mbGCkVlVnLOk6VvZmYLgAXFZZKqjzp0Go0eWCm0fb9aUr4F8Czep/wNScV/aEOBZ81sHvAY8B4wpHBQ0prA1sC9dYo5dCKNrqEfxvsOr5T0MzzBD8GTdjD+QXFUcnwCPrhyPPAzADP7SNJFwFmS5iXnT0he58YG/yyhCTU0oc1sqaTvAuPxtu3awBxgVzN7CEDS7sAFwCy8r3GUmV1V9DJjk7gnA6vgtfruZvZxo36O0Lwa3sthZu8Cw5OvcscfAbavcP0SfHRxdF0CDJ1a3JwUciUSOuRKJHQOffjhh1mHsJxGxZTFSGGos+7du9OlS5esw1hGo1YXiBo65EokdMiVSOiQK5HQIVcioUOuREKHXImEDrkSCR1yJRI65EokdMiVSOiQK5HQIVcioUOuREKHXImEDrkSCR1yJRI65EokdMiVSOiQK5HQIVcioUOuREKHXImEDrkSCR1yJRI65EokdMiVSOiQK5HQIVcioUOuREKHXImEDrkSCR1yJRI65EokdMiVSOiQK5HQIVcioUOuREKHXImEDrkSCR1yJRI65EokdMiVSOiQK5HQIVcioUOuREKHXImEDrkSCR1yJRI65EokdMiVSOiQK92yeFNJPwZOAjYBXgBON7P/SY71Ay4CdgI+AK4AxprZkqLrhwMjgV7ALOBYM5vVyJ8hNKeG19CSDgauBC4FNgf+CFwnaQdJKwLTgC7AYOBI4CjgtKLrDwF+A4wFtgGeA+6U1LORP0doTg1NaEldgHHAhWZ2oZm9YGYTgDuAocB+QD/gYDObY2ZTgdHACEmrJC9zCnCxmV1tZk8BhwPvA0c38mcJzanRNXR/PGGvLS40sz3N7ExgR2C2mb1bdPgeYA1ga0nrJa9xV9G1S4AZwM51jTx0Co1uQyv5vqqkacBWwEvAeDO7FegDzC255vXke19gUfK43DmDlnszqQfQo6S4T3sCD51DVTW0pNWKHv9A0vGSNq3iJdZMvk8BrgO+hbeZb5G0G7Aq8FHJNYXn3ZPjtHFO9zLvNwL/gyn+mlFFvKGTSVVDS/oy8CfgeuBUSacDp+If3sZJ2tXMHkzxUh8n3yeZ2e+Tx7MlbQOcgNfAK5dcU3i+kNYautw5C8u833n4H0+xPkRS51baJsfZeG3+J0ndgJ8D/xf/IDYFGA98M8XrFJoKc0rKnwS+B/wFb4YU61107WtFZXNKzilthmBmC4AFxWWSSk8LOZK2yTEEOCmphXcE1gEuM7N3gIsp035twyy8R6L0/C2B54HpwICk7VswNLlmlpm9CVgSDwCSuuJ91vemjCHkWNoaehXgzeTxHvi//unJ8yVlryjDzBZJOgdvtrwOPAgcCOwO7Ab8Da/tr5d0It4jMgGYaGaLk5eZCFwgyYCHgRPxtvXktHGE/Eqb0M8CQyQ9COwL3F2UYD/Ga81UzGy8pA/w/ugNgWeAH5jZXwEk7QFcAjwEzE8ejyu6frKktZKydYGZwG5m9nbaGEJ+pU3oX+MjeiOB1fARPCQ9DGwNHFTNm5rZJGBSG8eex3s/Kl1/LnBuNe8ZPh9StaHN7Dq83XomsIOZ3Z0cmg58x8xuqE94IVQn9cCKmf0Nb+MWl51Q84hC6IDUCS1pQ2AM3hzojd88dCDwmJldU5/wQqhOqiaHpP7AbPwD4UPASsmhnsAfJO1bl+hCqFLafuiJ+LDxxnivRhcAMzsMuAEYVZfoQqhSNQMrE8zsA6Cl5NgVwFdqGVQI7ZU2oZcmX+WsVuFYCA2VNqGnA2MkrVlU1pIMOw8nbvYJTSJtL8co4H58utO9eLNjNN7U2Bj4Rl2iC6FKaQdWngYGAnfiybsE2AUf8t7BzB6rW4QhVKGagZUXgIPrGEsIHVbNwMoK+Czr1ShTs5vZXctdFEKDpZ2xMhC/oX9Dkj7oREvyvAXoWvPoQqhS2hp6EvAJcCg+MyS66UJTSpvQg4CDzOymegYTQkel7YeeD3xYz0BCqIW0CT0F+EUykBJC00rb5FgMbAe8lMxS+aDkeIuZHVLTyEJoh7QJfSitywFsU+Z46Q1LIWQiVUKb2cb1DiSEWqh6bTtJffH+6DnAUjNb9BmXhNAw1YwU7oXPtO6PNzG2xe/Aexs42syibzpkLu0UrD2BW4AX8dtFC6OF9wI/IWashCaRttvuV8B1ZrYXvkJRYQrW+fjSBofWJboQqpQ2obfAF5op525go9qEE0LHpE3oBfiC4+X0o2SFzxCykjahbwbGS9qhqKxF0kbAL4Hbah1YCO2RNqFPxrd9mEHrFhE34DNWliTHQ8hc2ilYC/Ch76OAv+ILkz+Kr7q/TbJOdAiZq2YK1mLgd8lXCE0p7YyVsRUOL8X3N3keuKNo3egQGi5tDf0jfLOdlfGZK+/gi413o3UaFsBTkoaa2Vu1DjSENNJ+KDwV3zrtP4HuZtYLT+59gbeBA/C+6qXAWXWIM4RU0tbQpwO/NLPrCwVm1gLcLGl9YJyZbSbpLHxhxxAykbaG3ghvI5fzCj64Aj6Bdp0OxhRCu6VN6CeBn7Zx7AhaNw3qT2s/dQgNV02T42ZJs/H1Od4C1sc3y9wc+IGkrYBzgCtrHmUIKaUdWPkTvhXFe/gHxIvx0cF3gKFmNhXvBbkuOR5CJqoZWLkbuFvSSng7+c3im/rN7Fbg1tqHGEJ6VU/BSgZO5tUhlhA6LO2HwhA6hUjokCuR0CFXqpn1vRKwlZk9lDzfEN988xYz+6hO8YVQlbSzvr+ID678d1HxFng33YPJ8HcImUvb5PgN8DGwT6HAzKYBmwErAhNqH1oI1Uub0EOBk8xsTnGhmRlwGrBnrQMLoT3SJvSKFY4tBlavQSwhdFjahH4AOEHSysWFyfPjgQdrHVgI7ZG2l+NUfMb3S5KmAW8CXwB2B9YGdq5PeCFUJ+3NSTOBr+NJvTtwHLA3XjPvYGaP1C3CEKpQzc1JjwM/rGMsIXRYNQMr3fGb+YcAPfC5hDOA35tZ6RYVIWQi7cDK2sBDwAW07ia7HXAh8IikHvUKMIRqpK2hzwZ6AzuZ2X2FQkk7AjcC44Gf1z68EKqTNqH3AcYUJzOAmc1IFqE5lXYktKT+wCxghJn9LinrB1wE7ITvtnUFMNbMlhRdNxwYCfRKrj/WzGZV+/4hf9L2Q6+Or95fzov4ojNVkbQicA3efCkum4YvXDMYOBJfT++0onMOwYfix+LNn+eAOyX1rDaGkD9pE/pp4DttHNuHtpc4qOQMfI5isf3wJREONrM5yVzF0cAISask55wCXGxmV5vZU8DhwPvA0e2IIeRM2ibHucB1ycjgdfgUrA3wlZQOA46p5k0l7YTXvAOAV4sO7QjMNrN3i8ruAdYAtpb0HL5Uwl2Fg2a2RNIMfHBnfDVxhPxJu0/hDUl795d41x14s+AjfNWky9O+YdIj8ke83fuapOLDffDFaooV1vnoCxS2kCt3zqA23qtHSXGftLGGzqeagZXxki4CtseHu+cDD5XUpmlcCtxvZteWObYqPqxerDB5oHtyvLis+JzuZV5vBEXt75B/Vc36ThY+/3/tfTNJP8KbFVu2ccoifBHIYoXnC2mtocuds7DM650HTCkp64MPCIUcajOhJb1G+j28W8zsiynOOwxfcam0qXGRpOPwfQ+3Krmmd/J9LvBaUdmcknNKmyGFP8AFxWUl7xtyplIN/Vdqvyn9wcAqJWXP4T0e1+LNmZ9I6pEkI/jkgveBWWa2WJLhw+/TACR1xfusf1vjWEMn1GZCm9mhtX4zM/tHaVlSY75lZq9I+ifeU3G9pBPxLrwJwMSinQEmAhckif0wcCLetp5c63hD51NVG1rS3ix/c9LtyVrRHWZmH0raA7gEv3dkfvJ4XNE5kyWtlZStC8wEdjOzt2sRQ+jc0u6x0h2YCnwT38btbaAnXjveK2nP9i5lYGZdSp4/jy8MWemac/G+8RCWkXak8AxgB3yvlcKWFKvge3wPwoehQ8hc2ibHgcDpZnZNocDMPgH+mKzJcQw+6BJCptLW0D3xjTbLeZTWrrUQMpU2oZ+n7YmwQ2jtHw4hU2mbHL8FzpO0iOVvThqFb1kRQubSJvRl+L3HZ7LsHW1dgKuIpcBCk0h7t91S4AhJE/Gmxzp4H/G9ZvZ0HeMLoSpp+6G7mtmSJHmfTsq+bGbP1TW6EKpU8UOhpE0l/QUfQCkuXwPf13tGstRuCE2hzYSW1IvWu9/eKDncBfg1vpzufZK+ULcIQ6hCpRp6FD7MPcDMrio+YGbvmdkYfHmwlZJzQ8hcpYT+NnCOmbXZx2xmL+B3v+1V68BCaI9KCd0XeCLFa8wEoh0dmkKlhH6HdOttrAn8qzbhhNAxlRL6QWD/FK+xP/BUbcIJoWMqJfQlwAGSjm3rhOTYgcDvah1YCO1RaQrW3ZLOBs6XdCRwG/AS0BVvM++Jb+12hZld14hgQ/gsFUcKzWyMpNn48lujiw61AH8HDjSz/6lfeCFU5zOHvs3sRuDG5Eb+vsAnwKtmNr/ewYVQrbT3cowFfpfstVJ6rB9wgpnF+tAhc2lv8D8N2LCNY9vRut5dCJmqtHLS/fjQNvi9Gw9WWHXo7zWOK4R2qdTk+CneJdcF/1B4FVC6UMwS4F0gPhiGplCp2+5JfKsJJH2Mt6GXW/kohGaSdsbKGQCSNgR2wdvTU/A9Tp4ws4/rFWAI1Uj7oRBJE/CBlavwuYW98d2xZsX90KFZpN2ncCS+69QYYHO8XQ1wFr7n96/qEl0IVUpbQ/8MGG9m5wDPFgrNbDqe5HvXIbYQqpY2ofsA97Vx7Hm8lg4hc2kT+lV8K4lytmXZnaxCyEzahWYmA+MlLcaX1QVYS9IPgZPxjTBDyFw1+xT2wxcZLyw+/pfk+zXEykmhSaTth24Bhks6D++HXhffjGe6maWZdxhCQ1S7rdtz+CY/oQYmTpzIpEmT2jx+/PHHM3LkyAZG1PlVujmpmr7lFjOLDS6rNHLkyE8Tdr/99gPgxhtvzDKkTq9SDT0mxfXFmwVFQofMVbo56bPWvTsMX2RmJWI7itAkqmpDw6c3KE0Gdse3dTs8WUEphMxVu0/h4XgX3orACDO7sC5RhdBOaecU9sVr5W8B9+C18kt1jCuEdvnMhE7W5DgHHyYfbmaX1j2qENqpUrfdRviKSLviG9kfYWZxz0aJTz75hG7dqv4oEuqk0m/iCWA1fERwHn4vR1vntpjZIbUNrXPo1q0bZ5xxRodf5+WXXwaoyWuddtrntwe1UkLPx1cghbbvtCuoyeb1IXRUpX7ofg2MI4SaSD2nMITOIBI65EokdMiVSOiQK9GBmqGZM2cya9asZcouv/zyTx9vvfXWDBw4sNFhdWqVBla2KMxGkbQE2N7MHpbUFVgMDDKzWW1dHz7bwIEDI2FrrFIN/aik94AH8IVltpb0NPABrQvNhNBUKrWh18ZXHy0slTsJX2l0Dj6QcoCkIZJ61DXCEKrQZkKb2UIzu7NoatVQfJOgc/Eaeh98I6H5kl6pe6QhpFBp8/ojtOzNG0vN7Bl8sUaAYcAa+Ab2MZMzNIVKbehjgUslvYs3MQ6VtDbwaHK8JVnewJKvVCStgS/u+H2gJ/AM8Cszm5oc7wdcBOyEt9evAMaa2ZKi1xiO/xH1AmYBx8YH1ACVmxxfwxPuULyJsS1wLfBPPMHHSzpJ0h6SelXxnlOA7+A7BAwAbgFukrSLpBWBacn7DQaOBI6iaAKupEPwlZrGAtvgyyrcKalnFTGEnKo4sGJm/zKz25Onw82sJ/AVPOHeB3YDrgbmpnkzSRsA++LTt+40s+fNbBzJLBhgP3yFpoPNbE5Sa48GRkhaJXmZU4CLzexqM3sque594OiUP3PIsbQjha8AHyWPn02ejzOzXZMk3yjl6/wb34H23pLyFmAd/DbV2Wb2btGxe/C2+taS1gP6A3cVDiZNkRnAziljCDmWdimwjYseLwU2Ljmeau8VM3sf+HNxmaTt8OXFfoHPWSyt7V9PvvcFFiWPy50zqPT9ki7FHiXFfdLEGjqnTO/lkLQZcBPwEHAZsCqt/wkKCs+7J8dp45zuZd5iBL6NRvHXjI7GHZpXZgktaSd8EfU3gL2SjYcWASuXnFp4vpDWGrrcOQvLvM15+H+T4q/Pmn0TOrFMbk6SNAy4Em9L/yBpigC8BmxVcnrv5Pvc5HihbE7JOct9MDWzBficyOL37kDkodk1vIaWdBDwR+AGvGZ+v+jwdGBAyXD6ULwXY5aZvYn3eQ8per2ueJ916QfN8DnU0BpaUh98wZq7gVHAukU15mLgZmA8cL2kE/EuvAnARDNbnJw3EbhAkgEPAyfibevJjfkpQjNrdA29L558u+A9E28UfU01sw+BPfB+7oeAS4FLaN01ADObjO9wOw6YCWwK7GZmbzfuxwjNqqE1tJldAFzwGec8j3ffVTrnXPwmqRCWEVOwQq5EQodciYQOuRIJHXIlEjrkSiR0yJVI6JArkdAhVyKhQ65EQodciYQOuRIJHXIlEjrkSiR0yJVI6JArkdAhVyKhQ65EQodciYQOuRIJHXIlEjrkSiR0yJVI6JArkdAhVyKhQ65EQodciYQOuRIJHXIlEjrkSiR0yJVI6JArkdAhVyKhQ65EQodciYQOuRIJHXIlEjrkSiR0yJVI6JArkdAhVyKhQ65EQodciYQOuRIJHXIlEjrkSiR0yJVI6JArkdAhVyKhQ65EQodciYQOuRIJHXIlEjrkSiR0yJVI6JAr3bIOoD0krQCcBhwB9ACmA8PN7MUs4wrZ66w19KnAMcBRwPZACzBN0sqZRhUy1+kSWtJKwEjgdDO7zcweBw4EegP7ZxpcyFynS2hgALAGcFehwMzeA2YBO2cUU2gSnbEN3Sf5Prek/HWgb3GBpB54G7vYFwHmzZtXs4AWLlxYs9eqhblz59KtW3P9aufOLf11tU/R761ruePN9VOns2ry/aOS8o+A7iVlI/APj8sZNmxYbaNqIlOnTmWTTTbJOoxl7LrrrrV+yV7AC6WFnTGhFyXfVwYWF5WvDJRWlecBU0rKVgI2AZ4DltQ+vHbrA8wAdmT5/z5ZacaYuuLJ/Ei5g50xoV9LvvcGrKi8N/Bk8YlmtgBYUOY1nq1HYB0hqfBwrpm9nGEon2rGmBLL1cwFnfFD4WPAe8CQQoGkNYGtgXsziik0iU5XQ5vZR5IuAs6SNA94CZiA/0u8MdPgQuY6XUInxuKxTwZWwUcKdzezjzONKmSuUya0mS0BRidfebEAOIPybf6sLKD5YqqoS0tLS9YxhFAznfFDYQhtioQOuRIJHXIlEjrkSqfs5eisJG2U9lwze7WesRQ0Y0wdEQndWC/jkxHSKHs3WR28TPPF1G6R0I21W9HjLfGZN2cB9wMfA4OSsnGf85jaLfqhMyLp78BZZva/JeXfASaY2eYRU/XiQ2F2NgPmlCl/DujX2FA+1YwxVSUSOjuPA8clM9gBkLQicAowM2Jqn2hDZ+dEYBqwp6TZQBdgG2A1YJeIqX2iDZ0hSRsDRwJbJEWzgcvMLLPZIc0YUzUioZtAsp7IYjNrml9GM8aURiR0hiQdDYwCNgL6AycAr5vZ+IipfeJDYUYkHYTPtPkDrZN9nwZ+KWlkxNQ+kdDZOQH4P2Z2OsnsczO7EDgaX+IsYmqHSOjsCJ86Vmo6JQvmNFAzxlSVSOjszAP+o0z5YHwVqCw0Y0xViYTOzmXAJZL2xft7N5f0c+AC4PcRU/tEL0eGJJ2FL1dWWMLsE+C3wHHJROBmimmEmS3NIqZqREJnTNKqwFfw/5bPmNl7ktYzszebKaasYqlWDH1nRNISYAMze4ui+ySSG+6fAlbPIKbHgCuBq82sU9y7USoSuoEk/RDYPXnaBbhA0qKS0zZm+UUnG+Uu4CTgHEl/wpP79s7Q1CiID4WN9QDwJeDLyfONk8eFry8BHwCHZhGcmR0HbAh8D1/l9XrgH5LOkfSVLGKqVrShMyLpbuD7yQqpTUnS6vgf15nA6mbW9FOwoobOiJkNBXaQ9O1CmaRLJe2RYViFOFaVNAy4AZiEL2E8Ktuo0ok2dEYkHQpcjg83F6wG3CLpx2Z2fQYx7QUcBHwXn094HTDYzMouLt6MosmREUlPA/9lZpeXlB8J/NzMvppBTJ8Af8F3PbjJzEq3/Wh6UUNn54vAnWXK78S30sjCFOAMM3vts05sVtGGzs7LQLmddHYku/sm9sP3oOm0oobOzsXA+ZK+BDyEL/YyCDgWX5M5Cw8A3wfOzej9Oyza0BmSdAJwHL6rE3jNPMHMLsoonv/F+6Dfxbf6WGbQx8x2yiCsqkRCNwFJ6wIfZ33PhKSKd9SZ2U8aFUt7RUJnSNIq+P7k/4H/m98SeMLM3sk0sE4sEjojktbH26zr45uG9gfOBwYCQ83smYzi6gX8FP8jGwHsBMwxM6t0XbOIXo7sTASeAL5Aa1v1R/jqRedkEVDyAfUJfLh7P/yOvx8CMyV9PYuYqhUJnZ1dgLFm9kGhILmvYxTwjYximgjcBGxK617qBwI347PBm14kdHbWwO+sKyer7tTBwKTixWWSmTNn4jv1Nr1I6OxMB4YXPW+RtBK+FvPfsgmJrpTPibVIljVodjGwkp0TgOmShuDz9ybjywisAeycUUzTgDHJnXbgf2Q98Tb9XzOKqSpRQ2fEzJ4GvgpMxRNpMfDfwAAzeyyjsI7HmxZv4ltO3w68gq8NfULblzWP6LZrApLWA5Y0Q/9zMkH2P4Gt8ArvCXyOYaeYKBsJnSFJJwO/ANZLil4Dzjazy7KLqlWyAulWwFOR0KEiSacAv8QXcbkf/0A2GDgGX5fj8gqX1yumvvjE2DH41hQz8QGWd4HdzGxWo2OqVnwozM7PgKPM7OqispuTG/9PxmezNNr5wJrAP4ED8PXsBgOH4R8Mv5lBTFWJD4XZWRsoN7VpBj7zOgtD8T+yl4E98SUMHgB+A8RIYajoZvze51IHJ8ey0A2YL6kLPvmg0FW3Ar4kWNOLJkd23gSOkbQjXisXNrncHrhV0h8KJ5rZjxsU06PA4cAb+H+Q25PBntHJsaYXCZ2drfC77QAKG1p+jI8grpV8NdpI4DagJ/BrM5sr6RJ8FsvuFa9sEtHLEZaR7FG4ZmEBHEn9gXeaoY88jWhDZ0TSmDbK15d0e6PjKUjWsdtW0vGShgO9OksyQ9TQmZH0PvAwMMzM5iVl38e76xaY2ZcrXV+nmPoBt+AzZ97FK7y18GbQfmb2dqNjqlbU0NnZBu/zfVzS/pKuAG4ErgW+llFMFwL/Bvqb2bpmtnYSy1r4AFDTixo6Q5K64Vs9HIR3i+1tZuUWn2lUPP8GvmFmj5aUbwfcYWZrZhNZelFDZ0RSV3xT+P3x1ZJex/c3yWq2CsB8fH5jqRbanozQVCKhszML7989zsz2wG8lfRC4R9L5jQpC0gqFL2Ac8FtJWxYd3wRvioxrVEwdEU2OjEh6BDi4dDa1pAOAS81s3QbFsRSvgQu60FojL8UnyrYA881sveVfobnEwEp2tse3TbsSv6Ntf2AffH+VLStdWGOHsWxCd2qR0Nn5GnAfvq7dNnjbdQC+8uj3aNCCjWY2pfi5pL3x21q/io9cPoGPGt7aiHg6KtrQ2TkHmGhmQ0g2ijezI/HusdOzCEjSd/F+6LeAsfhs73/ht7V+u9K1zSJq6Oxsg98TXepSfLP4LIzFZ8wUj2L+RtKZ+E3/mY1gphU1dHYWU/4GpI3wwY0sbA5cVab8KrIb7KlKJHR2bgbOTlYeBV8yYAu8yXFbRjH9E+hTprwP2e2dWJVI6OycgC8V8Ca+WdDjwGP4B7ETM4rpFuAiSZ/WxpIG4H9kUzOKqSrRD50xSbuy7JIBf85q51ZJPYA78PZ9odmzGn5z/25m9m4WcVUjEjosIxkx/BbeF94F/89xR2fZHjkSOuRKtKFDrkRCh1yJhA65EgkdciUSOuTK/wdyQELJeL/uVwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 144x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot the observed and expected overlaps\n",
    "plt.figure(figsize=(2,5))\n",
    "plt.bar([0,1],[network_exp_mean_overlap,network_num_overlap],color=['gray','black'],width=1)\n",
    "plt.errorbar([0],[network_exp_mean_overlap],[2*network_exp_std_overlap],color='k',capsize=4)\n",
    "plt.ylabel('# NetColoc genes')\n",
    "plt.xticks([0,1],['expected','observed'],rotation='vertical')\n",
    "plt.xlim([-1,2])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 9. Compute network colocalization subnetwork."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "773\n",
      "26351\n"
     ]
    }
   ],
   "source": [
    "# select the genes in the network intersection, make a subgraph\n",
    "\n",
    "G_overlap = network_colocalization.calculate_network_overlap_subgraph(G_int,z_D1['z'],z_D2['z'],z_score_threshold=zthresh,\n",
    "                                                                     z1_threshold=1.5,z2_threshold=1.5)\n",
    "print(len(G_overlap.nodes()))\n",
    "print(len(G_overlap.edges()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10. Transform NetColoc subnetwork edges to cosine similarities\n",
    "\n",
    "Cosine similarity transformation (and network transformations in general) provide better visualization of clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing the adjacency matrix...\n",
      "computing the cosine similarity...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/churas/anaconda3/envs/netcoloc/lib/python3.9/site-packages/scipy/spatial/distance.py:699: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  dist = 1.0 - uv / np.sqrt(uu * vv)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rank transforming...\n",
      "number of transformed edges returned = \n",
      "7032\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/churas/anaconda3/envs/netcoloc/lib/python3.9/site-packages/netcoloc/network_colocalization.py:241: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.\n",
      "  sim_rank.values[[np.arange(sim_rank.shape[0])]*2] = 0\n"
     ]
    }
   ],
   "source": [
    "G_cosSim=network_colocalization.transform_edges(G_overlap,method='cosine_sim',edge_weight_threshold=0.95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Compute network colocalized systems map_\n",
    "\n",
    "### 11. Convert network colocalization subnetwork to form used in community detection module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d1_seeds</th>\n",
       "      <th>d2_seeds</th>\n",
       "      <th>z_d1</th>\n",
       "      <th>z_d2</th>\n",
       "      <th>z_both</th>\n",
       "      <th>d1_name</th>\n",
       "      <th>d2_name</th>\n",
       "      <th>sum_seeds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PTEN</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>17.775818</td>\n",
       "      <td>18.270646</td>\n",
       "      <td>324.775666</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNB1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15.369891</td>\n",
       "      <td>16.212593</td>\n",
       "      <td>249.185790</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POGZ</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.518539</td>\n",
       "      <td>13.035299</td>\n",
       "      <td>163.182897</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.964197</td>\n",
       "      <td>12.579514</td>\n",
       "      <td>163.083297</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSD1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.263667</td>\n",
       "      <td>12.748115</td>\n",
       "      <td>156.338630</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDM5B</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.795978</td>\n",
       "      <td>11.553692</td>\n",
       "      <td>147.840781</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHD4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.485093</td>\n",
       "      <td>17.863170</td>\n",
       "      <td>80.117979</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2A</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.267999</td>\n",
       "      <td>16.113881</td>\n",
       "      <td>68.774032</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDK13</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.335774</td>\n",
       "      <td>15.374536</td>\n",
       "      <td>66.660517</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDM5A</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.980719</td>\n",
       "      <td>13.264934</td>\n",
       "      <td>66.068913</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2D</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.060878</td>\n",
       "      <td>15.117007</td>\n",
       "      <td>61.388323</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMAD2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2.662487</td>\n",
       "      <td>22.308273</td>\n",
       "      <td>59.395477</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITSN2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3.470855</td>\n",
       "      <td>15.705446</td>\n",
       "      <td>54.511328</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KCNMA1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15.406589</td>\n",
       "      <td>3.343184</td>\n",
       "      <td>51.507060</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHD8</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>16.649729</td>\n",
       "      <td>2.935970</td>\n",
       "      <td>48.883109</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        d1_seeds  d2_seeds       z_d1       z_d2      z_both d1_name d2_name  \\\n",
       "PTEN           1         1  17.775818  18.270646  324.775666     ASD     CHD   \n",
       "CTNNB1         1         1  15.369891  16.212593  249.185790     ASD     CHD   \n",
       "POGZ           1         1  12.518539  13.035299  163.182897     ASD     CHD   \n",
       "KMT2C          1         1  12.964197  12.579514  163.083297     ASD     CHD   \n",
       "NSD1           1         1  12.263667  12.748115  156.338630     ASD     CHD   \n",
       "KDM5B          1         1  12.795978  11.553692  147.840781     ASD     CHD   \n",
       "CHD4           0         1   4.485093  17.863170   80.117979     ASD     CHD   \n",
       "KMT2A          0         1   4.267999  16.113881   68.774032     ASD     CHD   \n",
       "CDK13          0         1   4.335774  15.374536   66.660517     ASD     CHD   \n",
       "KDM5A          0         1   4.980719  13.264934   66.068913     ASD     CHD   \n",
       "KMT2D          0         1   4.060878  15.117007   61.388323     ASD     CHD   \n",
       "SMAD2          0         1   2.662487  22.308273   59.395477     ASD     CHD   \n",
       "ITSN2          0         1   3.470855  15.705446   54.511328     ASD     CHD   \n",
       "KCNMA1         1         0  15.406589   3.343184   51.507060     ASD     CHD   \n",
       "CHD8           1         0  16.649729   2.935970   48.883109     ASD     CHD   \n",
       "\n",
       "        sum_seeds  \n",
       "PTEN            2  \n",
       "CTNNB1          2  \n",
       "POGZ            2  \n",
       "KMT2C           2  \n",
       "NSD1            2  \n",
       "KDM5B           2  \n",
       "CHD4            1  \n",
       "KMT2A           1  \n",
       "CDK13           1  \n",
       "KDM5A           1  \n",
       "KMT2D           1  \n",
       "SMAD2           1  \n",
       "ITSN2           1  \n",
       "KCNMA1          1  \n",
       "CHD8            1  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compile dataframe of metadata for overlapping nodes\n",
    "node_df = pd.DataFrame(index=list(G_overlap.nodes))\n",
    "node_df['d1_seeds']=0\n",
    "node_df['d2_seeds']=0\n",
    "node_df['d1_seeds'].loc[list(np.intersect1d(seed_dict[d1_name],node_df.index.tolist()))]=1\n",
    "node_df['d2_seeds'].loc[list(np.intersect1d(seed_dict[d2_name],node_df.index.tolist()))]=1\n",
    "node_df['z_d1']=z_D1.loc[list(G_overlap.nodes)]['z']\n",
    "node_df['z_d2']=z_D2.loc[list(G_overlap.nodes)]['z']\n",
    "node_df['z_both']=node_df['z_d1']*node_df['z_d2']\n",
    "node_df['d1_name']=d1_name\n",
    "node_df['d2_name']=d2_name\n",
    "node_df['sum_seeds']=node_df['d1_seeds']+node_df['d2_seeds']\n",
    "\n",
    "node_df = node_df.sort_values('z_both',ascending=False)\n",
    "node_df.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 12. Run community detection on NetColoc subnetwork (recommend HiDef).\n",
    "\n",
    "Documentation for CDAPS utils to build multiscale systems map in notebook\n",
    "\n",
    "https://cdapsutil.readthedocs.io/en/latest/quicktutorial.html#example  \n",
    "https://cdapsutil.readthedocs.io/en/latest/cdapsutil.html#community-detection  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "773\n",
      "26351\n"
     ]
    }
   ],
   "source": [
    "print(len(G_overlap.nodes()))\n",
    "print(len(G_overlap.edges()))\n",
    "G_overlap_cx = ndex2.create_nice_cx_from_networkx(G_overlap)\n",
    "G_overlap_cx.set_name(d1_name+'_'+d2_name+'_NetColoc_subgraph') \n",
    "for node_id, node in G_overlap_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_overlap_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "cd = cdapsutil.CommunityDetection()\n",
    "\n",
    "# Run HiDeF on CDAPS REST service\n",
    "G_hier = cd.run_community_detection(G_overlap_cx, algorithm='hidefv1.1beta',arguments={'--maxres':'20'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hierarchy name: hidefv1.1beta_(none)_ASD_CHD_NetColoc_subgraph\n",
      "# nodes: 81\n",
      "# edges: 85\n"
     ]
    }
   ],
   "source": [
    "# Print information about hierarchy\n",
    "print('Hierarchy name: ' + str(G_hier.get_name()))\n",
    "print('# nodes: ' + str(len(G_hier.get_nodes())))\n",
    "print('# edges: ' + str(len(G_hier.get_edges())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sr = cdapsutil.ServiceRunner()\n",
    "# sr.get_algorithms() # this will print out available options for each algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 13. Convert the NetColoc hierarchy to networkx format, and write out features of the hierarchy to a pandas dataframe, for easier access in Python.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nodes:\n",
      "81\n",
      "\n",
      "number of edges:\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "G_hier = G_hier.to_networkx(mode='default')\n",
    "G_hier\n",
    "\n",
    "nodes = G_hier.nodes()\n",
    "\n",
    "# print the number of nodes and edges in the hierarchy for diagnostic purposes\n",
    "print('number of nodes:')\n",
    "print(len(G_hier.nodes()))\n",
    "print('\\nnumber of edges:')\n",
    "print(len(G_hier.edges()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>represents</th>\n",
       "      <th>name</th>\n",
       "      <th>system_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PHF12 PRR23B BRPF1 HIST1H2AC ZNF236 TAF3 TMOD4...</td>\n",
       "      <td>770</td>\n",
       "      <td>False</td>\n",
       "      <td>9.589</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130</td>\n",
       "      <td>C773</td>\n",
       "      <td>C773</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCOR2 ZCCHC12 HDAC9 PHF2 PHF12 HIST1H4F BRPF1 ...</td>\n",
       "      <td>273</td>\n",
       "      <td>False</td>\n",
       "      <td>8.093</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6</td>\n",
       "      <td>C774</td>\n",
       "      <td>C774</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RO60 TRIM52 ANKRD17 DOP1A SUPT20H MACF1 INTS6 ...</td>\n",
       "      <td>241</td>\n",
       "      <td>False</td>\n",
       "      <td>7.913</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13</td>\n",
       "      <td>C775</td>\n",
       "      <td>C775</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CNTN2 SLC6A1 ATP2B4 NEFM ATP1A2 TMOD4 KCNJ3 AF...</td>\n",
       "      <td>165</td>\n",
       "      <td>False</td>\n",
       "      <td>7.366</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86</td>\n",
       "      <td>C777</td>\n",
       "      <td>C777</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZCCHC12 HDAC4 HDAC9 AFDN FOXQ1 ZEB2 SALL1 NRIP...</td>\n",
       "      <td>109</td>\n",
       "      <td>False</td>\n",
       "      <td>6.768</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12</td>\n",
       "      <td>C779</td>\n",
       "      <td>C779</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       CD_MemberList  CD_MemberList_Size  \\\n",
       "0  PHF12 PRR23B BRPF1 HIST1H2AC ZNF236 TAF3 TMOD4...                 770   \n",
       "1  RCOR2 ZCCHC12 HDAC9 PHF2 PHF12 HIST1H4F BRPF1 ...                 273   \n",
       "2  RO60 TRIM52 ANKRD17 DOP1A SUPT20H MACF1 INTS6 ...                 241   \n",
       "3  CNTN2 SLC6A1 ATP2B4 NEFM ATP1A2 TMOD4 KCNJ3 AF...                 165   \n",
       "4  ZCCHC12 HDAC4 HDAC9 AFDN FOXQ1 ZEB2 SALL1 NRIP...                 109   \n",
       "\n",
       "  CD_Labeled CD_MemberList_LogSize CD_CommunityName CD_AnnotatedMembers  \\\n",
       "0      False                 9.589                                        \n",
       "1      False                 8.093                                        \n",
       "2      False                 7.913                                        \n",
       "3      False                 7.366                                        \n",
       "4      False                 6.768                                        \n",
       "\n",
       "  CD_AnnotatedMembers_Size CD_AnnotatedMembers_Overlap  \\\n",
       "0                        0                         0.0   \n",
       "1                        0                         0.0   \n",
       "2                        0                         0.0   \n",
       "3                        0                         0.0   \n",
       "4                        0                         0.0   \n",
       "\n",
       "  CD_AnnotatedMembers_Pvalue  HiDeF_persistence represents  name  system_ID  \n",
       "0                        0.0                130       C773  C773          0  \n",
       "1                        0.0                  6       C774  C774          1  \n",
       "2                        0.0                 13       C775  C775          2  \n",
       "3                        0.0                 86       C777  C777          3  \n",
       "4                        0.0                 12       C779  C779          4  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add node attributes to dataframe for easier access\n",
    "hier_df = pd.DataFrame.from_dict(dict(G_hier.nodes(data=True)), orient='index')\n",
    "hier_df['system_ID']=hier_df.index.tolist()\n",
    "# some columns are not the right type\n",
    "hier_df['CD_MemberList_Size']=[int(x) for x in hier_df['CD_MemberList_Size'].tolist()]\n",
    "hier_df['HiDeF_persistence']=[int(x) for x in hier_df['HiDeF_persistence'].tolist()]\n",
    "hier_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14. Remove systems with no seed genes (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/9v/g6w18hn97w9gj7qy8r70y9z80000gr/T/ipykernel_5628/1660269378.py:21: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  frac_both_seeds.append(num_both_temp/np.float(len(system_genes)))\n",
      "/var/folders/9v/g6w18hn97w9gj7qy8r70y9z80000gr/T/ipykernel_5628/1660269378.py:22: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  frac_d1_seeds.append((num_d1_temp-num_both_temp)/np.float(len(system_genes)))\n",
      "/var/folders/9v/g6w18hn97w9gj7qy8r70y9z80000gr/T/ipykernel_5628/1660269378.py:23: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  frac_d2_seeds.append((num_d2_temp-num_both_temp)/np.float(len(system_genes)))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>represents</th>\n",
       "      <th>name</th>\n",
       "      <th>system_ID</th>\n",
       "      <th>num_d1_seeds</th>\n",
       "      <th>num_d2_seeds</th>\n",
       "      <th>num_both_seeds</th>\n",
       "      <th>frac_d1_seeds</th>\n",
       "      <th>frac_d2_seeds</th>\n",
       "      <th>frac_both_seeds</th>\n",
       "      <th>frac_no_seeds</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C773</th>\n",
       "      <td>PHF12 PRR23B BRPF1 HIST1H2AC ZNF236 TAF3 TMOD4...</td>\n",
       "      <td>770</td>\n",
       "      <td>False</td>\n",
       "      <td>9.589</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130</td>\n",
       "      <td>C773</td>\n",
       "      <td>C773</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>0.053247</td>\n",
       "      <td>0.029870</td>\n",
       "      <td>0.007792</td>\n",
       "      <td>0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C774</th>\n",
       "      <td>RCOR2 ZCCHC12 HDAC9 PHF2 PHF12 HIST1H4F BRPF1 ...</td>\n",
       "      <td>273</td>\n",
       "      <td>False</td>\n",
       "      <td>8.093</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6</td>\n",
       "      <td>C774</td>\n",
       "      <td>C774</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0.058608</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.014652</td>\n",
       "      <td>0.901099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C775</th>\n",
       "      <td>RO60 TRIM52 ANKRD17 DOP1A SUPT20H MACF1 INTS6 ...</td>\n",
       "      <td>241</td>\n",
       "      <td>False</td>\n",
       "      <td>7.913</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13</td>\n",
       "      <td>C775</td>\n",
       "      <td>C775</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0.037344</td>\n",
       "      <td>0.029046</td>\n",
       "      <td>0.008299</td>\n",
       "      <td>0.925311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C777</th>\n",
       "      <td>CNTN2 SLC6A1 ATP2B4 NEFM ATP1A2 TMOD4 KCNJ3 AF...</td>\n",
       "      <td>165</td>\n",
       "      <td>False</td>\n",
       "      <td>7.366</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86</td>\n",
       "      <td>C777</td>\n",
       "      <td>C777</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.890909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C779</th>\n",
       "      <td>ZCCHC12 HDAC4 HDAC9 AFDN FOXQ1 ZEB2 SALL1 NRIP...</td>\n",
       "      <td>109</td>\n",
       "      <td>False</td>\n",
       "      <td>6.768</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12</td>\n",
       "      <td>C779</td>\n",
       "      <td>C779</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.055046</td>\n",
       "      <td>0.045872</td>\n",
       "      <td>0.009174</td>\n",
       "      <td>0.889908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CD_MemberList  CD_MemberList_Size  \\\n",
       "name                                                                          \n",
       "C773  PHF12 PRR23B BRPF1 HIST1H2AC ZNF236 TAF3 TMOD4...                 770   \n",
       "C774  RCOR2 ZCCHC12 HDAC9 PHF2 PHF12 HIST1H4F BRPF1 ...                 273   \n",
       "C775  RO60 TRIM52 ANKRD17 DOP1A SUPT20H MACF1 INTS6 ...                 241   \n",
       "C777  CNTN2 SLC6A1 ATP2B4 NEFM ATP1A2 TMOD4 KCNJ3 AF...                 165   \n",
       "C779  ZCCHC12 HDAC4 HDAC9 AFDN FOXQ1 ZEB2 SALL1 NRIP...                 109   \n",
       "\n",
       "     CD_Labeled CD_MemberList_LogSize CD_CommunityName CD_AnnotatedMembers  \\\n",
       "name                                                                         \n",
       "C773      False                 9.589                                        \n",
       "C774      False                 8.093                                        \n",
       "C775      False                 7.913                                        \n",
       "C777      False                 7.366                                        \n",
       "C779      False                 6.768                                        \n",
       "\n",
       "     CD_AnnotatedMembers_Size CD_AnnotatedMembers_Overlap  \\\n",
       "name                                                        \n",
       "C773                        0                         0.0   \n",
       "C774                        0                         0.0   \n",
       "C775                        0                         0.0   \n",
       "C777                        0                         0.0   \n",
       "C779                        0                         0.0   \n",
       "\n",
       "     CD_AnnotatedMembers_Pvalue  HiDeF_persistence represents  name  \\\n",
       "name                                                                  \n",
       "C773                        0.0                130       C773  C773   \n",
       "C774                        0.0                  6       C774  C774   \n",
       "C775                        0.0                 13       C775  C775   \n",
       "C777                        0.0                 86       C777  C777   \n",
       "C779                        0.0                 12       C779  C779   \n",
       "\n",
       "      system_ID  num_d1_seeds  num_d2_seeds  num_both_seeds  frac_d1_seeds  \\\n",
       "name                                                                         \n",
       "C773          0            41            23              23       0.053247   \n",
       "C774          1            16             7               7       0.058608   \n",
       "C775          2             9             7               7       0.037344   \n",
       "C777          3            12             6               6       0.072727   \n",
       "C779          4             6             5               5       0.055046   \n",
       "\n",
       "      frac_d2_seeds  frac_both_seeds  frac_no_seeds  \n",
       "name                                                 \n",
       "C773       0.029870         0.007792       0.909091  \n",
       "C774       0.025641         0.014652       0.901099  \n",
       "C775       0.029046         0.008299       0.925311  \n",
       "C777       0.036364         0.000000       0.890909  \n",
       "C779       0.045872         0.009174       0.889908  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hier_df.index=hier_df['name']\n",
    "hier_df.head()\n",
    "\n",
    "num_d1_seeds, num_d2_seeds, num_both_seeds = [],[],[]\n",
    "frac_d1_seeds, frac_d2_seeds, frac_both_seeds=[],[],[]\n",
    "\n",
    "systems_keep = []\n",
    "for c in hier_df.index.tolist():\n",
    "    system_genes = hier_df['CD_MemberList'].loc[c].split(' ')\n",
    "    d1_temp = list(np.intersect1d(system_genes,seed_dict[d1_name]))\n",
    "    d2_temp = list(np.intersect1d(system_genes,seed_dict[d2_name]))\n",
    "    num_d1_temp = len(d1_temp)\n",
    "    num_d2_temp = len(d2_temp)\n",
    "    if (num_d1_temp+num_d2_temp)>0: # keep the system if it has at least 1 seed gene\n",
    "        systems_keep.append(c)\n",
    "        num_both_temp = len(np.intersect1d(d1_temp,d2_temp))\n",
    "        num_both_seeds.append(num_both_temp)\n",
    "        num_d1_seeds.append(num_d1_temp-num_both_temp)\n",
    "        num_d2_seeds.append(num_d2_temp-num_both_temp)\n",
    "        \n",
    "        frac_both_seeds.append(num_both_temp/np.float(len(system_genes)))\n",
    "        frac_d1_seeds.append((num_d1_temp-num_both_temp)/np.float(len(system_genes)))\n",
    "        frac_d2_seeds.append((num_d2_temp-num_both_temp)/np.float(len(system_genes)))\n",
    "\n",
    "        \n",
    "frac_no_seeds = np.subtract(1.0,np.array([frac_d1_seeds,frac_d2_seeds,frac_both_seeds]).sum(axis=0))\n",
    "\n",
    "hier_df = hier_df.loc[systems_keep]\n",
    "hier_df['num_d1_seeds']=num_d1_seeds\n",
    "hier_df['num_d2_seeds']=num_d2_seeds\n",
    "hier_df['num_both_seeds']=num_d2_seeds\n",
    "hier_df['frac_d1_seeds']=frac_d1_seeds\n",
    "hier_df['frac_d2_seeds']=frac_d2_seeds\n",
    "hier_df['frac_both_seeds']=frac_both_seeds\n",
    "hier_df['frac_no_seeds']=frac_no_seeds\n",
    "print(len(hier_df))\n",
    "\n",
    "hier_df.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42\n",
      "43\n"
     ]
    }
   ],
   "source": [
    "# prune G_hier--> only keep systems with at least one seed gene\n",
    "\n",
    "nkeep=[]\n",
    "for n in list(G_hier.nodes()):\n",
    "    if G_hier.nodes(data=True)[n]['name'] in systems_keep:\n",
    "        nkeep.append(n)\n",
    "        \n",
    "\n",
    "G_hier = nx.subgraph(G_hier, nkeep)\n",
    "print(len(G_hier.nodes()))\n",
    "print(len(G_hier.edges()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 15. Annotate systems with gprofiler. \n",
    "\n",
    "Annotate moderately sized systems (between 50 to 1000 genes per system) if they are significantly enriched for a Gene Ontology biological process. Also require that the GO term is enriched with p<1E-5 and shares at least 3 genes with the system to annotate, to increase the stringency of the annotation. Label the system using the GO term that meets these criteria, and has the highest sum of precision and recall. Systems which have no GO terms meeting these criteria are labeled with their unique system ID.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C773\n",
      "770\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>26</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.120283e-16</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.060046</td>\n",
       "      <td>query_1</td>\n",
       "      <td>433</td>\n",
       "      <td>0.366197</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>91</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>3.123725e-57</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.210162</td>\n",
       "      <td>query_1</td>\n",
       "      <td>433</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>91</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>3.123725e-57</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.210162</td>\n",
       "      <td>query_1</td>\n",
       "      <td>433</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>10622</td>\n",
       "      <td>18</td>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>REAC:R-HSA-8986944</td>\n",
       "      <td>4.904561e-10</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.041570</td>\n",
       "      <td>query_1</td>\n",
       "      <td>433</td>\n",
       "      <td>0.305085</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>10622</td>\n",
       "      <td>16</td>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>REAC:R-HSA-5578775</td>\n",
       "      <td>8.247799e-09</td>\n",
       "      <td>[REAC:R-HSA-5576891]</td>\n",
       "      <td>0.036952</td>\n",
       "      <td>query_1</td>\n",
       "      <td>433</td>\n",
       "      <td>0.296296</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            description  effective_domain_size  \\\n",
       "8       PKMTs methylate histone lysines                  10622   \n",
       "0                Chromatin organization                  10622   \n",
       "1           Chromatin modifying enzymes                  10622   \n",
       "16  Transcriptional Regulation by MECP2                  10622   \n",
       "23                      Ion homeostasis                  10622   \n",
       "\n",
       "    intersection_size                                 name  \\\n",
       "8                  26      PKMTs methylate histone lysines   \n",
       "0                  91               Chromatin organization   \n",
       "1                  91          Chromatin modifying enzymes   \n",
       "16                 18  Transcriptional Regulation by MECP2   \n",
       "23                 16                      Ion homeostasis   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "8   REAC:R-HSA-3214841  1.120283e-16  [REAC:R-HSA-3247509]   0.060046   \n",
       "0   REAC:R-HSA-4839726  3.123725e-57        [REAC:0000000]   0.210162   \n",
       "1   REAC:R-HSA-3247509  3.123725e-57  [REAC:R-HSA-4839726]   0.210162   \n",
       "16  REAC:R-HSA-8986944  4.904561e-10   [REAC:R-HSA-212436]   0.041570   \n",
       "23  REAC:R-HSA-5578775  8.247799e-09  [REAC:R-HSA-5576891]   0.036952   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "8   query_1         433  0.366197         True   REAC         71  \n",
       "0   query_1         433  0.333333         True   REAC        273  \n",
       "1   query_1         433  0.333333         True   REAC        273  \n",
       "16  query_1         433  0.305085         True   REAC         59  \n",
       "23  query_1         433  0.296296         True   REAC         54  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C774\n",
      "273\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>22</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.171292e-21</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.135802</td>\n",
       "      <td>query_1</td>\n",
       "      <td>162</td>\n",
       "      <td>0.309859</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>72</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>7.242424e-71</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>query_1</td>\n",
       "      <td>162</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>72</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>7.242424e-71</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>query_1</td>\n",
       "      <td>162</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>10622</td>\n",
       "      <td>12</td>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>REAC:R-HSA-8943724</td>\n",
       "      <td>1.500079e-09</td>\n",
       "      <td>[REAC:R-HSA-6807070]</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>query_1</td>\n",
       "      <td>162</td>\n",
       "      <td>0.203390</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Formation of the beta-catenin:TCF transactivat...</td>\n",
       "      <td>10622</td>\n",
       "      <td>13</td>\n",
       "      <td>Formation of the beta-catenin:TCF transactivat...</td>\n",
       "      <td>REAC:R-HSA-201722</td>\n",
       "      <td>2.290174e-08</td>\n",
       "      <td>[REAC:R-HSA-201681]</td>\n",
       "      <td>0.080247</td>\n",
       "      <td>query_1</td>\n",
       "      <td>162</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "5                     PKMTs methylate histone lysines                  10622   \n",
       "0                         Chromatin modifying enzymes                  10622   \n",
       "1                              Chromatin organization                  10622   \n",
       "17              Regulation of PTEN gene transcription                  10622   \n",
       "18  Formation of the beta-catenin:TCF transactivat...                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "5                  22                    PKMTs methylate histone lysines   \n",
       "0                  72                        Chromatin modifying enzymes   \n",
       "1                  72                             Chromatin organization   \n",
       "17                 12              Regulation of PTEN gene transcription   \n",
       "18                 13  Formation of the beta-catenin:TCF transactivat...   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "5   REAC:R-HSA-3214841  1.171292e-21  [REAC:R-HSA-3247509]   0.135802   \n",
       "0   REAC:R-HSA-3247509  7.242424e-71  [REAC:R-HSA-4839726]   0.444444   \n",
       "1   REAC:R-HSA-4839726  7.242424e-71        [REAC:0000000]   0.444444   \n",
       "17  REAC:R-HSA-8943724  1.500079e-09  [REAC:R-HSA-6807070]   0.074074   \n",
       "18   REAC:R-HSA-201722  2.290174e-08   [REAC:R-HSA-201681]   0.080247   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "5   query_1         162  0.309859         True   REAC         71  \n",
       "0   query_1         162  0.263736         True   REAC        273  \n",
       "1   query_1         162  0.263736         True   REAC        273  \n",
       "17  query_1         162  0.203390         True   REAC         59  \n",
       "18  query_1         162  0.142857         True   REAC         91  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C775\n",
      "241\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>4.396872e-08</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.159664</td>\n",
       "      <td>query_1</td>\n",
       "      <td>119</td>\n",
       "      <td>0.069597</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>4.396872e-08</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.159664</td>\n",
       "      <td>query_1</td>\n",
       "      <td>119</td>\n",
       "      <td>0.069597</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "0  Chromatin modifying enzymes                  10622                 19   \n",
       "1       Chromatin organization                  10622                 19   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "0  Chromatin modifying enzymes  REAC:R-HSA-3247509  4.396872e-08   \n",
       "1       Chromatin organization  REAC:R-HSA-4839726  4.396872e-08   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-4839726]   0.159664  query_1         119  0.069597   \n",
       "1        [REAC:0000000]   0.159664  query_1         119  0.069597   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC        273  \n",
       "1         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C777\n",
      "165\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>10622</td>\n",
       "      <td>16</td>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>REAC:R-HSA-5578775</td>\n",
       "      <td>1.233108e-17</td>\n",
       "      <td>[REAC:R-HSA-5576891]</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>120</td>\n",
       "      <td>0.296296</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>15</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>1.482072e-15</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>120</td>\n",
       "      <td>0.254237</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>31</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>1.457429e-31</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.258333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>120</td>\n",
       "      <td>0.244094</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>27</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>1.170684e-26</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.225000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>120</td>\n",
       "      <td>0.226891</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Ion transport by P-type ATPases</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Ion transport by P-type ATPases</td>\n",
       "      <td>REAC:R-HSA-936837</td>\n",
       "      <td>5.597134e-09</td>\n",
       "      <td>[REAC:R-HSA-983712]</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>120</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              description  effective_domain_size  \\\n",
       "7                         Ion homeostasis                  10622   \n",
       "11  NCAM signaling for neurite out-growth                  10622   \n",
       "1                      Cardiac conduction                  10622   \n",
       "4                      L1CAM interactions                  10622   \n",
       "20        Ion transport by P-type ATPases                  10622   \n",
       "\n",
       "    intersection_size                                   name  \\\n",
       "7                  16                        Ion homeostasis   \n",
       "11                 15  NCAM signaling for neurite out-growth   \n",
       "1                  31                     Cardiac conduction   \n",
       "4                  27                     L1CAM interactions   \n",
       "20                 10        Ion transport by P-type ATPases   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "7   REAC:R-HSA-5578775  1.233108e-17  [REAC:R-HSA-5576891]   0.133333   \n",
       "11   REAC:R-HSA-375165  1.482072e-15   [REAC:R-HSA-422475]   0.125000   \n",
       "1   REAC:R-HSA-5576891  1.457429e-31   [REAC:R-HSA-397014]   0.258333   \n",
       "4    REAC:R-HSA-373760  1.170684e-26   [REAC:R-HSA-422475]   0.225000   \n",
       "20   REAC:R-HSA-936837  5.597134e-09   [REAC:R-HSA-983712]   0.083333   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "7   query_1         120  0.296296         True   REAC         54  \n",
       "11  query_1         120  0.254237         True   REAC         59  \n",
       "1   query_1         120  0.244094         True   REAC        127  \n",
       "4   query_1         120  0.226891         True   REAC        119  \n",
       "20  query_1         120  0.181818         True   REAC         55  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C779\n",
      "109\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Transcriptional regulation by RUNX3</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Transcriptional regulation by RUNX3</td>\n",
       "      <td>REAC:R-HSA-8878159</td>\n",
       "      <td>4.595415e-09</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.169492</td>\n",
       "      <td>query_1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.106383</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FOXO-mediated transcription</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>FOXO-mediated transcription</td>\n",
       "      <td>REAC:R-HSA-9614085</td>\n",
       "      <td>1.315692e-06</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>query_1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.106061</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Constitutive Signaling by NOTCH1 HD+PEST Domai...</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Constitutive Signaling by NOTCH1 HD+PEST Domai...</td>\n",
       "      <td>REAC:R-HSA-2894862</td>\n",
       "      <td>8.174264e-06</td>\n",
       "      <td>[REAC:R-HSA-2894858]</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>query_1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Signaling by NOTCH1 HD+PEST Domain Mutants in ...</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Signaling by NOTCH1 HD+PEST Domain Mutants in ...</td>\n",
       "      <td>REAC:R-HSA-2894858</td>\n",
       "      <td>8.174264e-06</td>\n",
       "      <td>[REAC:R-HSA-2644603]</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>query_1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Signaling by NOTCH1 in Cancer</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Signaling by NOTCH1 in Cancer</td>\n",
       "      <td>REAC:R-HSA-2644603</td>\n",
       "      <td>8.174264e-06</td>\n",
       "      <td>[REAC:R-HSA-5663202]</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>query_1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "5                 Transcriptional regulation by RUNX3                  10622   \n",
       "17                        FOXO-mediated transcription                  10622   \n",
       "23  Constitutive Signaling by NOTCH1 HD+PEST Domai...                  10622   \n",
       "24  Signaling by NOTCH1 HD+PEST Domain Mutants in ...                  10622   \n",
       "25                      Signaling by NOTCH1 in Cancer                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "5                  10                Transcriptional regulation by RUNX3   \n",
       "17                  7                        FOXO-mediated transcription   \n",
       "23                  6  Constitutive Signaling by NOTCH1 HD+PEST Domai...   \n",
       "24                  6  Signaling by NOTCH1 HD+PEST Domain Mutants in ...   \n",
       "25                  6                      Signaling by NOTCH1 in Cancer   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "5   REAC:R-HSA-8878159  4.595415e-09   [REAC:R-HSA-212436]   0.169492   \n",
       "17  REAC:R-HSA-9614085  1.315692e-06   [REAC:R-HSA-212436]   0.118644   \n",
       "23  REAC:R-HSA-2894862  8.174264e-06  [REAC:R-HSA-2894858]   0.101695   \n",
       "24  REAC:R-HSA-2894858  8.174264e-06  [REAC:R-HSA-2644603]   0.101695   \n",
       "25  REAC:R-HSA-2644603  8.174264e-06  [REAC:R-HSA-5663202]   0.101695   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "5   query_1          59  0.106383         True   REAC         94  \n",
       "17  query_1          59  0.106061         True   REAC         66  \n",
       "23  query_1          59  0.105263         True   REAC         57  \n",
       "24  query_1          59  0.105263         True   REAC         57  \n",
       "25  query_1          59  0.105263         True   REAC         57  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C787\n",
      "16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C802\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C804\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C815\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C816\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C830\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C836\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C778\n",
      "160\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>20</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>8.971185e-23</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.194175</td>\n",
       "      <td>query_1</td>\n",
       "      <td>103</td>\n",
       "      <td>0.281690</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>66</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>1.938361e-79</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.640777</td>\n",
       "      <td>query_1</td>\n",
       "      <td>103</td>\n",
       "      <td>0.241758</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>66</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>1.938361e-79</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.640777</td>\n",
       "      <td>query_1</td>\n",
       "      <td>103</td>\n",
       "      <td>0.241758</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>10622</td>\n",
       "      <td>11</td>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>REAC:R-HSA-8943724</td>\n",
       "      <td>1.500143e-10</td>\n",
       "      <td>[REAC:R-HSA-6807070]</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>query_1</td>\n",
       "      <td>103</td>\n",
       "      <td>0.186441</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RMTs methylate histone arginines</td>\n",
       "      <td>10622</td>\n",
       "      <td>11</td>\n",
       "      <td>RMTs methylate histone arginines</td>\n",
       "      <td>REAC:R-HSA-3214858</td>\n",
       "      <td>3.302657e-09</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>query_1</td>\n",
       "      <td>103</td>\n",
       "      <td>0.139241</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              description  effective_domain_size  \\\n",
       "2         PKMTs methylate histone lysines                  10622   \n",
       "0             Chromatin modifying enzymes                  10622   \n",
       "1                  Chromatin organization                  10622   \n",
       "11  Regulation of PTEN gene transcription                  10622   \n",
       "14       RMTs methylate histone arginines                  10622   \n",
       "\n",
       "    intersection_size                                   name  \\\n",
       "2                  20        PKMTs methylate histone lysines   \n",
       "0                  66            Chromatin modifying enzymes   \n",
       "1                  66                 Chromatin organization   \n",
       "11                 11  Regulation of PTEN gene transcription   \n",
       "14                 11       RMTs methylate histone arginines   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "2   REAC:R-HSA-3214841  8.971185e-23  [REAC:R-HSA-3247509]   0.194175   \n",
       "0   REAC:R-HSA-3247509  1.938361e-79  [REAC:R-HSA-4839726]   0.640777   \n",
       "1   REAC:R-HSA-4839726  1.938361e-79        [REAC:0000000]   0.640777   \n",
       "11  REAC:R-HSA-8943724  1.500143e-10  [REAC:R-HSA-6807070]   0.106796   \n",
       "14  REAC:R-HSA-3214858  3.302657e-09  [REAC:R-HSA-3247509]   0.106796   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "2   query_1         103  0.281690         True   REAC         71  \n",
       "0   query_1         103  0.241758         True   REAC        273  \n",
       "1   query_1         103  0.241758         True   REAC        273  \n",
       "11  query_1         103  0.186441         True   REAC         59  \n",
       "14  query_1         103  0.139241         True   REAC         79  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C824\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C797\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C806\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C776\n",
      "173\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C842\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C812\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C849\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C800\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>REAC:R-HSA-422475</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>[REAC:R-HSA-9675108]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.009107</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>REAC:R-HSA-9675108</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.008711</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  description  effective_domain_size  intersection_size  \\\n",
       "1               Axon guidance                  10622                  5   \n",
       "2  Nervous system development                  10622                  5   \n",
       "\n",
       "                         name              native   p_value  \\\n",
       "1               Axon guidance   REAC:R-HSA-422475  0.000003   \n",
       "2  Nervous system development  REAC:R-HSA-9675108  0.000003   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1  [REAC:R-HSA-9675108]        1.0  query_1           5  0.009107   \n",
       "2  [REAC:R-HSA-1266738]        1.0  query_1           5  0.008711   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        549  \n",
       "2         True   REAC        574  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C807\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C808\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C780\n",
      "45\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>REAC:R-HSA-5578775</td>\n",
       "      <td>7.347196e-14</td>\n",
       "      <td>[REAC:R-HSA-5576891]</td>\n",
       "      <td>0.256410</td>\n",
       "      <td>query_1</td>\n",
       "      <td>39</td>\n",
       "      <td>0.185185</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>1.227840e-08</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.179487</td>\n",
       "      <td>query_1</td>\n",
       "      <td>39</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Post NMDA receptor activation events</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Post NMDA receptor activation events</td>\n",
       "      <td>REAC:R-HSA-438064</td>\n",
       "      <td>2.359736e-09</td>\n",
       "      <td>[REAC:R-HSA-442755]</td>\n",
       "      <td>0.205128</td>\n",
       "      <td>query_1</td>\n",
       "      <td>39</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>12</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>3.123228e-13</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>query_1</td>\n",
       "      <td>39</td>\n",
       "      <td>0.094488</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Activation of NMDA receptors and postsynaptic ...</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Activation of NMDA receptors and postsynaptic ...</td>\n",
       "      <td>REAC:R-HSA-442755</td>\n",
       "      <td>8.009680e-09</td>\n",
       "      <td>[REAC:R-HSA-112314]</td>\n",
       "      <td>0.205128</td>\n",
       "      <td>query_1</td>\n",
       "      <td>39</td>\n",
       "      <td>0.089888</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "5                                     Ion homeostasis                  10622   \n",
       "17              NCAM signaling for neurite out-growth                  10622   \n",
       "14               Post NMDA receptor activation events                  10622   \n",
       "6                                  Cardiac conduction                  10622   \n",
       "15  Activation of NMDA receptors and postsynaptic ...                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "5                  10                                    Ion homeostasis   \n",
       "17                  7              NCAM signaling for neurite out-growth   \n",
       "14                  8               Post NMDA receptor activation events   \n",
       "6                  12                                 Cardiac conduction   \n",
       "15                  8  Activation of NMDA receptors and postsynaptic ...   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "5   REAC:R-HSA-5578775  7.347196e-14  [REAC:R-HSA-5576891]   0.256410   \n",
       "17   REAC:R-HSA-375165  1.227840e-08   [REAC:R-HSA-422475]   0.179487   \n",
       "14   REAC:R-HSA-438064  2.359736e-09   [REAC:R-HSA-442755]   0.205128   \n",
       "6   REAC:R-HSA-5576891  3.123228e-13   [REAC:R-HSA-397014]   0.307692   \n",
       "15   REAC:R-HSA-442755  8.009680e-09   [REAC:R-HSA-112314]   0.205128   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "5   query_1          39  0.185185         True   REAC         54  \n",
       "17  query_1          39  0.118644         True   REAC         59  \n",
       "14  query_1          39  0.105263         True   REAC         76  \n",
       "6   query_1          39  0.094488         True   REAC        127  \n",
       "15  query_1          39  0.089888         True   REAC         89  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C783\n",
      "24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>19</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>7.263732e-36</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.95</td>\n",
       "      <td>query_1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.159664</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>2.048648e-07</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.25</td>\n",
       "      <td>query_1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>3.779888e-14</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.50</td>\n",
       "      <td>query_1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.078740</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>REAC:R-HSA-6807878</td>\n",
       "      <td>1.806428e-09</td>\n",
       "      <td>[REAC:R-HSA-199977]</td>\n",
       "      <td>0.35</td>\n",
       "      <td>query_1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.068627</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>REAC:R-HSA-397014</td>\n",
       "      <td>2.583222e-12</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.50</td>\n",
       "      <td>query_1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              description  effective_domain_size  \\\n",
       "1                      L1CAM interactions                  10622   \n",
       "11  NCAM signaling for neurite out-growth                  10622   \n",
       "6                      Cardiac conduction                  10622   \n",
       "8     COPI-mediated anterograde transport                  10622   \n",
       "7                      Muscle contraction                  10622   \n",
       "\n",
       "    intersection_size                                   name  \\\n",
       "1                  19                     L1CAM interactions   \n",
       "11                  5  NCAM signaling for neurite out-growth   \n",
       "6                  10                     Cardiac conduction   \n",
       "8                   7    COPI-mediated anterograde transport   \n",
       "7                  10                     Muscle contraction   \n",
       "\n",
       "                native       p_value              parents  precision    query  \\\n",
       "1    REAC:R-HSA-373760  7.263732e-36  [REAC:R-HSA-422475]       0.95  query_1   \n",
       "11   REAC:R-HSA-375165  2.048648e-07  [REAC:R-HSA-422475]       0.25  query_1   \n",
       "6   REAC:R-HSA-5576891  3.779888e-14  [REAC:R-HSA-397014]       0.50  query_1   \n",
       "8   REAC:R-HSA-6807878  1.806428e-09  [REAC:R-HSA-199977]       0.35  query_1   \n",
       "7    REAC:R-HSA-397014  2.583222e-12       [REAC:0000000]       0.50  query_1   \n",
       "\n",
       "    query_size    recall  significant source  term_size  \n",
       "1           20  0.159664         True   REAC        119  \n",
       "11          20  0.084746         True   REAC         59  \n",
       "6           20  0.078740         True   REAC        127  \n",
       "8           20  0.068627         True   REAC        102  \n",
       "7           20  0.051282         True   REAC        195  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C790\n",
      "13\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Potassium Channels</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Potassium Channels</td>\n",
       "      <td>REAC:R-HSA-1296071</td>\n",
       "      <td>4.265582e-13</td>\n",
       "      <td>[REAC:R-HSA-112316]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.077670</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>REAC:R-HSA-5578775</td>\n",
       "      <td>1.852849e-06</td>\n",
       "      <td>[REAC:R-HSA-5576891]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ion transport by P-type ATPases</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Ion transport by P-type ATPases</td>\n",
       "      <td>REAC:R-HSA-936837</td>\n",
       "      <td>1.852849e-06</td>\n",
       "      <td>[REAC:R-HSA-983712]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Potential therapeutics for SARS</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Potential therapeutics for SARS</td>\n",
       "      <td>REAC:R-HSA-9679191</td>\n",
       "      <td>6.704152e-06</td>\n",
       "      <td>[REAC:R-HSA-9679506]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.050633</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neuronal System</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Neuronal System</td>\n",
       "      <td>REAC:R-HSA-112316</td>\n",
       "      <td>1.636934e-08</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "1               Potassium Channels                  10622                  8   \n",
       "3                  Ion homeostasis                  10622                  4   \n",
       "4  Ion transport by P-type ATPases                  10622                  4   \n",
       "5  Potential therapeutics for SARS                  10622                  4   \n",
       "2                  Neuronal System                  10622                  8   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "1               Potassium Channels  REAC:R-HSA-1296071  4.265582e-13   \n",
       "3                  Ion homeostasis  REAC:R-HSA-5578775  1.852849e-06   \n",
       "4  Ion transport by P-type ATPases   REAC:R-HSA-936837  1.852849e-06   \n",
       "5  Potential therapeutics for SARS  REAC:R-HSA-9679191  6.704152e-06   \n",
       "2                  Neuronal System   REAC:R-HSA-112316  1.636934e-08   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1   [REAC:R-HSA-112316]   0.666667  query_1          12  0.077670   \n",
       "3  [REAC:R-HSA-5576891]   0.333333  query_1          12  0.074074   \n",
       "4   [REAC:R-HSA-983712]   0.333333  query_1          12  0.072727   \n",
       "5  [REAC:R-HSA-9679506]   0.333333  query_1          12  0.050633   \n",
       "2        [REAC:0000000]   0.666667  query_1          12  0.020000   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        103  \n",
       "3         True   REAC         54  \n",
       "4         True   REAC         55  \n",
       "5         True   REAC         79  \n",
       "2         True   REAC        400  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C809\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Formation of the beta-catenin:TCF transactivat...</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Formation of the beta-catenin:TCF transactivat...</td>\n",
       "      <td>REAC:R-HSA-201722</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>[REAC:R-HSA-201681]</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>query_1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.043956</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "1  Formation of the beta-catenin:TCF transactivat...                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "1                  4  Formation of the beta-catenin:TCF transactivat...   \n",
       "\n",
       "              native   p_value              parents  precision    query  \\\n",
       "1  REAC:R-HSA-201722  0.000003  [REAC:R-HSA-201681]   0.571429  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1           7  0.043956         True   REAC         91  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C781\n",
      "28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>13</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>5.122962e-17</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>13</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>5.122962e-17</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "1  Chromatin modifying enzymes                  10622                 13   \n",
       "2       Chromatin organization                  10622                 13   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "1  Chromatin modifying enzymes  REAC:R-HSA-3247509  5.122962e-17   \n",
       "2       Chromatin organization  REAC:R-HSA-4839726  5.122962e-17   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1  [REAC:R-HSA-4839726]   0.764706  query_1          17  0.047619   \n",
       "2        [REAC:0000000]   0.764706  query_1          17  0.047619   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C782\n",
      "26\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HATs acetylate histones</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>HATs acetylate histones</td>\n",
       "      <td>REAC:R-HSA-3214847</td>\n",
       "      <td>1.729678e-11</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.056738</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>7.709298e-13</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.036630</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>7.709298e-13</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.036630</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Regulation of TP53 Activity</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Regulation of TP53 Activity</td>\n",
       "      <td>REAC:R-HSA-5633007</td>\n",
       "      <td>9.185246e-06</td>\n",
       "      <td>[REAC:R-HSA-3700989]</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "2      HATs acetylate histones                  10622                  8   \n",
       "0  Chromatin modifying enzymes                  10622                 10   \n",
       "1       Chromatin organization                  10622                 10   \n",
       "4  Regulation of TP53 Activity                  10622                  5   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "2      HATs acetylate histones  REAC:R-HSA-3214847  1.729678e-11   \n",
       "0  Chromatin modifying enzymes  REAC:R-HSA-3247509  7.709298e-13   \n",
       "1       Chromatin organization  REAC:R-HSA-4839726  7.709298e-13   \n",
       "4  Regulation of TP53 Activity  REAC:R-HSA-5633007  9.185246e-06   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "2  [REAC:R-HSA-3247509]   0.615385  query_1          13  0.056738   \n",
       "0  [REAC:R-HSA-4839726]   0.769231  query_1          13  0.036630   \n",
       "1        [REAC:0000000]   0.769231  query_1          13  0.036630   \n",
       "4  [REAC:R-HSA-3700989]   0.384615  query_1          13  0.031250   \n",
       "\n",
       "   significant source  term_size  \n",
       "2         True   REAC        141  \n",
       "0         True   REAC        273  \n",
       "1         True   REAC        273  \n",
       "4         True   REAC        160  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C784\n",
      "24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>2.558357e-27</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.197183</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>3.503510e-19</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>3.503510e-19</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "0  PKMTs methylate histone lysines                  10622                 14   \n",
       "1      Chromatin modifying enzymes                  10622                 14   \n",
       "2           Chromatin organization                  10622                 14   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "0  PKMTs methylate histone lysines  REAC:R-HSA-3214841  2.558357e-27   \n",
       "1      Chromatin modifying enzymes  REAC:R-HSA-3247509  3.503510e-19   \n",
       "2           Chromatin organization  REAC:R-HSA-4839726  3.503510e-19   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]   0.823529  query_1          17  0.197183   \n",
       "1  [REAC:R-HSA-4839726]   0.823529  query_1          17  0.051282   \n",
       "2        [REAC:0000000]   0.823529  query_1          17  0.051282   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         71  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C811\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C786\n",
      "17\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>1.860626e-11</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Regulation of insulin secretion</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Regulation of insulin secretion</td>\n",
       "      <td>REAC:R-HSA-422356</td>\n",
       "      <td>5.636173e-07</td>\n",
       "      <td>[REAC:R-HSA-163685]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Integration of energy metabolism</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Integration of energy metabolism</td>\n",
       "      <td>REAC:R-HSA-163685</td>\n",
       "      <td>1.981410e-06</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.046729</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>REAC:R-HSA-422475</td>\n",
       "      <td>1.981410e-06</td>\n",
       "      <td>[REAC:R-HSA-9675108]</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.014572</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>REAC:R-HSA-9675108</td>\n",
       "      <td>2.107978e-06</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             description  effective_domain_size  \\\n",
       "1  NCAM signaling for neurite out-growth                  10622   \n",
       "4        Regulation of insulin secretion                  10622   \n",
       "6       Integration of energy metabolism                  10622   \n",
       "5                          Axon guidance                  10622   \n",
       "7             Nervous system development                  10622   \n",
       "\n",
       "   intersection_size                                   name  \\\n",
       "1                  7  NCAM signaling for neurite out-growth   \n",
       "4                  5        Regulation of insulin secretion   \n",
       "6                  5       Integration of energy metabolism   \n",
       "5                  8                          Axon guidance   \n",
       "7                  8             Nervous system development   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "1   REAC:R-HSA-375165  1.860626e-11   [REAC:R-HSA-422475]   0.466667  query_1   \n",
       "4   REAC:R-HSA-422356  5.636173e-07   [REAC:R-HSA-163685]   0.333333  query_1   \n",
       "6   REAC:R-HSA-163685  1.981410e-06  [REAC:R-HSA-1430728]   0.333333  query_1   \n",
       "5   REAC:R-HSA-422475  1.981410e-06  [REAC:R-HSA-9675108]   0.533333  query_1   \n",
       "7  REAC:R-HSA-9675108  2.107978e-06  [REAC:R-HSA-1266738]   0.533333  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1          15  0.118644         True   REAC         59  \n",
       "4          15  0.064103         True   REAC         78  \n",
       "6          15  0.046729         True   REAC        107  \n",
       "5          15  0.014572         True   REAC        549  \n",
       "7          15  0.013937         True   REAC        574  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C789\n",
      "13\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>1.808162e-08</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>query_1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>1.808162e-08</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>query_1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "1  Chromatin modifying enzymes                  10622                  6   \n",
       "2       Chromatin organization                  10622                  6   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "1  Chromatin modifying enzymes  REAC:R-HSA-3247509  1.808162e-08   \n",
       "2       Chromatin organization  REAC:R-HSA-4839726  1.808162e-08   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1  [REAC:R-HSA-4839726]   0.857143  query_1           7  0.021978   \n",
       "2        [REAC:0000000]   0.857143  query_1           7  0.021978   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C795\n",
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C791\n",
      "13\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HATs acetylate histones</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>HATs acetylate histones</td>\n",
       "      <td>REAC:R-HSA-3214847</td>\n",
       "      <td>5.470672e-08</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.035461</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>5.083965e-07</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.018315</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>5.083965e-07</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.018315</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "0      HATs acetylate histones                  10622                  5   \n",
       "1  Chromatin modifying enzymes                  10622                  5   \n",
       "2       Chromatin organization                  10622                  5   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "0      HATs acetylate histones  REAC:R-HSA-3214847  5.470672e-08   \n",
       "1  Chromatin modifying enzymes  REAC:R-HSA-3247509  5.083965e-07   \n",
       "2       Chromatin organization  REAC:R-HSA-4839726  5.083965e-07   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]   0.833333  query_1           6  0.035461   \n",
       "1  [REAC:R-HSA-4839726]   0.833333  query_1           6  0.018315   \n",
       "2        [REAC:0000000]   0.833333  query_1           6  0.018315   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC        141  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C785\n",
      "18\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>15</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>5.608502e-28</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.126050</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>9</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>1.827283e-13</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.5625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.070866</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>10622</td>\n",
       "      <td>9</td>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>REAC:R-HSA-397014</td>\n",
       "      <td>8.067196e-12</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.5625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.046154</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>10622</td>\n",
       "      <td>15</td>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>REAC:R-HSA-422475</td>\n",
       "      <td>5.095604e-18</td>\n",
       "      <td>[REAC:R-HSA-9675108]</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.027322</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>10622</td>\n",
       "      <td>15</td>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>REAC:R-HSA-9675108</td>\n",
       "      <td>7.999296e-18</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.026132</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  description  effective_domain_size  intersection_size  \\\n",
       "1          L1CAM interactions                  10622                 15   \n",
       "6          Cardiac conduction                  10622                  9   \n",
       "7          Muscle contraction                  10622                  9   \n",
       "3               Axon guidance                  10622                 15   \n",
       "4  Nervous system development                  10622                 15   \n",
       "\n",
       "                         name              native       p_value  \\\n",
       "1          L1CAM interactions   REAC:R-HSA-373760  5.608502e-28   \n",
       "6          Cardiac conduction  REAC:R-HSA-5576891  1.827283e-13   \n",
       "7          Muscle contraction   REAC:R-HSA-397014  8.067196e-12   \n",
       "3               Axon guidance   REAC:R-HSA-422475  5.095604e-18   \n",
       "4  Nervous system development  REAC:R-HSA-9675108  7.999296e-18   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1   [REAC:R-HSA-422475]     0.9375  query_1          16  0.126050   \n",
       "6   [REAC:R-HSA-397014]     0.5625  query_1          16  0.070866   \n",
       "7        [REAC:0000000]     0.5625  query_1          16  0.046154   \n",
       "3  [REAC:R-HSA-9675108]     0.9375  query_1          16  0.027322   \n",
       "4  [REAC:R-HSA-1266738]     0.9375  query_1          16  0.026132   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        119  \n",
       "6         True   REAC        127  \n",
       "7         True   REAC        195  \n",
       "3         True   REAC        549  \n",
       "4         True   REAC        574  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C792\n",
      "12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>9.664742e-10</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.070423</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>2.965646e-07</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.018315</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>2.965646e-07</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.018315</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "0  PKMTs methylate histone lysines                  10622                  5   \n",
       "1      Chromatin modifying enzymes                  10622                  5   \n",
       "2           Chromatin organization                  10622                  5   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "0  PKMTs methylate histone lysines  REAC:R-HSA-3214841  9.664742e-10   \n",
       "1      Chromatin modifying enzymes  REAC:R-HSA-3247509  2.965646e-07   \n",
       "2           Chromatin organization  REAC:R-HSA-4839726  2.965646e-07   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]   0.833333  query_1           6  0.070423   \n",
       "1  [REAC:R-HSA-4839726]   0.833333  query_1           6  0.018315   \n",
       "2        [REAC:0000000]   0.833333  query_1           6  0.018315   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         71  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C803\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C805\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HATs acetylate histones</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>HATs acetylate histones</td>\n",
       "      <td>REAC:R-HSA-3214847</td>\n",
       "      <td>2.976194e-07</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.028369</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>1.423520e-06</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.014652</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>1.423520e-06</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.014652</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "0      HATs acetylate histones                  10622                  4   \n",
       "1  Chromatin modifying enzymes                  10622                  4   \n",
       "2       Chromatin organization                  10622                  4   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "0      HATs acetylate histones  REAC:R-HSA-3214847  2.976194e-07   \n",
       "1  Chromatin modifying enzymes  REAC:R-HSA-3247509  1.423520e-06   \n",
       "2       Chromatin organization  REAC:R-HSA-4839726  1.423520e-06   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]        1.0  query_1           4  0.028369   \n",
       "1  [REAC:R-HSA-4839726]        1.0  query_1           4  0.014652   \n",
       "2        [REAC:0000000]        1.0  query_1           4  0.014652   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC        141  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C813\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C841\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C853\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# gprofiler annotation of clusters\n",
    "\n",
    "# hier_df = hier_df.sort_values('name')\n",
    "\n",
    "system_name_list = []\n",
    "for p in hier_df.index.tolist():\n",
    "    focal_genes=hier_df['CD_MemberList'].loc[p].split(' ')\n",
    "    print(p)\n",
    "    print(len(focal_genes))\n",
    "    if len(focal_genes)>2:\n",
    "        gp_temp = pd.DataFrame(gp.profile(focal_genes,significance_threshold_method='fdr',\n",
    "                                               sources=['REAC']))\n",
    "        if len(gp_temp)>0: # make sure data is not empty\n",
    "            \n",
    "            # make sure terms are specific, and overlap with at least 3 genes\n",
    "            gp_temp = gp_temp[(gp_temp['term_size']<1000)&(gp_temp['term_size']>50)]\n",
    "            gp_temp = gp_temp[gp_temp['intersection_size']>=3]\n",
    "            \n",
    "            gp_temp = gp_temp[gp_temp['p_value']<1E-5] # set a stringent pvalue threshold\n",
    "            \n",
    "            gp_temp = gp_temp.sort_values('recall',ascending=False)\n",
    "            \n",
    "            if len(gp_temp)>1:\n",
    "                system_name_list.append(gp_temp.head(1)['name'].tolist()[0])\n",
    "            else:\n",
    "                system_name_list.append(p)\n",
    "        else:\n",
    "            system_name_list.append(p)\n",
    "            \n",
    "\n",
    "        display(gp_temp.head())\n",
    "        \n",
    "    else:\n",
    "        system_name_list.append(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Validate identified genes and systems_\n",
    "\n",
    "### 16. Load and parse mouse knockout database\n",
    "\n",
    "Requires DDOT for ontology parsing https://github.com/michaelkyu/ddot/blob/master/examples/Tutorial.ipynb\n",
    "\n",
    "Parse the ontology, data from http://www.informatics.jax.org/vocab/mp_ontology\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "querying 2001-3000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 7001-8000...done.\n",
      "querying 8001-9000...done.\n",
      "querying 9001-10000...done.\n",
      "querying 10001-11000...done.\n",
      "querying 11001-12000...done.\n",
      "querying 12001-13000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 14001-15000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 19001-20000...done.\n",
      "querying 20001-21000...done.\n",
      "querying 21001-22000...done.\n",
      "querying 22001-23000...done.\n",
      "querying 23001-24000...done.\n",
      "querying 24001-24280...done.\n",
      "Finished.\n",
      "36 input query terms found dup hits:\n",
      "\t[('Adam1a', 2), ('Adam1b', 2), ('Adam5', 2), ('Ccdc39', 2), ('Chaserr', 2), ('Dnm3os', 2), ('Eldr', \n",
      "12264 input query terms found no hit:\n",
      "\t['+|Cav1', '+|Cm', '+|Del(10Prmt2-Pdxk)4Yey', '+|Del(11Mpo-Chad)18Brd', '+|Del(11Stat5a-Stat5b)1Mam'\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "24324\n",
      "12060\n",
      "12016\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI_Allele_Accession_ID</th>\n",
       "      <th>Allele symbol</th>\n",
       "      <th>involves</th>\n",
       "      <th>MP</th>\n",
       "      <th>PMID</th>\n",
       "      <th>MGI_marker_accession_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>human_ortholog</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000600</td>\n",
       "      <td>12529408</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001716</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001698</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001092</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000961</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           MGI_Allele_Accession_ID Allele symbol               involves  \\\n",
       "gene_name                                                                 \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "\n",
       "                   MP      PMID MGI_marker_accession_ID gene_name  \\\n",
       "gene_name                                                           \n",
       "Rb1        MP:0000600  12529408               MGI:97874       Rb1   \n",
       "Rb1        MP:0001716  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0001698  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0001092  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0000961  16449662               MGI:97874       Rb1   \n",
       "\n",
       "          human_ortholog  \n",
       "gene_name                 \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_df = validation.load_MGI_mouseKO_data()\n",
    "mgi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14470\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0000001</th>\n",
       "      <td>mammalian phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000002</th>\n",
       "      <td>obsolete Morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000003</th>\n",
       "      <td>abnormal adipose tissue morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000011</th>\n",
       "      <td>abnormal adipose tissue morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000005</th>\n",
       "      <td>increased brown adipose tissue amount</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      description\n",
       "MP                                               \n",
       "MP:0000001                    mammalian phenotype\n",
       "MP:0000002                    obsolete Morphology\n",
       "MP:0000003     abnormal adipose tissue morphology\n",
       "MP:0000011     abnormal adipose tissue morphology\n",
       "MP:0000005  increased brown adipose tissue amount"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Child</th>\n",
       "      <th>Relation</th>\n",
       "      <th>Namespace</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MP:0005375</td>\n",
       "      <td>MP:0000003</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MP:0001778</td>\n",
       "      <td>MP:0000005</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MP:0001781</td>\n",
       "      <td>MP:0000008</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MP:0005334</td>\n",
       "      <td>MP:0000010</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MP:0000003</td>\n",
       "      <td>MP:0000013</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Parent       Child Relation        Namespace\n",
       "0  MP:0005375  MP:0000003     is_a  MPheno.ontology\n",
       "1  MP:0001778  MP:0000005     is_a  MPheno.ontology\n",
       "2  MP:0001781  MP:0000008     is_a  MPheno.ontology\n",
       "3  MP:0005334  MP:0000010     is_a  MPheno.ontology\n",
       "4  MP:0000003  MP:0000013     is_a  MPheno.ontology"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "0 genes, 13337 terms, 0 gene-term relations, 16846 term-term relations\n",
       "node_attributes: ['description']\n",
       "edge_attributes: ['Relation', 'Namespace']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MPO = validation.load_MPO()\n",
    "MPO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 17. Identify phenotype(s) of interest. \n",
    "\n",
    "Recommend including a negative control, a phenotype that is not expected to overlap with the two phenotypes of interest.\n",
    "\n",
    "#### here we find MPO terms related to brain or heart\n",
    "\n",
    "Modify as needed for each specific project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "179\n",
      "['MP:0000778', 'MP:0000811', 'MP:0000937', 'MP:0000938', 'MP:0000939', 'MP:0000940', 'MP:0000958', 'MP:0000965', 'MP:0000966', 'MP:0000968']\n"
     ]
    }
   ],
   "source": [
    "# find terms related to brain\n",
    "# ---- modify this part as needed for your project -----\n",
    "MP_focal_brain_list = []\n",
    "for t in MPO.node_attr.index.tolist():\n",
    "    descr_temp = MPO.node_attr.loc[t]['description']\n",
    "    if descr_temp.find('nervous')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "    elif descr_temp.find('neuron')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "    elif descr_temp.find('synapt')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "        \n",
    "print(len(MP_focal_brain_list))\n",
    "print(MP_focal_brain_list[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "283\n",
      "['MP:0000266', 'MP:0000267', 'MP:0000269', 'MP:0000270', 'MP:0000274', 'MP:0000275', 'MP:0000276', 'MP:0000277', 'MP:0000278', 'MP:0000285']\n"
     ]
    }
   ],
   "source": [
    "# find terms related to heart\n",
    "MP_focal_heart_list = []\n",
    "for t in MPO.node_attr.index.tolist():\n",
    "    descr_temp = MPO.node_attr.loc[t]['description']\n",
    "    if descr_temp.find('cardi')>-1:\n",
    "        MP_focal_heart_list.append(t)\n",
    "    elif descr_temp.find('heart')>-1:\n",
    "        MP_focal_heart_list.append(t)\n",
    "        \n",
    "print(len(MP_focal_heart_list))\n",
    "print(MP_focal_heart_list[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 18. Compute the enrichment of selected phenotype(s) in NetColoc subnetwork as a whole."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal innate immunity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 593\n",
      "number overlapping genes = 18\n",
      "0.18898423692536037\n",
      "(-0.7929957036010307, 0.1565796241030007)\n",
      "-0.318208039749015\n",
      "\n",
      "abnormal nervous system tract morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 206\n",
      "number overlapping genes = 28\n",
      "1.166793312457235e-10\n",
      "(0.9268853651714772, 1.7372391108779637)\n",
      "1.3320622380247205\n",
      "\n",
      "hippocampal neuron degeneration\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 1\n",
      "0.9189768413841188\n",
      "(-2.1023314783704894, 1.8948754275403965)\n",
      "-0.10372802541504633\n",
      "\n",
      "abnormal motor neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 165\n",
      "number overlapping genes = 12\n",
      "0.041364430833055195\n",
      "(0.02413346010214379, 1.2082139854390608)\n",
      "0.6161737227706023\n",
      "\n",
      "motor neuron degeneration\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 4\n",
      "0.08544461374532926\n",
      "(-0.1264844171389794, 1.939123443287731)\n",
      "0.9063195130743757\n",
      "\n",
      "decreased motor neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 75\n",
      "number overlapping genes = 7\n",
      "0.0262071269040612\n",
      "(0.10491304243542132, 1.667628855843513)\n",
      "0.8862709491394671\n",
      "\n",
      "abnormal motor neuron innervation pattern\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 4\n",
      "0.08544461374532926\n",
      "(-0.1264844171389794, 1.939123443287731)\n",
      "0.9063195130743757\n",
      "\n",
      "abnormal sensory neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 470\n",
      "number overlapping genes = 41\n",
      "6.0245930068745e-07\n",
      "(0.5084475855405233, 1.166128549132203)\n",
      "0.8372880673363632\n",
      "\n",
      "decreased sensory neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 189\n",
      "number overlapping genes = 17\n",
      "0.0009012057475945441\n",
      "(0.3493741872174866, 1.356585462963848)\n",
      "0.8529798250906673\n",
      "\n",
      "abnormal sensory neuron innervation pattern\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 91\n",
      "number overlapping genes = 12\n",
      "4.021777561425395e-05\n",
      "(0.6697107090905872, 1.892869367330672)\n",
      "1.2812900382106296\n",
      "\n",
      "abnormal sympathetic neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.4686886847279028\n",
      "(-1.2910822233877126, 2.805692458195041)\n",
      "0.7573051174036642\n",
      "\n",
      "abnormal cholinergic neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 2\n",
      "0.09117092388718655\n",
      "(-0.20586139930296832, 2.7750385445448416)\n",
      "1.2845885726209367\n",
      "\n",
      "abnormal enteric nervous system morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 4\n",
      "0.02768853399344521\n",
      "(0.1292272696803709, 2.2249915001418223)\n",
      "1.1771093849110965\n",
      "\n",
      "abnormal enteric neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 4\n",
      "0.004772594146946692\n",
      "(0.4732235062552608, 2.625120961790083)\n",
      "1.549172234022672\n",
      "\n",
      "impaired synaptic plasticity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 4\n",
      "0.13541832664261494\n",
      "(-0.24466956101260018, 1.8095254127230833)\n",
      "0.7824279258552416\n",
      "\n",
      "abnormal CNS synaptic transmission\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 882\n",
      "number overlapping genes = 77\n",
      "4.6971779643013324e-12\n",
      "(0.6213915041901008, 1.1130298370422156)\n",
      "0.8672106706161582\n",
      "\n",
      "abnormal nervous system electrophysiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 330\n",
      "number overlapping genes = 44\n",
      "1.5772658499720259e-15\n",
      "(0.9996244641309734, 1.6515295852402376)\n",
      "1.3255770246856056\n",
      "\n",
      "abnormal autonomic nervous system morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 91\n",
      "number overlapping genes = 13\n",
      "5.199943035033099e-06\n",
      "(0.7838102991863312, 1.9670847958627644)\n",
      "1.3754475475245478\n",
      "\n",
      "abnormal somatic nervous system morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1008\n",
      "number overlapping genes = 82\n",
      "8.083630026343453e-11\n",
      "(0.5519579080890997, 1.0286269758100355)\n",
      "0.7902924419495676\n",
      "\n",
      "abnormal neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1512\n",
      "number overlapping genes = 142\n",
      "1.7808346280534757e-25\n",
      "(0.8224136195711838, 1.2029594768760772)\n",
      "1.0126865482236305\n",
      "\n",
      "abnormal AMPA-mediated synaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 28\n",
      "number overlapping genes = 4\n",
      "0.011582294135060888\n",
      "(0.3056605842131268, 2.42759701608229)\n",
      "1.3666288001477085\n",
      "\n",
      "abnormal NMDA-mediated synaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 23\n",
      "number overlapping genes = 3\n",
      "0.042175504372980434\n",
      "(0.04454549491452364, 2.4758255649739302)\n",
      "1.260185529944227\n",
      "\n",
      "abnormal excitatory postsynaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 137\n",
      "number overlapping genes = 16\n",
      "1.9991111066222797e-05\n",
      "(0.6191323109509653, 1.6720288915100574)\n",
      "1.1455806012305114\n",
      "\n",
      "abnormal excitatory postsynaptic potential\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 100\n",
      "number overlapping genes = 10\n",
      "0.003997381382735099\n",
      "(0.3080183248934888, 1.6227003375857016)\n",
      "0.9653593312395952\n",
      "\n",
      "abnormal PNS synaptic transmission\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 36\n",
      "number overlapping genes = 3\n",
      "0.20918907501630768\n",
      "(-0.42542072540947085, 1.9427986483096162)\n",
      "0.7586889614500727\n",
      "\n",
      "abnormal synaptic depression\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 116\n",
      "number overlapping genes = 15\n",
      "6.230562580131939e-06\n",
      "(0.714296403131482, 1.808706831008652)\n",
      "1.261501617070067\n",
      "\n",
      "increased synaptic depression\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 2\n",
      "0.09117092388718655\n",
      "(-0.20586139930296832, 2.7750385445448416)\n",
      "1.2845885726209367\n",
      "\n",
      "decreased synaptic depression\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.8633436480619967\n",
      "(-3.066067008423299, 2.571030402078807)\n",
      "-0.24751830317224588\n",
      "\n",
      "abnormal inhibitory postsynaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 80\n",
      "number overlapping genes = 6\n",
      "0.12953053770924394\n",
      "(-0.18918330508101033, 1.4810221065260702)\n",
      "0.64591940072253\n",
      "\n",
      "abnormal neuron specification\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 2\n",
      "0.07430692903898481\n",
      "(-0.13400132024790046, 2.863374760947753)\n",
      "1.3646867203499262\n",
      "\n",
      "abnormal neuron apoptosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 226\n",
      "number overlapping genes = 17\n",
      "0.010068603346204361\n",
      "(0.15640322788115102, 1.155733417451279)\n",
      "0.656068322666215\n",
      "\n",
      "increased neuron apoptosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 163\n",
      "number overlapping genes = 12\n",
      "0.0372693873758096\n",
      "(0.03712252445828379, 1.2217645788961236)\n",
      "0.6294435516772037\n",
      "\n",
      "decreased neuron apoptosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 7\n",
      "0.016102667431425258\n",
      "(0.17870028362341006, 1.7471436065911803)\n",
      "0.9629219451072952\n",
      "\n",
      "neuron degeneration\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 400\n",
      "number overlapping genes = 32\n",
      "9.52485170313906e-05\n",
      "(0.3653561753765317, 1.102667008615613)\n",
      "0.7340115919960724\n",
      "\n",
      "loss of hippocampal neurons\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 2\n",
      "0.09117092388718655\n",
      "(-0.20586139930296832, 2.7750385445448416)\n",
      "1.2845885726209367\n",
      "\n",
      "loss of cortex neurons\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 3\n",
      "0.014538170261144479\n",
      "(0.30645049307197403, 2.789839266000591)\n",
      "1.5481448795362827\n",
      "\n",
      "abnormal dopaminergic neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 48\n",
      "number overlapping genes = 4\n",
      "0.14692116683608336\n",
      "(-0.2667257429900529, 1.7854914872489742)\n",
      "0.7593828721294607\n",
      "\n",
      "loss of dopaminergic neurons\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 1\n",
      "0.754742690502002\n",
      "(-1.6964227773513678, 2.339731533642189)\n",
      "0.32165437814541065\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal GABAergic neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 1\n",
      "0.8612857732176978\n",
      "(-2.173908471404186, 1.818014537632454)\n",
      "-0.17794696688586598\n",
      "\n",
      "loss of GABAergic neurons\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.4686886847279028\n",
      "(-1.2910822233877126, 2.805692458195041)\n",
      "0.7573051174036642\n",
      "\n",
      "premature neuronal precursor differentiation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 3\n",
      "0.035084862207620124\n",
      "(0.09174340829672722, 2.5313251635248353)\n",
      "1.3115342859107813\n",
      "\n",
      "abnormal synaptic transmission\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1084\n",
      "number overlapping genes = 88\n",
      "1.7512019088295157e-11\n",
      "(0.5616854182580336, 1.0237144984462494)\n",
      "0.7926999583521415\n",
      "\n",
      "abnormal nervous system development\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1219\n",
      "number overlapping genes = 120\n",
      "9.139889766740165e-24\n",
      "(0.8423766045790744, 1.2505144626353852)\n",
      "1.0464455336072298\n",
      "\n",
      "abnormal spinal cord interneuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 2\n",
      "0.629347798334702\n",
      "(-1.0771880196727275, 1.7810298627957146)\n",
      "0.35192092156149357\n",
      "\n",
      "abnormal brain interneuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 38\n",
      "number overlapping genes = 3\n",
      "0.24565093798337312\n",
      "(-0.4815600992301081, 1.8810349972173763)\n",
      "0.6997374489936341\n",
      "\n",
      "abnormal sympathetic nervous system physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 3\n",
      "0.07898575866161935\n",
      "(-0.12475907874974035, 2.280043270888264)\n",
      "1.0776420960692619\n",
      "\n",
      "neuronal intranuclear inclusions\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 2\n",
      "0.034721976572630577\n",
      "(0.11865993758771443, 3.186410251469998)\n",
      "1.6525350945288562\n",
      "\n",
      "abnormal proprioceptive neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.6654057948463779\n",
      "(-1.5785997727551222, 2.4724566182146925)\n",
      "0.44692842272978517\n",
      "\n",
      "abnormal synaptic glutamate release\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.9013454248419377\n",
      "(-3.000261872406575, 2.643321901442657)\n",
      "-0.17846998548195891\n",
      "\n",
      "abnormal miniature excitatory postsynaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 123\n",
      "number overlapping genes = 16\n",
      "2.723223588380637e-06\n",
      "(0.7389442027248575, 1.7997015282403144)\n",
      "1.2693228654825859\n",
      "\n",
      "abnormal synaptic vesicle morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 55\n",
      "number overlapping genes = 6\n",
      "0.014653589706503148\n",
      "(0.20874156957230006, 1.9103663761109255)\n",
      "1.0595539728416128\n",
      "\n",
      "abnormal synaptic vesicle recycling\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 2\n",
      "0.19882179353281548\n",
      "(-0.5037657581367763, 2.421543609728108)\n",
      "0.9588889257956659\n",
      "\n",
      "abnormal synaptic vesicle number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 6\n",
      "0.0047605879820646095\n",
      "(0.37849153137079616, 2.098001519183564)\n",
      "1.2382465252771802\n",
      "\n",
      "abnormal neuron physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 616\n",
      "number overlapping genes = 68\n",
      "3.930977631733078e-17\n",
      "(0.8663225881902028, 1.392416318405912)\n",
      "1.1293694532980574\n",
      "\n",
      "abnormal nervous system regeneration\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 0\n",
      "0.7085007748865393\n",
      "(-3.3426457184957474, 2.2716904129671445)\n",
      "-0.5354776527643015\n",
      "\n",
      "abnormal synaptic plasticity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 7\n",
      "0.016102667431425258\n",
      "(0.17870028362341006, 1.7471436065911803)\n",
      "0.9629219451072952\n",
      "\n",
      "abnormal neuronal precursor proliferation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 95\n",
      "number overlapping genes = 17\n",
      "1.0468361353239116e-09\n",
      "(1.1193559712709358, 2.1786631482426295)\n",
      "1.6490095597567827\n",
      "\n",
      "abnormal neuronal precursor cell number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 14\n",
      "3.1602646716505507e-09\n",
      "(1.193303545293485, 2.3736122760617953)\n",
      "1.7834579106776403\n",
      "\n",
      "increased neuronal precursor cell number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 6\n",
      "5.581269551400956e-06\n",
      "(1.239566937139967, 3.1216672666232417)\n",
      "2.1806171018816043\n",
      "\n",
      "decreased neuronal precursor cell number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 7\n",
      "0.0004597098230593974\n",
      "(0.6360371622651932, 2.2516197036913397)\n",
      "1.4438284329782665\n",
      "\n",
      "increased motor neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.4165110534089509\n",
      "(-1.2042232960573682, 2.9095647582912183)\n",
      "0.852670731116925\n",
      "\n",
      "abnormal somatic nervous system physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 75\n",
      "number overlapping genes = 12\n",
      "1.9700596126340186e-06\n",
      "(0.8868995253339194, 2.130086657630213)\n",
      "1.508493091482066\n",
      "\n",
      "increased sensory neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 53\n",
      "number overlapping genes = 10\n",
      "1.3370953429020265e-06\n",
      "(1.0146647573208636, 2.3985000310583597)\n",
      "1.7065823941896117\n",
      "\n",
      "abnormal neuronal migration\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 129\n",
      "number overlapping genes = 9\n",
      "0.1014287164930537\n",
      "(-0.11192106412184066, 1.250581150705173)\n",
      "0.5693300432916661\n",
      "\n",
      "abnormal retinal neuronal layer morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 422\n",
      "number overlapping genes = 32\n",
      "0.00032456122397812653\n",
      "(0.3068465444781935, 1.0425590819888315)\n",
      "0.6747028132335124\n",
      "\n",
      "abnormal olfactory sensory neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 1\n",
      "0.9851380836945554\n",
      "(-1.984378165036651, 2.022459620171136)\n",
      "0.0190407275672424\n",
      "\n",
      "abnormal autonomic nervous system physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 4\n",
      "0.09441254388581086\n",
      "(-0.15125979215093022, 1.9118368053243513)\n",
      "0.8802885065867105\n",
      "\n",
      "abnormal parasympathetic nervous system physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 2\n",
      "0.04603562074698575\n",
      "(0.027221081354414034, 3.0670172147152694)\n",
      "1.5471191480348416\n",
      "\n",
      "increased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 0\n",
      "0.49445077887491684\n",
      "(-3.76934514766549, 1.8206594193334773)\n",
      "-0.9743428641660064\n",
      "\n",
      "decreased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 2\n",
      "0.5108091515947554\n",
      "(-0.9533608614479969, 1.916070143336135)\n",
      "0.481354640944069\n",
      "\n",
      "abnormal synaptic bouton morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 1\n",
      "0.9851380836945554\n",
      "(-1.984378165036651, 2.022459620171136)\n",
      "0.0190407275672424\n",
      "\n",
      "abnormal neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 498\n",
      "number overlapping genes = 60\n",
      "1.119500732971176e-17\n",
      "(0.943154685767809, 1.5032382116040668)\n",
      "1.223196448685938\n",
      "\n",
      "increased neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 84\n",
      "number overlapping genes = 13\n",
      "1.325331430193573e-06\n",
      "(0.8741393579024077, 2.0655924743157286)\n",
      "1.4698659161090681\n",
      "\n",
      "decreased neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 416\n",
      "number overlapping genes = 46\n",
      "5.08243035851535e-12\n",
      "(0.7952341701675512, 1.4258384991459383)\n",
      "1.1105363346567447\n",
      "\n",
      "abnormal miniature inhibitory postsynaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 53\n",
      "number overlapping genes = 8\n",
      "0.00019663070265918834\n",
      "(0.6797223857463817, 2.1907521588653074)\n",
      "1.4352372723058444\n",
      "\n",
      "abnormal synaptonemal complex\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.9815840310025946\n",
      "(-2.0252556806614357, 1.9781069591869644)\n",
      "-0.023574360737235622\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal neuron differentiation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 322\n",
      "number overlapping genes = 36\n",
      "7.237618580955353e-10\n",
      "(0.7595623433515886, 1.468332287515879)\n",
      "1.1139473154337338\n",
      "\n",
      "abnormal hippocampus neuron morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 117\n",
      "number overlapping genes = 16\n",
      "1.025588650547447e-06\n",
      "(0.7949801178967061, 1.8597509210007073)\n",
      "1.3273655194487066\n",
      "\n",
      "abnormal neuron proliferation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 124\n",
      "number overlapping genes = 18\n",
      "5.603945938169333e-08\n",
      "(0.8942549828409467, 1.904159849227655)\n",
      "1.3992074160343009\n",
      "\n",
      "abnormal sensory neuron physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 64\n",
      "number overlapping genes = 9\n",
      "0.00018160667147324575\n",
      "(0.6446336515369168, 2.0615808579734685)\n",
      "1.3531072547551926\n",
      "\n",
      "abnormal synaptic dopamine release\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.9013454248419377\n",
      "(-3.000261872406575, 2.643321901442657)\n",
      "-0.17846998548195891\n",
      "\n",
      "increased nervous system tumor incidence\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 66\n",
      "number overlapping genes = 7\n",
      "0.01039269595885618\n",
      "(0.24188867596710673, 1.8155944954364047)\n",
      "1.0287415857017557\n",
      "\n",
      "reduced AMPA-mediated synaptic currents\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.5198727619873533\n",
      "(-1.3710330526001422, 2.711509659464158)\n",
      "0.6702383034320079\n",
      "\n",
      "abnormal excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 3\n",
      "0.19184783106560221\n",
      "(-0.3961286327466911, 1.9751608761717168)\n",
      "0.7895161217125128\n",
      "\n",
      "increased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 2\n",
      "0.05924079198808309\n",
      "(-0.05661240005508428, 2.960119468698246)\n",
      "1.4517535343215808\n",
      "\n",
      "decreased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 23\n",
      "number overlapping genes = 2\n",
      "0.2775253488348519\n",
      "(-0.6476233706369432, 2.2567670930990436)\n",
      "0.8045718612310502\n",
      "\n",
      "decreased dopaminergic neuron number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 1\n",
      "0.9496398420701603\n",
      "(-2.0645332523105107, 1.9356295878691223)\n",
      "-0.06445183222069417\n",
      "\n",
      "increased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.9423205387074528\n",
      "(-2.9298359644723018, 2.72122282407428)\n",
      "-0.1043065701990109\n",
      "\n",
      "decreased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.9653376112196761\n",
      "(-2.7721624251913255, 2.897879208194592)\n",
      "0.06285839150163319\n",
      "\n",
      "ectopic neuron\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 65\n",
      "number overlapping genes = 6\n",
      "0.04231127345195354\n",
      "(0.03031364154827665, 1.7162486544365927)\n",
      "0.8732811479924347\n",
      "\n",
      "impaired neuron differentiation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 3\n",
      "0.006146523540344024\n",
      "(0.504241586526222, 3.0386679339507925)\n",
      "1.7714547602385071\n",
      "\n",
      "abnormal neuronal stem cell morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.4165110534089509\n",
      "(-1.2042232960573682, 2.9095647582912183)\n",
      "0.852670731116925\n",
      "\n",
      "abnormal neuronal stem cell physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.4165110534089509\n",
      "(-1.2042232960573682, 2.9095647582912183)\n",
      "0.852670731116925\n",
      "\n",
      "nervous system inclusion bodies\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 62\n",
      "number overlapping genes = 3\n",
      "0.7663432428410305\n",
      "(-0.9860318718789389, 1.3384617823424452)\n",
      "0.17621495523175312\n",
      "\n",
      "abnormal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.6654057948463779\n",
      "(-1.5785997727551222, 2.4724566182146925)\n",
      "0.44692842272978517\n",
      "\n",
      "abnormal hippocampal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.9866227844466524\n",
      "(-2.8540920034898085, 2.8056751585497657)\n",
      "-0.024208422470021418\n",
      "\n",
      "increased neuronal precursor proliferation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.4165110534089509\n",
      "(-1.2042232960573682, 2.9095647582912183)\n",
      "0.852670731116925\n",
      "\n",
      "decreased neuronal precursor proliferation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 17\n",
      "number overlapping genes = 2\n",
      "0.12991996784343027\n",
      "(-0.3357966656582818, 2.618550344953685)\n",
      "1.1413768396477018\n",
      "\n",
      "abnormal postsynaptic density morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 1\n",
      "0.797007135690424\n",
      "(-1.7505239152141234, 2.279404932969622)\n",
      "0.2644405088777493\n",
      "\n",
      "abnormal synaptic physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1089\n",
      "number overlapping genes = 88\n",
      "2.3631243909130323e-11\n",
      "(0.5564262248201381, 1.0183721897493725)\n",
      "0.7873992072847553\n",
      "\n",
      "impaired synaptic physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 4\n",
      "0.13541832664261494\n",
      "(-0.24466956101260018, 1.8095254127230833)\n",
      "0.7824279258552416\n"
     ]
    }
   ],
   "source": [
    "# add a negative control phenotype: abnormal innate immunity: MP:0008171\n",
    "# negative controls are tough here because we're dealing with development... which impacts almost everything\n",
    "MP_focal_list = ['MP:0002419']+MP_focal_brain_list\n",
    "root_KO_brain_df=validation.MPO_enrichment_root(hier_df,MPO,mgi_df,MP_focal_list,G_int,verbose=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0002419</th>\n",
       "      <td>1.889842e-01</td>\n",
       "      <td>-0.318208</td>\n",
       "      <td>-0.792996</td>\n",
       "      <td>0.156580</td>\n",
       "      <td>593</td>\n",
       "      <td>abnormal innate immunity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000778</th>\n",
       "      <td>1.166793e-10</td>\n",
       "      <td>1.332062</td>\n",
       "      <td>0.926885</td>\n",
       "      <td>1.737239</td>\n",
       "      <td>206</td>\n",
       "      <td>abnormal nervous system tract morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000811</th>\n",
       "      <td>9.189768e-01</td>\n",
       "      <td>-0.103728</td>\n",
       "      <td>-2.102331</td>\n",
       "      <td>1.894875</td>\n",
       "      <td>27</td>\n",
       "      <td>hippocampal neuron degeneration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000937</th>\n",
       "      <td>4.136443e-02</td>\n",
       "      <td>0.616174</td>\n",
       "      <td>0.024133</td>\n",
       "      <td>1.208214</td>\n",
       "      <td>165</td>\n",
       "      <td>abnormal motor neuron morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000938</th>\n",
       "      <td>8.544461e-02</td>\n",
       "      <td>0.906320</td>\n",
       "      <td>-0.126484</td>\n",
       "      <td>1.939123</td>\n",
       "      <td>42</td>\n",
       "      <td>motor neuron degeneration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0002419  1.889842e-01 -0.318208        -0.792996         0.156580   \n",
       "MP:0000778  1.166793e-10  1.332062         0.926885         1.737239   \n",
       "MP:0000811  9.189768e-01 -0.103728        -2.102331         1.894875   \n",
       "MP:0000937  4.136443e-02  0.616174         0.024133         1.208214   \n",
       "MP:0000938  8.544461e-02  0.906320        -0.126484         1.939123   \n",
       "\n",
       "            num_genes_in_term                            MP_description  \n",
       "MP:0002419                593                  abnormal innate immunity  \n",
       "MP:0000778                206  abnormal nervous system tract morphology  \n",
       "MP:0000811                 27           hippocampal neuron degeneration  \n",
       "MP:0000937                165          abnormal motor neuron morphology  \n",
       "MP:0000938                 42                 motor neuron degeneration  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_brain_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal heart morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1803\n",
      "number overlapping genes = 152\n",
      "2.3646592611862763e-21\n",
      "(0.7087056357138845, 1.077756875909922)\n",
      "0.8932312558119033\n",
      "\n",
      "abnormal heart development\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 364\n",
      "number overlapping genes = 52\n",
      "5.322907151036442e-20\n",
      "(1.1123272942687739, 1.7181436490514623)\n",
      "1.415235471660118\n",
      "\n",
      "abnormal heart looping\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 130\n",
      "number overlapping genes = 14\n",
      "0.000227621602923038\n",
      "(0.49259285315341916, 1.6111659345292884)\n",
      "1.0518793938413538\n",
      "\n",
      "abnormal heart tube morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 81\n",
      "number overlapping genes = 12\n",
      "6.809631925587996e-06\n",
      "(0.7998738149572537, 2.0345015518027605)\n",
      "1.417187683380007\n",
      "\n",
      "enlarged heart\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 785\n",
      "number overlapping genes = 65\n",
      "3.619679924079847e-09\n",
      "(0.5323586417835797, 1.0619066337822554)\n",
      "0.7971326377829175\n",
      "\n",
      "heart right ventricle hypertrophy\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 1\n",
      "0.5802887302068613\n",
      "(-2.545070542400306, 1.424998293820308)\n",
      "-0.5600361242899989\n",
      "\n",
      "abnormal heart shape\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 39\n",
      "number overlapping genes = 4\n",
      "0.06165919863088523\n",
      "(-0.04826613977386485, 2.0257144459180227)\n",
      "0.988724153072079\n",
      "\n",
      "abnormal myocardial fiber morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 202\n",
      "number overlapping genes = 17\n",
      "0.0023445740712928565\n",
      "(0.2773458702256568, 1.2814362530453587)\n",
      "0.7793910616355078\n",
      "\n",
      "abnormal heart valve morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 176\n",
      "number overlapping genes = 19\n",
      "1.6382533709127552e-05\n",
      "(0.5773096148718158, 1.5405951370199058)\n",
      "1.0589523759458608\n",
      "\n",
      "abnormal pericardium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 223\n",
      "number overlapping genes = 31\n",
      "5.029404499437299e-12\n",
      "(0.9749568346785731, 1.7478447204940748)\n",
      "1.361400777586324\n",
      "\n",
      "enlarged pericardium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 7\n",
      "0.0002986447643499877\n",
      "(0.6855030714985748, 2.3076633356018075)\n",
      "1.4965832035501911\n",
      "\n",
      "distended pericardium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 11\n",
      "2.7484429900715973e-07\n",
      "(1.0737795438329356, 2.3974231228200034)\n",
      "1.7356013333264695\n",
      "\n",
      "absent myocardial trabeculae\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 4\n",
      "0.0021144137933004053\n",
      "(0.6202711456432393, 2.80344393241747)\n",
      "1.7118575390303548\n",
      "\n",
      "abnormal cardiac stroke volume\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 48\n",
      "number overlapping genes = 2\n",
      "0.9789698648846425\n",
      "(-1.3984870597927908, 1.4366176304647935)\n",
      "0.01906528533600138\n",
      "\n",
      "abnormal response to cardiac infarction\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 82\n",
      "number overlapping genes = 6\n",
      "0.1457993574892621\n",
      "(-0.2151445108373935, 1.4534242868244356)\n",
      "0.619139887993521\n",
      "\n",
      "dextrocardia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 40\n",
      "number overlapping genes = 1\n",
      "0.6148462380042299\n",
      "(-2.4961589605049497, 1.4763296519084828)\n",
      "-0.5099146542982336\n",
      "\n",
      "mesocardia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 1\n",
      "0.5698206576402062\n",
      "(-1.445090549827524, 2.6253708612335607)\n",
      "0.5901401557030184\n",
      "\n",
      "abnormal cardiovascular system physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1854\n",
      "number overlapping genes = 143\n",
      "5.47971421162651e-16\n",
      "(0.5900392693262477, 0.9667106962049876)\n",
      "0.7783749827656177\n",
      "\n",
      "cardiac hypertrophy\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 198\n",
      "number overlapping genes = 18\n",
      "0.0005355483743151943\n",
      "(0.37557851143072835, 1.3555358709884469)\n",
      "0.8655571912095876\n",
      "\n",
      "abnormal cardiac output\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 60\n",
      "number overlapping genes = 2\n",
      "0.766973291630497\n",
      "(-1.6248480268907017, 1.1980418879007133)\n",
      "-0.2134030694949942\n",
      "\n",
      "abnormal heart rate\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 431\n",
      "number overlapping genes = 41\n",
      "2.8770240882927235e-08\n",
      "(0.6046106052010929, 1.2650075342587206)\n",
      "0.9348090697299067\n",
      "\n",
      "irregular heartbeat\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 68\n",
      "number overlapping genes = 12\n",
      "3.789665109352423e-07\n",
      "(0.998970693373189, 2.2543597494018988)\n",
      "1.6266652213875439\n",
      "\n",
      "pericardial edema\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 83\n",
      "number overlapping genes = 11\n",
      "7.909147453073436e-05\n",
      "(0.6474985207927064, 1.9247643662473513)\n",
      "1.2861314435200288\n",
      "\n",
      "heart inflammation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 3\n",
      "0.345365388586066\n",
      "(-0.6095477984379047, 1.7414047411517433)\n",
      "0.5659284713569193\n",
      "\n",
      "myocarditis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 1\n",
      "0.5698206576402062\n",
      "(-1.445090549827524, 2.6253708612335607)\n",
      "0.5901401557030184\n",
      "\n",
      "small heart\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 365\n",
      "number overlapping genes = 36\n",
      "6.278274025848929e-08\n",
      "(0.6195576508041176, 1.3233579792990193)\n",
      "0.9714578150515685\n",
      "\n",
      "abnormal myocardial trabeculae morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 146\n",
      "number overlapping genes = 30\n",
      "1.2054213668804359e-18\n",
      "(1.4272807398567697, 2.24354053188237)\n",
      "1.8354106358695699\n",
      "\n",
      "disorganized myocardium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 8\n",
      "2.3870552341918836e-05\n",
      "(0.8893715767551296, 2.4283895430064053)\n",
      "1.6588805598807674\n",
      "\n",
      "heart left ventricle hypertrophy\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 66\n",
      "number overlapping genes = 7\n",
      "0.01039269595885618\n",
      "(0.24188867596710673, 1.8155944954364047)\n",
      "1.0287415857017557\n",
      "\n",
      "increased heart rate\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 166\n",
      "number overlapping genes = 13\n",
      "0.016627462348492907\n",
      "(0.12666562163867368, 1.2684075036594484)\n",
      "0.697536562649061\n",
      "\n",
      "thin myocardium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 93\n",
      "number overlapping genes = 14\n",
      "8.692507477071168e-07\n",
      "(0.8650984080907332, 2.0110670469107186)\n",
      "1.438082727500726\n",
      "\n",
      "heart hypoplasia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 102\n",
      "number overlapping genes = 18\n",
      "4.849644823787423e-10\n",
      "(1.1187903055974169, 2.147319688186909)\n",
      "1.633054996892163\n",
      "\n",
      "dilated heart left ventricle\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 107\n",
      "number overlapping genes = 4\n",
      "0.853362910184875\n",
      "(-1.095882218626697, 0.9070051229845773)\n",
      "-0.09443854782105987\n",
      "\n",
      "dilated heart right ventricle\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 52\n",
      "number overlapping genes = 1\n",
      "0.44082103834278596\n",
      "(-2.759246688797551, 1.201556443896569)\n",
      "-0.7788451224504911\n",
      "\n",
      "dilated cardiomyopathy\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 83\n",
      "number overlapping genes = 5\n",
      "0.37633613294823665\n",
      "(-0.4976467606547653, 1.3164938048708112)\n",
      "0.4094235221080229\n",
      "\n",
      "increased heart weight\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 261\n",
      "number overlapping genes = 26\n",
      "3.5100759438255913e-06\n",
      "(0.5625939510193204, 1.385938183149015)\n",
      "0.9742660670841676\n",
      "\n",
      "decreased heart weight\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 168\n",
      "number overlapping genes = 8\n",
      "0.6597690849238667\n",
      "(-0.5534375503350264, 0.87407424485928)\n",
      "0.1603183472621268\n",
      "\n",
      "dystrophic cardiac calcinosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.8633436480619967\n",
      "(-3.066067008423299, 2.571030402078807)\n",
      "-0.24751830317224588\n",
      "\n",
      "abnormal cardiovascular development\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 762\n",
      "number overlapping genes = 87\n",
      "6.056466191551197e-23\n",
      "(0.9514983426567536, 1.4234781346317593)\n",
      "1.1874882386442565\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal cardiac muscle contractility\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 288\n",
      "number overlapping genes = 26\n",
      "3.564460670607235e-05\n",
      "(0.4543689627807999, 1.2736117393349617)\n",
      "0.8639903510578808\n",
      "\n",
      "increased myocardial infarct size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 3\n",
      "0.10248674821888423\n",
      "(-0.19978189766462506, 2.194758760447531)\n",
      "0.9974884313914529\n",
      "\n",
      "decreased myocardial infarct size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.9815840310025946\n",
      "(-2.0252556806614357, 1.9781069591869644)\n",
      "-0.023574360737235622\n",
      "\n",
      "abnormal epicardium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 17\n",
      "number overlapping genes = 4\n",
      "0.0005475775877472561\n",
      "(0.8573609153793456, 3.1033257177965794)\n",
      "1.9803433165879625\n",
      "\n",
      "abnormal heart atrium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 234\n",
      "number overlapping genes = 27\n",
      "4.04228711466471e-08\n",
      "(0.7340883028147984, 1.5494888600904617)\n",
      "1.14178858145263\n",
      "\n",
      "dilated heart atrium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 58\n",
      "number overlapping genes = 6\n",
      "0.020844714891266103\n",
      "(0.15181361701194251, 1.8481141451360585)\n",
      "0.9999638810740006\n",
      "\n",
      "cardiac fibrosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 169\n",
      "number overlapping genes = 13\n",
      "0.019841797167911467\n",
      "(0.10750291698546066, 1.2483987567787262)\n",
      "0.6779508368820935\n",
      "\n",
      "abnormal cardiomyocyte apoptosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 61\n",
      "number overlapping genes = 7\n",
      "0.005600531151053308\n",
      "(0.3268984041341355, 1.9082473182011541)\n",
      "1.1175728611676448\n",
      "\n",
      "increased cardiomyocyte apoptosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 53\n",
      "number overlapping genes = 6\n",
      "0.011365427267018726\n",
      "(0.24856712507072087, 1.9541084109474771)\n",
      "1.101337768009099\n",
      "\n",
      "decreased cardiac output\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 2\n",
      "0.8529352270241201\n",
      "(-1.2867276495157638, 1.555552234771599)\n",
      "0.13441229262791765\n",
      "\n",
      "increased cardiac output\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.9866227844466524\n",
      "(-2.8540920034898085, 2.8056751585497657)\n",
      "-0.024208422470021418\n",
      "\n",
      "abnormal fetal cardiomyocyte proliferation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 54\n",
      "number overlapping genes = 13\n",
      "2.871073740103841e-10\n",
      "(1.3925654442326203, 2.648716454246848)\n",
      "2.020640949239734\n",
      "\n",
      "decreased heart right ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 2\n",
      "0.3053362294731352\n",
      "(-0.6913331765663779, 2.207325932295369)\n",
      "0.7579963778644956\n",
      "\n",
      "abnormal heart right ventricle morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 178\n",
      "number overlapping genes = 13\n",
      "0.032411152130711614\n",
      "(0.05208877628865627, 1.1906277873445954)\n",
      "0.6213582818166259\n",
      "\n",
      "abnormal heart left ventricle morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 334\n",
      "number overlapping genes = 28\n",
      "9.755548841692963e-05\n",
      "(0.3891916660360123, 1.1769540832551684)\n",
      "0.7830728746455904\n",
      "\n",
      "abnormal heart right atrium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 41\n",
      "number overlapping genes = 3\n",
      "0.30420862318349906\n",
      "(-0.5602914894315691, 1.794957107423431)\n",
      "0.6173328089959309\n",
      "\n",
      "abnormal heart left atrium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 1\n",
      "0.754742690502002\n",
      "(-1.6964227773513678, 2.339731533642189)\n",
      "0.32165437814541065\n",
      "\n",
      "increased heart rate variability\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 4\n",
      "0.0011232866363722824\n",
      "(0.7318307560595021, 2.9424324111699547)\n",
      "1.8371315836147284\n",
      "\n",
      "decreased heart rate variability\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 50\n",
      "number overlapping genes = 4\n",
      "0.17136173155533907\n",
      "(-0.30943726542151406, 1.7390773240275719)\n",
      "0.7148200293030289\n",
      "\n",
      "abnormal endocardium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 28\n",
      "number overlapping genes = 3\n",
      "0.09031776033585186\n",
      "(-0.16297124870511626, 2.236500497876723)\n",
      "1.0367646245858033\n",
      "\n",
      "abnormal cardiac cell glucose uptake\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.6654057948463779\n",
      "(-1.5785997727551222, 2.4724566182146925)\n",
      "0.44692842272978517\n",
      "\n",
      "abnormal myocardium compact layer morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 73\n",
      "number overlapping genes = 17\n",
      "1.3189222553704169e-12\n",
      "(1.4339764603151557, 2.5292036866813192)\n",
      "1.9815900734982375\n",
      "\n",
      "thin myocardium compact layer\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 14\n",
      "3.656298216752572e-13\n",
      "(1.7121645457269041, 2.9766486299678645)\n",
      "2.3444065878473843\n",
      "\n",
      "abnormal cardiac muscle relaxation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 30\n",
      "number overlapping genes = 2\n",
      "0.4809453628210191\n",
      "(-0.9198618123750102, 1.9528647124391114)\n",
      "0.5165014500320506\n",
      "\n",
      "abnormal heartbeat\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 480\n",
      "number overlapping genes = 43\n",
      "1.2281944992924212e-07\n",
      "(0.5468186106071233, 1.1906473649863745)\n",
      "0.8687329877967489\n",
      "\n",
      "absent heartbeat\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 23\n",
      "number overlapping genes = 5\n",
      "0.0002095284906672438\n",
      "(0.8856446687669821, 2.8725424452149344)\n",
      "1.8790935569909584\n",
      "\n",
      "cardia bifida\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 3\n",
      "0.018573845980655108\n",
      "(0.24831858977024668, 2.7187832284716005)\n",
      "1.4835509091209236\n",
      "\n",
      "abnormal myocardial fiber physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 136\n",
      "number overlapping genes = 14\n",
      "0.0004356770419378219\n",
      "(0.44328523242584306, 1.55894261594062)\n",
      "1.0011139241832314\n",
      "\n",
      "failure of heart looping\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 7\n",
      "0.0004597098230593974\n",
      "(0.6360371622651932, 2.2516197036913397)\n",
      "1.4438284329782665\n",
      "\n",
      "abnormal direction of heart looping\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 40\n",
      "number overlapping genes = 2\n",
      "0.7719157747125434\n",
      "(-1.2131752365284585, 1.6343047747431585)\n",
      "0.21056476910735\n",
      "\n",
      "abnormal response of heart to induced stress\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 97\n",
      "number overlapping genes = 8\n",
      "0.04296636315380799\n",
      "(0.023767824859932674, 1.4778654468490888)\n",
      "0.7508166358545107\n",
      "\n",
      "increased response of heart to induced stress\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 51\n",
      "number overlapping genes = 3\n",
      "0.520477220408005\n",
      "(-0.7854855226856448, 1.5518107591056325)\n",
      "0.38316261820999387\n",
      "\n",
      "decreased response of heart to induced stress\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 5\n",
      "0.0020335177237590703\n",
      "(0.5510943994510294, 2.4707556459452977)\n",
      "1.5109250226981636\n",
      "\n",
      "enlarged myocardial fiber\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 7\n",
      "0.007245759233250766\n",
      "(0.2920278716033624, 1.8701602784221802)\n",
      "1.0810940750127713\n",
      "\n",
      "small myocardial fiber\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.9013454248419377\n",
      "(-3.000261872406575, 2.643321901442657)\n",
      "-0.17846998548195891\n",
      "\n",
      "myocardial fiber degeneration\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 1\n",
      "0.759041519112583\n",
      "(-2.303361201080962, 1.6799605308929926)\n",
      "-0.31170033509398465\n",
      "\n",
      "decreased myocardial fiber number\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 2\n",
      "0.1745890514713454\n",
      "(-0.4508370467882683, 2.4830426369149263)\n",
      "1.016102795063329\n",
      "\n",
      "abnormal cardinal vein morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 32\n",
      "number overlapping genes = 8\n",
      "4.714163261496223e-07\n",
      "(1.261582986929956, 2.8684402025932405)\n",
      "2.0650115947615983\n",
      "\n",
      "abnormal anterior cardinal vein morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 4\n",
      "9.60149892420955e-05\n",
      "(1.1682807912011117, 3.5282988857906536)\n",
      "2.348289838495883\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal heart weight\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 423\n",
      "number overlapping genes = 34\n",
      "5.059217685387154e-05\n",
      "(0.3824805948739928, 1.0988497250564286)\n",
      "0.7406651599652108\n",
      "\n",
      "dilated heart\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 229\n",
      "number overlapping genes = 15\n",
      "0.0614560884656363\n",
      "(-0.024206433487610313, 1.0328676386450124)\n",
      "0.5043306025787011\n",
      "\n",
      "decreased cardiac muscle contractility\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 226\n",
      "number overlapping genes = 21\n",
      "0.00012123220700204534\n",
      "(0.4372691601320303, 1.3472350618520843)\n",
      "0.8922521109920574\n",
      "\n",
      "hemopericardium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 4\n",
      "0.13541832664261494\n",
      "(-0.24466956101260018, 1.8095254127230833)\n",
      "0.7824279258552416\n",
      "\n",
      "abnormal heart ventricle morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 711\n",
      "number overlapping genes = 68\n",
      "4.570182316391412e-13\n",
      "(0.7029841777538363, 1.2251303823769364)\n",
      "0.9640572800653864\n",
      "\n",
      "pericardial effusion\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 114\n",
      "number overlapping genes = 17\n",
      "7.249674895151047e-08\n",
      "(0.9094822181320055, 1.9504170791155317)\n",
      "1.4299496486237686\n",
      "\n",
      "abnormal myocardium layer morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 420\n",
      "number overlapping genes = 52\n",
      "3.5708839073811525e-16\n",
      "(0.9471930152471373, 1.5467943114129064)\n",
      "1.2469936633300218\n",
      "\n",
      "cardiomyopathy\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 127\n",
      "number overlapping genes = 9\n",
      "0.09186199674769863\n",
      "(-0.09540049570010645, 1.2678978962685026)\n",
      "0.5862487002841981\n",
      "\n",
      "decreased heart rate\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 189\n",
      "number overlapping genes = 28\n",
      "5.7329936141955674e-12\n",
      "(1.0254131884690556, 1.8413711711600949)\n",
      "1.4333921798145752\n",
      "\n",
      "abnormal heart size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 1088\n",
      "number overlapping genes = 93\n",
      "1.1056803042949771e-13\n",
      "(0.6304057531737921, 1.0823626026964785)\n",
      "0.8563841779351353\n",
      "\n",
      "increased cardiac muscle contractility\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 8\n",
      "0.0004260699804799821\n",
      "(0.5989683665890039, 2.1007461909575316)\n",
      "1.3498572787732677\n",
      "\n",
      "cardiac interstitial fibrosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 59\n",
      "number overlapping genes = 4\n",
      "0.3022170200117279\n",
      "(-0.48193643663051233, 1.5531928899504308)\n",
      "0.5356282266599592\n",
      "\n",
      "abnormal heart position or orientation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 60\n",
      "number overlapping genes = 2\n",
      "0.766973291630497\n",
      "(-1.6248480268907017, 1.1980418879007133)\n",
      "-0.2134030694949942\n",
      "\n",
      "myocardium necrosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 23\n",
      "number overlapping genes = 0\n",
      "0.6366986181605931\n",
      "(-3.478108193076542, 2.1272961964069683)\n",
      "-0.6754059983347869\n",
      "\n",
      "abnormal heart septum morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 375\n",
      "number overlapping genes = 40\n",
      "5.524814033904181e-10\n",
      "(0.7277425136013178, 1.4000002972906949)\n",
      "1.0638714054460063\n",
      "\n",
      "abnormal cardiac outflow tract development\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 147\n",
      "number overlapping genes = 15\n",
      "0.00030897124797637365\n",
      "(0.45310956983594974, 1.5310736350127268)\n",
      "0.9920916024243382\n",
      "\n",
      "congestive heart failure\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 83\n",
      "number overlapping genes = 9\n",
      "0.002948166627077781\n",
      "(0.35961643766373763, 1.751021442607319)\n",
      "1.0553189401355283\n",
      "\n",
      "dilated heart ventricle\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 149\n",
      "number overlapping genes = 6\n",
      "0.968147236452596\n",
      "(-0.8366749333758992, 0.8032629866526722)\n",
      "-0.016705973361613502\n",
      "\n",
      "enlarged heart atrium\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 3\n",
      "0.3246011713800361\n",
      "(-0.5852214325411866, 1.767825037557718)\n",
      "0.5913018025082657\n",
      "\n",
      "increased heart ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 190\n",
      "number overlapping genes = 13\n",
      "0.05744432272821188\n",
      "(-0.017398747508121093, 1.118364436789399)\n",
      "0.5504828446406389\n",
      "\n",
      "abnormal heart ventricle pressure\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 62\n",
      "number overlapping genes = 2\n",
      "0.7310285781406287\n",
      "(-1.658078453614856, 1.1632468778843585)\n",
      "-0.24741578786524876\n",
      "\n",
      "abnormal fetal cardiomyocyte morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 34\n",
      "number overlapping genes = 7\n",
      "2.0759037523003292e-05\n",
      "(0.9777709878153221, 2.6466675137033424)\n",
      "1.8122192507593322\n",
      "\n",
      "abnormal cardiac epithelial to mesenchymal transition\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 5\n",
      "0.0005470364879184568\n",
      "(0.7467923613975828, 2.702760623455101)\n",
      "1.7247764924263418\n",
      "\n",
      "ventricular tachycardia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 5\n",
      "0.00014701840580985638\n",
      "(0.9365602238641685, 2.936054628653071)\n",
      "1.9363074262586197\n",
      "\n",
      "delayed heart looping\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 3\n",
      "0.0013223091669410333\n",
      "(0.8482736718907788, 3.5060095237844258)\n",
      "2.1771415978376023\n",
      "\n",
      "cardiac valve regurgitation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 2\n",
      "0.33373540366373877\n",
      "(-0.7332189647462175, 2.160197253716539)\n",
      "0.7134891444851608\n",
      "\n",
      "heart hemorrhage\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 6\n",
      "0.018601702876747044\n",
      "(0.1704337742072376, 1.8684412800261385)\n",
      "1.019437527116688\n",
      "\n",
      "abnormal heart ventricle outflow tract morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 96\n",
      "number overlapping genes = 10\n",
      "0.0026324657039468896\n",
      "(0.3521950969666727, 1.6698936983689079)\n",
      "1.0110443976677903\n",
      "\n",
      "heart right ventricle hypoplasia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.4686886847279028\n",
      "(-1.2910822233877126, 2.805692458195041)\n",
      "0.7573051174036642\n",
      "\n",
      "abnormal heart and great vessel attachment\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 211\n",
      "number overlapping genes = 20\n",
      "0.00012328230426930789\n",
      "(0.4472556028045922, 1.380041398792418)\n",
      "0.9136485007985051\n",
      "\n",
      "abnormal heart and great artery attachment\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 209\n",
      "number overlapping genes = 20\n",
      "0.00010346728159337719\n",
      "(0.45766591443692517, 1.390907877300028)\n",
      "0.9242868958684767\n",
      "\n",
      "abnormal heart right ventricle outflow tract morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 58\n",
      "number overlapping genes = 7\n",
      "0.003699628641152253\n",
      "(0.3815817245463, 1.9682142051057179)\n",
      "1.174897964826009\n",
      "\n",
      "abnormal heart left ventricle outflow tract morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 73\n",
      "number overlapping genes = 8\n",
      "0.004619950340778111\n",
      "(0.32847384277091174, 1.804328301683643)\n",
      "1.0664010722272774\n",
      "\n",
      "abnormal heart atrium and ventricle connection\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 85\n",
      "number overlapping genes = 9\n",
      "0.0037109064344590095\n",
      "(0.3338194112577292, 1.723259443555313)\n",
      "1.0285394274065212\n",
      "\n",
      "heart left ventricle outflow tract stenosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 2\n",
      "0.25041978531040454\n",
      "(-0.6019234700523313, 2.308758450163624)\n",
      "0.8534174900556462\n",
      "\n",
      "heart right ventricle outflow tract stenosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 2\n",
      "0.19882179353281548\n",
      "(-0.5037657581367763, 2.421543609728108)\n",
      "0.9588889257956659\n",
      "\n",
      "abnormal atrium myocardium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 2\n",
      "0.034721976572630577\n",
      "(0.11865993758771443, 3.186410251469998)\n",
      "1.6525350945288562\n",
      "\n",
      "abnormal ventricle myocardium morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 104\n",
      "number overlapping genes = 19\n",
      "6.287359533850072e-11\n",
      "(1.1738654147210412, 2.1792601261981743)\n",
      "1.6765627704596078\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "myocardium hypoplasia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 87\n",
      "number overlapping genes = 16\n",
      "1.7233013258163125e-09\n",
      "(1.1342059709554007, 2.228746361678207)\n",
      "1.6814761663168039\n",
      "\n",
      "ventricle myocardium hypoplasia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 14\n",
      "3.1602646716505507e-09\n",
      "(1.193303545293485, 2.3736122760617953)\n",
      "1.7834579106776403\n",
      "\n",
      "myocardial trabeculae hypoplasia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 62\n",
      "number overlapping genes = 13\n",
      "4.552502743034969e-09\n",
      "(1.226170891531459, 2.4577259020768727)\n",
      "1.8419483968041659\n",
      "\n",
      "abnormal heart electrocardiography waveform feature\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 326\n",
      "number overlapping genes = 26\n",
      "0.00047379461971250607\n",
      "(0.3190651395855106, 1.1337625072859943)\n",
      "0.7264138234357524\n",
      "\n",
      "abnormal heart layer morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 429\n",
      "number overlapping genes = 53\n",
      "2.1001408068865133e-16\n",
      "(0.9483707412246236, 1.5425829420564892)\n",
      "1.2454768416405564\n",
      "\n",
      "thin ventricle myocardium compact layer\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 4\n",
      "0.00024187486348255127\n",
      "(1.0008539508628171, 3.294162605714398)\n",
      "2.1475082782886075\n",
      "\n",
      "heart block\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 111\n",
      "number overlapping genes = 17\n",
      "3.996384804611502e-08\n",
      "(0.9398527393269295, 1.9832131284493544)\n",
      "1.461532933888142\n",
      "\n",
      "increased heart right ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 97\n",
      "number overlapping genes = 3\n",
      "0.6198753289272847\n",
      "(-1.4432601648172352, 0.8602807746825782)\n",
      "-0.29148969506732847\n",
      "\n",
      "abnormal heart right ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 117\n",
      "number overlapping genes = 5\n",
      "0.9205480919578747\n",
      "(-0.8530516515442328, 0.9445313595250673)\n",
      "0.045739853990417245\n",
      "\n",
      "abnormal heart left ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 150\n",
      "number overlapping genes = 12\n",
      "0.017549020175898183\n",
      "(0.12585528180193695, 1.314536192732091)\n",
      "0.720195737267014\n",
      "\n",
      "increased heart left ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 114\n",
      "number overlapping genes = 11\n",
      "0.003724268321541965\n",
      "(0.30040785459626873, 1.5522776578084927)\n",
      "0.9263427562023807\n",
      "\n",
      "decreased heart left ventricle size\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 1\n",
      "0.7137418132123223\n",
      "(-2.362323277476059, 1.6174513511112825)\n",
      "-0.37243596318238836\n",
      "\n",
      "abnormal cardiac muscle tissue morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 485\n",
      "number overlapping genes = 57\n",
      "3.479618929406704e-16\n",
      "(0.9050370802022847, 1.4776630243279805)\n",
      "1.1913500522651326\n",
      "\n",
      "cardiac muscle necrosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 30\n",
      "number overlapping genes = 0\n",
      "0.5092280872865361\n",
      "(-3.7372386558493864, 1.8542435640343853)\n",
      "-0.9414975459075006\n",
      "\n",
      "sinus bradycardia\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 4\n",
      "5.8227051527538736e-05\n",
      "(1.263724486790903, 3.668532117041193)\n",
      "2.466128301916048\n",
      "\n",
      "abnormal heart left ventricle pressure\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 1\n",
      "0.5639567456646528\n",
      "(-2.568661639025704, 1.400283228620813)\n",
      "-0.5841892052024455\n",
      "\n",
      "abnormal heart right ventricle pressure\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 1\n",
      "0.8376810631135032\n",
      "(-1.8018574382215609, 2.2224930964301284)\n",
      "0.21031782910428376\n",
      "\n",
      "absent heart\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.4165110534089509\n",
      "(-1.2042232960573682, 2.9095647582912183)\n",
      "0.852670731116925\n",
      "\n",
      "abnormal fetal cardiomyocyte physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 59\n",
      "number overlapping genes = 14\n",
      "8.204025528020357e-11\n",
      "(1.3985642997052583, 2.606952179791504)\n",
      "2.002758239748381\n",
      "\n",
      "decreased fetal cardiomyocyte proliferation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 17\n",
      "number overlapping genes = 8\n",
      "4.0544131034679097e-10\n",
      "(2.0915468171271896, 4.001798449092356)\n",
      "3.0466726331097727\n",
      "\n",
      "thick heart valve cusps\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 4\n",
      "0.004772594146946692\n",
      "(0.4732235062552608, 2.625120961790083)\n",
      "1.549172234022672\n",
      "\n",
      "abnormal heart echocardiography feature\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 50\n",
      "number overlapping genes = 7\n",
      "0.0010056973907009424\n",
      "(0.5438680505126527, 2.148067653980946)\n",
      "1.3459678522467993\n",
      "\n",
      "abnormal heart valve physiology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 2\n",
      "0.36261254166938983\n",
      "(-0.7734268821999262, 2.1151749945612917)\n",
      "0.6708740561806827\n",
      "\n",
      "increased cardiac stroke volume\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.9013454248419377\n",
      "(-3.000261872406575, 2.643321901442657)\n",
      "-0.17846998548195891\n",
      "\n",
      "decreased cardiac stroke volume\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 34\n",
      "number overlapping genes = 2\n",
      "0.5999611705575835\n",
      "(-1.0476330209516944, 1.8131291845995583)\n",
      "0.38274808182393194\n",
      "\n",
      "delayed heart development\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 1\n",
      "0.797007135690424\n",
      "(-1.7505239152141234, 2.279404932969622)\n",
      "0.2644405088777493\n",
      "\n",
      "cardiac edema\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.9423205387074528\n",
      "(-2.9298359644723018, 2.72122282407428)\n",
      "-0.1043065701990109\n",
      "\n",
      "abnormal pericardial cavity morphology\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 127\n",
      "number overlapping genes = 14\n",
      "0.0001610910772977707\n",
      "(0.5181770857730242, 1.6383209787717745)\n",
      "1.0782490322723994\n",
      "\n",
      "prenatal lethality prior to heart atrial septation\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 2\n",
      "0.713695622696696\n",
      "(-1.6742912824246567, 1.146289926827953)\n",
      "-0.2640006777983519\n",
      "\n",
      "abnormal cardiac muscle glycogen level\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 1\n",
      "0.754742690502002\n",
      "(-1.6964227773513678, 2.339731533642189)\n",
      "0.32165437814541065\n",
      "\n",
      "increased cardiac muscle glycogen level\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 1\n",
      "0.5698206576402062\n",
      "(-1.445090549827524, 2.6253708612335607)\n",
      "0.5901401557030184\n",
      "\n",
      "abnormal heart ventricle wall thickness\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 185\n",
      "number overlapping genes = 24\n",
      "9.540597026639145e-09\n",
      "(0.8387962701528111, 1.7089340157490631)\n",
      "1.2738651429509371\n",
      "\n",
      "abnormal heart left ventricle weight\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 3\n",
      "0.19184783106560221\n",
      "(-0.3961286327466911, 1.9751608761717168)\n",
      "0.7895161217125128\n",
      "\n",
      "abnormal cardiac thrombosis\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 4\n",
      "0.02768853399344521\n",
      "(0.1292272696803709, 2.2249915001418223)\n",
      "1.1771093849110965\n",
      "\n",
      "abnormal circardian behavior entrainment\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.5198727619873533\n",
      "(-1.3710330526001422, 2.711509659464158)\n",
      "0.6702383034320079\n",
      "\n",
      "increased cardiac cell glucose uptake\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.9866227844466524\n",
      "(-2.8540920034898085, 2.8056751585497657)\n",
      "-0.024208422470021418\n",
      "\n",
      "myocardial fiber disarray\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.8280090140884289\n",
      "(-3.1278201308170908, 2.503595583641881)\n",
      "-0.31211227358760496\n",
      "\n",
      "abnormal response to myocardial ischemic injury\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 51\n",
      "number overlapping genes = 3\n",
      "0.520477220408005\n",
      "(-0.7854855226856448, 1.5518107591056325)\n",
      "0.38316261820999387\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "decreased susceptibility to myocardial ischemic injury\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.9815840310025946\n",
      "(-2.0252556806614357, 1.9781069591869644)\n",
      "-0.023574360737235622\n",
      "\n",
      "increased susceptibility to myocardial ischemic injury\n",
      "number of genes in root node = 770\n",
      "number of genes in focal MPO term = 30\n",
      "number overlapping genes = 3\n",
      "0.11547585843475806\n",
      "(-0.23529026205721792, 2.154675502635877)\n",
      "0.9596926202893297\n"
     ]
    }
   ],
   "source": [
    "MP_focal_list = MP_focal_heart_list\n",
    "root_KO_heart_df=validation.MPO_enrichment_root(hier_df,MPO,mgi_df,MP_focal_list,G_int,verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0000266</th>\n",
       "      <td>2.364659e-21</td>\n",
       "      <td>0.893231</td>\n",
       "      <td>0.708706</td>\n",
       "      <td>1.077757</td>\n",
       "      <td>1803</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000267</th>\n",
       "      <td>5.322907e-20</td>\n",
       "      <td>1.415235</td>\n",
       "      <td>1.112327</td>\n",
       "      <td>1.718144</td>\n",
       "      <td>364</td>\n",
       "      <td>abnormal heart development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000269</th>\n",
       "      <td>2.276216e-04</td>\n",
       "      <td>1.051879</td>\n",
       "      <td>0.492593</td>\n",
       "      <td>1.611166</td>\n",
       "      <td>130</td>\n",
       "      <td>abnormal heart looping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000270</th>\n",
       "      <td>6.809632e-06</td>\n",
       "      <td>1.417188</td>\n",
       "      <td>0.799874</td>\n",
       "      <td>2.034502</td>\n",
       "      <td>81</td>\n",
       "      <td>abnormal heart tube morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000274</th>\n",
       "      <td>3.619680e-09</td>\n",
       "      <td>0.797133</td>\n",
       "      <td>0.532359</td>\n",
       "      <td>1.061907</td>\n",
       "      <td>785</td>\n",
       "      <td>enlarged heart</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0000266  2.364659e-21  0.893231         0.708706         1.077757   \n",
       "MP:0000267  5.322907e-20  1.415235         1.112327         1.718144   \n",
       "MP:0000269  2.276216e-04  1.051879         0.492593         1.611166   \n",
       "MP:0000270  6.809632e-06  1.417188         0.799874         2.034502   \n",
       "MP:0000274  3.619680e-09  0.797133         0.532359         1.061907   \n",
       "\n",
       "            num_genes_in_term                  MP_description  \n",
       "MP:0000266               1803       abnormal heart morphology  \n",
       "MP:0000267                364      abnormal heart development  \n",
       "MP:0000269                130          abnormal heart looping  \n",
       "MP:0000270                 81  abnormal heart tube morphology  \n",
       "MP:0000274                785                  enlarged heart  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_heart_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/churas/anaconda3/envs/netcoloc/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "      <th>MPO_term_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0002882</th>\n",
       "      <td>1.780835e-25</td>\n",
       "      <td>1.012687</td>\n",
       "      <td>0.822414</td>\n",
       "      <td>1.202959</td>\n",
       "      <td>1512</td>\n",
       "      <td>abnormal neuron morphology</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003861</th>\n",
       "      <td>9.139890e-24</td>\n",
       "      <td>1.046446</td>\n",
       "      <td>0.842377</td>\n",
       "      <td>1.250514</td>\n",
       "      <td>1219</td>\n",
       "      <td>abnormal nervous system development</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002925</th>\n",
       "      <td>6.056466e-23</td>\n",
       "      <td>1.187488</td>\n",
       "      <td>0.951498</td>\n",
       "      <td>1.423478</td>\n",
       "      <td>762</td>\n",
       "      <td>abnormal cardiovascular development</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000266</th>\n",
       "      <td>2.364659e-21</td>\n",
       "      <td>0.893231</td>\n",
       "      <td>0.708706</td>\n",
       "      <td>1.077757</td>\n",
       "      <td>1803</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000267</th>\n",
       "      <td>5.322907e-20</td>\n",
       "      <td>1.415235</td>\n",
       "      <td>1.112327</td>\n",
       "      <td>1.718144</td>\n",
       "      <td>364</td>\n",
       "      <td>abnormal heart development</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0002882  1.780835e-25  1.012687         0.822414         1.202959   \n",
       "MP:0003861  9.139890e-24  1.046446         0.842377         1.250514   \n",
       "MP:0002925  6.056466e-23  1.187488         0.951498         1.423478   \n",
       "MP:0000266  2.364659e-21  0.893231         0.708706         1.077757   \n",
       "MP:0000267  5.322907e-20  1.415235         1.112327         1.718144   \n",
       "\n",
       "            num_genes_in_term                       MP_description  \\\n",
       "MP:0002882               1512           abnormal neuron morphology   \n",
       "MP:0003861               1219  abnormal nervous system development   \n",
       "MP:0002925                762  abnormal cardiovascular development   \n",
       "MP:0000266               1803            abnormal heart morphology   \n",
       "MP:0000267                364           abnormal heart development   \n",
       "\n",
       "           MPO_term_type  \n",
       "MP:0002882         brain  \n",
       "MP:0003861         brain  \n",
       "MP:0002925         heart  \n",
       "MP:0000266         heart  \n",
       "MP:0000267         heart  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join brain and heart results together\n",
    "root_KO_brain_df['MPO_term_type']='brain'\n",
    "root_KO_heart_df['MPO_term_type']='heart'\n",
    "root_KO_df = root_KO_brain_df.append(root_KO_heart_df)\n",
    "root_KO_df['MPO_term_type'].loc['MP:0002419']='neg_ctrl'\n",
    "root_KO_df = root_KO_df.sort_values('OR_p')\n",
    "root_KO_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "heart       156\n",
       "brain       100\n",
       "neg_ctrl      1\n",
       "Name: MPO_term_type, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_df['MPO_term_type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot top performing brain and heart terms + negative control term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot top performing brain and heart terms + negative control term (MP:0002419')\n",
    "\n",
    "# plot 5 best and 5 worst (for terms which have at least 150 genes)\n",
    "\n",
    "heart_terms_plot = root_KO_heart_df[root_KO_heart_df['num_genes_in_term']>150]\n",
    "heart_terms_plot = heart_terms_plot.sort_values('OR_p',ascending=True).head(5).index.tolist() #+heart_terms_plot.sort_values('OR_p',ascending=True).tail(5).index.tolist()\n",
    "\n",
    "brain_terms_plot = root_KO_brain_df[root_KO_brain_df['num_genes_in_term']>150]\n",
    "brain_terms_plot = brain_terms_plot.sort_values('OR_p',ascending=True).head(5).index.tolist() #+brain_terms_plot.sort_values('OR_p',ascending=True).tail(5).index.tolist()\n",
    "\n",
    "neg_ctrl_terms_plot=['MP:0002419']\n",
    "\n",
    "terms_plot = brain_terms_plot+heart_terms_plot +neg_ctrl_terms_plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAesAAAF/CAYAAABpBOH0AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAACdAklEQVR4nOydeXwV1fXAv6wRCJugsroRPaCAa1zYtMW1YluRqoBVbEVbt1pRRPyBirYiasWlqK1WVBShCC64oVAhKgqoiCAcCJvsm2wBBELy++PcgcnLS/KyvpfH/X4++by8mTv3nntn3px7lrlTJTc3F4/H4/F4PIlL1XgL4PF4PB6Pp3C8svZ4PB6PJ8Hxytrj8Xg8ngTHK2uPx+PxeBIcr6w9Ho/H40lwqsdbAI/H46koRCQFSAfWAPviLI7HE0k1oCkwU1V3h3d4Ze3xeA4m0oGMeAvh8RRBZ+Cz8AavrD0ez8HEGoDXXnuNJk2axFuWMmfcuHEA9OjRI86SFI89e2D2bJg1y/6vKJo1g3PPhcMPr7g2C2Pt2rX07t0b3HUaxitrj8dzMLEPoEmTJrRo0SLespQ5DRo0AKhUfcvNhW++AVWoWdP+Kopt2+CHH+DYY8ENXaKQL0TjE8w8Ho/HEzdWr4avvoLdu4suWx4sWWJW/b4Ez2Dwytrj8Xg8cWHnTpgxA9avj58MOTlm2S9dGj8ZYsG7wT0ejydJOPHEE+MtQrFYuBDmz4+3FOYOnz7dYtf16sVbmuh4Ze3xeDxJQnp6erxFiJkdO+D77ys2oawwVqwwl3yiKmvvBvd4PJ4kYe/evezduzfeYsTEmjWmHBOFPXtg3rzEmTxE4pW1x+PxJAmvvfYar732WrzFKJJ9+8z9vWtXvCXJy48/wrp18ZYiOl5Zezwej6dC2bwZVq6MtxT52bo1saz9MF5ZJxkicr+IJODPoGSISB8RyRURn19RwZTF2IvIua6OtLKUzVO52bEDtm+PtxTRWbPGnv1ONLyy9ng8Hk+FsnWrPbaViGzeDFlZ8ZYiP95a8Xg8Hk+FsibfYpolIzNzPDNnDiUrazWpqc1ITx9AWlr3UtW5fbtNJOrWLRsZywqvrCsZItIWGAp0BOoAq4BnVXVYRLl7gduBFGAicJuqbnT7coEbgCtdPVtcHUNCx18EDAbaAbuAt4H+qrrZ7V8GvAlchL0lphcwAJgONAKuAvYATwFjgOeBM4HlTpZJrp6WwDCgK9AQWA+8DgxQ1SLXFBKRkdibatYC1wKpwBTgBlVd7co0Bx53su4DPgf6qeqiUB1pqtopVO9DwNWqenRozB4ErgFqAecBi1yfrwZaAEuAf6jqC+6Yc4H/Ab8CHgWOA5YCd6vq2wX0537gXDe2A4F62Pn7C3beLweygMdV9bHQcb2BOwEBfnJjOFhVfy5E/neBfwOdXJurgUdU9V8RYl0jIgNdH+cBt6jqdFfvIYWNQZT+VQVuAf4EHOva/BcwTFVzXJlWwNPYywy2Y+fuz8BDwLfAbKCrqk4J1fsCcJSqnh+t3YOFk08+Od4ixMSWLaWvIzNzPBkZ/cnOtiy1rKxVZGT0ByiVwv755/itplYYXllXIkSkNvAxMBnoAGQDfwQeEZEpqjrLFW0OnA9cANTFbsgvA5eEqnsUu2n2dXU8ICKfquo0EfkNMAH4G3AdcCR285wkImcGN1XgZuDXwEZgAXbTvgO7qZ6MKYYHMCV6J6DAE06Wpq6OdzEFfQGwDbgUGA58BYyLcWiuAF4DugDNMEX3EPAHEakDfIrd4M/BlPUdwFci0k5VV8XYBsBNwMXY5GCuk+8sNw7zgG7ACBGprapPhY57FLgNWAE8CbwiIs1VtSBn29nAJuAXmIKfgCnXR4C/Y+drmIi8p6rzReRWTKHdBbwPtAVGAGlA+K4VKT/AIOBhbGJ3MfCsiOxQ1XBKcV9sMrYDU6z/FZGWqpqLjXssYxAei+vdeHyOKeQngcOBO9w1PhmbCHXGJl/PYoodVf1ORL7Grq0psP938TtsArAfEWkANIhov/Isml0CKoOy3rcvtqU9J04s/GUk69Z9TU5O3uessrN3MXVqPxYseL3QY7t1K/jWkpNjf4mGV9aVizqYIntWVbcBiMhg4G7MAg6U9R7gSlVd58rcgina1qq6wJV5VVVHuf2DsJtnR2AacA8wUVUHubLqLLdZwIXAB277B4GF7OoBmK+qD7rvTwD3AWNUdYLb9hwwXkQOwyzEV4FxqrrcVfOkiPR3/YlVWWcBN6rqXmCBiIzClD+Yhd8Y6OX2IyLXY4qwL3B/jG2AjdlMV0drTBFerqpvuf1PiMixwEAReTp03H2BFejGepbr3/QC2qkO9FXVn1x/vgd2Bd4TERkK9AfaisgCzAJ/TlWfdMcvchbsOBFpo6rBGlH75Xf1AExW1QfcJhWRszDFHVbWN6jq9+6Yx4DxQFMRqVeMMcCVvxXzmrzsNmeKyKHAwyLygKvvCOD0kCeoN/BdqKoXscnKTaq6E7gMyMEmNWFux66/g4adLhBcu3btOEtS/kQq6qK2V3a8sq5EqOoGERkB9BSRUzDL6SS3u1qo6OJAUTu+cp9tMQuY0Ceqmisi24DgfTftMRd4uO2vRSTL7QuU9ZJoYob+3+E+M0PbgicrU1x/ngF6iEg/15/2mHUc7k9RLAkUsWNrqC+nYq7kzU45BRwCtClGG5C3v+3d57SIMlMxj0X4pXsLQv9vdZ+FvVtoo1PUATsi2t4/hsBhQJMC5AjkDJR1tPM1NeL7F8BvI7YtDP2/2X3WonhjANAaqFFA+Rpu/6nAokBRA6jqHBHZGir/OuZJuAybVPQBRgcu/xDDgZER21qQxO+zHjt2LAB9+vSJryCFUK0aVI9B8xRm/QKMHn0GWVn5HWOpqc2LPLYwqla1v0TDK+tKhIg0Ab4ENmAx5EnATMy9GibSyRQovnAkJlpUpkrEZyRVgbBSjLakQbTlk6I6lZyLehrmMRiLucdnkP9mXhSF9aUqNlm4JEqZwnI+o/02wv0tbIwg7zgUJl80Yh7DYsoRy/mqRv7rJ5rTskox2w6OKap8diHlAFDVrSLyJnC1iHwK/BLzBkWW24LlY+wnYsLmiROHHlr6OtLTB+SJWQNUr16L9PQBpaq3Vi1ISSmtdGVPAs4fPIXQC7OkOqrqQ8613NDtC9/gWolIOJexE5CLxRRjYQ4W/92PiKQDtTkQ6ywLLsQsqa6qOlhVx2CW5xEUccMuBnOxmPs2Vc1U1UxgGZasFfRxD2Z9hzmuiHrnuM8uEdvPwZKmNlMBOA/KugLkgKLPV+Ri0h2Ab2JsvrhjMB9TyNHK78YmVd8BaSLSKNgppmHrRxzzIpaUeB3wQyhfw1MJOOIIqFLKX3haWnc6dx5GampzoAqpqc3p3HlYqbPB69aF1NTSyVYeeMu6crECcz9eKSLTsMzfJ9y+8FwwBRjrMngbYclhr6tqNDdoNIZiceUhmJuxJfAMFmudUtiBxSRYvKW3iIxx7TyMuUTLam47Ckt8e1NE7sIsrXuxZKogJv85cL2I9MGS0X6Nxbw3FVSpS+x6G3haRHI4kFzVF7jLhRbKqAtFMhSL4S4G3sPi4U8Cb6nqwkKPhCtEZDrwIeb+7o71v0iKOwaqus2Fce4TkY2Yy70Tdh6edftHY0mJo0RkAHa9B7Hv8FIVU4EfsXj9vbHI60kc6teH2rVtcZTSkJbWvdTKOZKGDaFOnTKtskzwlnXlYhz2mNOjmJUyHLMwppHXQvoWs46muGM+Am6MtRFnsV+O3XjnYElgU4DzVTW7tJ0ItTMDy8y+BYvrjsRuwqPJb/GVtI2tmCW3HlNIMzFL+/xQ4tVrmHJ7DLNEOwL/F0P1PYE3sMzruZiVd2Mo0atCUNXhWOLWn4EfsOviJSdfUbyCZeB/j2XtX6Gq7xWj+eKOQT/sGh7iZB3o/u/n+rIbe8SuJhbyeZMDcef9mUMuE30kZnCMKoa8ngSgTp3Ee445oFmzeEsQnSq5ibiumsfjKXfcs/KjVDWWiUmFICJHA8dHPGXQDFtPoIuqZoS2Pw80UtXCn/HJX//SyZMn06JF8j3FNXeuRT3atm0bZ0kKZ98+mDQJvvqq6LIVScOG0KMHNG8en/ZXrlxJ165dAY5R1WXhfd4N7vF4EolawPvOBf4m9pz0Q9hz118CiMj5WOb4tRx4RM9D4ivpgGrVQMTeZ51Iy44edRQcHvkMQ4Lg3eAejydhcKGJq4DeWAz8Eyxr/7zQ43nXYYvD/F1Vi/vkQFKzdetWtm7dWnTBBKBJE0gk50ZKCpxwAtSoEW9JouMta4/nICVYSjXRUNVxFLIgjqr2qkBxKhUTJti6MIn8nHVA7drQvj0sX54Yy3sefXTixqvBW9Yej8fjiRNpaZAInvsGDeCssxLzka0Ab1l7PB6PJy4ccgicfjqsXl28N3FVr26Pf9WtaxZ6jRp5n9vet8+s9awse4tWYe/OrlYN0tPhyCNL3o+KwCtrj8fj8cSNpk3Nqv3oo6KTzVJTrXzdulZ27VqYP9/eQb1nj72Ao1o1K9e4sbm1jznGFPnGjVY+8iUixx8P7drZcYmMV9Yej8fjiSsnnGDKd+pUe0VlJLVqwbHHmkWdmQnz5sGPP8b2dqw6dUwht28PJ51kSnvFCsjNtTrPPRfqRa5fmIB4Ze3xeDxJwtlnnx1vEUpEjRpw6qmmQDMyYFdoFfvmzW150vnzYfr04r8Le8cO+PZb+O47U87nnGNx8pwc6NTJ6q4MeGXt8Xg8SUJlflFJSorFjlNSTClv3mzPYmdnw3vvgWrRdRRGTo5Z5WvWwK9+BRdfbC71yoJX1h6Px5MkbNxobxZt3LhxnCUpGYGFfcQRsHIlzJkD77xjiWJlQe3aZqmvXg2TJ1ss+5hjyqbu8sY/uuXxeDxJwsSJE5k4cWK8xSgVVarAYYeZ+3rOnLJ5v/Qhh9giLK1bmzVdowasWmVJbcuWlYnY5Y63rD0ej8eTMOTmWnx5xgzL6K5Txx692rABtm2zrO9YXmlRrZpZ0o0bH3iTVmTG99q18Mkn0K2bKfNExitrj8fj8SQMK1bYCz6CVc1q1bK/Qw+1jPHdu83qzso6oLhzc80iDxR03bpmTR9yiB1bGCtXwqxZcN55Vj5R8cra4/F4PAlBVhZ8+SVsivIm+erVDzxiddhhljC2d++Bx7cCZV2Stb2//x5atrRHuxIVr6w9Ho/HkxBkZsLChbGVrVrVMsfLgt27YeZMe7FIo0ZlU2dZ45W1x+PxJAldunSJtwglZscOmDvXHtWKB2vWmEvcK2uPx+PxlCvHHntsvEUoMWvXWoZ2vNi3z1ZGO/74ouPc8aDSP7olIveLyMp4y1FWiEgfEckVET+RKmdEpJOIdEoAOa4XkRjyW8usvaPdNXZeRbVZXohIDRG5I95yJApr165l7dq18RajRCxcmHflsniwejWsXx9fGQqi0itrj6ckiEg1IANIi7csnlJxDfB4vIVIFD788EM+/PDDeItRbHbuTAwlmZVV/OVMKwqvrD0HK1WKLuKpBPjzmAQEr7JMBIrzqs6KJOFdrSLSFhgKdATqAKuAZ1V1WES5e4HbgRRgInCbqm50+3KBG4ArXT1bXB1DQsdfBAwG2gG7gLeB/qq62e1fBrwJXAQ0BXoBA4DpQCPgKmAP8BQwBngeOBNY7mSZ5OppCQwDugINgfXA68AAVY14eVvU8RgJVAPWAtcCqcAU4AZVXe3KNMesjYuAfcDnQD9VXRSqox5QG+gA/BfoA/xWVd8NtfUEcJ6qthORqsAtwJ+AY4HVwL+AYaqaIyJHA0uB81X1E3d8dWAvcJ2qjhSR2m58ugENgAXAg6r6ZgF9LbC8iPwGeAtIU9XFoWM+AZaqal8RuRh4EDgB2AG8D9zuzuled8hLItJHVc8VkfrAo8BlQE3ga+wamOXqvh84F7sOBroxnAj8BbtGLweygMdV9bFofXL1XAYMAY4DvgI+jdhf08l9NVAfmAsMVtVJbkzWAP+nqk+HjvkLMAhopqp7ROQ6oD9wNLAMeA54WlWjvqdIRHoDdwIC/IRdk4NV9We3Pxe4Cfg9cCqQ6WR4qzRjU5icoWvqSqAfcLLr+99V9V8i0gf4d0i+61R1ZNRB9yQ0pVXWmZnjmTlzKFlZq0lNbUZ6+gDS0rqXqK4NGyw7vKwyzcuKhLas3Y3pY0y5dgBOxBThIyJyeqhoc+B84ALsxn4a8HJEdY8CI7Eb94vAAyLSxbXzG+xGPhk4HegNdAYmOSUVcDPwV+A8YJrbdgewAruRPAM84Op60smxPEKWdzHlfgF2YxyG3SQvi3lg4ApXRxfX347AQ64vdbCbfzXgHPe3EfjKKfGAy4D/ORmHuL73CnY6N3FP4CW36VFMgTwCtHX/DwQKVEpReBBoD/wKaION0xgRKWh13sLKvwesw5RHIHNL4BfAf0SkMTAB+I879rfYOQ3kbeE+bwe6i0gVV/9xwKXYROtL4HMROSUk09mYUvoFNkG7HJgH/ICN5SvAMBFpE61DItIBU2gTXN9eA+6OKDYSm2hdDZwCjAUmisglqroT+w30ijjm98DrTlHfgE3WhmC/mf/DJpZDC5DpVuw8jwROAm4N6osoOszJexJ2HY8Xkc4lHZtiyPk48DfsPE4BnnXXwBhMiYNNoMdE658n8dmzx/5KQmbmeDIy+pOVtQrIJStrFRkZ/cnMHF+i+nbvPrAgSyKR6JZ1HWA4ZgVvAxCRwdjNrR0wy5XbA1ypqutcmVswRdtaVRe4Mq+q6ii3fxBwG6bkpgH3ABNVdZArq87SmAVcCHzgtn8QWMiuHoD5qvqg+/4EcB8wRlUnuG3PYTe1wzDL4lVgnKoud9U8KSL9XX/GxTguWcCNqroXWCAiozDlD3aTbAz0cvsRkeuxG2hf4H5XbrOqPhLqy0vACyKSqqpZrr5GwCgRqYfdwAeoajDxyBSRQ4GHReSBGOVuBWwHlqjqFncepgGbi1teVbNF5GVMoQV9ugZYqKrTReRkzMvyoxvr5SJyKQeu+XXuc6uq/iQiXbEJ4eGqusHtGygiHTHrsI/bVh3oq6o/YWP/PbAr8PSIyFDMUmwLzI/Sp1uBr1R1sPu+UETaYxNBRCQNmySlBxY98A8ROQm4C5ukvAR8ISKtVHWxU36nYecXzML+u6qOdt+XuHM4wv1+9uMmKQOB51T1Sbd5kZukjhORNqoa9OMVVf2n+/8eEfkF9jvKKOHYxCrncFV9x9VxF3AdcJaqjhaRbQCqmi+rSkQaYB6ZMC0iy3niT05O9CVEJ07sUeSx69Z9TU5OXk2fnb2LqVP7sWBB5HwzP9265b3t5ubG9p7siiahlbWqbhCREUBPZ92kYbN6MMsxYHGgqB1fuc+2mOuU0Ceqmut+5DXdpvaYCzzc9tcikuX2Bcp6STQxQ//vcJ+ZoW1BfmOK688zQA8R6ef60x5oFtGfolgSKGLH1lBfTsVckJsjXpd3CGaZBCyNqHMCMAKzuF/FLKv3VHW9iJwB1OCANyFgqtvemgPKrzAewSyyDSLyFTAJGK2qW0pY/j9AfxHpqKqfY2GBFwBUdbaIjAbeFZE1wCeYW3ZCAW2d6j6XRoxbCjZ2ARudMgrYQd7rYv/5LqCddph1GOYLnLLGLGmATyPkqIF5mHCTEcWs6wexc/Wdqn7rJoUtgAedazqgquvHMSEZAQ4DmhD93IJdn/MjtoXl/lXoe8xjU0w5F4T2b3WfNSma27HJ80FD165d4y1CiSjNizoiFXVR24uiShX7SzQSWlmLSBPMFbkBiyFPAmZibucwkbHeQPGFnRnRHBtVIj4jqcqB2CbkvckF7I2yraC4YB3splgHc22+DMwg/42yKArrS1VssnBJlDLhF83l6Yuq/iwibwC9ReQtzG3cM6LuSIKfWLQxgIjryymZlljI4jzMEh7k3LuTIg8uqryqqoh8BlwtIvuwWPoroeN7Oav/YlfHq8CXInJuAX3ZiYUzIgmPd8znuxAixzNcZzCm5+KUc4jwdf4ScJ2IPISFbf4RcfydwEdR2v4RmxwWJEukHGHZIvteLUKm4oxNceQs7HovjOGYaz9MCw54ApKOli1bxluEElG9ui0TujfiCoq0eqMxevQZzgWel9TU5jEdX5AsiUZCK2vMcjgMOF5V9wCISDu3L/xjbSUidVU1SFHoBORi8bJYmIPFf8OJL+lYAtbckoufjwsxC66Fqq5y7RwKHEHZZbXOxVy221R1vWujOvAGNkEYW8ixL2HJaH0w9/N7bvt87EbchQOhB7B4+G5scpDqttUL7T8uXLlTnJ85l+Y7InI7Fs/sgU3EKEH5F7EY5y4sTLHWHXsWFhr5K+b9GC4ivbCYa1PyewLmYue7tqrOCcnwL+z6eCZSvhIyG4vthgnnXwTXW7OQGxynlHM44AF6FYvj/gnrz2tu+3pscttKVTNDx18BdMe8D/tR1XUisg47t+E72zkR8gCkk9cz0QH4poB+FkWx5CyEAp9Pdx6YLeFtEd6KpGPFCrNjKpvSrlMHUlNhc0EBsUJITx9ARkZ/srMP2B/Vq9ciPX1AiWQ59FB7GUiikejKegVQC7hSRKZhCVlPuH1hN2MKMFZEBmJx1qexZJtobutoDMXiykOwm15L7OY8i/wuy9IQLN7SW0TGuHYexlycZZV7OApL0nnTxfe2APdi1uWgQo5DVWeIyEJMCfxbVbPd9m0uHHGfiGzE3J+dXH3Puv3bMZfn7SKygAP5BmGrKA2zgvsCi7EkrqNcfdGIpfx/sYzxP2EWZsA24GYR2YO5xmti8fwlwDpV3Sciu4ATRORw4ENMkY52CVc/ujr/wIF8gLLgCcy6fwzLpj8D+HOwU1XnichE4J8unvw9prwGYrHaoNxqEfkIFyoInnxwIZ5HgL+LyFIshHMi8CzwjqrujqKwhmKJX4uxCVo7LEHyLVUNr9R8m4jMw7xbN2Ahqb6RlcVCCeWMxnbYP7me7/ItDlomT54MQJ8+feIrSDGpU8felFUSZR1kfZdVNnizZkWXiQcJnQ2OzfSHYZnI87Gb/4uY2zg9VO5bbIY/xR3zEXBjrI24ZLDLsczqOZjVMgV7DKnMVqpV1RlY9vgtWBxuJBYHHE3e/pSmja2YlbQeU0AzgSOxvkRLeIrkJaAuB7LAA/ph52IIZt0OdP/3c+3mYsleh2Dn43Vs0hReXe7PWOx4FLAQy2C/p5DHbYosr6o7MG9BFhaTDrb/gCm5Xzh5vsCssAv1wCNyD2OPI33ktp2PhSXGYNfBucBlqlpmEzZV/RoLUfzStXE78PeIYldi1/Gz2FhfB/wxlNwXEPVcqerj2FMLN7vjn8YecfpTATINxxLf/uzKD3d19owo+hyWUPYddo1doKqzi+hygRRXzgKYhHmDPscmEJ5KSGqqvXO6pKSldadnzxn07buSnj1nlFhRp6SYZZ2IVMmN5S3eHk8C4yzM71X1znjLkqyIPcfcV1VfiLcspUHcs9uTJ0+mRYvkSwwfOXIkUPksa4AffoAJE/LHrSuSo46C7t2hfv34tL9y5cogSfAYVV0W3pfobnCPp0DEno9vj1nAN8VXGo/HUxqaNYMmTWBFZPpwBVGlCrRpEz9FXRReWXsqM3djj43drKFVzDweT+WjQQNTlitXRn/murxp1Mgs60TFK2tPpUVVO8RbhoMFVU3AJ089kVx00UXxFqFUHH88LFgAP/5Yse1WqwYnnwyHH16x7RYHr6w9Ho8nSWjSpEm8RSgVjRtDhw6waRPs2FF0+bKiVSto186UdqKS6NngHo/H44mRJUuWsGRJrE+sJiatWsEpp5RuVbPi0LAhnHVW4saqA7xl7fF4PEnCtGm2GOKxxx4bZ0lKTo0acMYZZlnPnl2+8eu6deHcc+Hoo8uvjbLCK2uPx+PxJBT16kGXLpahPXt2+bxYo359U9Rt21acFV8avLL2eDweT8LRsKEp09RU+OYbe+d1WVClCrRoYbHx449P7Dh1GK+sPR6Px5OQ1KsHnTubcp0+HZYvL52VXbu2ZX2feqols1UmvLL2eDweT8JSo4ZZwIcdBosXw7x5sHo17I72LrYCOPRQi0ufcAIceSTUjOUFqwmGV9Yej8eTJHTr1i3eIpQbDRvC6adD69awZg0sWQLr18P27eYiz862ZLSqVeGQQyx5rF49U85HHmnPUFevxBqvEovu8Xg8njCNK5tvtwSkpsJxx9nf7t2mrHfuzKusU1LM5V23buWJSReFV9Yej8eTJKgqkPzv7Q5ISbG/gwGvrD0ejydJmD59OnDwKOvikJUFP/0EW7eaNQ5medevbzHt1NT4ylcUXll7PB6PJ2nZvh0WLrTEtHXrzGUeLLRSpYq5y484wp63Pv74xFXaXll7PB6PJylZtQo+/9yUdXZ2/v25ubZS2pIl9vKQxYvt+evmzSte1qLwytrj8Xg8SceKFfDhh6awYyE726zvrVvhwguhZcvyla+4VIJF1jwej8fjiZ3Nm+HTT2NX1GFWrrRjt2wpY6FKiVfWnlIhIveLyMp4y1FWiEgfEckVEe91KkeS7bpJFC677DIuu+yyeIsRV3JzYc4cWLq05HUsXWp1lOdLRIqLV9Yej8eTJNSvX5/6if6ux3Jm40b44YfSLUuak2N1bNxYdnKVFm89eDweT5Iwd+5cANq2bRtnSeLH2rWwYUPB+zMzxzNz5lCyslaTmtqM9PQBpKV1z1du/Xqr67DDylHYYuCVtadQRKQtMBToCNQBVgHPquqwiHL3ArcDKcBE4DZV3ej25QI3AFe6era4OoaEjr8IGAy0A3YBbwP9VXWz278MeBO4CGgK9AIGANOBRsBVwB7gKWAM8DxwJrDcyTLJ1dMSGAZ0BRoC64HXgQGqui+G8RgJVAPWAtcCqcAU4AZVXe3KNAced7LuAz4H+qnqolAdaaraKVTvQ8DVqnp0aMweBK4BagHnAYtcn68GWgBLgH+o6gvumHOB/wG/Ah4FjgOWAner6tsF9Od+4FzgfeA2oDHwFfAnVZ0fkqVv0I7bthJ4QVXvF5E+wP1Af+ARoAnwGdAH6Oc+c4BXVPWOiPYLu27qu35cBtQEvsauiVkh2c/DzvGlwH9V9Y/R+nmwMGvWLODgVtarVxdsVWdmjicjoz/Z2bsAyMpaRUZGf4B8Cjsnx5Y1bdeuXMWNGa+sPQUiIrWBj4HJQAcgG/gj8IiITAlumkBz4HzgAqAu8G/gZeCSUHWPArcAfV0dD4jIp6o6TUR+A0wA/gZcBxwJPA1MEpEzVTX46d0M/BrYCCzAFNcdwEPAyZhiewBToncCCjzhZGnq6ngXU9AXANuwm/xwTEGNi3ForgBeA7oAzbBJxEPAH0SkDvApMBs4B1PWdwBfiUg7VS1OystNwMXY5GCuk+8sNw7zgG7ACBGprapPhY57FFO8K4AngVdEpLmqFvSSwbOBHdj5quHa+Sfwy2LI2szJ9VugHjbOc4BXnMwXA8NFZJKqfuiOKfC6EZEq2ARiD3aOtmDn93MROUtVv3V1dAS+xc5/jbBAItIAaBAhZ4ti9MmToPTo0SPq9txcs4h37Yp+3Lp1X5OTsyfPtuzsXUyd2o8FC17PV37yZHj+eXseOxrjxsV6yyg9Xll7CqMOpsieVdVtACIyGLgbs4ADZb0HuFJV17kyt2CKtrWqLnBlXlXVUW7/IEyZdASmAfcAE1V1kCurItLb1X8h8IHb/kFgIbt6AOar6oPu+xPAfcAYVZ3gtj0HjBeRw4As4FVgnKoud9U8KSL9XX9i/eVlATeq6l5ggYiMwhQOmIXfGOjl9iMi1wO/wCYq98fYBtiYzXR1tAa6A5er6ltu/xMiciwwUESeDh13n6pOcccNwsaxHeaFiEZN4BpV3eSOeRr4ezHkBFOUf1HV71wd/wPSMY9CDrBQRIY4OQJlXeB1gynyDsDhqho4NQeKSEfgL5i1HvB/qro1iky3Y9eD5yChqISwSEVd1PagzoKUdUXilbWnQFR1g4iMAHqKyClAGnCS2x1eHn9xcMN1fOU+22IWMKFPVDVXRLZhSgKgPeYCD7f9tYhkuX2Bsl4STczQ/zvcZ2ZoWzDHTnH9eQboISL9XH/aY1ZhcZb7XxIoYsfWUF9OxSzLzRFLPh4CtClGG5C3v+3d57SIMlMxj8XhoW0LQv8HSqywlwJuDBR16JiSvEQw8lwsCXlFwM5FeCXnwq6bY9z/SyPGMQUby4CfClDUYBPNkRHbWgAZBZT3VBIKsmhzcmD8eHCh+3yMHn0GWVn5nVupqc3p1i1/ne3aQffuXll7EhwRaQJ8CWzAYsiTgJmYezVMZKw3UHzhN85Ge/tslYjPSKoCYaUYzbm1N8q2qBEr56KehnkMxmIu1xnkV4BFUVhfqmKThUuilCnIDQ3Rf4vh/hY2RpB3HAqTLxrFeDPwfqLJG3kuisrHLey6qQrsxNzbkYTlLcDhCaq6BXOf78evmZ3cVK0KTZsWrKzT0wfkiVkDVK9ei/T0AVHLN2uWGIoavLL2FE4v4DDgeFXdAyAiQbpF+BJuJSJ1VdUtj08nIBeLq8bCHCz++1iwQUTSgdpYrLasuBCzfFsEsWMRORQ4gsKVWXGYi7lot6nqetdGdeANbIIwFnP/1os47rgi6p3jPrsA40PbzwFWA5tLJXXh7CUkr4g0xFz9paWw6yYHO/+1VTXoOyLyL2wsnimD9pOOK664It4ixJ2mTW2975078+8LkshiyQavXRuaNClvaWPHK2tPYazAMpGvFJFpgGAJW5DXnZkCjBWRgVhm9tPA66oazW0djaFYXHkIlrjVErsZz8IyrcuKYBGO3iIyxrXzMBZvLasX7Y3CEt/eFJG7MMvuXizBKojJfw5c77KoP8WS5i4ANlEAqjpfRN4GnhaRHA4kmPUF7nKhhTLqQj4+B/qKyFRsovF391laCrxuRGQ5lqQ3WkRuBX4E/gT8gQP5AZ4IateuHW8R4k6TJra296JF0fenpXWPqpwjadnSXvCRKPhFUTyFMQ57zOlRYD4WA3wRcxunh8p9C3yDKdZxwEfAjbE24pLBLseUzxwsCWwKcL6qRll+v2So6gwsM/sWLK47Eov5jiZvf0rTxlbM+l2PJVLNxLLbzw8ehcImJE9inoS5WKLd/8VQfU/MQh/hjrsOS3R7sixkL4Q/A2swpf0+8AnwRRnUW+B14x6jOx8LU4zBrotzgcuC5DlPfmbPns3s2bPjLUZcqV0b0tNL9/asunXh9NOtrkShSm4irafm8Xg85YiIHA0snTx5Mi1aJN9TXCNHjgSgT58+cZUj3uzdC198AZ99Zv8Xhxo1oFMn6NgRqlew73nlypV07doV4BhVXRbe593gHo/H40kqatQw63rfPpgxA37+ObbjatWCM86wv4pW1EWRYOJ4PB6Px1N6atc26/iww0xhr15tyjsa1apZ5veZZ8Jxx0FKWWWwlCFeWXs8Ho8nKUlJsWelmze3V1+q2vuqA0v7kEOgfn0QgRYt4NBD4ytvYXhl7fF4PJ6k5tBD7e/EE2H7dtjjnmVISbFEtGrFWRIpTnhl7fF4PElC79694y1CQlOtGjRoEG8pSoZX1h6Px5Mk1KhRo+hCnkqJf87a4/F4koSZM2cyc+bMeIvhKQe8svZ4PJ4kYd68ecybF+sqv57KhFfWHo/H4/EkOF5Zezwej8eT4Hhl7fF4PB5PguOzwT0ej8eTNGRlwZYtsG2bLYCSkwM1a0LDhvaCjkMPteVIKxteWXs8Hk+ScDC/wOOnn2DJEpg3DzZuhB07TFEH1KgB9erZqy9PPBGOPDIxlxUtCK+sPR6Px1Npyc6GxYth+nT48ce8CjrM3r2waZP9qdoypKefDocfXrHylhSvrD0ejydJ+OILe814hw4d4ixJxbB3L3zzDWRkmPs7Vnbtspd7rF0L551nVnai4xPMPB6PJ0lYuHAhCxcujLcYFUJuLnz3HUydWjxFHebHH2HSJFizpmxlKw+8svZ4PB5PpWPFCnN979xZunpWroSvvip9PeWNV9YHESJyv4isjLccZYWI9BGRXBHx4RyHG4/rE6X+8pbHc3Cya5cp2E2byqa+H36ARHdI+Jucx+MpDk2BrfEWwnNws3YtLFtWdvXt2QNz58Jxx0GdOmVXb1nilbXH44kZVV0bbxk8BXOwvHVr4UJ7NKssWbUK1q+HY44p23rLCq+skwgRaQsMBToCdYBVwLOqOiyi3L3A7UAKMBG4TVU3un25wA3Ala6eLa6OIaHjLwIGA+2AXcDbQH9V3ez2LwPeBC7CLLFewABgOtAIuArYAzwFjAGeB84EljtZJrl6WgLDgK5AQ2A98DowQFX3xTAeI4FqwFrgWiAVmALcoKqrXZnmwONO1n3A50A/VV0UqiNNVTuF6n0IuFpVjw6N2YPANUAt4Dxgkevz1UALYAnwD1V9wR1zLvA/4FfAo8BxwFLgblV9u4D+3O/q/gD4K/b7fduNWdjaPU5EJgGdgc3AP1X1byJyGHZN/ElV/xOq9/+Aa1T1eBE5DngaOMuN3ZduPOaE+to31I/ewJ2AAD9h52ewqv5cQB+KunZqu/PxO6AmMBaojV0vfYGVwPOqOjhU59XAP4EmqrorWrsHCwfD+6x37jTLuigyM8czc+ZQsrJWk5rajPT0AaSldS+w/K5d9nx2oiprH7NOEtxN7mNMuXYATsQU4SMicnqoaHPgfOACoBtwGvByRHWPAiOBE4AXgQdEpItr5zfA+8Bk4HSgN6YUJolI+Hq6GVMo5wHT3LY7gBXAycAzwAOuriedHMsjZHkXU+4XYMpgGKYYLot5YOAKV0cX19+OwEOuL3WATzGldI772wh85ZR4cbgJUzC/AeYCr2HK5U5MMf0bGCEit0Uc9yhwG9AWU+iviEhqIe2kAxdjY/JboBMwLqLMzcCr2Pn7J/CQiFygqhuAd7BJRZhrgZfc/29gCj0dm0DtBcZHE0REbnXHjQROAm4Ffo8p7GjlY7l2XsYmTj2x67geNrlDVfe6fl0tIlVCVfcB3jjYFfXBwo4dsH174WUyM8eTkdGfrKxVQC5ZWavIyOhPZmbUS3k/iZwV7i3r5KEOMByzgrcBiMhg4G5MWcxy5fYAV6rqOlfmFuxm2VpVF7gyr6rqKLd/EKZMOmJK9x5goqoOcmXVWVezgAsxqw/gg8BCdvUAzFfVB933J4D7gDGqOsFtew4Y7yzALOzGPE5Vl7tqnhSR/q4/kQqqILKAG92NfoGIjMIUHZgSaAz0cvtxyVC/wBTt/TG2ATZmM10drYHuwOWq+pbb/4SIHAsMFJGnQ8fdp6pT3HGDsHFsh3khCuJKVV3ljrkZ+EBETlTV4N2Iz6vqq+7/v4nIXZhynIRNvt4TkaNUdbmIdASO4cAkqRU26VumqntF5A9AaxGpFvZmOGU5EHhOVZ90mxc5pTtORNqo6vwIuQu9dkRkAdAD6KaqH7t2rsEmJAEvAv3ctgznffkFcG/kIIlIA6BBxOYWUcYzaZg6dSoA55xzTpwlKTt69OiR5/vu3eau3leIb23duq/JydmTZ1t29i6mTu3HggVR55IATJ4ML78MVUJTwXHjYr3VlC/esk4SnNU0ArhKRJ4VkY8xlyGY5RiwOFDUjq/cZ9vQtkBpo6q5wDbMJQnQngOWclDma0wptg9tXhJNzND/QcQpM7QtsIxSnJX0DNBJRJ4SkfddJnuziP4UxZJAETu2hvpyKma5bRaRLBHJwvraEmhTjDYgb3+DcZgWUWYqcAQQXjNpQej/wJVdk4JZHChqxxfus11oW2Re6xbMPQ/wEWY5X+2+9wEmBWEBTOn1AzaJyLvA5cCcKGGHw4AmRO8j5L0WCG0r7No51W2eHtr/MzAj9H2+2x94B64BFqhqcB2HuR0LLYT/MqKUSxqWLl3K0qVL4y1G3IlU1EVtrwx4yzpJEJEmWHxxAxYHnATMxNzOYSJvuoHi2x3atpv8VIn4jKQq5jINiOaS3BtlW9TFAZ2LehrmMRiLWX4zyK8ciqKwvlTFJguXRClT2DIL0X434f4WNkaQdxwKky8akWMYnL/weY1mc1QBUNUcF4e/WkQew8IEfwwKqeo/ReS/WCy9KzAEGCQip6jqmsj6ohCtj7EcsxfILqJcwIvA484r9HvghQLKDcdc9GFakOQKO9mItGw3boSxY826LojRo89wLvC8pKY2p1u3gi3lM86AX/2qxKKWK96yTh56YdZOR1V9yLmWG7p94ZtfKxGpG/reCcgF5hEbc7D4735EJB1LAppbEsEL4ELM0uqqqoNVdQxmeR5B0TfzWJkLHAlsU9VMVc0ElmFJekEf92DWd5jjiqh3jvvsErH9HGA1lvRVUtJEpGHoe7Cu5DfFqOMlLAfgdkxJvgMgIkeIyDNATVUdqaq/xyz2IzBX836cd2Yd0fsI0a+Foq6dOdi1eGZofw0OWNwBY7BJyu2Y2/5VoqCqW1R1WfiPA94mTyWlbl37K4z09AFUr14rz7bq1WuRnj6g0OOaNi2tdOWHt6yThxWYq/NKEZmG3YyfcPvC75ZJAcaKyEAs8epp4HVVjea2jsZQLK48BEuiaom5q2dhmdZlRXBT7S0iY1w7DwM1yNuf0jAKy9h+08V1t2Bu4IuBIK76OXC9iPTBktF+jcW8C1yOQVXni8jbwNMikoNNhLphcfC7VDXXxfBLQm0sCe0eTIk+A/xXVRfHWoGqLhGR/2E5A/9W1cA3uMnJ2crVvw1zk+/lQM5DmKHAMBFZDLyHKfYngbdUNdoSE4VeO6qaLSJjgadEZDc2sbnblcsNyZ/lyt0HvB8R1vEkOSkp9uasxYVc8UHWd3GywevWhUaNylrassNb1snDOCxb+lFgPuYCfBFzG6eHyn2LWWFT3DEfATfG2oiz2C/HbupzMKtmCnC+qmYXdmxxUNUZWPb4LVhcdyQWDx1N3v6Upo2tmKW3HvgQCxscifUlSI56DVNAj2HWX0fg/2KovieWWT3CHXcdluj2ZKFHFc1qJ2eGq/9t8md3x8JL2ORu/yNc7vxdjIUmJgM/YB6ObtGUr6oOxzLA/+zKDnf19ozWYIzXzg3AZ9ijf18CO7EYdWSwMZ/8Hqhduza1a9eOtxjlTqtWUL9+4WXS0rrTs+cM+vZdSc+eMwpV1AAtWsARR5ShkGVMldzc3KJLeTyeuOOes75eVUud0ews58tV9fQiC1cQInII9tjWZFXdHtquwKjgSQK3rSfwD6BlcSaJInI0sHTy5Mm0aJHUieFJzZ498PHHMHNm2dRXuzZceim0KW5aaRmzcuVKunbtCnCMC9vsx7vBPZ6DCBHphMXc7wL+EmdxItmNPRf+qVt4JhtLfjsK+C+AiJyKhXgeBkaUpTfHU3moWdPeRb1qFaxeXXT5wqhSBU4+2az1RMa7wT2eg4tumEL8LxazTxjcY4KXYImS07GQTQfggtAaAGdh4Z1vsZCPJ8Qnn3zCJ598Em8xKoQjjoBOnaBBg9LVIwKnnWYTgETGW9YeTyVBVe+neAu1RKtjAJZUl5Co6mwOLFoTbf8ILA/AE4WVKw+uZPfWre291p9+Chs2FO/YatXs+HPOSezEsgCvrD0ej8dTKalaFU44wZLNZswAVVvhrCgOPdSs6XbtoF7kg5kJilfWHo/H46m0VKlimdyHHmrK94cfLI6dlWUv/cjNNSs6NdWU+nHHwbHHQpMmtr2y4JW1x+PxeCo9tWubIj76aNi2zV74sWsX5ORA9er2nuo6dYp+5CtR8cra4/F4koR6lcWnW47UqGEx6MoQhy4OXll7PB5PktC9e+ELf3gqL/7RLY/H4/F4EhyvrD0ejydJ+PDDD/nwww/jLYanHPBucI/H40kS1q5dG28RPOWEt6w9Ho/H40lwvLL2eDwejyfB8cra4/F4PJ4Ex8esPR6PJ0lolGwPF3v245W1x+PxJAmXXnppvEXwlBNeWXs8Ho/noCY3FzZtOrCWeI0a0LhxYr020ytrj8fjSRLeffddwFvYsZKbC6tW2du6MjNtTfGcHDjkEGjWDNq2haOOsnXH402ZKGsRuR+4XlVblEV98UZE+gAvATVUNTvO4iQ1ItIJQFU/i7csYURkGTBKVf+vjOprCxylqu+VRX2FtDMSSFPVTolQf3nL48nLpk2b4i1CpSEnB+bOhalTzaoOs2sXbN4MixbBKadAx47xf5Wmzwb3xA0RqQZkAGnxlqUC+ABIj7cQZcBfgF/HWwiPp7QsWgSTJ+dX1GH27IGZM+HLL2N7T3Z54t3gnnhSJd4CVCBJ0VdV3RpvGTye0rJrF3zzDWyN4WrOyYE5c0DEXOLxIiZl7Vx4Q4GOQB1gFfCsqg6LKHcvcDuQAkwEblPVjW5fLnADcKWrZ4urY0jo+IuAwUA7YBfwNtBfVTe7/cuAN4GLgKZAL2AAMB1oBFwF7AGeAsYAzwNnAsudLJNcPS2BYUBXoCGwHngdGKCq+2IYj5FANWAtcC2QCkwBblDV1a5Mc+BxJ+s+4HOgn6ouCtVRD6gNdAD+C/QBfquq74baegI4T1XbiUhV4BbgT8CxwGrgX8AwVc0RkaOBpcD5qvqJO746sBe4TlVHikhtNz7dgAbAAuBBVX2zgL4WWF5EfgO8hbk5F4eO+QRYqqp9ReRi4EHgBGAH8D5wuzune90hL4lIH1U9V0TqA48ClwE1ga+xa2CWq/t+4FzsOhjoxnAiZvENBS4HsoDHVfWxaH1y9XRw5dOBDcC7wD2quq0k5d043wP8ATgCUGCQqk50121z4D7Xz6MLuJYnEdv5vRq77tOAOcCdqpoREre6iDwCXIf9XicDN6rqGhGZADRS1S6hvrUElgFdgC+Bvzt5Dnfbn1DV51zZkYTc2iLS2pXvDNQCpmHna24B49jclT8f++3NBAaGwyAi0gsYBBzj+vcaMFxVq7jfw2XAMaqa68rXBtYAN6nqa9Ha9XjCrFsHK1fm3ZaZOZ6ZM4eSlbWa1NRmpKcPIC3N3mKWlQULF0LLllA1Tv7oIpt1P4SPMeXaATgRU4SPiMjpoaLNsR/gBdiN/TTg5YjqHgVGYjfuF4EHRKSLa+c32I18MnA60Bu7AUxySirgZuCvwHnYjQHgDmAFcDLwDPCAq+tJJ8fyCFnexZT7BYBgivtO7CYQK1e4Orq4/nYEHnJ9qQN8iin0c9zfRuArd7MKuAz4n5NxiOt7r2CncxP3xOLnYOP3IPAI0Nb9PxAoUClF4UGgPfAroA02TmNE5JgSlH8PWAf8PiRzS+AXwH9EpDEwAfiPO/a32DkN5A1yHG4HuotIFVf/ccCl2ETrS+BzETklJNPZmML+BTZBuxyYB/yAjeUrwDARaROtQyLSHhvrSa5vPd1xk5wMJSk/3PVjADbZnABMEJGTMQW/Bpu8hV3hkddyrOd3OKbwTgXmOznCoYQzgcbYWF/ivj/u9r0IdIo439cCi1T1c+Am4HfYpPp44GngWRHpHGVcjsQmytWw31JnIAfIcPsiy9fFJq1p2DlLBxYCk4N7iYh0A17Frvn22LUzNFTNi8BR2G8qoDuQC4yPbPNgo0mTJjRp0iTeYiQ8q1bBjh0Hvmdmjicjoz9ZWauAXLKyVpGR0Z/MzAOX1PLlsH17xcsaEItlXQe7OTwbsiIGA3djN6VZrtwe4EpVXefK3ILdRFqr6gJX5lVVHeX2DwJuw5TcNMwqmaiqg1xZFZHerv4LsZgfwAeBhezqAZivqg+6708A9wFjVHWC2/YcMF5EDsOsrleBcaq63FXzpIj0d/0ZF8OY4Oq5UVX3AgtEZBR2wwJTII2BXm4/InI9plz6Ave7cptV9ZFQX14CXhCRVFXNcvU1AkaJSD3gVsz6DyYemSJyKPCwiDwQo9ytgO3AElXd4s7DNGBzccuraraIvIxZekGfrgEWqup0p6hSgB/dWC8XkUs5cN2tc59bVfUnEemKTQgPV9UNbt9AEemIWc593LbqQF9V/Qkb+++BXYGnR0SGAv0xhTc/Sp/uAiar6kPu+yIR6QkswZTAp8UpLyJfY+f1DlUd48oMcZOtOqq6QURygKxQvyB0LRfz/A5T1dHuuL7AL4EbnZxgnqIbnJdIReQN4OKgTWzicDU2GQA7Zy+6/1thHpClqroGeEZEFmBKNZKbgN3AVaq6y8lzhRuXm7F7RJjeQBPgbFc3InIDprTvwiYIdwHjQ167hSJyPDapQVXnisgMbIL4qStzLTA6kCFARBpg3qAwSZEEWxAXXXRRvEWocHr06FHsYzZvzusCX7fua3Jy9uQpk529i6lT+7FgweuAPcb17rv2WFdxGDcuVpVSOEVa1u7mMgK4SkSeFZGPgcCBUC1UdHGgqB1fuc+2oW2B0sa5sLZhrk6wWfS0UFlU9WtMKbYPbV4STczQ/8F8KTO0LfgRp7gf9DOYdfGUiLwvIiuBZhH9KYolgSJ2bA315VTMPbtZRLJEJAvra0vMwgxYGlHnBGzSE1j4vwfeU9X1QGugBhFjBEx121vHKPcjwEnABhH5DLgXO3dbSlj+P0Arp1DBbpwvAajqbGA08K6IrBaRVzDPzA8FtHWq+1wajJsbuw7kHbeNTlEH7KCA811IOxdEtPG92xfNGi+qvGDn/svwQap6n7NWCyJ8LRfn/E4NtbEXm9C2C9cbEc7ZjLmocdtfxnlD3Hk7FvNGAPwTu3ZXisgsEXkY2BTx2w5oD8wKK0lV3enkaV9A+cxAUbvyOViSYVD+VMxaDxM5Ji8CPUSklvPk/JID3qcwt2O/sfBfRpRynoOMKhH+s0hFXdD2yOMqkiItaxFpgt2ENmAx5ElYnGlFRNHIWG+g+MI5dNHy6apEfEZSlQOxTThwIw6zN8q2nGiVORf1NMxjMBa7cc0g/w2hKArrS1VMeVwSpUxW6P88fVHVn50V1FtE3sLcxj0j6o4kmHBFGwOIOMfO4m2JhSzOw6yqQSJySdhjEWt5VVWnxK8WkX3kvfGjqr2cVXixq+NV4EsRObeAvuzEwhmRhMc75vNdAFWBN7DQQyQbomwrqvzR7v/cYsgAec9/cc5vZP+rkff3Fy3vIlz/f4B7RORMTGl/GChQVV3kXOrnYuf7EmCAiFyvqi8WUmekzNHOUSzlswspF/AG8ASWlX4M5lmbEaXccCzsFqYFSaywx483t2337t3jLEnFURLLddEiePNN+Pln+z569BnOBZ6X1NTmdOtm9Z9+Olx8MVQrjklXhsQSKu8FHAZ0VNWHnGu5odsX/lG1cjGpgE7YzWtejLLMweK/+xGRdCwBK2qySgm5EJu9d1XVwc5tuRVLCiqredNc4Ehgm6pmqmomlqgzlIg+RuElLPGtD+Z+Dp7LnY/d0CKPPwdTZJmYVQ5mGQUcFy7sFGcnVX1HVW/D4pKLgai+pBjLv4h5A67AXLtr3bFnicgTagxX1UuwpKcuWFJVpHKbi53v2sG4ubHrD/wmmnwlZC5m4S8OtVEVUwAtS1B+EXZu8jyaJSL/E5HgOe2iFHks5zdgfzsiUhOLn39TRP37cfJPxc7XZYSsUhG5DbhcVT9W1btVtT2Ws3JllKrmAKeLSK3Q8XWwnJNov9k5QJqINA2Vr4rFuoPy32Ex9jB5vrtw3H+xuHcPolvVqOoWVV0W/uOAVzAp2bZtG9u2Rc2R9IRo2tQWPQlITx9A9eq18pSpXr0W6ekDAKhVC9q0iZ+ihthi1iswF9qVIjINc/k94faF3YwpwFgRGYjFWZ8GXlfVaG7raAzF4spDsOzPlpi7ehaWaV1WBD/W3iIyxrXzMOZqLMhtWlxGYYlGb4rIXVhy3r2YdTmokONQ1RkishD4G/BvdYuyqOo2ERmBZRRvBL7AJkSDcPkEIrIdc63e7uKMQb5B2CpNw6zgvpjSPRNL2PmiAJFiKf9fLGP8T1hcMmAbcLOI7AFewFzFVzkZ16nqPhHZBZwgIocDHwKzgdEicivwo6vzDxzIBygLHsesq2dF5CmgLhbqqUP02Gyh5VV1j4g8jSVMrsNc5L0w9/0dro7twPEi0kzdEwNhYjy/h7riQ0RkNXY+7nVyPFvMMXgRe1piB/BOaPsRwGAR2YkpztbAKZh7PJIRWKz8DbEsfTDvQ01XdySvYbkp40TkTuz6+Ct2T/mDKzMUeE9E7nBydcJyW6LJPwnzKvyq6O56PAdITYWTT4a1a22J0SDrO1o2eJUq0Lq1Kfh4EotlPQ7Lln4Um/0Px34o08hrSXyLze6nuGM+wn7IMeEs9suxzOo5mLt0CvYYUpmtIubcZXdgj8gswNxkU7HYapksWqH2LGoXLNHnQyxscCTWl2gJT5G8hCmESIuhH3YuhmBx34Hu/36u3VwscegQ7Hy8jk2awtbEn4FPsAnFQiyD/R5VHVmALEWWV9UdWEghC3uMKtj+A5ap+wsnzxeYhXlhKKb6MJao9JHbdj4WlhiDXQfnApepaplN2FT1S8zD0g57NGwiZh2fp6r5glcxlr8Hu5b+iVmJvwYuVdVv3f7HsMnaHMn7dEOYQs9viBGu3NdYrsU50SYARTAOc5ePisi9uB+bWD2Nne9/YYr3ocgKXNJgF2ySm4ElfOViCWTLo5TfhlnRq7HfxVdY2ORcVZ3pynyIPeJ5M+aVuwF4jgNeo6CuDOy6DnI6PJ5i0aYNdOkCderY97S07vTsOYO+fVfSs+cM0tK6U7UqnHgidOoU/yVHq+TmFjfM5vHkR0Q+Ar5X1TvjLUuyIlGeoy9FXS2w0MzJWsAz0fFARM4B1qqqhrYNBP6oqq1C22phSv/3qjoxf00F1n80sHTy5Mm0aJF8ieEjR44EoE+fPnGVo7KQnQ1LlsC8efZo1rZttl54Sgo0aQInnADHHw8NGlSMPCtXrqRr165g6wgsC+/zK5h5SoXY8/HtMQv4pvhK4ykKlyx4BhZemJpIitpxIRai6oO5+U/GsrpHAIhIQyz7uweW3PdBtEoOVpJxAlKeVK9uyviYY2D9env2OifHHtM67DCoW7foOioKr6w9peVuLK55s4ZWMfMkLI0xd/0SLESRaDyAxeBfxRJbV2A5MsFz19WxMNwmoKfGsOLgwcR5550XbxEqJTVqQPPmRZeLJ94N7vF4DhqS3Q3uqdwU5gb3b93yeDyeJGHs2LGMHTs23mJ4ygHvBvd4PJ4kYefOnfEWwVNOeMva4/F4PJ4Exytrj8fj8XgSHK+sPR6Px+NJcHzM2uPxeJKEY44p6LX0nsqOV9Yej8eTJJxzzjnxFiExyc6GrCzYt8/ec1mjhi0QHs93XhYTr6w9Ho/Hk3zs2GFv6li5Etasge3bTWlXqWJLlDVsaCuhNG8Ohx9uCjyB8cra4/F4koTXXnsNgN69exdRMonZudNeWD1nDqxeDbt2RS/344/w3XemtI86yl7D1bJlfN+DWQheWXs8Hk+SsHfv3qILJTOrV8PMmfZmjj35XqAXnc2b7W/ZMkhPh/btE2tRcIdX1h6Px+Op/CxdCp98AqtWlez4LVtgyhTYtAk6dzaLO4Hwj255PB6Pp3Lz448waVLJFXXAvn3w7bcwbZq9LzOB8Mra4/F4PJWXrVvhs88siawsyM21ePecOaa8EwTvBvd4PJ4k4fjjj4+3CBXLvn2mVBeX8dt59+2DWbMsUzxBnl33ytrj8XiShA4dOsRbhIplw4bys4C3bDGXePPm9qhXnDko3OAicr+IrIy3HGWFiPQRkVwRietkS0TOc3Ic7b5/KiKj4ilTWVJR4ywi1V07fcqznYg2k+ZciUgnEekUbzk8cWD5cti4sfzq//FHWL++/OovBgeFsvZUGN2Bm+MthOfgQUSqARlAWrxlSQRGjhzJyJEj4y1GxbBjByxYYDHm8mLLFlPYCYB3g3vKDFX9Kd4yeA46Ks96kZ6yJSurWBnb4zMzGTpzJquzsmiWmsqA9HS6p8Uwx1u1ytzscV4sJSmUtYi0BYYCHYE6wCrgWVUdFlHuXuB2IAWYCNymqhvdvlzgBuBKV88WV8eQ0PEXAYOBdsAu4G2gv6pudvuXAW8CFwFNgV7AAGA60Ai4CtgDPAWMAZ4HzgSWO1kmuXpaAsOArkBDYD3wOjBAVWMKzojIla7t1sA615dHYqnfuWPvB8YDfwBmqOoFItIZeAxoDywAXolo81Ngpape7b6fAfwNSAdygUnAnaq6QkR6Ay8BTcJKXkS+BT5T1VuLOq8iUtuNZTeggZPpQVV9M1TfzcAtwNFunB9R1Zfcvlygr6q+ECq/EnhBVe+PMqYlGrco9bQA/gn8ErvOBkQp0w14ADjB9Xs08JCq7haRfwOnquppofKNgdVAT1V9U0TaAI8DXYDtwBSgn6qujWzLHd8a+DvQGagFTMOu7blu/0jsd7MB6APsBka5vu92oZClwKVuDNoCi4E/AicCg4DGwP+Aa4NzXpSc7pqa5cb7cswb+C7wJ1XdDgSrgLwkIn1U9dxo/fMkIVlZtoRoDIzPzKR/Rga7srMBWJWVRf+MDICiFfbWrdZW/fqlEre0VHpl7W7YHwOTgQ5ANnaDeEREpqjqLFe0OXA+cAFQF/g38DJwSai6R7Ebe19XxwMi8qmqThOR3wATMOVzHXAk8DQwSUTOVNUcV8fNwK+BjZjyGADcATwEnAxcg92ErwXuBBR4wsnS1NXxLqYILgC2YTfA4cBXwLgYxqQHdnO/15VvD7wsIjtU9ZkY6z/K/Z0K1BaRYzBl+6qTvR022ShIhtOxG/4r2ASpPnZTzhCRdphCGwH8LqjHKeeTgT/GeF4fdH37FbAZO29jROQ4VV0qIne6MrcDnwDnAf8WkQ2qOrGocYxCscctyrhUBz7EFNO5mAJ8LqLMRcB/setmEtAKu9YEuAKb5FwvIqKq6g67ysn0rog0w1zDbwD9MOU7BJguIm1VdUdEe0diE8pprm9g45YhIiepauAH7A68j52PY4EXgVRs3AOeAa4HVmLXyvvAt9hvookbp/7AgGLIeSt27aQDJwGvYb+bIUAL19btrr1wvxpgk7gwLfAkBT1uucUSzGLg63Xr2JOTk2fbruxs+k2dyusLFhR+cI0a8M47jHvnnZKKWiZUemWNWVzDMYtrG4CIDAbuxhRKoKz3AFeq6jpX5hZM0bZW1eBsvaqqo9z+QcBtmFU3DbgHmKiqg1xZddbhLOBC4AO3/YPAQnb1AMxX1Qfd9yeA+4AxqjrBbXsOGC8ihwFZ2E1nnKoud9U8KSL9XX+KVNbYje9NVX3YfV8kIqlArojUKkb9Q1R1iZPxYcyquklVs4EFInIUNsGJxp3YZOVGVc11dVwOLAGuVtVnReQNoDcHlP7vgTmq+o0bi+EUfl5bYUpviapucedsGrBZRKpgyu4ZVQ3qX+z6X2xKOm5R6IpZmq0DRSsifwBmhsrcC7yoqs+G5P4TMEVEjlbVL0RkITZ2g12Z3wOvqeoeEfkzsEZVbwnJfwU2gfwdMDJCppswS/kqVd0VKr8Em3ze7cptBXqr6k5grhvvp0XkrlBdw1X1E1fHKOwc/llVFwJzROR/brwAYpVTVXWg+3+RiFyL/S7BvEYAW6OEYW7Hfmueg5xIRV3U9kSk0itrVd0gIiOAniJyCpZocpLbHQ4yLA4UteMr99kWUyqEPlHVXBHZBgQ5++05cGMMynwtIlluX6Cso92kNfR/YC1khrYFK82nuP48A/QQkX6uP+2BZhH9KYx8Sl1VXw7+L0b94b60A2Y7RR3wRSEytAemBIraybBaRDLdPjAL8Qtn2a3ElM9jrmws5/URzNrdICJfYVboaKe4G2Oeii8jxmF4ITIXiKruKuG4RdIO2B6yiAG+xiaTAacCZ0Rkhwex2TbAMpx1DQwWkeOBM4AbQ8e3cddmmEPc8ZG0B2YFihpAVXeKyCwOnCtcmZ2h719g95DWQOBej+Vab1xMOTVi/1bMU1YUw8k/MWmBWfNJyYknnhhvESqMccOHw5gxEIPCPWP0aFZlRV5m0Dw1lXHduhV+8BFHwBVXlFDKsqPSK2sRaYLdkDdgMeRJmJWyIqJoZKw3uMHuDm3bTX6qRHxGUpUDcTM4oHjDRFtdP+oVJiJ1MOuwDjAWc4/PcNtiZS8WIy5t/ZF9iRyDwt4aUOR4qeqXIqJYbH8mcDjm4ozpvKrqdBdHPh9zcV8DDBKRSzgwGStuqmjU30Qpxy2SPGPjJobhsawK/ANzM0cSLNP0CvCQiJyFhXJmq+rs0PHTsByMSLYUJU+EHGG5Is938BvaV0gZQiGiaPXHImdhv8sCUdUtEfUEnq6kJT09Pd4iVBx16thfDHHrAenpeWLWALWqV2dALOOVmmp/cSYZHt3qBRwGdFTVh5xrOViBPfyDbiUi4VepdMJu5PNibGcOlgSzHxFJx+KSc0sieAFciFkcXVV1sKqOwSyJI4g98/UHLL4XlvUhEfmgFPXPBk4TkZTQttMLKT8H6Ozc0YEMzTHXdXi8XsJcnj2xMEMQhCryvIrIA0AnVX1HVW8DjseSmnqo6lYs4SpyHF4RkcAtvheoF9rXkANWXyRlcV7AxjHVxe2Ddttgk4CAuYCoambwh1nwj2L5FqjqamwC8zssXv1S5PFYsl9w/CbM0mxHfuYAp4dDBG5ycjp5z9UpkveZ8w7Az8D8mHufl+LKGY1yfG6n8rF3796D581bderE/Has7mlpDOvcmeapqVTBLOphnTvHlg1++OGQklJ0uXKm0lvWmKVVC7hSRKZhP/4n3L7wCKcAY0VkIJaZ/TTweiGxxUiGYnHlIZj11xJLppmFZbCWFcHiLb1FZIxr52GgBnn7U5SsE0RkBvAOcAoWv+tfivqfxxJ9XhSRv2Fu4MGFlH8c+Ax4XkSe5ECC2RosEz7gVSxp7zjMDR4Qy3lNA64Wkb6Ykj4TS+4K3PNDgYdFZAHm+jwPmxT81u3/HOgrIlMxN/TfyeuODlMW5wUsG3oWlvB3E5Y490/yeloeAf4rIvdj2eYtgBeAZRHZ3C9h1ncKziPhGIG5xF9312sulsV+MtEnlkH5N1ybYMlbNcmbRNgSeE5EHsPc1A8AT6tqlgs7FJfiypkPl4W/CzhBRA5X1cRYwSJOBO+z7tOnT3wFqQjq1YNmzey1mDHQPS0tNuUcpmZNe9d1ApAMlvU47Af+KDbDH47dwKaR16r6FvgGU6zjgI84EOMrEmfZXY49JjQHUzJTgPMj4rilQlVnYIlRt2Ax9JHAVCy7OyYfl6q+i8Uz+2JW9lDg/1R1REnrV9UV2KNGR2HjOAzLcC+sH+djjx7NwmLLizFLeWuo3BrsXOzkQNwfYjuvf8ayvEcBC50896jqSLf/GUyh3OfG4TbssaH3QsevwZT2+66uqHH4sjgvrp59WPb6QswynohdS5tDZcZhWd+/Ab7HFPHHWDZ2mHcw78C7qropdPxS4BzMWv/MyZkN/CLkuQjLtBzzGqVgk5pPMcV5diiZDiwM8TN2Pp8BniTKY2exUlw5C+FhLEnuo5LK4qmEVKsGbdpArRLljMZGkybQtGnR5SqAKrnlufqLx+NJCsSes05T1Uq9rKe458EnT55MixbJ9xRXsHrZQWFZA+zcCe++C/NLGokphOrV4YIL4Iwzyr7uAli5ciVdu3YFOEZVl+URp8Kk8Hg8Ho8nFnJzbSGSHTtg794DGd9Vq9pzz3XqWNJX7drQoQOsXQubNxdeZ3E57jiz3BMEr6w9Ho/HE3+ysmDdOlvoZM0aU75ZWbB7d15lnZJiirphQ3NRN24M550HX311YGnQ0nLYYXDWWTEnsFUEXll7PJ4iUdU+8ZbBUzQnn3xyvEUoHjk59tasH3+EefNMSf/8c+HH7NgBP/1kx3z3HRxyiD0L3aSJZW5v2WJv4yppVnyjRtC1a8IklgV4Ze3xeDxJQqVS1j/9BLNn25uzNm6MaXGTqPz8synnpUvte/Pm0K4drFgR83KkAFSpAkceCZ07Q6tWJZOlHPHK2uPxeJKEnTttgbnatfMtS584ZGfD4sUwfbop2bJKcq5a1epasMAe5zrpJHORZ2YWbWU3aABt28Ipp5hlnYB4Ze3xeDxJwtixY4EEzgbfuhVmzYJvv7V4dFlTpYq9HWvPHmunTRuLPc+fD5s2HZgYVKtmce969SAtzSzpI46w5LUExStrj8fj8ZQ/mzbBtGnw/fcld3nHSk33SofFi00BX3KJWdc7d5pCP+QQSx6rUyfur76MFa+sPR6Px1O+/PQTTJkCP/xQdm7vWFE1S7prVzj00IptuwxJhhXMPB6Px5OobNsGU6fGR1GDtfnDD2bVb9tW8e2XEV5Zezwej6d82LfP4tPffx8fRR2Qmwtz5lj2eVk8hx0HvBvc4/F4koTTTy/sRXhxYOlS+Prr8o9Rx0JOjiWdNW+ekI9mFYVX1h6Px5MktG3bNt4iHGDLFns8K5Fcz9u2mUyNGtnjWpUI7wb3eDyeJGHr1q1s3bq16IIVwdy5BxYqSSSWLLHV0ioZXll7PB5PkjBhwgQmTJgQbzEs+3v+/MRwf0eSk2MJZz/9FG9JioVX1h6Px+MpW1atspdyJCrr1tkqZ5UIr6w9Ho/HU3bs2mVu5uzseEtSMNnZJmNRLw1JILyy9ng8Hk/ZsXGjWdaJzqpVJmslwStrDyJyv4isjLccZYWI9BGRXBGJ+rSDiIwUkc8qWi7X9lEiclU82i4uRY1jjHWc6+pIK0vZPAnMli3ls+53WbN9u8laSfCPbnk8FcsoYDHwRrwF8SQfZ599drxFgPXr47sASqzk5lrsOpEedysEr6w9noqlSrwF8CQvIhJfAfbsgbVr4ytDcVi3zl7wkcBv2wrwyvogQETaAkOBjkAdYBXwrKoOiyh3L3A7kAJMBG5T1Y1uXy5wA3Clq2eLq2NI6PiLgMFAO2AX8DbQX1U3u/3LgDeBi4CmQC9gADAdaARcBewBngLGAM8DZwLLnSyTXD0tgWFAV6AhsB54HRigqrGuJVhdRB4BrnNjMhm4UVXXuDaaA487WfcBnwP9VHWR258CPAT0AJoDO1wdN6nqehE5GlgK3AvcCuwFtgEnAh1F5FpVzae4ReR+4Fw3TgOBeti5+At2Di8HsoDHVfWx0HG9gTsBAX5y4zFYVX92+3OBB4FrgFrAecC7wL+BTq7N1cAjqvqvCLGuEZGBQAtgHnCLqk539R6CncOr3f4lwD9U9YVogy4iVYFbgD8Bx7o2/wUMU9UcV6YV8DTQGdjuzsOf3Xh/C8wGuqrqlFC9LwBHqer50do9WNjoYrCNGzeOjwBZWWXuAh+fmcnQmTNZnZVFs9RUBqSn0z2tjKIq27fbXyV4wYePWSc5IlIb+BhTrh0wZTEGeEREwmsTNgfOBy4AugGnAS9HVPcoMBI4AXgReEBEurh2fgO8jyms04He2M12krtBB9wM/BVTFtPctjuAFcDJwDPAA66uJ50cyyNkeRdT7hdgymkYpqgui3lgbBLQ2Ml4ifv+uOtLHeBToBpwjvvbCHzllDiuzSswZX8cpgTPxSYrYa51fe0BdAFmAGOxyUpBnO3q+gU2gbkcU5I/YOPxCjBMRNo4eW8FXsLOzUnY5OD3mMIOcxPwO+A3wFy3bRDwJTb2TwPPOsUfpi82sToN+Bn4r4gEE43X3P47sUnav4ERInJbAX17FJs0PAK0df8PBB5zfamNXUM1sHNzBdAHU+yo6nfA19h4Ezrmd8B/CmjzoGHixIlMnDgxfgJkZ5t1XUaMz8ykf0YGq7KyyAVWZWXRPyOD8ZmZZdPA3r2JnbUewlvWyU8dYDhmBW8DEJHBwN3YzXWWK7cHuFJV17kyt2CKtrWqLnBlXlXVUW7/IOA2zMqeBtwDTFTVQa6supv+LOBC4AO3/YPAQnb1AMxX1Qfd9yeA+4AxqjrBbXsOGC8ih2FW5avAOFVd7qp5UkT6u/6Mi3Fc1gM3OEtcReQN4GK37ypMkfdS1b1Ohusx5dkXuB+YCYxX1anumOUi8pGTIcwIVd2/XJKI7AV2qWphvsLqQF9V/QlYICLfu2OGuTqGAv2BtiKyAFN2z6nqk+74RW6CNE5E2qjqfLf9VVWdGZIFYLKqPuA2qYichXlXXgvJc4Oqfu+OeQwYDzQVkXpAd+ByVX3LlX1CRI4FBorI0+FOufK3Yh6QYPKVKSKHAg+LyAOuviOA00Nend7Ad6GqXsQmKzep6k5skpYD5FkNREQaAA0ixrYFnvIjN7dY8eoeRUwsvl63jj0RC6vsys6m39SpvL5gQQFHwbhu3WITICencsTX8ZZ10qOqG4ARwFUi8qyIfAwEmd/VQkUXB4ra8ZX7DGdf7P91qGou5tZ1b3mnPQcs5aDM15hybR/avCSamKH/d7jP8NR5l/tMUdVdmPXdSUSeEpH3XSZ7s4j+FMWSCJf5Zsw9DHAq5n7eLCJZIpKF9bUl0Mb1bRRQQ0SGish4EfkBU/KRMkTrb1FsdIo6YAcFjAdwGNCEiLEHgklEUWM/NeL7F+SfcCwM/b/ZfdYK1R2t7SOAwyO2t8Ys5mjla7j9pwKLAkUNoKpzgPAamq9j4xx4UvoAowOXf4jbsVBE+C8DT6UhUlEXtT2Z8ZZ1kiMiTTA35wYshjwJswpXRBSNjPUGSmd3aNtu8lMl4jOSqli8NmBXlDJ7o2yL+mt0LuppmMdgLOYen0F+BVAU0WLbQR+qYsrxkihlspwcz2Eu2pcxt/yDQD/g6Ijy0fpbFDGPB4WPe2RdsYx9NfKPTUFjVZy2g2OKKp9dSDkAVHWriLwJXC0inwK/xDw7kQzHQgNhWuAVdvlRpQpUjd0GLMoCPmP0aFZFiYE3T02N3XoujKpViyVvPPHKOvnphVlfx6vqHgARCSyn8E2xlYjUVdXt7nsnIBeLlcbCHCwmG056SgdqcyA+WhZciFlfLVR1lWvnUMySK6tM67mYtbZNVde7Nqpjj1uNFZHJwI3A1aq6313sYshFLYlUpj43VV0nIuuwsQ+HAM5xn0WNfXrE9w7ANzE2P8d9dsFc4+G2V3PACg+YjynkLhwIvwTld2MTpO+AP4lII1XdBCDmr68fUdeL2MTzOuAHVZ0VsR9V3YLlauwn7tnSyU7NmpCSUmbVDUhPp39GBrtCceVa1aszID3ysi0hhxxSKTLBwSvrg4EVmMvyShGZhiVkPeH2hX9VKZgiGoglbz0NvK6qsbpxh2Jx5SFYvLMl5q6eBUwp7MBiErjwe4vIGNfOw5gbtazuEqOwDOc3ReQu7IZ/LxbTHoS5xLcCvxaRGcAhWCz2VCz5qTC2A8eIyFGhmHtpGYrFcBcD72Fu7CeBt1R1YaFHwhUiMh34EPgtFjP+dSyNqup8EXkbeFpEcrCJXTcsrn+XquaGlaOqbhOREcB9IrIRc7l3wsb0Wbd/NJZgOEpEBmDXbhD7Dk90pgI/YvH6e2OR92CgS5cu8RWgbl2oX7/MVjALsr7LLRu8Xj2TuRJQOex/T2kYh2UuP4pZNsMxq2Qaea2qbzGLaoo75iPMeowJlwx2OXaznoMlgU0BzlfVMku3VNUZWPb4LVgMfSR24x5NfiuxpG1sxay/9ZgSmwkcifVlvks6+x0Wv57jytTGFEdrEUktpPqnsdjsfBeiKAt5h2OThT9jGePDsezwnjEc/gpwKfA9lrl+haq+V4zme2IehxGYFX8d9gjckwWU74ddj0OcrAPd//1cX3Zjj8vVxMI3b3LAlb0/zdjlTIzEDI5RxZA3qTn22GM59thj4ydAtWrQtLAHHYpP97Q0ZvTsycq+fZnRs2fZKWqAZs1M5kpAldxKkgnn8XjKFvfc+yhV/b94yxLgnk8/PuKJgWbY2gBdVDUjtP15oJGq9ihm/UsnT55MixbJlxi+1i1I0qRJmcwDS8aCBfDf/8K+WJc8iBPVqsEVV0AChUZWrlxJ165dAY5R1WXhfd4N7vF4EolawPvOBf4m9ujVQ8AizNJGRM7HvBPXYs/aexwffvghAH369ImfEPXrQ4MGsGlT/GSIhQYNTNZKgneDezyehME9E34VtqjOPOATLAP/vOCZd8zV/nfg76pa3KcAPOXNYYfBUUfFW4qiOfpoiNdKbyXAW9Yez0GKqh4dbxmioarjKGRxG1XtVYHieIpL9epwwgnwww+J+77oWrWgTRuTtZLgLWuPx+PxlC1NmkDz5kWXixfNmpV5Ilx545W1x+PxeMqW1FQ46aQyfea6zEhJMdnq1Im3JMWi8vgAPB6Px1MoLpM4MTj+eFi2DL6JdY2dCqJtW5OtkuGVtcfj8SQJLVu2jLcIBzjkEEhPh9WrE+cd102awOmnm2yVDO8G93g8niRhxYoVrFgRuex/HGnaFM46KzGU4yGHwNlnV7pYdYBX1h6Px5MkTJ48mcmTJ8dbjLy0aQMdOsR3De6aNU2G1q3jJ0Mp8cra4/F4POVHSoq5w88+25RmRVOzpln36emJmfAWIz5m7fF4PJ7ypVatA9b1F1/ArpK8ObYU7aanJ4YrvhR4Ze3xeDye8ueQQ+DMM6FhQ5g+vczezFUgzZubNX/88fGx6MsYr6w9Ho/HUzHUrGmPTh1+OMyaBXPmlP0qZ7VqQfv2cNpp1k6S4JW1x+PxJAkXXXRRvEWIjcMPh65dbX3u77+HlSth+/bS1Vm3LrRoYYr62GMrdXw6Gl5ZezweT5IQ11djFpeUFFtDvFUrew570SLIzIStW2OPadeqZW/OSkuD446z56iTTEkHeGXt8Xg8ScKSJUsAOPbYY+MsSTFISbG3dB15JJxyilnYW7bAmjWwYQPs2QM5OVa2alVzpR92mD0v3aCBWdSHHgpVqpS9bDk5kJUF2dn2/uvateP2CJpX1h6Px5MkTJtmbwytVMo6oEoVaNTI/gBOPtni2dGUdXlndv/8s1n7CxfCihWmrKtWNdlOOMFeBFKvXvnKEIFX1h6Px+NJTA45pOIfudqwAT7/3FzyWVl5961aBQsWWGy8Uyc45pjyseij4BdFOQgQkftFZGW85SgrRKSPiOSKSNTJpoiMFJHPKlou1/ZRInJVMY95SESWlZNI0do7141fWkW1WV6ISKqI3BJvOTxJwsaN8PHHMHt2fkUdsGcPLFkCH3xgnxWEV9YeT9kyCqgkKblJwd3AnfEWwpME7N1rz38vXBhb+Q0bYOpU2LSpfOVyeGXt8ZQtFeMT8wT48faUDevWweLFxTtm1Sp77KwC8DHrJEBE2gJDgY5AHWAV8KyqDosody9wO5ACTARuU9WNbl8ucANwpatni6tjSOj4i4DBQDtgF/A20F9VN7v9y4A3McuyKdALGABMBxoBVwF7gKeAMcDzwJnAcifLJFdPS2AY0BVoCKwHXgcGqOq+GIeluog8AlznxmQycKOqrnFtNAced7LuAz4H+qnqIrc/BXgI6AE0B3a4Om5S1fUicjSwFLgXuBXYC2wDTgQ6isi1qhpVkYjIDUB/V+9HwMqI/fWBR4HLgJrA126cZ7mxWQb8VlXfDR3zBHCeqrYTkSrAXcCfgCbAQuBRVX2tAHmqAre48scCq4F/AcNUNSfU16ux85kGzAHuVNUMV8dITHGudPXUwLwMfwNGAOdj5/HeQI6i5BSRc4H/Ab9y43Gck+NuVX1bRO534x9cv79Q1U+j9fFgoVu3bvEWofKyeLFloUcwPjOToTNnsjori2apqQxIT6d7mosg7dsH8+bZKmm1apWreN6yruSISG3gY0y5dsCUxRjgERE5PVS0OXbDvADoBpwGvBxR3aPASOAE4EXgARHp4tr5DfA+prBOB3oDnYFJ7mYfcDPwV+A8YJrbdgewAjgZeAZ4wNX1pJNjeYQs72LK/QJAMMV9J6a8YuVMoLGT8RL3/XHXlzrAp0A14Bz3txH4yilxXJtXYMr+OOAa4FxsshLmWtfXHkAXYAYwFpus5ENEemLKazhwEvAVcFNofxVsbI4DLnVyfwl8LiKnqOoK7Bz0Ch1TDegJvOQ2/Q07D7dhE6sngWdFZH87ETwKPAg8ArR1/w8EHosoNxz4O3AqMB879+G491XY5OpMoB/wZ2yi8RZwCjAVeEFEGhRTzkddmbbAEuAVEUl18g3HJghNgS8K6N9BQ+PGjWncuHG8xah87NljWd8RjM/MpH9GBquyssgFVmVl0T8jg/GZmQcKbdoE27aVu4jesq781MFuWM+q6jYAERmMxfLaAbNcuT3Alaq6zpW5BbvZtlbVBa7Mq6o6yu0fhN0gO2JK9x5goqoOcmVVRHq7+i8EPnDbPwgsZFcPwHxVfdB9fwK4DxijqhPctueA8SJyGJAFvAqMU9XlrponRaS/68+4GMdlPXCDs8RVRN4ALnb7rsIUeS9V3etkuB74BdAXuB+YCYxX1anumOUi8pGTIcwIVZ0X6u9eYJeqri1ArtuA/6rqM+77wyLSEVNEAL/EJl2Hq+oGt22gK/MXoA+mlF8QkVRVzcImNY2AUW4i8lfg96o60R2/2FnH/bGJwn5EpB7mGRigqsGEKVNEDnWyPRAqPkxVR7vj+jpZb8SsY4DtwK1uzDOdZ2OKqr7kjhmOTXqOE5EfiiHnfao6xdUxCLvm2qnqdBHZAeyLNt5uUtAgYnOLyHLJhKoC+393Bz09evSIreC+fRaDjlj69Ot169gTPDbm2JWdTb+pU3l9gbttVqsGkybFvBjLuHGx3sLy4pV1JUdVN4jICKCniJyCuShPcrurhYouDhS14yv32RYIlHXwiarmisg2zA0L0J4Iq1JVvxaRLLcvUNbR0iM19P8O9xmamhIsV5Ti+vMM0ENE+rn+tAeaRfSnKJZEuMw3A4Gf6lSgHrA54qZ2CNDG9W2UiJwnIkOB44HWmJU/PbKdYsgEpuz/G7HtCw4o61Pd59II2VKcfAATMGV2GTax+T3wnnPPp7tyrzjXdEB1IEVEIn11rTGX9bSI7VPd9tbAutA2AFR1r4jMIu/kZXHEmO+kgPOMeW9ilXNB6P+t7jOWNzPcjk0MDxqmT7fL0yvrYlLA41eRijrq9ipVKuTxLa+sKzki0gRzk27AYsiTMKsw0qcTGesNFN/u0Lbd5KdKxGckVbF4bUC0dQL3RtkW9VfgLMNpmMdgLOYen0F+ZVIU0WLbQR+qYkrkkihlspwcz2Fu8Jcxt/yDmGv36IjyJXnXX+RYhsenKqbkTo5y3G4AVf3ZeQp6i8hbwG8xN3hwPJibfG5BdRQiS1iOSNkiz2M18o5zzOeZ4slZ2HVZGMOxsE6YFkBGDMd6koBiWbH/+59ld4c4Y/RoVkV5hKt5airjgvyAtDTo3t1WNytHvLKu/PQCDgOOV9U9ACISWDvhG1orEamrqsFq+Z2AXGAesTEHi8nuj2M6K6420W+2JeVCzLpsoaqrXDuHAkdQdpm/czF38jZVXe/aqA68AYwVkcmYe/fqcFKWiLQBinpFUG4R+2djbu7HQ9vCuQVzsTGtrapzQm3/CzsHgfv8JSwprg/mfn7PbV8AZANHqupboeNvBU5U1T9FWF3zMSXbhQMhE7A4/m5sUnOo25YOfOvqq4nlG7xEySiunAVR4Hir6hYsl2M/3uL0FEhamj1fvXXr/k0D0tPpn5HBruzs/dtqVa/OgPR0+1Ktmr1FrJwVNXhlnQyswNy7V4rINMxV+4TbFw6ipGCKaCAW33waeF1VY3XjDsXiykOA14CWmOKYBUwpdS8OEGRG9xaRMa6dhzGXbFmt0D8Ky2p+U0Tuwm7o92Ix7UFYVvdW4NciMgNz196KTSK+LqLu7cAxInJUKOYe5lFsHO/CEq9+hbmzV7v9H2IKfbRTXD9i2dJ/wGLTAKjqDBFZiCVp/VtVs932rc4rMERENmMKvbNr9+FIYVR1mwuj3CciGzGXfCc3Ds+6/YGyHiIiq4HFbrzqAM8WMR5RKa6chbAdaCimhZerahm/b9Fz0HDEEbYi2ezZ+zcFWd8FZoM3bWqrmVUAPhu88jMOy1x+FLOShmOZ3NMwSyjgW+AbTLGOwx4ZujHWRlwy2OVYJvkcLFY6BTg/UBRlgarOwLLHb8Gsr5FYrHQ0eftTmja2Ypbkekw5zgSOxPoy3yWd/Q6LX89xZWpjGdKtXSZyQTyNxXnnuxBFZNtvYzHmPwLfA78B/hHavw/L2p+BZfXPwbLQLwuSrEK8BNQlv3X7V1fnEOyaGITlGwwhOv2wa2gI8IPr5xC3PcwIV+5rLIfgHFVdTckprpzRGINNWOcQPazh8cRGzZrQoYO9BSxE97Q0ZvTsycq+fZnRs+cBRd2oEXTpAhWUfV8lN7cor53H4zmYCT1nfb6qfhJncUpF0JfJkyfTooIsoopkq3Ph1q9fP86SVGLWrYOMDHvuOtqrOmvUsBd5dOpkrvMyTC5buXIlXbt2BThGVZeF93k3uMfj8SQJXkmXAUccAb/6lb11a8ECW6Vs716LTzdsCCeeaMq6YcMKFcsra4/H40kS5s61XM+2bdsWUdJTKLVrw7HH2nu2t22z57CrVoU6dWJ+nrqs8cra4/EUinPH+TW4KwGzZllCv1fWZURgTScAPsHM4/F4PJ4Exytrj8fj8XgSHK+sPR6Px+NJcLyy9ng8Ho8nwfEJZh6Px5MkXHHFFfEWwVNOeGXt8Xg8SULtClij2hMfvBvc4/F4koTZs2czO7S2tSd58Mra4/F4kgSvrJMXr6w9Ho/H40lwvLL2eDwejyfB8cra4/F4PJ4Ex2eDezwej+fgYfNmew3m9u32go6aNW397yZNoFateEtXIF5ZezweT5LQu3fveIuQuKxZA0uW2Gsv1661114G1KkDzZtD27Zw9NFQr17cxCwIr6w9Ho8nSahRo0a8RUg8cnNh4UL49FNT2NHYscPKLF4MrVvDOefA4YdXqJhF4WPWHo/HkyTMnDmTmTNnxluMxCIzEz76qGBFHWbfPpg3Dz75BDZtKn/ZioFX1hWIiNwvIivjLUdZISJ9RCRXRKJ6aERkpIh8VtFyubaPEpGrCtmfVOeiPBCRT0VkVCnriNs1cDAyb9485s2bF28xEoeNG2HqVPjpp+Idt2gRfPklZGeXj1wlwCtrT7IyCrgo3kJ4PJ44smIFrF5d/ONyc80lvn592ctUQnzM2pOsVIm3AB6PJ47s2GEu7ZycfLvGZ2YydOZMVmdl0Sw1lQHp6XRPS8tb6KefLCGtWbMKErhwvLIuQ0SkLTAU6AjUAVYBz6rqsIhy9wK3AynAROA2Vd3o9uUCNwBXunq2uDqGhI6/CBgMtAN2AW8D/VV1s9u/DHgTsyybAr2AAcB0oBFwFbAHeAoYAzwPnAksd7JMcvW0BIYBXYGGwHrgdWCAqu6LcViqi8gjwHVuTCYDN6rqGtdGc+BxJ+s+4HOgn6oucvtTgIeAHkBzYIer4yZVXS8iRwNLgXuBW4G9wDbgRKCjiFyrqkUq7sLOnYgc5r7/SVX/Ezrm/4BrVPV4EakC3AX8CWgCLAQeVdXXXNlzgf+5MvcAq4FTVTWUkrr/3A0HOgMXY+f/PuB74J/ACcB84HpV/cYd0xC4H/gtcAQwDxiiqm+7/X3c/vHAH4AZwN+dPJcBj2HXyVfAraoa9qOmisgLbvyrAm8BN6vqdld3a1dXZ6AWMA27FucWMM7NXfnzsWtqJjBQVT8LlekFDAKOAeYArwHDVbWKiDzhZD5GVXNd+drAGuyaeC1au56DkE2bosapx2dm0j8jg13Oxb0qK4v+GRkA+RX2okVw0klQt265i1sUXlmXEe6G8TGmSDoA2cAfgUdEZIqqznJFm2M3qguAusC/gZeBS0LVPQrcAvR1dTwgIp+q6jQR+Q0wAfgbpgCPBJ4GJonImaoaTCNvBn4NbAQWYMr6DkzxnQxcAzwAXAvcCSjwhJOlqavjXUxBX4ApwEsxRfIVMC7GoTkTUx6dXb1jMOXcS0TqAJ8Cs4FzMGV9B/CViLRT1VXYZOG3Ts6lQHvgJWyyckuonWuB8zBFmwl8ACwD/lKUgLGcOxF5Bxuz/4QOvTb0/W9Ab2zc5wNdgGdFpL6qjggd81vgLCA1UlGH+Bs2mQsU+wjs/NwObABGum1niUg14CPsWvojsMLJOUFEfquq77g6j3J/pwK1gcZu+xNO5hXYWP9PRCSY+AG/wZTradgEaAywEhgoIkdiE8Bp2DUC8CCQISInqeqPEeNcF5uMrQIuB7a7Pk0WkY5unLsBr7p+vwX80skY8KI75hzs2gHoDuRik5Fwew2ABhFj2wJPpadHjx5FF9q5M6ob++t169gTYW3vys6m39SpvL5gQd7CNWrAW2/ZZyGMGxfr7bDkeGVddtTBFNmzqroNQEQGA3djFnCgrPcAV6rqOlfmFkzRtlbV4Ep5VVVHuf2DgNswi28adhObqKqDXFkVkd6u/gsxJQXwQWAhu3oA5qvqg+77E5jFNkZVJ7htzwHjnSWZhd00x6nqclfNkyLS3/Un1qtzPXCDs8RVRN7ALEYwC78x0CtQXCJyPfALbKJyP2Z5jVfVqe6Y5SLykZMhzIiwRSgie4Fdqro2BhljOXcvAu+JyFGqulxEOmKW38tu0vFX4PeqOtHVudhZ/f0xxRrwaOA1KISPVfXfTo5ngOuBp1T1E7ftFWzSBTbxSwdOCyxt4F4RaY95G94J1TtEVZe4Os512/qr6vtuW29MaffCrHiAb1T13lCfJgGnu+83AbuBq1R1l6vjCmAJNgG4O6JfvTGvw9khz8oNTv67MG/SXdj5DrxRC0XkeGx8UdW5IjID+D0HlPW1wOhAhhC3Y9f4QUOfPn3iLULCE6moi9qeKHhlXUao6gYRGQH0FJFTgDTgJLe7Wqjo4kBRO75yn20xC5jQJ6qaKyLbgJpuU3vMqgy3/bWIZLl9gbJeEk3M0P873GdmaFtws0tx/XkG6CEi/Vx/2gPNIvpTFEsiXOabMXcpmJVXD9jsJhMBhwBtXN9Gich5IjIUOB5oDQhm0eVppxgy5SHGc/cRZhFejVm+fYBJqrpaRNKdzK+IyMhQ1dWBFBEJL4sUi5yxnKcU9397YEdIUQdMxazcMNHaDiZBqOpPIqLknQgtjCi/GTg01PassJJU1Z0iMsvti6Q9kBkoalc+R0QysFAL2DURqWCn4ZS140XgUTfRbYxZ3/eSn+GYFyJMCyAjSllPJSImS3bJEhgzBnbvzrP5jNGjWZWVla9489RUxnXrlndjs2ZwxRXQoEEppC0bvLIuI0SkCfAl5qZ8G5iEWYUrIopGxnoDZRC+onaTnyoRn5FUxeK1AZFWBhH7A6JOJ521OA2zOsdi7vEZbltxiBbbDvpQFVNCl0Qpk+XkeA64wrX/LqaA+gFHR5SP1t+YiOXcOaUyErhaRB5zMv0x1A8wizRarDZ8PmORM+bzROzXQ0FtR5apRt5zVtj5K07bsZbPLqRcwBuYa/zXmHdjvqrOiCykqluwmP9+IiaFSccXX3wBQIcOHeIsSQJQvz40apQvG3xAenqemDVArerVGZCenr+OZs0SIl4N/tGtsqQXcBjQUVUfcq7lhm5f+ObTysXuAjph8bZYH46cg8VD9+Msu9pEVxQl5ULMyumqqoNVdQywFUtgKqtM67lYzH2bqmaqaiYWZx4KdBGRRsCNWNLTX1X1JVX9FrO6i5IhtxhyxHruXsKs+tsx5RK4mBdgSubIoB+uLxcDd4byCMqDOUAdETk1Yvs5xHY97L9DiUhj4Dgg0kovrO3Tw54DN8k7vYC25wBpItI0VL4qls8QlP8Oy3MIk+e7C1X8F4t798DOiwdYuHAhCxdGOkMOUho1guOPz7e5e1oawzp3pnlqKlUwi3pY5875k8tq1YI2baBacRyJ5Ye3rMuOFZh790oRmYbd1IPEmJRQuRRgrIgMxDKznwZeD2KJMTAUiysPwbJkWwLPYHHVKaXuxQGCBUN6i8gY187DQA3y9qc0jMIS394UkbswK+heTMkNwpLatgK/dnHKQ7CM71OBr4uoeztwTBBjLqJsTOdOVZeIyP8wN+2/VXWP277VeQCGiMhmLImqM5Yo+HCRo1A6JmHKdZSI3Iqdt2uAX2FZ00XxtIsbb8HkXYdZrrEwAptMvSEi97ttQ7CQzfNRyr+G5VyME5E7sfP7V2y8/+DKDMVyA+7AJkOdsJyNSF7E+l4N66vHk59WrWD2bNiyJc/m7mlp+ZVzJM2b28s9EgRvWZcd47Bs2kexbODh2A1lGiHrBfgWu7lOccd8hN3wYsJZfZcD3TBL5VVX1/mqWmbL7Ti34h1YxvUCLPY3FRhN3v6Upo2tmJdgPfAh5no+EuvLfJd09jvMkp7jytQGBgKtRSS1kOqfxuLb852buzBiPXdgVlwt8maFgymdf2DKaj422RjsvpcbLh/gAiyG/waWWd8V+LWqvhVDFf/CQgyfY27yc1R1R+GH7G97OXb+UrA48KeYR+PsaBMkZxF3xh5b+xDL1zgWOFdVZ7oyH2KPLt6MeZtuAJ7DEjPDdWVgE5P3VDVxVq7wJBZNmkB6ur1ZqzgceiicdZa94CNBqJKbWxxvocdzcCMi9wCXq+rpRRZOYELPfR/nXPYJgYicA6xVVQ1tGwj8UVVbhbbVwpR+OAM/lvqPBpZOnjyZFi2S7ymukSNHAj4rPA+7dsFXX8H06fmSzaLSqBH88pdwwglQpWLXVlq5ciVdu3YFW0dgWXifd4N7PDEgIp2weO5dxPDstqfEXIiFXvoAi7E1AW7HPf7mFoD5JRar3sCBpx88+LduRaVWLTj7bEs4mzPHFkr5+ef85erXh5Yt4bTT7DWZFayoi8Ira48nNrphsdNXsVi7p3x4AHsC4VUs6W8Flj8QPHddHQtRbAJ6FmMlvYMC/z7rAkhJgVNOgeOOM2U9fz5s3WpLkVavblnfaWlwxBHFd5lXEN4N7vF4DhqS3Q3uiZHcXNi7116JWaOGKewEwLvBPR6P5yBg6lRb4+acc86JsyQJTpUqCWtBF4TPBvd4PJ4kYenSpSxdujTeYnjKAa+sPR6Px+NJcLyy9ng8Ho8nwfHK2uPxeDyeBMcnmHk8Hk+SULt27XiL4CknvLL2eDyeJOGKK66ItwiecsK7wT0ej8fjSXC8svZ4PJ4k4ZNPPuGTTz6JtxiecsC7wT0ejydJWLlyZdGFDmb2udVpE+Qd1cXBK2uPx+PxJCf79sGGDbB2ra0Jvnmzba9Xz9YDP+IIOPxwW3I0wfHK2uPxeDzJx8aNMHs2qNr/0d6D0bAhtGplb9pq2rTCRSwOXll7PB6PJ7lYvBimToUVK6Ir6YDNm2HWLFi5Ejp1sndYV03MVC6vrD0ejydJqFevXrxFiD/LlsFHH8H69bEfs3YtfPKJKeoTTig30UpDqacQInK/iCRNVoOI9BGRXBGJ60RGRM5zchztvn8qIgfVe5RFZJmIPFSK42uIyB2h7wlxbisLbqyuL2UdpTqHnuLRvXt3unfvHm8x4seWLfDZZ8VT1JHHrltX1lKVCYlp73ui0R24Od5CVDKuAR4PfR8DNFXV7DjJ4/F4ypP582HJkpIfv3o1fPcd5OSUnUxlhLcwKgmq+lO8ZaiEVAl/UdVdwK44yeLxlDsffvghABdddFGcJYkDW7easi6tos3MhJNOskzxBKJIZS0ibYGhQEegDrAKeFZVh0WUuxe4HUgBJgK3qepGty8XuAG40tWzxdUxJHT8RcBgoB12Q30b6K+qm93+ZcCbwEVAU6AXMACYDjQCrgL2AE9hFtTzwJnAcifLJFdPS2AY0BVoCKwHXgcGqOq+GMYMEbnStd0aWOf68kgs9YtIH+B+YDzwB2CGql4gIp2Bx4D2wALglYg2PwVWqurV7vv1wG3AcUAO8C1wh6rOcPurA/e4No4AFBikqhML6NMyYDjQGbgYO0f3Ad8D/wROAOYD16vqN+6YXKCvqr4Qqmcl8IKq3i8i1YC/Y+fqcGAZ8ISqPhfLWEaRsQN2LaYDG4B3gXtUdVuUsn2Af4fkvM7tegmooarZInIx8KDr2w7gfeD20DV3J/BnoAWwBvgP8KCq5orI/W4sWoTavB74t6pWcd9ruvqvBuoDc4HBwbUYReajgaXApdg10hZYDPwROBEYBDQG/gdcG0zgRKS1G+fOQC1gGvbbmev2jwTqAbWBDsA/XJPnAR8Af8XuBW9jv5WtIbGOE5FJru7NwD9V9W8hmQv93Ubp4xnA37BzmAtMAu5U1RVufzXgAey6beDkWwmcpKrnisi3wPeqek2ozk7AVODooJ6DlbVr18ZbhPixebM9plUA4zMzGTpzJquzsmiWmsqA9HS6p6XlL7hxI/z0U8Ip60Ld4CJSG/gYu3F3wG4YY4BHROT0UNHmwPnABUA34DTg5YjqHgVGYjfGF4EHRKSLa+c32I1yMnA60Bu7OUwSkbCMN2M3lvOwGxLAHcAK4GTgGeyH/j7wpJNjeYQs72LK/QJAMMV6J3BZYWMRGpMewGhgLKZY+wGDROSWYtR/lPs7FbhDRI7BblrfAadgN95BhchwGaZAH8OUXFegJqZMAoZjk6cB2I10AjBBRE4upHt/Az7ElMR7wAjgBeBu4CzMUh1RyPGR3AT8DpukHQ88DTzrJiaxjGW4z+2x62OSK9sTO7+TRKRKZHnsOu3n/m/qvofra4yNyX+ANsBvsWvuMbf/UmAg8CdsQnQ3cC+meGNlJDa5vBo7r2OBiSJySRHHPYOdt5OBndj13Av4NRYOOQfo7+Q8EpuwVsOuuc7Y5C3D7Qu4DFPyp2ETFjCFebE77rdAJ2BchCw3A69iv9t/Ag+JyAWu7Vh/t7jyp2O/26XYpP0S4Egna11XbCg25rdiv4/V7v+AF4HuIlIntK0P8MnBrqgPerZsgV3RHWfjMzPpn5HBqqwscoFVWVn0z8hgfGZm/sI5OSWLeZczRVnWdbCb/rOB9SIig7EbVztgliu3B7hSVde5MrdgP9jWqrrAlXlVVUe5/YMwq7Aj9uO9B5ioqoGCUhHp7eq/EJtdA3wQtkpEBGC+qj7ovj+BWYNjVHWC2/YcMF5EDgOysBvPOFVd7qp5UkT6u/5E3qii0Q94U1Ufdt8XiUgqkCsitYpR/xBVXeJkfBizFG9y8dQFInIUNsGJxibMqnvVfV8uIv8G/uUUVyrQF7O0AyU1xFktdfJXt5+PVTWwRp8BrgeeUtVP3LZXgOIkC7XCLNalqroGeEZEFgAL3f4CxzJKXXcBk1X1oVDZnsASTHl9Gi6sqrtEZJv7f62TP1ykBeYF+tGdq+VOQQe/iVbAbmCZqv4I/Cgiq4AfY+m4iKRhE4p0VQ1+J/8QkZNcX94r5PDhoTEfhf0G/6yqC4E5IvI/7HoCmxDtBq5ybn5E5Ao3Ljdjv1WAzWGPRWgsrlTVVW7bzcAHInKiqs5z+58PXWd/E5G7MMU8idh/twF3Yl6jG1U117V5uZP1ahF52cl8l6q+6fbfhhkKAa9hv4vuwKvuN/c7zHOXBxFpgFnnYVpElvNUHnr06FHwzq1bDyx6EsHX69axJ8I9vis7m35Tp/L6ggX5D5g6FRo1KrCpceNiURVlS6HKWlU3iMgIoKeInAKkASe53eH12hYHitrxlftsi/04CX3i3IjbMGsQzFIaHNH21yKS5fYFP/pomQMa+n+H+wxPl4KpVorrzzNADxHp5/rTHmgW0Z/CyKfUVXW/5V6M+sN9aQfMjkh8+qIgAVR1moi0cZOe1pjl197troZZ9DWBLyOOu6+IvsUylilF1BHmn5hFt9K5Lz8GxoaulULHMoJTMZdsVpR9bYhQ1kWhqrNFZDTwroisAT7BwjcTXJFRmCt2oYj84GR/0ynuWDjFfX4aMUmogXmqChUv9H9B56Gx+789MCtQ1ACqulNEZnHgmgCzZiNZHChqR3DNtQMCZb0w7yFswVztQdux/G4JlZ8SKGpXfrWIZLp9bVzd00P7c0UkA/MyoKqbRWQCljz4Kqa09wFvRenf7djk3XOQE6moi9qeiBSqrEWkCXbD34DFoiYBMzG3c5jIWG+gmHaHtu0mP1UiPiOpCuwNfY/m49gbZVvUM+BcZ9Mw63Is5h6fwQGXeizsJbrlV9z6I/sSOQbR+hW008vV/Tp2Y/sXFqL4Z8SxhawGEJWYx7IQ9l9TqrrIWZjnYqGLS4ABInK9qr5IIWMZharAG8CQKPsKDlQVgqr2EpEHMFfw+djN/0sROVdVN7qQwdmYm/hC4C8i8qCqDi6gyvDvKXADn0t+5VxUbkS+86CqBZ2HsvztBL/bsHzRZC3u7zbyuILKZxdRLuBFzHPXDPg98LqqRru/DMdCEWFaABlF1F9paVSINZgMFGrR/vADjBsXNcHsjNGjWZWVf57fPDWVcd265a/r4ovhzDNLI2qZU5QbvBdwGHC8qu4BEJHA/Rb+QbUSkbqqut1974TdhOcRG3OALrh4oWsnHUuImRtjHbFwIWahtQi5/g7FErCKukEE/IDF+vYj9hzpaVhCU0nqnw38QURSQjed0wspPwAYqap9QzL82v1bBViE3fzSgW9CZf5HXldyadmLJS4F9TfkgMUXuDDXq+obmGV6t0tWuhK74RY4lqp6cURbc7EJyeKQC/V47NGse4Ct5KfAiYCInIW5gP+KWbLD3SToNaCpiPwSqK+q/wQ+B+4TkX9hiYyDsdBP3Yhqj4uQF6BZyA0e9C+HCIu0FMwBrhWRWiE3eB3s+nmh0CMhTUQahpLBAnfzNwUdEKXt4vxu5wCdRaRK6Bw2x0IOw7HrdheWGPp16LgzyTvZn4LlolyH5WvcG004Vd1CxEQpwsuRdFx66aXxFiF+1K8PdeuaOzyCAenp9M/IYFf2AedlrerVGZCenq8sKSnQuHH+7XGmKGW9AnNLXSki0zD36hNuX9gdmgKMFZGBWHLV09hsN9YH3oZiceUh2M2yJZZkMwv7YZYVweItvUVkjGvnYcw1Gat7dyiWqDUDeAdzd96OJfyUtP7nsSSaF0Xkb5j7vLCb+QrgbBE5DbsZ/RoIkrJSVDVLRJ7GkvjWYRndvbCb8R1R6ispnwN9RWQqprz+7j4DjgAGi8hOLHmuNTZegQegsLGM5HHMInpWRJ7CFOUIzIsR6aoN2A77Fcj8iH3b/r+9e4+Tqi4fOP5ZYLmKwi4ECBaL4IOhZCX2Ay0kvJCh5T2voUmWty5mpnkhbyload4yszbzkoqorWmACAgYiqUSKg+gLEHEZbktq+gK7O+P5ztyOMzMzu7O7swuz/v14jXsme98z3NmZ+c538v5HuBCEanGklpbLBG/h81K7wDcKiKbsZ6RPlgrOdFVPAfYU0SuwCbJHYp1zQKgqm+JyLPA3WEewb+xLtsr2TEzPRvuAc4H/iI2Qx2s96Et9rlKpyPwYDiGHtjf3BOq+m6G+67r3+1twGzgPhG5A5shfxs20/6x0H3/G2CciKzETubOxSY3zkhUErrG/4gl6bdV9Z84162b3ZwjSbJOzPrOaDZ4z57QvXtjR1tntS2KMhGbzTwB+7K7HWsRvcTOLaLXsbPxF8NrJmNfIBkJk8FOxGaSz8e6I18EjtQsLmARLmv6MZbYFmJdZDOxL9skp1hJ6yjDJl6Nxb5MbgauUtV76lt/mMX6VWyG+L+w9zxd6/ci7AtuBjY/YDQ7EkViP1eE/d+NtXKOA45V1dczOc4MfT/EMQebFfwCO4+1j8MS4Z1YQv0dlkBugPTvZXxHqjoX6xk5EGt1PYu1xI5I9PokMSXENofYBCRVfRtLniOwz+/LWEv8aFXdFibaXYPNyl8IPIH1DlwUXj8TSxYXY38bZ7HridCp2N/DveH4zgG+k2Zcvs7C5LivYCeDs7DPRA0wNDLJMZWV2LDWLGyI4RkiJxwZ7LtOf7fh7+NIbGb5a9iVE+8Ch0YuF7samy9wP9bjVBLiiv+OS7ETqj/gPlFWVkZZWVmuw8iNdu1g8GDo0CHp0yf078+rp53GirFjefW005In6sJCOPBAuytXnimoSbfIuXOuRZIk14nnA7HLEmer6trItinAclX9TmTbUOxStN6quq4O9fcFlk6bNo0+ffLq0LOitLQUgDFjxuQ0jpyprobp02Hu3PQ38EjlwANh1CjolO6imcazYsUKRo4cCVCiquXR53wFM+dcPrkMuCBcIrYJu/77q1iLPLEAzIFYz9HDdUnUbjfQti0ccgh88AHMn1+3hD1ggN15K0eJuja+NrhzLp+cjs01eAGboHo6cLKqTg/PD8C6wDeTfG6D29117QojRsChh2aWeNu1g4MPhqOOyrtVy6K8Ze3cbkhVx2FzCvJK6PpLeduoMM8hP5s+Ln906QLDh0PfvrBggd2vurLSusnBxqY7d4ZevWDQIOjXD9q3z2XEtfJk7ZxzLUTPnj1zHUL+KCyE/v3hM5+xtb6rqqx7HGwSWqdOUFRkLetmwJO1c861ELvl3bZqU1ho3dt53MWdCR+zds455/KcJ2vnnGshJk2axKRJk3IdhmsE3g3unHMtRGXlLrd2dy2Et6ydc865POfJ2jnnnMtznqydc865POdj1s4510K0xPXOnfFk7ZxzTaiqqoq1a9eyYcMG1qxZw0cffURhYSHdunWjuLiYoqIiunbtWq+6jzjiiCxH6/KFJ2vnnGsCmzdvZtGiRSxYsIBVq1axZcuWXcq0bduWHj16MHDgQAYOHEhxcXEOInX5yJO1c841smXLlvGPf/yDxYsXs23btpTlqqurWb58OStWrGDx4sUMHTqU/v3707p164z28/jjjwNwyimnZCVulz88WTvnXCNavHgxU6dOZc2aNRm/pqamhvLyctavX8+IESMYPHhwRgn7g8Ta167F8dngzjnXSJYvX84LL7xQp0QdVVlZyfTp01m8eHGWI3PNjSdr55xrBFVVVbzyyiusXr26QfVUVlby8ssvU1FRkaXIXHPkydrVmYiME5EVuY4jW0RkjIjUiEjSYSERKRWR2U0c0x4iclED65ghIg9lK6amEv98icgBIvL1XMZUH0uXLmXhwoVZqWv58uW89dZbWanLNU+erJ2r3Q+A45p4n5cDP2lgHScAF2YhlqZ2K/D5yM/PA0NyFEu9bNmyhQULFrB169as1FdTU8OiRYtYt25d2nIlJSWUlJRkZZ8uv/gEM+dqoaqbcrDbgoZWoKrrsxFIU1PVKqAqsqnB70VTW7duHf/73/9qLbdkyRLmzZtHVVUVe+yxB0OGDKF///5Jy65evZqKioq0l3MNHz683jG7/ObJ2u1CRA4AbgYOBToB/wXuVdXxsXI/B34ItAOeBS5R1YrwXA3wXeDUUM/GUMd1kdePAq4BDgS2AM8AP1XVDeH5cuBJYBTQCzgd+BnwD6AY+BZQDfwGeAy4D/gSsCzEMiXUsw8wHhgJdAXWAI8AP1PV1NfR7IizFOivqoeJyOHAdOAYYAIwAFgKXK6qz4TyM4DXwr5OxHqwyoDvqermUOYbwJXAAUBr4G3gKlV9TkTGAT+PvI8jVHWGiIwGfgF8NvxOHgVuUNWPUsQ9A1ihqmeKyBhgHPBT4BagJzAbGANcGh63Aw+q6o/D68cBh4ffwZXAntjv+QfY5+NELKnepqq3xt+rSBw3AGeqat/IMaX8bIT9nqeqfcJnoDdwbTiGXwM3AT1CUk/sYzFQqqo3JnsvmlplZSVVVVVpyyxZsoRZs2Z90vquqqpi1qxZAEkT9tatW6moqEBEsh+wy3uerN1ORKQjMBWYBgwDtgLfAW4RkRdV9bVQtDdwJHAU0Bm4H/gTEB1bnABcBIwNdfxCRGao6kshWT0F3AicA3wauBOYIiJfUtXtoY4LsS7oCmAhlqx/DNwAHAScjSWwb2Pdxop9of8JS/BgiXJNiLUSOBa4HXgFmFjPt2oCcAmwHLgDeFBEekcSyMXAbVj37eeAh0Ns14nIF4FJwGXYCcdeWAJ6UER6Y93AewEnhdevDyc2T4RjnwLsG94vATK9qHZv7P38JpZ4y4D5wIPA/wFfA24XkSmq+vfwmqHAOmAEdmLyFHAElvBvwn6v40Xkb6r6ToZxQJrPRqzcEOBN7OTqFuyEYjz23pQCiMhhQD/sd/4JEekCdInVV+t6nCeddFIdDiO5TZs2sWHDhrRlVq9ezfbt23fatnXrVmbOnJlyrHv69Ol07949ZZ19+/YFoLy8PO2+J06s78fe5YqPWbu4Tlgiu0BV31HVxVjrF6wFnFANnKqqr4cv2IuAY0RkYKTMn1X1IVVdClwNbMZaUgBXAM+q6tVqpgJnAAcDR0fqeF5Vp6jqv1Q1cRHpO6p6vaq+hyVmgMdU9SlVfRv4LdBTRLqLSAfgz8BYVX1DVd9T1TuAlbHjqatrVfXF8P5cjSW/aH2qqleq6mJVnYgl2MSxb8Na/r9S1aWq+kY4jmJg75Dw3we2qeoqVa3GWtoPqOq9qvpu6DX4HnCyiPTNMOZC4Aeq+qaqzsJ6CLYAl6rqovC+VMaOow323i1U1TLg3+HYxqvqIqyFXYD1ENRFus/GJ1R1LZagq1R1raquA54GzooU+zYwRVXjkx5/iPV6RP/NqmOc9VJTU1NrmXiirm17JvUWFBRQUNDsRg1cBrxl7XaiqmtF5B7gNBH5PNAfaxmCddcmvKuq0WtSXgmPB2AtYCKPqGqNiFQCbcOmwew4CUiU+aeIVIXnng+b30sWZuT/74fHJZFtiXUc24XjuQs4SUQuDcczGGtlZrYsVHLRpk9iTLttZFs0xkSZ3gCq+oaIrBeRy4H9Q0wHhXKpYvoCcEjoCk5IfCvvD5RnGHf8vXsv0osB9t61i/xcERv7fp+dfyefvNcZ7j8h3WejNg8Az4fhjQqsZ2FsknK3E1rfEX2oJWFno9U5Z84cpk6dmrbMo48+mrSrfI899mD06NFJX/O5z32O448/PmWdpaWlAEyYMCHzYF2z4Mna7UREegJzgbXYGPIUYB7W3RsVH+tNJJno+GmysdSC2GNcK+DjyM+7LqC88/MJSZsjItIJeAnrMXgc6yp9NWxriHTHlvZ5ERkOTAaewxLHw0BHrMWYSivgV1iiiqt9JtMO8fcudTMueflMXhOX7HumtvcvnReAFdgchmXYUM0z8UKquhEbD/9EU433FhUV0apVq7St5CFDhuw0Zg3Qpk0bhgxJPfG9R48eWY3TNR+erF3c6UB3YL/Q/YqIJLpFo1+m+4pI58SEKeAwoAbI9GLQ+cBXsPFZwn6GYElrQf3D38XRWKu0j6r+N+ynCOhB7mYZXwrMUtUTEhtE5OLw30RM8f7OBYCo6pLIa74C/Aj4Pjt6GHKtGhsSiBrQwDp3ei9UdbuI/BGb4PYf4JFUk+xyZc8996Rz585s2pT6QoLEJLJMZ4MXFhamHa92LZsnaxe3HOgAnCoiL2ETmBLjwtGuznbA4yJyJTbWeif2pZms2zqZm4FJInId1rLcB7gLm0X9YoOPYofEOOYZIvJY2M8vsfHbunbdZsty4AQR+XL4/wjg+vBcIqbNQFexpuAybHLVE2Gm9CNYd+7vgXJVXdWEsddmDnBe6K6fgU0OPAqbpFZfm4H9RGRvVV0Ztv0RuAobXx/WgLobRXFxMb169UqbrMESdqrkHNejRw+6deuWtsx+++2XcYyuefEJZi5uIjbbdgLwDjbu9wDWbRztn3sd+BeWWCdi3brnZ7oTVX0KaxmNxlrZfw51Hamq2VlJwvbzKjaD+iJsnLQUmIld9pSrhTauAV7GZmO/gY23not1+SdiegxL5POBr4dJaqcA38AmeT2Mzdo/gfzyMDY7/lasN+BQLKk2xK3YTPX5ItIKQFWXYScDC1X19QbWn3Xt27fngAMOoLCwMCv1FRQUMHDgwFrvcz1s2DCGDcu7cxeXBQWZzFp0zrl8IyILgbtU9a46vKYvsHTatGn06VPrVVwN8sEHHzB58mTefPPNBtdVUlLCscceS1FRURYic/lqxYoVjBw5EqBEVcujz3k3uHOu2RCRQqxrfTg27+DB3EaUWseOHTnkkENYs2ZNRquZpdKlSxeGDRuWUaJOzAYfM2ZMvffn8pN3gzvnmg1V/Rhbse4k4CxVrcxxSGn17t2bI488kl69etVeOImuXbsycuRI9t133yxH5pobb1k755oVVe2d6xjqol+/fowePZq5c+eycOFCPv442dVwO2vVqhX9+vVj6NChlJSU0KqVt6t2d56snXOukfXu3ZtRo0YxYMAAFixYwMqVK3n//fd3WZGsQ4cO9OjRg0GDBjFgwAC6dOmSm4Bd3vFk7ZxzTaBTp04MHjyY/v37s27dOjZu3EhFRQXV1dW0adOG4uJiunTpQnFxMZ07d851uC7PeLJ2zrkm1LFjRzp27Mg+++yT9boHDRqU9TpdfvBk7ZxzLUS6pUpd8+bJ2jm3O2kNsGpVPi36lj2JdcbbtPGv9uYo8rnc5YY+/ht1zu1OegGcccYZuY7DuXR6Ae9GN3iyds7tTuYBX8buVBa/c1xLkLgF6JfZsS5+S9bSjrc1lqjnxZ/w5Uadc66FSCynSpLlKlui3el4/Up755xzLs95snbOOefynCdr55xzLs95snbOuZZjI/CL8Lg72Mhucrw+wcw555zLc96yds455/KcJ2vnnHMuz/miKM4518yJSCvgWuA8oAvwEnChqr6Xy7iaiohcDhyrqoflOpbG4i1r55xr/q4GLgDOB4YCNcBkEWmX06iagIhcANyU6zgam7esnXOuGRORtsClwBWq+mzY9i1sSdWTgYdyGF6jEZG9gfuAEcCiHIfT6Lxl7ZxzzdtBQGfgxcQGVa0E/gUMz1FMTeGLQDUwGHglx7E0Om9ZO+dc89YnPMZvZLES2KeJY2kyqloGlAGISI6jaXzesnbOueatY3j8KLb9I6B9E8fiGokna+eca962hMf4ZLJ2QFUTx+IaiSdr55xr3paHx71j2/emZdzj2eHJ2jnnmrs3gUrg8MQGEdkT+AIwM0cxuSzzCWbOOdeMqepHInIXcJOIrAKWAjdjreqJOQ3OZY0na+eca/6uwb7P7wc6YCuYHa2qH+c0Kpc1ftct55xzLs/5mLVzzjmX5zxZO+ecc3nOk7VzzjmX5zxZO+ecc3nOk7VzzjmX5zxZO+ecc3nOr7N2zrk6EJFyYLaqntlI9X8amA0MUdXVke3/B/wYOAwoxu5XPQu4Q1Vfi9VRCnw7SfUfAP8BngCuU9WtofxY4DhVPTbrB1QLERmNXSc+GKgIsV2tqlWRMt8Enkry8smqOiqU2R8oBfbHVm47L/b+HQfcBeynqh9mGNtewMXYfcH7AR8CbwF3quqTsbI1wI2qelUmddeVt6ydcy5PiEgB8AfgrliiuRR4GSgCLgOOAq4GPg3MFZFLklS3Fhga+/ct4PXw2hsjZR8A9hGRc7N9TOmIyPHAX7EbjpwC/ABbNnWaiEQbkweR/Hh+GCnzILAaOBHoBtwe2U9rbFW3q+uQqAdg9wQ/H/gzcDzwHWwt9okiMr4Oh9pg3rJ2zrn88Q3gYOC4xAYROQa4FbheVa+JFhaRh4DfALeLyFuqOi3ydLWqzk2yjzIRKQHOAS4HUNXtInIjcLeI/EVVP6gtUBEZA/xRVQvqdIQ7GwcoMEpVq0O9s4D3Qnz3h3IHAa+nOJ5EC/hg4GBV/aeI7AH8LlLkHGArlnRrFU4UHge2AV9U1TWRp/8qIquBy0TkWVV9KZM6G8qTtXPONYCItAd+ApwJ9MXW5P49MF5Vt0fK/QS4AOiFtdhuxlqVI1R1Rih2JfBkLFleiyW0a+P7VtWaUO9x4flp8TIpbALiy1eWhbjPw04AmsL+wL2JRA2gqmtE5B3g6+ycrB9LU0/iWBK3C60GWgOISEfgF8DY6O+jFseEfZ4aS9QJN2DLum7LsL4G82TtnHP1FLqty7Au2euwLubDsS/z/YBzQ7lrsGQ6AXgROJpY8hERAYYAV0W2dQvbblPVpGtDhxt5PAVcIiLFqrou8vrod3wr4FPYScVRwG2xej4UkbLwfNJkLSKt2DF82irJPralijOFCuwEJ7qPQqx7v134uQvwGWA/Efk3INh4/Z2E90VVK0XkbWCMiNwUjmF2qPJHwCJVfa4OcR2DJeKkr1HVjcCFdaivwTxZO+dc/Y0CjgDOUtWHwrapIrIFuF5EbgfeBX4G3KOqPwtlpohIJ2w8NOGr4XFeZFtfoAC7k1Y6S0K5TwOJZN0bSHYjj2XYmHWyMdd5wGkispeqbkry/B/YdeJadB8jgBm1xBqv7+cicgXWqu+AnejshY1jg7VwwY5tHLAR+GaIvwjrjQDr6n4MG9N/DTg5nOxcBhwlIl8Afh1e8ztVvTNNXPsAFdFJbrnmydo55+rvcGA7Nr4Z9RBwfXj+U1gSejJW5lF2Ttb9gM2quiGyLTEeXNvds7bGygOswbqSwZLfT7FW+oWq+rcU9ZRjLebPAPOTPD8Om1ENMBrrLRgSeV5Db0Pr6IsSs85T1NcG66a+CTvO+4FngM+GMvPDccxW1cqwbVpi+EFEJqjqBlV9FSgRkU6q+j6AiPwGmAK8gZ3w/BqbKf6siCxR1edTxLU1fgy55rPBnXOu/oqA9dEx12BVeOwCdA//Xxsrszr2817YpVVR5eGxpJY49g2P/4ls+1hVXwv/pmHJVYFJInJoinrej8SyC1UtT9SZiC2yj9dUdTPW8v44+k9E+qaob2vobdgTGAR8SlUvxMb114cy61X1uUiiTvgbUMiOpJ6oM5Go98Vmb18JHIr9Ln6tqvOASdjs81SWAUVholpSItInzeuzzpO1c87V33rsS71tbHuv8FiBTTgD6BEr86nYzxVYQvmEqq7FLtk6MYwX7yJclnQCMEdVK1IFGu5tfTbWE/BgaJnGdY3EUl9lWGs7+m9lsoIiMlxERqnqh6r6tqpuDGPgB2KT8BCRQ0Tk/CQv71BLrDcCpaq6BHvvN6pqYkLYeqBnmmOYjOXHUSni7gwsEpGMZpdngydr55yrv5nY92i8lZZYMGU21gW7CbtON+qE2M/lQDsR6R7bfg0wALglRQzjsZb3DbUFq6qLsVno/bBx9Lg+WDJfkeS5jKjqulhr+7UkPQ8JJwG/D5PKEs7FTlqeDj9/EfitiBwWe+2p2DXP78YrFZEh2CSx68Km1ex8UrU3u/ZsRE0G/g3cGMa94xKzwZssWfuYtXPO1d/zwHQsmfTCZoMPxxLhI6o6HyAsoHG9iGwO5YcD3w91JC4nmhIeDyOyWpeqThORHwG3ichB2KSsFVhiPReb1HWZqv49w5hvwbqqLxeRUlUtjzx3GNZC31xbJapaiq0Y1hC/BcYCfxKRB7BVzG4BHlPVmaHMQ9iM7kdF5CosyZ6BXa52aorx8PHYTPFEQn4F2AyMF5E52MIpZ6UKSlW3isjZ2O/knyJyB3bSVRRedxxws6pOSVVHtnnL2jnn6ilcpjQaSzqXYOOop2CXX50dKfpLrIV8JvAs8BXCgiSEWc8hab6KtQjj+7kDGIZNGrsZmIoltf8BQ1X1tvhr0sT8IbZSWHt2XuWrPZb4n8i0roZS1bewyWMDsGvOL8a6r8+KlNlMWNUMm4T2NDAQOF5Vd4k1LCKzP5FL08J166cAXwPuwcaun6kltjewLvyngO9h3fv3AJ2Bb6rqFXU/4vorqKmpyyVxzjnn6iKMwZ4OTFfV5ZHtF2DXCheH63YRkW9gLcm9M2ndZjnOs7EWab9MVjBzTcuTtXPONTIReQv4CBvrrMAmUN0APK2q58TKzgKeU9VfNmF8rbEu/LtV9b6m2q/LnHeDO+dc4xsNLALuxbqwf4hd8/vdJGXPBi4UkXSzlbPtPGCFJ+r85S1r55xzLs95y9o555zLc56snXPOuTznydo555zLc56snXPOuTznydo555zLc56snXPOuTz3/11ClSRLYT9nAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 216x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(3,6))\n",
    "\n",
    "plt.errorbar(root_KO_df.loc[terms_plot]['log_OR'],np.arange(len(terms_plot)),\n",
    "            xerr=[np.subtract(root_KO_df.loc[terms_plot]['log_OR'],root_KO_df.loc[terms_plot]['log_OR_CI_lower']),\n",
    "                   np.subtract(root_KO_df.loc[terms_plot]['log_OR_CI_upper'],root_KO_df.loc[terms_plot]['log_OR'])],color='k',fmt='o')\n",
    "\n",
    "color_temp = root_KO_df.loc[terms_plot]['MPO_term_type'].map({'brain':'blue','heart':'red','neg_ctrl':'black'})\n",
    "\n",
    "sns.scatterplot(x=root_KO_df.loc[terms_plot]['log_OR'],\n",
    "                y=np.arange(len(terms_plot)),size=root_KO_df.loc[terms_plot]['num_genes_in_term'],sizes=(200, 2000),\n",
    "                alpha=.5,\n",
    "               hue=color_temp.tolist(),palette={'blue':'blue','red':'red','black':'black'},legend=False)\n",
    "\n",
    "plt.yticks(np.arange(len(terms_plot)),root_KO_df.loc[terms_plot]['MP_description'])\n",
    "plt.xticks([0,1,2])\n",
    "plt.xlabel('log(OR) +- 95% CI')\n",
    "\n",
    "plt.plot([0,0],[-.5,len(terms_plot)-.5],'--',color='gray')\n",
    "plt.ylim([-.5,len(terms_plot)-.5])\n",
    "\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "\n",
    "\n",
    "# plt.savefig('ASD_CHD_brain_heart_MPO_root_enrichment_210813.png',dpi=300,bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 19. Compute the enrichment of phenotype(s) in NetColoc subsystems."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abnormal neuron morphology\n",
      "Genes and Terms to keep: 262\n",
      "abnormal nervous system development\n",
      "Genes and Terms to keep: 191\n",
      "abnormal cardiovascular development\n",
      "Genes and Terms to keep: 139\n",
      "abnormal heart morphology\n",
      "Genes and Terms to keep: 369\n",
      "abnormal heart development\n",
      "Genes and Terms to keep: 39\n",
      "abnormal myocardial trabeculae morphology\n",
      "Genes and Terms to keep: 7\n",
      "abnormal neuron number\n",
      "Genes and Terms to keep: 55\n",
      "abnormal neuron physiology\n",
      "Genes and Terms to keep: 47\n",
      "abnormal heart layer morphology\n",
      "Genes and Terms to keep: 49\n",
      "abnormal cardiac muscle tissue morphology\n",
      "Genes and Terms to keep: 62\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abnormal neuron morphology:-log(OR_p)</th>\n",
       "      <th>abnormal neuron morphology:log_OR</th>\n",
       "      <th>abnormal neuron morphology:num_genes</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal cardiovascular development:-log(OR_p)</th>\n",
       "      <th>abnormal cardiovascular development:log_OR</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal neuron physiology:num_genes</th>\n",
       "      <th>abnormal neuron physiology:gene_ids</th>\n",
       "      <th>abnormal heart layer morphology:-log(OR_p)</th>\n",
       "      <th>abnormal heart layer morphology:log_OR</th>\n",
       "      <th>abnormal heart layer morphology:num_genes</th>\n",
       "      <th>abnormal heart layer morphology:gene_ids</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:-log(OR_p)</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:log_OR</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:num_genes</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C773</th>\n",
       "      <td>24.749376</td>\n",
       "      <td>1.012687</td>\n",
       "      <td>142</td>\n",
       "      <td>AFDN AFF1 ANK2 ANK3 ARID1B ATM ATP1A3 ATP1B2 A...</td>\n",
       "      <td>23.039059</td>\n",
       "      <td>1.046446</td>\n",
       "      <td>120</td>\n",
       "      <td>AFDN AGO2 ANKRD17 APC ARHGEF7 ARID1A ARID1B BC...</td>\n",
       "      <td>22.217781</td>\n",
       "      <td>1.187488</td>\n",
       "      <td>...</td>\n",
       "      <td>68</td>\n",
       "      <td>ARHGEF2 ATM ATP1A3 BRAF BRPF1 CACNA1C CACNA1D ...</td>\n",
       "      <td>15.677752</td>\n",
       "      <td>1.245477</td>\n",
       "      <td>53</td>\n",
       "      <td>AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...</td>\n",
       "      <td>15.458468</td>\n",
       "      <td>1.191350</td>\n",
       "      <td>57</td>\n",
       "      <td>AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C774</th>\n",
       "      <td>4.152585</td>\n",
       "      <td>0.688834</td>\n",
       "      <td>40</td>\n",
       "      <td>ARID1B BCL11A BRD7 BRPF1 CHD7 CLOCK CTNNB1 DNM...</td>\n",
       "      <td>17.974257</td>\n",
       "      <td>1.350002</td>\n",
       "      <td>56</td>\n",
       "      <td>ARID1A ARID1B BCL11A BRD1 BRD2 BRPF1 CHD7 CREB...</td>\n",
       "      <td>21.708259</td>\n",
       "      <td>1.618705</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>ARHGEF2 BRPF1 CHD7 CTNNB1 EHMT2 EPAS1 FOXG1 HI...</td>\n",
       "      <td>8.647302</td>\n",
       "      <td>1.362651</td>\n",
       "      <td>22</td>\n",
       "      <td>ARID1A BRD2 CHD7 CREBBP CTNNB1 EP300 EZH2 FOXO...</td>\n",
       "      <td>8.043151</td>\n",
       "      <td>1.281307</td>\n",
       "      <td>23</td>\n",
       "      <td>ARID1A BRD2 CHD7 CREBBP EP300 EPAS1 EZH2 FOXO1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C775</th>\n",
       "      <td>2.532538</td>\n",
       "      <td>0.570380</td>\n",
       "      <td>32</td>\n",
       "      <td>AFF1 ATM ATRX ATXN1 ATXN2 BRAF CDC14A CLEC16A ...</td>\n",
       "      <td>4.119965</td>\n",
       "      <td>0.770561</td>\n",
       "      <td>31</td>\n",
       "      <td>AGO2 ANKRD17 APC ARHGEF7 BRAF C2CD3 CDK13 CHD8...</td>\n",
       "      <td>3.488283</td>\n",
       "      <td>0.831981</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>ATM BRAF CHD8 CREB1 GSK3B KIDINS220 PTEN SPG7 ...</td>\n",
       "      <td>2.222506</td>\n",
       "      <td>0.825185</td>\n",
       "      <td>12</td>\n",
       "      <td>AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 NFAT5 N...</td>\n",
       "      <td>2.170585</td>\n",
       "      <td>0.782679</td>\n",
       "      <td>13</td>\n",
       "      <td>AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 FNIP1 N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C777</th>\n",
       "      <td>23.585124</td>\n",
       "      <td>1.718719</td>\n",
       "      <td>53</td>\n",
       "      <td>ANK2 ANK3 ATP1A3 ATP1B2 CACNA1A CACNA1D CACNA1...</td>\n",
       "      <td>2.428448</td>\n",
       "      <td>0.697194</td>\n",
       "      <td>20</td>\n",
       "      <td>CACNA1A CNTN1 CNTN2 DSCAM FGFR2 ISL1 LGI1 MAPT...</td>\n",
       "      <td>2.330480</td>\n",
       "      <td>0.797321</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>ATP1A3 CACNA1C CACNA1D CACNA1F CACNA1G CAMK2A ...</td>\n",
       "      <td>6.131843</td>\n",
       "      <td>1.404871</td>\n",
       "      <td>14</td>\n",
       "      <td>AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 ISL1 PPP3R1 ...</td>\n",
       "      <td>6.070921</td>\n",
       "      <td>1.353035</td>\n",
       "      <td>15</td>\n",
       "      <td>AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 FKBP1B ISL1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C779</th>\n",
       "      <td>6.442488</td>\n",
       "      <td>1.184215</td>\n",
       "      <td>24</td>\n",
       "      <td>AFDN ATXN1 BCL11A CTNNB1 FOXG1 FOXN4 GLI2 HIPK...</td>\n",
       "      <td>15.792268</td>\n",
       "      <td>1.768521</td>\n",
       "      <td>31</td>\n",
       "      <td>AFDN BCL11A CREBBP CTBP1 CTBP2 CTNNB1 EP300 FO...</td>\n",
       "      <td>21.110560</td>\n",
       "      <td>2.136095</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>CTNNB1 EPAS1 FOXG1 HDAC4 HIPK2 MYCN NCOR2 NOTC...</td>\n",
       "      <td>6.121006</td>\n",
       "      <td>1.591720</td>\n",
       "      <td>11</td>\n",
       "      <td>CREBBP CTNNB1 EP300 FOXO1 GATA4 MEF2C MYCN NOT...</td>\n",
       "      <td>7.498114</td>\n",
       "      <td>1.654940</td>\n",
       "      <td>13</td>\n",
       "      <td>CREBBP EP300 EPAS1 FOXO1 GATA4 HDAC9 MEF2C MYC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      abnormal neuron morphology:-log(OR_p)  \\\n",
       "C773                              24.749376   \n",
       "C774                               4.152585   \n",
       "C775                               2.532538   \n",
       "C777                              23.585124   \n",
       "C779                               6.442488   \n",
       "\n",
       "      abnormal neuron morphology:log_OR  abnormal neuron morphology:num_genes  \\\n",
       "C773                           1.012687                                   142   \n",
       "C774                           0.688834                                    40   \n",
       "C775                           0.570380                                    32   \n",
       "C777                           1.718719                                    53   \n",
       "C779                           1.184215                                    24   \n",
       "\n",
       "                    abnormal neuron morphology:gene_ids  \\\n",
       "C773  AFDN AFF1 ANK2 ANK3 ARID1B ATM ATP1A3 ATP1B2 A...   \n",
       "C774  ARID1B BCL11A BRD7 BRPF1 CHD7 CLOCK CTNNB1 DNM...   \n",
       "C775  AFF1 ATM ATRX ATXN1 ATXN2 BRAF CDC14A CLEC16A ...   \n",
       "C777  ANK2 ANK3 ATP1A3 ATP1B2 CACNA1A CACNA1D CACNA1...   \n",
       "C779  AFDN ATXN1 BCL11A CTNNB1 FOXG1 FOXN4 GLI2 HIPK...   \n",
       "\n",
       "      abnormal nervous system development:-log(OR_p)  \\\n",
       "C773                                       23.039059   \n",
       "C774                                       17.974257   \n",
       "C775                                        4.119965   \n",
       "C777                                        2.428448   \n",
       "C779                                       15.792268   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C773                                    1.046446   \n",
       "C774                                    1.350002   \n",
       "C775                                    0.770561   \n",
       "C777                                    0.697194   \n",
       "C779                                    1.768521   \n",
       "\n",
       "      abnormal nervous system development:num_genes  \\\n",
       "C773                                            120   \n",
       "C774                                             56   \n",
       "C775                                             31   \n",
       "C777                                             20   \n",
       "C779                                             31   \n",
       "\n",
       "           abnormal nervous system development:gene_ids  \\\n",
       "C773  AFDN AGO2 ANKRD17 APC ARHGEF7 ARID1A ARID1B BC...   \n",
       "C774  ARID1A ARID1B BCL11A BRD1 BRD2 BRPF1 CHD7 CREB...   \n",
       "C775  AGO2 ANKRD17 APC ARHGEF7 BRAF C2CD3 CDK13 CHD8...   \n",
       "C777  CACNA1A CNTN1 CNTN2 DSCAM FGFR2 ISL1 LGI1 MAPT...   \n",
       "C779  AFDN BCL11A CREBBP CTBP1 CTBP2 CTNNB1 EP300 FO...   \n",
       "\n",
       "      abnormal cardiovascular development:-log(OR_p)  \\\n",
       "C773                                       22.217781   \n",
       "C774                                       21.708259   \n",
       "C775                                        3.488283   \n",
       "C777                                        2.330480   \n",
       "C779                                       21.110560   \n",
       "\n",
       "      abnormal cardiovascular development:log_OR  ...  \\\n",
       "C773                                    1.187488  ...   \n",
       "C774                                    1.618705  ...   \n",
       "C775                                    0.831981  ...   \n",
       "C777                                    0.797321  ...   \n",
       "C779                                    2.136095  ...   \n",
       "\n",
       "      abnormal neuron physiology:num_genes  \\\n",
       "C773                                    68   \n",
       "C774                                    21   \n",
       "C775                                     9   \n",
       "C777                                    27   \n",
       "C779                                    13   \n",
       "\n",
       "                    abnormal neuron physiology:gene_ids  \\\n",
       "C773  ARHGEF2 ATM ATP1A3 BRAF BRPF1 CACNA1C CACNA1D ...   \n",
       "C774  ARHGEF2 BRPF1 CHD7 CTNNB1 EHMT2 EPAS1 FOXG1 HI...   \n",
       "C775  ATM BRAF CHD8 CREB1 GSK3B KIDINS220 PTEN SPG7 ...   \n",
       "C777  ATP1A3 CACNA1C CACNA1D CACNA1F CACNA1G CAMK2A ...   \n",
       "C779  CTNNB1 EPAS1 FOXG1 HDAC4 HIPK2 MYCN NCOR2 NOTC...   \n",
       "\n",
       "      abnormal heart layer morphology:-log(OR_p)  \\\n",
       "C773                                   15.677752   \n",
       "C774                                    8.647302   \n",
       "C775                                    2.222506   \n",
       "C777                                    6.131843   \n",
       "C779                                    6.121006   \n",
       "\n",
       "      abnormal heart layer morphology:log_OR  \\\n",
       "C773                                1.245477   \n",
       "C774                                1.362651   \n",
       "C775                                0.825185   \n",
       "C777                                1.404871   \n",
       "C779                                1.591720   \n",
       "\n",
       "      abnormal heart layer morphology:num_genes  \\\n",
       "C773                                         53   \n",
       "C774                                         22   \n",
       "C775                                         12   \n",
       "C777                                         14   \n",
       "C779                                         11   \n",
       "\n",
       "               abnormal heart layer morphology:gene_ids  \\\n",
       "C773  AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...   \n",
       "C774  ARID1A BRD2 CHD7 CREBBP CTNNB1 EP300 EZH2 FOXO...   \n",
       "C775  AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 NFAT5 N...   \n",
       "C777  AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 ISL1 PPP3R1 ...   \n",
       "C779  CREBBP CTNNB1 EP300 FOXO1 GATA4 MEF2C MYCN NOT...   \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:-log(OR_p)  \\\n",
       "C773                                          15.458468      \n",
       "C774                                           8.043151      \n",
       "C775                                           2.170585      \n",
       "C777                                           6.070921      \n",
       "C779                                           7.498114      \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:log_OR  \\\n",
       "C773                                          1.191350   \n",
       "C774                                          1.281307   \n",
       "C775                                          0.782679   \n",
       "C777                                          1.353035   \n",
       "C779                                          1.654940   \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:num_genes  \\\n",
       "C773                                                 57     \n",
       "C774                                                 23     \n",
       "C775                                                 13     \n",
       "C777                                                 15     \n",
       "C779                                                 13     \n",
       "\n",
       "     abnormal cardiac muscle tissue morphology:gene_ids  \n",
       "C773  AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...  \n",
       "C774  ARID1A BRD2 CHD7 CREBBP EP300 EPAS1 EZH2 FOXO1...  \n",
       "C775  AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 FNIP1 N...  \n",
       "C777  AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 FKBP1B ISL1 ...  \n",
       "C779  CREBBP EP300 EPAS1 FOXO1 GATA4 HDAC9 MEF2C MYC...  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP_focal_top = root_KO_df.head(10).index.tolist() # record the top 10 overall\n",
    "MP_full_results_df = validation.MPO_enrichment_full(hier_df,MPO,mgi_df,MP_focal_top,G_int)\n",
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MP:0002882',\n",
       " 'MP:0003861',\n",
       " 'MP:0002925',\n",
       " 'MP:0000266',\n",
       " 'MP:0000267',\n",
       " 'MP:0002189',\n",
       " 'MP:0008946',\n",
       " 'MP:0004811',\n",
       " 'MP:0010545',\n",
       " 'MP:0010630']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP_focal_top"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 20. Annotate the NetColoc Systems Map with mouse knockout data, input genes, and enriched GO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/churas/anaconda3/envs/netcoloc/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abnormal neuron morphology:-log(OR_p)</th>\n",
       "      <th>abnormal neuron morphology:log_OR</th>\n",
       "      <th>abnormal neuron morphology:num_genes</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal cardiovascular development:-log(OR_p)</th>\n",
       "      <th>abnormal cardiovascular development:log_OR</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:log_OR</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:num_genes</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:gene_ids</th>\n",
       "      <th>gprofiler_name</th>\n",
       "      <th>num_d1_seeds</th>\n",
       "      <th>num_d2_seeds</th>\n",
       "      <th>frac_d1_seeds</th>\n",
       "      <th>frac_d2_seeds</th>\n",
       "      <th>frac_both_seeds</th>\n",
       "      <th>frac_no_seeds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C773</th>\n",
       "      <td>24.749376</td>\n",
       "      <td>1.012687</td>\n",
       "      <td>142</td>\n",
       "      <td>AFDN AFF1 ANK2 ANK3 ARID1B ATM ATP1A3 ATP1B2 A...</td>\n",
       "      <td>23.039059</td>\n",
       "      <td>1.046446</td>\n",
       "      <td>120</td>\n",
       "      <td>AFDN AGO2 ANKRD17 APC ARHGEF7 ARID1A ARID1B BC...</td>\n",
       "      <td>22.217781</td>\n",
       "      <td>1.187488</td>\n",
       "      <td>...</td>\n",
       "      <td>1.191350</td>\n",
       "      <td>57</td>\n",
       "      <td>AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...</td>\n",
       "      <td>ASD-CHD systems map</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>0.053247</td>\n",
       "      <td>0.029870</td>\n",
       "      <td>0.007792</td>\n",
       "      <td>0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C774</th>\n",
       "      <td>4.152585</td>\n",
       "      <td>0.688834</td>\n",
       "      <td>40</td>\n",
       "      <td>ARID1B BCL11A BRD7 BRPF1 CHD7 CLOCK CTNNB1 DNM...</td>\n",
       "      <td>17.974257</td>\n",
       "      <td>1.350002</td>\n",
       "      <td>56</td>\n",
       "      <td>ARID1A ARID1B BCL11A BRD1 BRD2 BRPF1 CHD7 CREB...</td>\n",
       "      <td>21.708259</td>\n",
       "      <td>1.618705</td>\n",
       "      <td>...</td>\n",
       "      <td>1.281307</td>\n",
       "      <td>23</td>\n",
       "      <td>ARID1A BRD2 CHD7 CREBBP EP300 EPAS1 EZH2 FOXO1...</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>0.058608</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.014652</td>\n",
       "      <td>0.901099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C775</th>\n",
       "      <td>2.532538</td>\n",
       "      <td>0.570380</td>\n",
       "      <td>32</td>\n",
       "      <td>AFF1 ATM ATRX ATXN1 ATXN2 BRAF CDC14A CLEC16A ...</td>\n",
       "      <td>4.119965</td>\n",
       "      <td>0.770561</td>\n",
       "      <td>31</td>\n",
       "      <td>AGO2 ANKRD17 APC ARHGEF7 BRAF C2CD3 CDK13 CHD8...</td>\n",
       "      <td>3.488283</td>\n",
       "      <td>0.831981</td>\n",
       "      <td>...</td>\n",
       "      <td>0.782679</td>\n",
       "      <td>13</td>\n",
       "      <td>AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 FNIP1 N...</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>0.037344</td>\n",
       "      <td>0.029046</td>\n",
       "      <td>0.008299</td>\n",
       "      <td>0.925311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C777</th>\n",
       "      <td>23.585124</td>\n",
       "      <td>1.718719</td>\n",
       "      <td>53</td>\n",
       "      <td>ANK2 ANK3 ATP1A3 ATP1B2 CACNA1A CACNA1D CACNA1...</td>\n",
       "      <td>2.428448</td>\n",
       "      <td>0.697194</td>\n",
       "      <td>20</td>\n",
       "      <td>CACNA1A CNTN1 CNTN2 DSCAM FGFR2 ISL1 LGI1 MAPT...</td>\n",
       "      <td>2.330480</td>\n",
       "      <td>0.797321</td>\n",
       "      <td>...</td>\n",
       "      <td>1.353035</td>\n",
       "      <td>15</td>\n",
       "      <td>AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 FKBP1B ISL1 ...</td>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.890909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C779</th>\n",
       "      <td>6.442488</td>\n",
       "      <td>1.184215</td>\n",
       "      <td>24</td>\n",
       "      <td>AFDN ATXN1 BCL11A CTNNB1 FOXG1 FOXN4 GLI2 HIPK...</td>\n",
       "      <td>15.792268</td>\n",
       "      <td>1.768521</td>\n",
       "      <td>31</td>\n",
       "      <td>AFDN BCL11A CREBBP CTBP1 CTBP2 CTNNB1 EP300 FO...</td>\n",
       "      <td>21.110560</td>\n",
       "      <td>2.136095</td>\n",
       "      <td>...</td>\n",
       "      <td>1.654940</td>\n",
       "      <td>13</td>\n",
       "      <td>CREBBP EP300 EPAS1 FOXO1 GATA4 HDAC9 MEF2C MYC...</td>\n",
       "      <td>Transcriptional regulation by RUNX3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0.055046</td>\n",
       "      <td>0.045872</td>\n",
       "      <td>0.009174</td>\n",
       "      <td>0.889908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      abnormal neuron morphology:-log(OR_p)  \\\n",
       "C773                              24.749376   \n",
       "C774                               4.152585   \n",
       "C775                               2.532538   \n",
       "C777                              23.585124   \n",
       "C779                               6.442488   \n",
       "\n",
       "      abnormal neuron morphology:log_OR  abnormal neuron morphology:num_genes  \\\n",
       "C773                           1.012687                                   142   \n",
       "C774                           0.688834                                    40   \n",
       "C775                           0.570380                                    32   \n",
       "C777                           1.718719                                    53   \n",
       "C779                           1.184215                                    24   \n",
       "\n",
       "                    abnormal neuron morphology:gene_ids  \\\n",
       "C773  AFDN AFF1 ANK2 ANK3 ARID1B ATM ATP1A3 ATP1B2 A...   \n",
       "C774  ARID1B BCL11A BRD7 BRPF1 CHD7 CLOCK CTNNB1 DNM...   \n",
       "C775  AFF1 ATM ATRX ATXN1 ATXN2 BRAF CDC14A CLEC16A ...   \n",
       "C777  ANK2 ANK3 ATP1A3 ATP1B2 CACNA1A CACNA1D CACNA1...   \n",
       "C779  AFDN ATXN1 BCL11A CTNNB1 FOXG1 FOXN4 GLI2 HIPK...   \n",
       "\n",
       "      abnormal nervous system development:-log(OR_p)  \\\n",
       "C773                                       23.039059   \n",
       "C774                                       17.974257   \n",
       "C775                                        4.119965   \n",
       "C777                                        2.428448   \n",
       "C779                                       15.792268   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C773                                    1.046446   \n",
       "C774                                    1.350002   \n",
       "C775                                    0.770561   \n",
       "C777                                    0.697194   \n",
       "C779                                    1.768521   \n",
       "\n",
       "      abnormal nervous system development:num_genes  \\\n",
       "C773                                            120   \n",
       "C774                                             56   \n",
       "C775                                             31   \n",
       "C777                                             20   \n",
       "C779                                             31   \n",
       "\n",
       "           abnormal nervous system development:gene_ids  \\\n",
       "C773  AFDN AGO2 ANKRD17 APC ARHGEF7 ARID1A ARID1B BC...   \n",
       "C774  ARID1A ARID1B BCL11A BRD1 BRD2 BRPF1 CHD7 CREB...   \n",
       "C775  AGO2 ANKRD17 APC ARHGEF7 BRAF C2CD3 CDK13 CHD8...   \n",
       "C777  CACNA1A CNTN1 CNTN2 DSCAM FGFR2 ISL1 LGI1 MAPT...   \n",
       "C779  AFDN BCL11A CREBBP CTBP1 CTBP2 CTNNB1 EP300 FO...   \n",
       "\n",
       "      abnormal cardiovascular development:-log(OR_p)  \\\n",
       "C773                                       22.217781   \n",
       "C774                                       21.708259   \n",
       "C775                                        3.488283   \n",
       "C777                                        2.330480   \n",
       "C779                                       21.110560   \n",
       "\n",
       "      abnormal cardiovascular development:log_OR  ...  \\\n",
       "C773                                    1.187488  ...   \n",
       "C774                                    1.618705  ...   \n",
       "C775                                    0.831981  ...   \n",
       "C777                                    0.797321  ...   \n",
       "C779                                    2.136095  ...   \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:log_OR  \\\n",
       "C773                                          1.191350   \n",
       "C774                                          1.281307   \n",
       "C775                                          0.782679   \n",
       "C777                                          1.353035   \n",
       "C779                                          1.654940   \n",
       "\n",
       "     abnormal cardiac muscle tissue morphology:num_genes  \\\n",
       "C773                                                 57    \n",
       "C774                                                 23    \n",
       "C775                                                 13    \n",
       "C777                                                 15    \n",
       "C779                                                 13    \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:gene_ids  \\\n",
       "C773  AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...    \n",
       "C774  ARID1A BRD2 CHD7 CREBBP EP300 EPAS1 EZH2 FOXO1...    \n",
       "C775  AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 FNIP1 N...    \n",
       "C777  AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 FKBP1B ISL1 ...    \n",
       "C779  CREBBP EP300 EPAS1 FOXO1 GATA4 HDAC9 MEF2C MYC...    \n",
       "\n",
       "                           gprofiler_name  num_d1_seeds num_d2_seeds  \\\n",
       "C773                  ASD-CHD systems map            41           23   \n",
       "C774      PKMTs methylate histone lysines            16            7   \n",
       "C775          Chromatin modifying enzymes             9            7   \n",
       "C777                      Ion homeostasis            12            6   \n",
       "C779  Transcriptional regulation by RUNX3             6            5   \n",
       "\n",
       "      frac_d1_seeds  frac_d2_seeds  frac_both_seeds frac_no_seeds  \n",
       "C773       0.053247       0.029870         0.007792      0.909091  \n",
       "C774       0.058608       0.025641         0.014652      0.901099  \n",
       "C775       0.037344       0.029046         0.008299      0.925311  \n",
       "C777       0.072727       0.036364         0.000000      0.890909  \n",
       "C779       0.055046       0.045872         0.009174      0.889908  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the best gprofiler annotation\n",
    "MP_full_results_df['gprofiler_name']=pd.Series(system_name_list,index=hier_df.index.tolist())\n",
    "# don't annotate the root node\n",
    "root_node = hier_df['CD_MemberList_Size'].sort_values(ascending=False).head(1).index.tolist()[0]\n",
    "MP_full_results_df['gprofiler_name'].loc[root_node]=d1_name+'-'+d2_name+' systems map'\n",
    "\n",
    "# also add the frac_seeds/num_seeds data here\n",
    "MP_full_results_df=MP_full_results_df.join(hier_df[['num_d1_seeds','num_d2_seeds','frac_d1_seeds','frac_d2_seeds','frac_both_seeds','frac_no_seeds']],\n",
    "                                          how='left')\n",
    "\n",
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 21. Export the NetColoc systems map to NDEx, with default style. \n",
    "\n",
    "Default style maps the fraction of seed genes from input set 1 (red) and input set 2 (blue) to node pie charts. The fraction of genes in each system that are in neither input set, but that are implicated by the network propagation are indicated in white."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert G_hier to nice cx network\n",
    "node_id_to_node_name = nx.get_node_attributes(G_hier, 'name')\n",
    "for node_id in list(G_hier.nodes):\n",
    "    del G_hier.nodes[node_id]['name']\n",
    "\n",
    "G_hier_cx = ndex2.create_nice_cx_from_networkx(G_hier)\n",
    "\n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    node['n'] = node_id_to_node_name[node_id]\n",
    "\n",
    "G_hier_cx.set_name(d1_name+'_'+d2_name+'_systems_map') \n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    data = MP_full_results_df.loc[node['r']]\n",
    "    for row, value in data.items():\n",
    "        if (row.find('gene_ids')>-1) or (row=='gprofiler_name'):\n",
    "            data_type = \"string\"\n",
    "            value=str(value)\n",
    "        else:\n",
    "            data_type = \"double\"\n",
    "            value = str(value) # nice cx can only accept strings as values...\n",
    "            if value=='inf': # check if inf, set to -1 if so\n",
    "                value='-1'\n",
    "            \n",
    "        G_hier_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "# Restore some hierarchy properties to their state before networkx conversion.\n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    for i in np.arange(len(G_hier_cx.nodeAttributes[node_id])):\n",
    "        dict_temp = G_hier_cx.nodeAttributes[node_id][i]\n",
    "        if dict_temp['n'] in ['CD_MemberList_Size','CD_MemberList_LogSize','HiDeF_persistence']:\n",
    "            G_hier_cx.set_node_attribute(node_id, dict_temp['n'], dict_temp['v'], type='double',overwrite=True)\n",
    "            \n",
    "# this is required so we can easily make subgraphs from systems\n",
    "G_hier_cx.set_network_attribute('__CD_OriginalNetwork',\n",
    "                                       values='0', type='long')\n",
    "            \n",
    "          \n",
    "# use apply_style_from_network-- this should overwrite the existing style\n",
    "netcoloc_template = ndex2.create_nice_cx_from_server('ndexbio.org',\n",
    "            uuid='f338dea0-117c-11ec-9e8e-0ac135e8bacf')\n",
    "G_hier_cx.apply_style_from_network(netcoloc_template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDEx server (probably ndexbio.org): ndexbio.org\n",
      "NDEx user name: cbass\n",
      "NDEx password: ········\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "#Upload to NDEx\n",
    "G_hier_cx.set_name(d1_name+'_'+d2_name+'_systems_map') \n",
    "SERVER = input('NDEx server (probably ndexbio.org): ')\n",
    "USERNAME = input('NDEx user name: ')\n",
    "PASSWORD = getpass.getpass('NDEx password: ')\n",
    "network_uuid_hier = G_hier_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 22. Apply another template style to NetColoc systems map for mouse knockout view, and export to NDEx. \n",
    "\n",
    "Select the property to be mapped to system node colors (should be one of the mouse knockout phenotypes we identified above). In this style, the log odds ratio is mapped to the system node color. Systems which are not significantly enriched for the phenotype are white (p<0.05).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------ modify this based on your project. Should be a system identified above -----\n",
    "# set the property we should map to system node colors\n",
    "mouse_KO_mapping_property = 'abnormal neuron morphology'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating CX\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "# apply a template style \n",
    "G_hier_cx.set_name(d1_name+'_'+d2_name+'_systems_map_mouse_KO:'+mouse_KO_mapping_property) \n",
    "# use apply_style_from_network-- this should overwrite existing style\n",
    "netcoloc_template = ndex2.create_nice_cx_from_server('ndexbio.org',\n",
    "            uuid='4958993c-df46-11eb-b666-0ac135e8bacf')\n",
    "raw_cx_st = json.dumps(netcoloc_template.to_cx())\n",
    "\n",
    "# replace the default template values with mouse_KO_mapping_property\n",
    "updated_raw_cx = re.sub('COL=abnormal heart development:log_OR', 'COL='+mouse_KO_mapping_property+':log_OR', raw_cx_st)\n",
    "updated_raw_cx = re.sub('COL=abnormal heart development:-log', \n",
    "                        'COL='+mouse_KO_mapping_property+':-log', updated_raw_cx)\n",
    "updated_raw_cx=json.loads(updated_raw_cx)\n",
    "netcoloc_template_updated = ndex2.create_nice_cx_from_raw_cx(updated_raw_cx)\n",
    "G_hier_cx.apply_style_from_network(netcoloc_template_updated)\n",
    "\n",
    "network_uuid_hier_mouse_KO = G_hier_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 23. Add genes associated with mouse knockout phenotypes to NetColoc subnetwork, export to NDEx. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/churas/anaconda3/envs/netcoloc/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d1_seeds</th>\n",
       "      <th>d2_seeds</th>\n",
       "      <th>z_d1</th>\n",
       "      <th>z_d2</th>\n",
       "      <th>z_both</th>\n",
       "      <th>d1_name</th>\n",
       "      <th>d2_name</th>\n",
       "      <th>sum_seeds</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal cardiovascular development:gene_ids</th>\n",
       "      <th>abnormal heart morphology:gene_ids</th>\n",
       "      <th>abnormal heart development:gene_ids</th>\n",
       "      <th>abnormal myocardial trabeculae morphology:gene_ids</th>\n",
       "      <th>abnormal neuron number:gene_ids</th>\n",
       "      <th>abnormal neuron physiology:gene_ids</th>\n",
       "      <th>abnormal heart layer morphology:gene_ids</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PTEN</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>17.775818</td>\n",
       "      <td>18.270646</td>\n",
       "      <td>324.775666</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNB1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15.369891</td>\n",
       "      <td>16.212593</td>\n",
       "      <td>249.185790</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POGZ</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.518539</td>\n",
       "      <td>13.035299</td>\n",
       "      <td>163.182897</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.964197</td>\n",
       "      <td>12.579514</td>\n",
       "      <td>163.083297</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSD1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.263667</td>\n",
       "      <td>12.748115</td>\n",
       "      <td>156.338630</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        d1_seeds  d2_seeds       z_d1       z_d2      z_both d1_name d2_name  \\\n",
       "PTEN           1         1  17.775818  18.270646  324.775666     ASD     CHD   \n",
       "CTNNB1         1         1  15.369891  16.212593  249.185790     ASD     CHD   \n",
       "POGZ           1         1  12.518539  13.035299  163.182897     ASD     CHD   \n",
       "KMT2C          1         1  12.964197  12.579514  163.083297     ASD     CHD   \n",
       "NSD1           1         1  12.263667  12.748115  156.338630     ASD     CHD   \n",
       "\n",
       "        sum_seeds  abnormal neuron morphology:gene_ids  \\\n",
       "PTEN            2                                    1   \n",
       "CTNNB1          2                                    1   \n",
       "POGZ            2                                    0   \n",
       "KMT2C           2                                    0   \n",
       "NSD1            2                                    0   \n",
       "\n",
       "        abnormal nervous system development:gene_ids  \\\n",
       "PTEN                                               1   \n",
       "CTNNB1                                             1   \n",
       "POGZ                                               0   \n",
       "KMT2C                                              0   \n",
       "NSD1                                               0   \n",
       "\n",
       "        abnormal cardiovascular development:gene_ids  \\\n",
       "PTEN                                               1   \n",
       "CTNNB1                                             1   \n",
       "POGZ                                               0   \n",
       "KMT2C                                              0   \n",
       "NSD1                                               0   \n",
       "\n",
       "        abnormal heart morphology:gene_ids  \\\n",
       "PTEN                                     1   \n",
       "CTNNB1                                   1   \n",
       "POGZ                                     0   \n",
       "KMT2C                                    0   \n",
       "NSD1                                     0   \n",
       "\n",
       "        abnormal heart development:gene_ids  \\\n",
       "PTEN                                      1   \n",
       "CTNNB1                                    0   \n",
       "POGZ                                      0   \n",
       "KMT2C                                     0   \n",
       "NSD1                                      0   \n",
       "\n",
       "        abnormal myocardial trabeculae morphology:gene_ids  \\\n",
       "PTEN                                                    0    \n",
       "CTNNB1                                                  0    \n",
       "POGZ                                                    0    \n",
       "KMT2C                                                   0    \n",
       "NSD1                                                    0    \n",
       "\n",
       "        abnormal neuron number:gene_ids  abnormal neuron physiology:gene_ids  \\\n",
       "PTEN                                  1                                    1   \n",
       "CTNNB1                                0                                    1   \n",
       "POGZ                                  0                                    0   \n",
       "KMT2C                                 0                                    0   \n",
       "NSD1                                  0                                    0   \n",
       "\n",
       "        abnormal heart layer morphology:gene_ids  \\\n",
       "PTEN                                           1   \n",
       "CTNNB1                                         1   \n",
       "POGZ                                           0   \n",
       "KMT2C                                          0   \n",
       "NSD1                                           0   \n",
       "\n",
       "        abnormal cardiac muscle tissue morphology:gene_ids  \n",
       "PTEN                                                    1   \n",
       "CTNNB1                                                  0   \n",
       "POGZ                                                    0   \n",
       "KMT2C                                                   0   \n",
       "NSD1                                                    0   "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add fields to node_df for genes in each mouse_KO phenotype of interest\n",
    "MP_genes_columns = [c for c in MP_full_results_df.columns.tolist() if c.find(':gene_ids')>-1]\n",
    "\n",
    "# look up overlapping genes in the root node, add them to node_df\n",
    "for MP in MP_genes_columns:\n",
    "    focal_genes = MP_full_results_df.loc[root_node][MP].split(' ')\n",
    "    node_df[MP]=0\n",
    "    node_df[MP].loc[focal_genes]=1\n",
    "node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "773\n",
      "26351\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "print(len(G_overlap.nodes()))\n",
    "print(len(G_overlap.edges()))\n",
    "G_overlap_cx = ndex2.create_nice_cx_from_networkx(G_overlap)\n",
    "G_overlap_cx.set_name(d1_name+'_'+d2_name+'_NetColoc_subgraph') \n",
    "for node_id, node in G_overlap_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_overlap_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "# apply a template style (834b6ad4-d2ea-11eb-b666-0ac135e8bacf)\n",
    "G_overlap_cx.apply_template('ndexbio.org','834b6ad4-d2ea-11eb-b666-0ac135e8bacf')\n",
    "\n",
    "network_uuid_NetColoc = G_overlap_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 24. Upload cosine-similarity transformed NetColoc subnetwork to NDEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "773\n",
      "7032\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "#Annotate network\n",
    "print(len(G_cosSim.nodes()))\n",
    "print(len(G_cosSim.edges()))\n",
    "G_cosSim_cx = ndex2.create_nice_cx_from_networkx(G_cosSim)\n",
    "G_cosSim_cx.set_name(d1_name+'_'+d2_name+'_NetColoc_subgraph_CosSim95') \n",
    "for node_id, node in G_cosSim_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_cosSim_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "        \n",
    "        \n",
    "\n",
    "# apply a template style (834b6ad4-d2ea-11eb-b666-0ac135e8bacf)\n",
    "G_cosSim_cx.apply_template('ndexbio.org','2cbed84b-e5c3-11eb-b666-0ac135e8bacf')\n",
    "\n",
    "network_uuid_NetColoc_CosSim = G_cosSim_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 25. Add 4 networks from above to network set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ndexbio.org/v2/networkset/414ae1c4-14e7-11ec-9e8e-0ac135e8bacf'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# append the datestring to the network set to guarantee uniqueness\n",
    "datestr = str(datetime.now())\n",
    "networkSetURL=ndex2.client.Ndex2(host=SERVER,username=USERNAME,password=PASSWORD).create_networkset(d1_name+'-'+d2_name+' network set: '+datestr,'network set for '+d1_name+'-'+d2_name+' NetColoc subgraph and systems map')\n",
    "networkSetURL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ndexbio.org/v2/networkset/414ae1c4-14e7-11ec-9e8e-0ac135e8bacf/members'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse out UUID from URL strings\n",
    "networkSetUUID = networkSetURL.split('/')[-1]\n",
    "networkSetUUID\n",
    "\n",
    "networkURLs = [network_uuid_NetColoc,network_uuid_NetColoc_CosSim,network_uuid_hier,network_uuid_hier_mouse_KO]\n",
    "networkUUIDs = [n.split('/')[-1] for n in networkURLs]\n",
    "\n",
    "ndex2.client.Ndex2(host=SERVER,username=USERNAME,password=PASSWORD).add_networks_to_networkset(networkSetUUID,\n",
    "                                                                                              networkUUIDs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
